{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bde4ba36",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c0337403",
   "metadata": {},
   "source": [
    "Question1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "41a58126-40d1-415a-9ea1-c5aea14333e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting py2neo\n",
      "  Using cached py2neo-2021.2.4-py2.py3-none-any.whl.metadata (9.9 kB)\n",
      "Requirement already satisfied: certifi in c:\\users\\yasser\\anaconda3\\envs\\my_env\\lib\\site-packages (from py2neo) (2024.6.2)\n",
      "Collecting interchange~=2021.0.4 (from py2neo)\n",
      "  Using cached interchange-2021.0.4-py2.py3-none-any.whl.metadata (1.9 kB)\n",
      "Requirement already satisfied: monotonic in c:\\users\\yasser\\anaconda3\\envs\\my_env\\lib\\site-packages (from py2neo) (1.6)\n",
      "Requirement already satisfied: packaging in c:\\users\\yasser\\anaconda3\\envs\\my_env\\lib\\site-packages (from py2neo) (23.2)\n",
      "Collecting pansi>=2020.7.3 (from py2neo)\n",
      "  Downloading pansi-2024.11.0-py2.py3-none-any.whl.metadata (3.1 kB)\n",
      "Requirement already satisfied: pygments>=2.0.0 in c:\\users\\yasser\\anaconda3\\envs\\my_env\\lib\\site-packages (from py2neo) (2.15.1)\n",
      "Requirement already satisfied: six>=1.15.0 in c:\\users\\yasser\\anaconda3\\envs\\my_env\\lib\\site-packages (from py2neo) (1.16.0)\n",
      "Requirement already satisfied: urllib3 in c:\\users\\yasser\\anaconda3\\envs\\my_env\\lib\\site-packages (from py2neo) (2.2.1)\n",
      "Requirement already satisfied: pytz in c:\\users\\yasser\\anaconda3\\envs\\my_env\\lib\\site-packages (from interchange~=2021.0.4->py2neo) (2024.1)\n",
      "Requirement already satisfied: pillow in c:\\users\\yasser\\anaconda3\\envs\\my_env\\lib\\site-packages (from pansi>=2020.7.3->py2neo) (10.3.0)\n",
      "Using cached py2neo-2021.2.4-py2.py3-none-any.whl (177 kB)\n",
      "Using cached interchange-2021.0.4-py2.py3-none-any.whl (28 kB)\n",
      "Downloading pansi-2024.11.0-py2.py3-none-any.whl (26 kB)\n",
      "Installing collected packages: pansi, interchange, py2neo\n",
      "Successfully installed interchange-2021.0.4 pansi-2024.11.0 py2neo-2021.2.4\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  WARNING: The script pansi-image.exe is installed in 'c:\\Users\\Yasser\\anaconda3\\envs\\my_env\\Scripts' which is not on PATH.\n",
      "  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.\n"
     ]
    }
   ],
   "source": [
    "!pip install py2neo # installing th py2neo for geenrating the graphs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a7879a7f-8cb6-4a6e-ada3-df19812ad5b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing the required librarries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5d2609da-900c-437e-ac54-8f5c44ae3e2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\Yasser\\\\Desktop\\\\ABIN_ASS_5'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cheking the current directory\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6bc92619-f8c8-4bb5-a6dd-9a00c5d1d1e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>HEAD</th>\n",
       "      <th>RELATION</th>\n",
       "      <th>TAIL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>DB00477</td>\n",
       "      <td>DRUG_TARGET</td>\n",
       "      <td>P21917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>DB11606</td>\n",
       "      <td>DRUG_TARGET</td>\n",
       "      <td>P04275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>DB15035</td>\n",
       "      <td>DRUG_TARGET</td>\n",
       "      <td>P04626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>DB09089</td>\n",
       "      <td>DRUG_TARGET</td>\n",
       "      <td>Q13936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>DB00331</td>\n",
       "      <td>DRUG_TARGET</td>\n",
       "      <td>P21695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2068667</th>\n",
       "      <td>2068667</td>\n",
       "      <td>DB14530</td>\n",
       "      <td>SIMILAR_TO</td>\n",
       "      <td>DB15533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2068668</th>\n",
       "      <td>2068668</td>\n",
       "      <td>DB14540</td>\n",
       "      <td>SIMILAR_TO</td>\n",
       "      <td>DB14544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2068669</th>\n",
       "      <td>2068669</td>\n",
       "      <td>DB14546</td>\n",
       "      <td>SIMILAR_TO</td>\n",
       "      <td>DB15533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2068670</th>\n",
       "      <td>2068670</td>\n",
       "      <td>DB14936</td>\n",
       "      <td>SIMILAR_TO</td>\n",
       "      <td>DB15381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2068671</th>\n",
       "      <td>2068671</td>\n",
       "      <td>DB15495</td>\n",
       "      <td>SIMILAR_TO</td>\n",
       "      <td>DB15496</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2068672 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Unnamed: 0     HEAD     RELATION     TAIL\n",
       "0                 0  DB00477  DRUG_TARGET   P21917\n",
       "1                 1  DB11606  DRUG_TARGET   P04275\n",
       "2                 2  DB15035  DRUG_TARGET   P04626\n",
       "3                 3  DB09089  DRUG_TARGET   Q13936\n",
       "4                 4  DB00331  DRUG_TARGET   P21695\n",
       "...             ...      ...          ...      ...\n",
       "2068667     2068667  DB14530   SIMILAR_TO  DB15533\n",
       "2068668     2068668  DB14540   SIMILAR_TO  DB14544\n",
       "2068669     2068669  DB14546   SIMILAR_TO  DB15533\n",
       "2068670     2068670  DB14936   SIMILAR_TO  DB15381\n",
       "2068671     2068671  DB15495   SIMILAR_TO  DB15496\n",
       "\n",
       "[2068672 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Reading the association.csv as dataframe\n",
    "#printing the association file\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "df=pd.read_csv(\"associations.csv\")\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7eb1bfaf-abfa-4add-8657-a7c1d441082c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2068672, 4)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking the shape of the association file\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d0a3fa8e-d24c-4a9f-b245-44ed4f545005",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['DRUG_TARGET', 'DRUG_CARRIER', 'DRUG_ENZYME', 'DRUG_TRANSPORTER',\n",
       "       'RELATED_GENETIC_DISORDER', 'DRUG_DISEASE_ASSOCIATION',\n",
       "       'PROTEIN_PATHWAY_ASSOCIATION', 'DISEASE_PATHWAY_ASSOCIATION',\n",
       "       'COMPLEX_IN_PATHWAY', 'DISEASE_GENETIC_DISORDER',\n",
       "       'MEMBER_OF_COMPLEX', 'PPI', 'DDI', 'DRUG_PATHWAY_ASSOCIATION',\n",
       "       'PROTEIN_DISEASE_ASSOCIATION', 'COMPLEX_TOP_LEVEL_PATHWAY', 'DPI',\n",
       "       'SIMILAR_TO'], dtype=object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#printing the unique relationships in the association file\n",
    "df['RELATION'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d8ece6ff-68c1-41c5-a434-65e1130fdaad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Unnamed: 0     HEAD     RELATION    TAIL\n",
      "0               0  DB00477  DRUG_TARGET  P21917\n",
      "1               1  DB11606  DRUG_TARGET  P04275\n",
      "2               2  DB15035  DRUG_TARGET  P04626\n",
      "3               3  DB09089  DRUG_TARGET  Q13936\n",
      "4               4  DB00331  DRUG_TARGET  P21695\n",
      "...           ...      ...          ...     ...\n",
      "15226       15226  DB04315  DRUG_TARGET  P18085\n",
      "15227       15227  DB06245  DRUG_TARGET  O15230\n",
      "15228       15228  DB00472  DRUG_TARGET  Q12809\n",
      "15229       15229  DB00675  DRUG_TARGET  P03372\n",
      "15230       15230  DB07664  DRUG_TARGET  P49761\n",
      "\n",
      "[15231 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "# Display rows where 'RELATION' column has the value 'TARGET'\n",
    "filtered_rows = df[df['RELATION'] == 'DRUG_TARGET']\n",
    "print(filtered_rows)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8c0e9e83-2900-4d6a-b70d-7d53bd39e3b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Unnamed: 0    HEAD                  RELATION        TAIL\n",
      "24152       24152  P61981  RELATED_GENETIC_DISORDER  MIM:617665\n",
      "24153       24153  Q14738  RELATED_GENETIC_DISORDER  MIM:616355\n",
      "24154       24154  P30153  RELATED_GENETIC_DISORDER  MIM:616362\n",
      "24155       24155  Q00005  RELATED_GENETIC_DISORDER  MIM:604326\n",
      "24156       24156  Q9H2F3  RELATED_GENETIC_DISORDER  MIM:607765\n",
      "...           ...     ...                       ...         ...\n",
      "28041       28041  Q9UDY2  RELATED_GENETIC_DISORDER  MIM:615878\n",
      "28042       28042  P60852  RELATED_GENETIC_DISORDER  MIM:615774\n",
      "28043       28043  Q05996  RELATED_GENETIC_DISORDER  MIM:618353\n",
      "28044       28044  P21754  RELATED_GENETIC_DISORDER  MIM:617712\n",
      "28045       28045  Q9HCJ5  RELATED_GENETIC_DISORDER  MIM:617865\n",
      "\n",
      "[3894 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "# Display rows where 'RELATION' column has the value 'RELATED_GENETIC_DISORDER'\n",
    "filtered_rows = df[df['RELATION'] == 'RELATED_GENETIC_DISORDER']\n",
    "print(filtered_rows)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "d9fb01fa-45e6-4b13-982b-23952a378589",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         Unnamed: 0     HEAD RELATION    TAIL\n",
      "2039965     2039965  DB11606      DPI  P04275\n",
      "2039966     2039966  DB15035      DPI  P04626\n",
      "2039967     2039967  DB01776      DPI  P01308\n",
      "2039968     2039968  DB05688      DPI  P23975\n",
      "2039969     2039969  DB13998      DPI  O14832\n",
      "...             ...      ...      ...     ...\n",
      "2067993     2067993  DB04272      DPI  Q9HB21\n",
      "2067994     2067994  DB04839      DPI  P11511\n",
      "2067995     2067995  DB07559      DPI  Q02127\n",
      "2067996     2067996  DB04856      DPI  P20815\n",
      "2067997     2067997  DB01229      DPI  Q16678\n",
      "\n",
      "[28033 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "# Display rows where 'RELATION' column has the value 'DPI'\n",
    "filtered_rows = df[df['RELATION'] == 'DPI']\n",
    "print(filtered_rows)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "24b399a0-6de3-4ceb-b63c-1c4a278b334e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Unnamed: 0    HEAD                  RELATION        TAIL\n",
      "24252       24252  Q9NX46  RELATED_GENETIC_DISORDER  MIM:618170\n",
      "24253       24253  P22570  RELATED_GENETIC_DISORDER  MIM:617717\n",
      "24254       24254  P12235  RELATED_GENETIC_DISORDER  MIM:617184\n",
      "24255       24255  Q8IUX7  RELATED_GENETIC_DISORDER  MIM:618000\n",
      "24256       24256  P51816  RELATED_GENETIC_DISORDER  MIM:309548\n",
      "...           ...     ...                       ...         ...\n",
      "24347       24347  Q4KMQ2  RELATED_GENETIC_DISORDER  MIM:262890\n",
      "24348       24348  P20594  RELATED_GENETIC_DISORDER  MIM:616255\n",
      "24349       24349  Q6UB99  RELATED_GENETIC_DISORDER  MIM:148050\n",
      "24350       24350  Q9UPS8  RELATED_GENETIC_DISORDER  MIM:188000\n",
      "24351       24351  P01008  RELATED_GENETIC_DISORDER  MIM:613118\n",
      "\n",
      "[100 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "print(filtered_rows.iloc[100:200, :])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "eb243d75-bba6-4c1f-bbc1-5d9f43bd724e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          DRUG_TARGET\n",
       "1          DRUG_TARGET\n",
       "2          DRUG_TARGET\n",
       "3          DRUG_TARGET\n",
       "4          DRUG_TARGET\n",
       "              ...     \n",
       "2068667     SIMILAR_TO\n",
       "2068668     SIMILAR_TO\n",
       "2068669     SIMILAR_TO\n",
       "2068670     SIMILAR_TO\n",
       "2068671     SIMILAR_TO\n",
       "Name: RELATION, Length: 2068672, dtype: object"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['RELATION']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "0eade85f-7c3e-4c97-93b7-e4a118cba21f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading the jsin files and loading them as list of dictioanries\n",
    "with open(\"diseases.json\") as file:\n",
    "  diseases=json.load(file)\n",
    "\n",
    "with open(\"drugs.json\") as file:\n",
    "  drugs=json.load(file)\n",
    "\n",
    "with open(\"proteins.json\") as file:\n",
    "  proteins=json.load(file)\n",
    " \n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "96f1373e-39b2-49e4-bf97-4a7803d0acf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of Disease:  11347\n",
      "Length of Proteins:  122750\n",
      "Length of Drugs:  42937\n"
     ]
    }
   ],
   "source": [
    "#Printing the shapes of all three json files\n",
    "print(\"Length of Disease: \",len(diseases))\n",
    "print(\"Length of Proteins: \",len(proteins))\n",
    "print(\"Length of Drugs: \",len(drugs))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "328ba2ac-85e9-4591-b33d-0736fc322085",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DISEASE': 'C000601856',\n",
       " 'NAME': 'Non ketotic hyperglycinemia syndrome',\n",
       " 'TYPE': 'SCR_DISEASE',\n",
       " 'DISEASE_SUPERGRP': nan}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Exploring the type of data for each of the json or given data\n",
    "diseases[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "da5916c5-38ba-44f3-91f4-baf200f8cd66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'PROTEIN': 'A0A023PXA5',\n",
       " 'FULL_NAME': 'Putative uncharacterized protein YAL019W-A {ECO:0000305}',\n",
       " 'NAME': 'YA19A',\n",
       " 'ORGANISM_CLASS': ['Eukaryota',\n",
       "  ' Fungi',\n",
       "  ' Dikarya',\n",
       "  ' Ascomycota',\n",
       "  ' Saccharomycotina',\n",
       "  ' Saccharomycetes',\n",
       "  ' Saccharomycetales',\n",
       "  ' Saccharomycetaceae',\n",
       "  ' Saccharomyces.'],\n",
       " 'OTHER_ID': {'$numberDouble': 'NaN'},\n",
       " 'RELATED_KEYWORD': {'$numberDouble': 'NaN'},\n",
       " 'RELATED_PUBMED_ID': ['pubmed:7731988', ' pubmed:24374639'],\n",
       " 'SHORT_NAME': {'$numberDouble': 'NaN'},\n",
       " 'SPECIES': 'YEAST',\n",
       " 'ACTIVE_SITE': {'$numberDouble': 'NaN'},\n",
       " 'BINDING_SITE': {'$numberDouble': 'NaN'},\n",
       " 'CONSERVED_SITE': {'$numberDouble': 'NaN'},\n",
       " 'DOMAIN': {'$numberDouble': 'NaN'},\n",
       " 'FAMILY': {'$numberDouble': 'NaN'},\n",
       " 'GO_BP': {'$numberDouble': 'NaN'},\n",
       " 'GO_CC': {'$numberDouble': 'NaN'},\n",
       " 'GO_MF': {'$numberDouble': 'NaN'},\n",
       " 'HOMOLOGOUS_SUPERFAMILY': {'$numberDouble': 'NaN'},\n",
       " 'PROTEIN_EXPRESSED_IN': {'$numberDouble': 'NaN'},\n",
       " 'PTM': {'$numberDouble': 'NaN'},\n",
       " 'REPEAT': {'$numberDouble': 'NaN'}}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Exploring the type of data for each of the json or given data\n",
    "proteins[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "924e25c6-44c5-401a-b59c-ed2c3f1bf4d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DRUG': 'DB00040',\n",
       " 'ALTERNATIVE_PARENT': nan,\n",
       " 'CLASS': 'Carboxylic_Acids_and_Derivatives',\n",
       " 'DIRECT_PARENT': 'Peptides',\n",
       " 'KINGDOM': 'Organic_Compounds',\n",
       " 'NAME': 'Glucagon',\n",
       " 'PRODUCT': ['Baqsimi',\n",
       "  ' Glucag0n',\n",
       "  ' GlucaGen',\n",
       "  ' GlucaGen_HypoKit',\n",
       "  ' Glucagon',\n",
       "  ' Glucagon_Ampoule_666_1mg',\n",
       "  ' Gvoke_HypoPen_0_5_mg_Auto_Injector',\n",
       "  ' Gvoke_HypoPen_1_mg_Auto_Injector',\n",
       "  ' Gvoke_PFS_0_5_mg_Pre_filled_Syringe',\n",
       "  ' Gvoke_PFS_1_mg_Pre_filled_Syringe'],\n",
       " 'PRODUCT_STAGE': 'approved',\n",
       " 'PUBMED_ARTICLE': ['12626323', ' 24370975', ' 659632'],\n",
       " 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues',\n",
       " 'SUBSTITUENT': nan,\n",
       " 'SUPERCLASS': 'Organic_Acids',\n",
       " 'SYNONYM': ['Glucagon',\n",
       "  ' Glucagon__recombinant_dna_origin_',\n",
       "  ' Glucagon_recombinant',\n",
       "  ' Glucagon__human',\n",
       "  ' Glucagon__porcine',\n",
       "  ' Glucagone'],\n",
       " 'TYPE': 'DRUG',\n",
       " 'DRUG_ATC_CODE': ['ATC:H04AA',\n",
       "  ' ATC:H04AA01',\n",
       "  ' ATC:H04A',\n",
       "  ' ATC:H',\n",
       "  ' ATC:H04'],\n",
       " 'DRUG_INDICATION_ASSOCIATION': nan,\n",
       " 'DRUG_SIDEEFFECT_ASSOCIATION': nan,\n",
       " 'DRUGBANK_NAME': nan,\n",
       " 'C_SMILES': None,\n",
       " 'MolecularWeight': nan,\n",
       " 'LogP': nan,\n",
       " 'NumHAcceptors': nan,\n",
       " 'NumHDonors': nan,\n",
       " 'TPSA': nan,\n",
       " 'NumRotatableBonds': nan,\n",
       " 'LipinskiViolationCount': nan,\n",
       " 'SOURCE': 'BIOKG',\n",
       " 'DRUGINCH': None,\n",
       " 'DRUGINCHKE': None,\n",
       " 'DRUGCOMP': nan,\n",
       " 'DRUGINKE': nan,\n",
       " 'COMPCLAS': nan}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Exploring the type of data for each of the json or given data\n",
    "drugs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "74a6113d-3254-40ea-86ea-90ea88f58c38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered_associations saved as csv file\n",
      "Filtered identifiers saved as JSON files.\n"
     ]
    }
   ],
   "source": [
    "\"\"\" Here we are cleaning the data ,Basically \n",
    "Clean the data by removing the edges where the node metadata information\n",
    "is not present. Simply prune any nodes that are not part of the associations\n",
    "file and vice versa donâ€™t put any edges in the associations file if there is no\n",
    "node metadata available.\"\"\"\n",
    "import pandas as pd\n",
    "\n",
    "# Define the filter function\n",
    "def filter_identifiers(items_list, key):\n",
    "    \"\"\"\n",
    "    Extract identifiers from a list of dictionaries based on the specified key.\n",
    "\n",
    "    Args:\n",
    "        items_list (list): List of dictionaries containing identifiers.\n",
    "        key (str): Key to extract identifiers.\n",
    "\n",
    "    Returns:\n",
    "        set: A set of identifiers for the specified key.\n",
    "    \"\"\"\n",
    "    return {item[key] for item in items_list if key in item}\n",
    "\n",
    "# Sample data: lists of dictionaries for diseases, proteins, and drugs\n",
    "\n",
    "# Filter identifiers\n",
    "disease_identifiers = filter_identifiers(diseases, \"DISEASE\")\n",
    "protein_identifiers = filter_identifiers(proteins, \"PROTEIN\")\n",
    "drug_identifiers = filter_identifiers(drugs, \"DRUG\")\n",
    "\n",
    "# Combine all identifiers into one set\n",
    "valid_identifiers = disease_identifiers | protein_identifiers | drug_identifiers\n",
    "\n",
    "# Step 1: Read the associations.csv\n",
    "association_file_path = \"associations.csv\"  # Replace with the actual path\n",
    "df = pd.read_csv(association_file_path)\n",
    "\n",
    "# Step 2: Filter associations based on valid_identifiers\n",
    "filtered_df = df[\n",
    "    (df['HEAD'].isin(valid_identifiers)) &\n",
    "    (df['TAIL'].isin(valid_identifiers))\n",
    "]\n",
    "\n",
    "# Save the filtered associations\n",
    "filtered_associations_path = \"filtered_associations.csv\"\n",
    "filtered_df.to_csv(filtered_associations_path, index=False)\n",
    "print(\"Filtered_associations saved as csv file\")\n",
    "print(f\"Filtered associations saved to {filtered_associations_path}.\")\n",
    "\n",
    "# Step 3: Filter dictionaries based on relations in filtered associations\n",
    "# Get unique HEAD and TAIL values in the filtered associations\n",
    "unique_identifiers_in_associations = set(filtered_df['HEAD']).union(set(filtered_df['TAIL']))\n",
    "\n",
    "# Filter diseases\n",
    "filtered_diseases = [d for d in diseases if d['DISEASE'] in unique_identifiers_in_associations]\n",
    "\n",
    "# Filter proteins\n",
    "filtered_proteins = [p for p in proteins if p['PROTEIN'] in unique_identifiers_in_associations]\n",
    "\n",
    "# Filter drugs\n",
    "filtered_drugs = [d for d in drugs if d['DRUG'] in unique_identifiers_in_associations]\n",
    "\n",
    "# Save the filtered dictionaries\n",
    "\n",
    "\n",
    "\n",
    "def clean_data(proteins, drugs, diseases):\n",
    "    # Function to clean invalid property values\n",
    "    def sanitize_value(value):\n",
    "        if isinstance(value, dict) and '$numberDouble' in value:\n",
    "            return None  # Replace invalid maps with None\n",
    "        return value\n",
    "\n",
    "    # Clean proteins\n",
    "    for protein in proteins:\n",
    "        for key in protein:\n",
    "            protein[key] = sanitize_value(protein[key])\n",
    "\n",
    "    # Clean drugs\n",
    "    for drug in drugs:\n",
    "        for key in drug:\n",
    "            drug[key] = sanitize_value(drug[key])\n",
    "\n",
    "    # Clean diseases\n",
    "    for disease in diseases:\n",
    "        for key in disease:\n",
    "            disease[key] = sanitize_value(disease[key])\n",
    "\n",
    "    return proteins, drugs, diseases\n",
    "filtered_proteins, filtered_drugs, filtered_diseases = clean_data(filtered_proteins, filtered_drugs, filtered_diseases)\n",
    "\n",
    "\n",
    "\n",
    "with open(\"filtered_diseases.json\", \"w\") as f:\n",
    "    json.dump(filtered_diseases, f)\n",
    "\n",
    "with open(\"filtered_proteins.json\", \"w\") as f:\n",
    "    json.dump(filtered_proteins, f)\n",
    "\n",
    "with open(\"filtered_drugs.json\", \"w\") as f:\n",
    "    json.dump(filtered_drugs, f)\n",
    "print(\"filtered_associations saved as csv file\")\n",
    "print(\"Filtered identifiers saved as JSON files.\")\n",
    "\n",
    "# Saving all the data as json files and filtered_ssociations as csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "1cacca69-320b-4b51-b3ab-d8be0cb010b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered Associations\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>HEAD</th>\n",
       "      <th>RELATION</th>\n",
       "      <th>TAIL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>DB00477</td>\n",
       "      <td>DRUG_TARGET</td>\n",
       "      <td>P21917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>DB11606</td>\n",
       "      <td>DRUG_TARGET</td>\n",
       "      <td>P04275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>DB15035</td>\n",
       "      <td>DRUG_TARGET</td>\n",
       "      <td>P04626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>DB09089</td>\n",
       "      <td>DRUG_TARGET</td>\n",
       "      <td>Q13936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>DB00331</td>\n",
       "      <td>DRUG_TARGET</td>\n",
       "      <td>P21695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1676899</th>\n",
       "      <td>2068667</td>\n",
       "      <td>DB14530</td>\n",
       "      <td>SIMILAR_TO</td>\n",
       "      <td>DB15533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1676900</th>\n",
       "      <td>2068668</td>\n",
       "      <td>DB14540</td>\n",
       "      <td>SIMILAR_TO</td>\n",
       "      <td>DB14544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1676901</th>\n",
       "      <td>2068669</td>\n",
       "      <td>DB14546</td>\n",
       "      <td>SIMILAR_TO</td>\n",
       "      <td>DB15533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1676902</th>\n",
       "      <td>2068670</td>\n",
       "      <td>DB14936</td>\n",
       "      <td>SIMILAR_TO</td>\n",
       "      <td>DB15381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1676903</th>\n",
       "      <td>2068671</td>\n",
       "      <td>DB15495</td>\n",
       "      <td>SIMILAR_TO</td>\n",
       "      <td>DB15496</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1676904 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Unnamed: 0     HEAD     RELATION     TAIL\n",
       "0                 0  DB00477  DRUG_TARGET   P21917\n",
       "1                 1  DB11606  DRUG_TARGET   P04275\n",
       "2                 2  DB15035  DRUG_TARGET   P04626\n",
       "3                 3  DB09089  DRUG_TARGET   Q13936\n",
       "4                 4  DB00331  DRUG_TARGET   P21695\n",
       "...             ...      ...          ...      ...\n",
       "1676899     2068667  DB14530   SIMILAR_TO  DB15533\n",
       "1676900     2068668  DB14540   SIMILAR_TO  DB14544\n",
       "1676901     2068669  DB14546   SIMILAR_TO  DB15533\n",
       "1676902     2068670  DB14936   SIMILAR_TO  DB15381\n",
       "1676903     2068671  DB15495   SIMILAR_TO  DB15496\n",
       "\n",
       "[1676904 rows x 4 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df=pd.read_csv(\"filtered_associations.csv\")\n",
    "print(\"Filtered Associations\")\n",
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "f569e629-56d0-4400-bb9d-c2623e0a3e49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered_associations shape:  (1676904, 4)\n"
     ]
    }
   ],
   "source": [
    "#Now we are evaluating the size of the data before cleaning and after the cleaning \n",
    "print(\"Filtered_associations shape: \",filtered_df.shape)\n",
    "# This is filtered_association.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "99201223-0617-43be-8c96-62e849d9ad77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "associations shape:  (2068672, 4)\n"
     ]
    }
   ],
   "source": [
    "print(\"associations shape: \",df.shape)\n",
    "#This is original associations.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "fff2021e-d995-49b6-af8a-c4e1cd1eba92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total unique nodes: 53210\n"
     ]
    }
   ],
   "source": [
    "unique_nodes = set(filtered_df[\"HEAD\"]).union(set(filtered_df[\"TAIL\"]))\n",
    "print(f\"Total unique nodes: {len(unique_nodes)}\")\n",
    "# Total unique nodes after cleaning the three json files drugs,diseases,proteins\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "649b586f-8f22-4966-9724-fb6cb151569e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['DRUG_TARGET', 'DRUG_CARRIER', 'DRUG_ENZYME', 'DRUG_TRANSPORTER',\n",
       "       'DRUG_DISEASE_ASSOCIATION', 'PPI', 'DDI',\n",
       "       'PROTEIN_DISEASE_ASSOCIATION', 'DPI', 'SIMILAR_TO'], dtype=object)"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df['RELATION'].unique()\n",
    "#These are the relationships after cleaning the data in the the assocations.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "616c9de1-4663-4bb2-9aaf-0bb7833409be",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Here we are reading the cleaned data\n",
    "with open(\"filtered_diseases.json\") as file:\n",
    "  filtered_diseases=json.load(file)\n",
    "    \n",
    "with open(\"filtered_drugs.json\") as file:\n",
    "  filtered_drugs=json.load(file)\n",
    "\n",
    "with open(\"filtered_proteins.json\") as file:\n",
    "  filtered_proteins=json.load(file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "2feb6973-c0cc-4af9-9d82-65fdf0f48722",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of Filtered Disease:  5806\n",
      "Length of Filtered Proteins:  38618\n",
      "Length of Filtered Drugs:  8786\n"
     ]
    }
   ],
   "source": [
    "# Here we are printing the shapes of different data aftter filtering\n",
    "print(\"Length of Filtered Disease: \",len(filtered_diseases))\n",
    "print(\"Length of Filtered Proteins: \",len(filtered_proteins))\n",
    "print(\"Length of Filtered Drugs: \",len(filtered_drugs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "63390ada-41ea-4c68-a397-c2b19e0d5214",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of Disease:  11347\n",
      "Length of Proteins:  122750\n",
      "Length of Drugs:  42937\n"
     ]
    }
   ],
   "source": [
    "# Here we are printing the shapes of different data before filtering\n",
    "print(\"Length of Disease: \",len(diseases))\n",
    "print(\"Length of Proteins: \",len(proteins))\n",
    "print(\"Length of Drugs: \",len(drugs))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ce672db0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "#initialing the unique keys for neo4j\n",
    "\n",
    "url = \"bolt://localhost:7687\"\n",
    "username = \"neo4j\"\n",
    "password = \"password\"  # Replace with your actual password\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e124144-865f-47d7-a541-541a8d9974e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from neo4j import GraphDatabase\n",
    "#Initialis the graph parametres and assiging some fucntions\n",
    "class Neo4jManager:\n",
    "    def __init__(self, uri, username, password):\n",
    "        self.driver = GraphDatabase.driver(uri, auth=(username, password))\n",
    "\n",
    "    def close(self):\n",
    "        self.driver.close()\n",
    "\n",
    "    def run_query(self, query, parameters=None):\n",
    "        with self.driver.session() as session:\n",
    "            return session.run(query, parameters)\n",
    "\n",
    "    def get_graph_stats(self):\n",
    "        query = \"\"\"\n",
    "        MATCH (n)\n",
    "        WITH COUNT(n) AS total_nodes\n",
    "        MATCH ()-[r]->()\n",
    "        RETURN total_nodes, COUNT(r) AS total_edges;\n",
    "        \"\"\"\n",
    "        with self.driver.session() as session:\n",
    "            result = session.run(query)\n",
    "            return result.single()\n",
    "\n",
    "    def fetch_node_by_id(self, node_id):\n",
    "        query = \"\"\"\n",
    "        MATCH (n)\n",
    "        WHERE ID(n) = $node_id\n",
    "        RETURN n\n",
    "        \"\"\"\n",
    "        parameters = {\"node_id\": node_id}\n",
    "        with self.driver.session() as session:\n",
    "            result = session.run(query, parameters)\n",
    "            return result.single()  # Return the node found\n",
    "\n",
    "\n",
    "# Example usage:\n",
    "neo4j_manager = Neo4jManager(url, username, password)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "c78e1948-18d6-47c6-ba76-e003c133273d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Graph created successfully\n"
     ]
    }
   ],
   "source": [
    "#INSERTING THE DATA INTO THE GRAPH\n",
    "def insert_nodes_and_relationships():\n",
    "    # Insert Disease nodes\n",
    "    for disease in filtered_diseases:\n",
    "        query = \"\"\"\n",
    "        MERGE (d:Disease {id: $id})\n",
    "        SET d += $properties\n",
    "        \"\"\"\n",
    "        properties = {key: disease[key] for key in disease if key != 'DISEASE'}\n",
    "        parameters = {\"id\": disease['DISEASE'], \"properties\": properties}\n",
    "        neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "    # Insert Drug nodes\n",
    "    for drug in filtered_drugs:\n",
    "        query = \"\"\"\n",
    "        MERGE (d:Drug {id: $id})\n",
    "        SET d += $properties\n",
    "        \"\"\"\n",
    "        properties = {key: drug[key] for key in drug if key != 'DRUG'}\n",
    "        parameters = {\"id\": drug['DRUG'], \"properties\": properties}\n",
    "        neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "    # Insert Protein nodes\n",
    "    for protein in filtered_proteins:\n",
    "        query = \"\"\"\n",
    "        MERGE (p:Protein {id: $id})\n",
    "        SET p += $properties\n",
    "        \"\"\"\n",
    "        properties = {key: protein[key] for key in protein if key != 'PROTEIN'}\n",
    "        parameters = {\"id\": protein['PROTEIN'], \"properties\": properties}\n",
    "        neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "    # Insert Relationships based on the RELATION column\n",
    "    for _, row in filtered_df.iterrows():\n",
    "        relation = row['RELATION']\n",
    "        head = row['HEAD']\n",
    "        tail = row['TAIL']\n",
    "\n",
    "        # Handle all specified relations dynamically\n",
    "        if relation == 'DRUG_TARGET':\n",
    "            query = \"\"\"\n",
    "            MATCH (d:Drug {id: $drug_id}), (p:Protein {id: $protein_id})\n",
    "            MERGE (d)-[:DRUG_TARGET]->(p)\n",
    "            \"\"\"\n",
    "            parameters = {\"drug_id\": head, \"protein_id\": tail}\n",
    "            neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "        elif relation == 'DRUG_CARRIER':\n",
    "            query = \"\"\"\n",
    "            MATCH (d:Drug {id: $drug_id}), (p:Protein {id: $protein_id})\n",
    "            MERGE (d)-[:DRUG_CARRIER]->(p)\n",
    "            \"\"\"\n",
    "            parameters = {\"drug_id\": head, \"protein_id\": tail}\n",
    "            neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "        elif relation == 'DRUG_ENZYME':\n",
    "            query = \"\"\"\n",
    "            MATCH (d:Drug {id: $drug_id}), (p:Protein {id: $protein_id})\n",
    "            MERGE (d)-[:DRUG_ENZYME]->(p)\n",
    "            \"\"\"\n",
    "            parameters = {\"drug_id\": head, \"protein_id\": tail}\n",
    "            neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "        elif relation == 'DRUG_TRANSPORTER':\n",
    "            query = \"\"\"\n",
    "            MATCH (d:Drug {id: $drug_id}), (p:Protein {id: $protein_id})\n",
    "            MERGE (d)-[:DRUG_TRANSPORTER]->(p)\n",
    "            \"\"\"\n",
    "            parameters = {\"drug_id\": head, \"protein_id\": tail}\n",
    "            neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "        elif relation == 'DRUG_DISEASE_ASSOCIATION':\n",
    "            query = \"\"\"\n",
    "            MATCH (d:Drug {id: $drug_id}), (dis:Disease {id: $disease_id})\n",
    "            MERGE (d)-[:DRUG_DISEASE_ASSOCIATION]->(dis)\n",
    "            \"\"\"\n",
    "            parameters = {\"drug_id\": head, \"disease_id\": tail}\n",
    "            neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "        elif relation == 'PPI':\n",
    "            query = \"\"\"\n",
    "            MATCH (p1:Protein {id: $protein1_id}), (p2:Protein {id: $protein2_id})\n",
    "            MERGE (p1)-[:PPI]->(p2)\n",
    "            \"\"\"\n",
    "            parameters = {\"protein1_id\": head, \"protein2_id\": tail}\n",
    "            neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "        elif relation == 'DDI':\n",
    "            query = \"\"\"\n",
    "            MATCH (d1:Drug {id: $drug1_id}), (d2:Drug {id: $drug2_id})\n",
    "            MERGE (d1)-[:DDI]->(d2)\n",
    "            \"\"\"\n",
    "            parameters = {\"drug1_id\": head, \"drug2_id\": tail}\n",
    "            neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "        elif relation == 'PROTEIN_DISEASE_ASSOCIATION':\n",
    "            query = \"\"\"\n",
    "            MATCH (p:Protein {id: $protein_id}), (d:Disease {id: $disease_id})\n",
    "            MERGE (p)-[:PROTEIN_DISEASE_ASSOCIATION]->(d)\n",
    "            \"\"\"\n",
    "            parameters = {\"protein_id\": head, \"disease_id\": tail}\n",
    "            neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "        elif relation == 'DPI':\n",
    "            query = \"\"\"\n",
    "            MATCH (p:Drug {id: $drug_id}), (d:Protein {id: $protein_id})\n",
    "            MERGE (p)-[:DPI]->(d)\n",
    "            \"\"\"\n",
    "            parameters = {\"drug_id\": head, \"protein_id\": tail}\n",
    "            neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "        elif relation == 'SIMILAR_TO':\n",
    "            query = \"\"\"\n",
    "            MATCH (d1:Drug {id: $drug1_id}), (d2:Drug {id: $drug2_id})\n",
    "            MERGE (d1)-[:SIMILAR_TO]->(d2)\n",
    "            \"\"\"\n",
    "            parameters = {\"drug1_id\": head, \"drug2_id\": tail}\n",
    "            neo4j_manager.run_query(query, parameters)\n",
    "\n",
    "# Insert data\n",
    "insert_nodes_and_relationships()\n",
    "\n",
    "print(\"Graph created successfully\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "2947ee12-74c0-4697-9c7e-31f574f437e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'PROTEIN': 'Q12229',\n",
       " 'FULL_NAME': 'UBX domain-containing protein 3',\n",
       " 'NAME': 'UBX3',\n",
       " 'ORGANISM_CLASS': ['Eukaryota',\n",
       "  ' Fungi',\n",
       "  ' Dikarya',\n",
       "  ' Ascomycota',\n",
       "  ' Saccharomycotina',\n",
       "  ' Saccharomycetes',\n",
       "  ' Saccharomycetales',\n",
       "  ' Saccharomycetaceae',\n",
       "  ' Saccharomyces.'],\n",
       " 'OTHER_ID': None,\n",
       " 'RELATED_KEYWORD': ['Cytoplasm',\n",
       "  ' Reference proteome',\n",
       "  ' Ubl conjugation pathway'],\n",
       " 'RELATED_PUBMED_ID': ['pubmed:8923743',\n",
       "  ' pubmed:9169867',\n",
       "  ' pubmed:24374639',\n",
       "  ' pubmed:14562095',\n",
       "  ' pubmed:14562106',\n",
       "  ' pubmed:15258615',\n",
       "  ' pubmed:25078903',\n",
       "  ' pubmed:29355480'],\n",
       " 'SHORT_NAME': None,\n",
       " 'SPECIES': 'YEAST',\n",
       " 'ACTIVE_SITE': None,\n",
       " 'BINDING_SITE': None,\n",
       " 'CONSERVED_SITE': None,\n",
       " 'DOMAIN': ['IPR006577', ' IPR001012'],\n",
       " 'FAMILY': None,\n",
       " 'GO_BP': 'GO:0030276',\n",
       " 'GO_CC': ['GO:0030136', ' GO:0005737', ' GO:0044695', ' GO:0005783'],\n",
       " 'GO_MF': ['GO:0072583', ' GO:0033523'],\n",
       " 'HOMOLOGOUS_SUPERFAMILY': ['IPR036249', ' IPR029071'],\n",
       " 'PROTEIN_EXPRESSED_IN': None,\n",
       " 'PTM': None,\n",
       " 'REPEAT': None}"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#exploring the filtered data\n",
    "filtered_proteins[17785]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "05edf578-d2ca-4ca8-ba51-0e922b32a7c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Nodes: 53210\n",
      "Total Edges: 1676903\n"
     ]
    }
   ],
   "source": [
    "# Instantiate the manager\n",
    "neo4j_manager = Neo4jManager(url, username, password)\n",
    "\n",
    "# Get total nodes and edges\n",
    "stats = neo4j_manager.get_graph_stats()\n",
    "if stats:\n",
    "    print(f\"Total Nodes: {stats['total_nodes']}\")\n",
    "    print(f\"Total Edges: {stats['total_edges']}\")\n",
    "\n",
    "neo4j_manager.close()\n",
    "\n",
    "#HERE ARE THE TOTAL NODES AND TOTAL RELATIONSHIPS INTHE GRAPH FORMED\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "e383817c-a1fb-4d39-8f0a-13ec9909c245",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered_associations shape:  (1676904, 4)\n"
     ]
    }
   ],
   "source": [
    "# comparing the shape of filtered assocaitons and the relationship in the graph formed\n",
    "print(\"Filtered_associations shape: \",filtered_df.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "f878f168-61d0-42da-9f4d-6d61e6e838ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Nodes: 53210\n",
      "Node Categories: [['Disease'], ['Drug'], ['Protein']]\n",
      "Number of Node Categories: 3\n",
      "Total Relationships: 1676903\n",
      "Relationship Types: ['DRUG_TARGET', 'DRUG_DISEASE_ASSOCIATION', 'DDI', 'DPI', 'DRUG_ENZYME', 'SIMILAR_TO', 'DRUG_TRANSPORTER', 'DRUG_CARRIER', 'PROTEIN_DISEASE_ASSOCIATION', 'PPI']\n",
      "Number of Relationship Types: 10\n",
      "Sample Relationships:\n",
      "DB00040 [['Drug']] -[DRUG_TARGET]-> O95838 [['Protein']]\n",
      "DB00040 [['Drug']] -[DRUG_TARGET]-> P43220 [['Protein']]\n",
      "DB00040 [['Drug']] -[DRUG_TARGET]-> P47871 [['Protein']]\n",
      "DB00040 [['Drug']] -[DRUG_DISEASE_ASSOCIATION]-> D000419 [['Disease']]\n",
      "DB00040 [['Drug']] -[DRUG_DISEASE_ASSOCIATION]-> D001145 [['Disease']]\n",
      "DB00040 [['Drug']] -[DRUG_DISEASE_ASSOCIATION]-> D001919 [['Disease']]\n",
      "DB00040 [['Drug']] -[DRUG_DISEASE_ASSOCIATION]-> D003928 [['Disease']]\n",
      "DB00040 [['Drug']] -[DRUG_DISEASE_ASSOCIATION]-> D005355 [['Disease']]\n",
      "DB00040 [['Drug']] -[DRUG_DISEASE_ASSOCIATION]-> D006327 [['Disease']]\n",
      "DB00040 [['Drug']] -[DRUG_DISEASE_ASSOCIATION]-> D006331 [['Disease']]\n"
     ]
    }
   ],
   "source": [
    "#EXPLORING THE FORMED GRAPH\n",
    "driver = GraphDatabase.driver(url, auth=(username, password))\n",
    "\n",
    "def fetch_data():\n",
    "    with driver.session() as session:\n",
    "        # Total nodes, their names, and categories\n",
    "        nodes_result = session.run(\"\"\"\n",
    "            MATCH (n)\n",
    "            RETURN COUNT(n) AS TotalNodes, \n",
    "                   COLLECT(n.id) AS NodeNames,\n",
    "                   COLLECT(DISTINCT labels(n)) AS NodeCategories\n",
    "            LIMIT 10\n",
    "        \"\"\")\n",
    "        for record in nodes_result:\n",
    "            print(f\"Total Nodes: {record['TotalNodes']}\")\n",
    "            print(f\"Node Categories: {record['NodeCategories']}\")\n",
    "            print(f\"Number of Node Categories: {len(record['NodeCategories'])}\")\n",
    "        \n",
    "        # Total relationships and their types\n",
    "        relationships_result = session.run(\"\"\"\n",
    "            MATCH ()-[r]->()\n",
    "            RETURN COUNT(r) AS TotalRelationships, \n",
    "                   COLLECT(DISTINCT type(r)) AS RelationshipTypes\n",
    "        \"\"\")\n",
    "        for record in relationships_result:\n",
    "            print(f\"Total Relationships: {record['TotalRelationships']}\")\n",
    "            print(\"Relationship Types:\", record['RelationshipTypes'])\n",
    "            print(\"Number of Relationship Types:\", len(record['RelationshipTypes']))\n",
    "        \n",
    "        # Sample relationships\n",
    "        sample_result = session.run(\"\"\"\n",
    "            MATCH (n)-[r]->(m)\n",
    "            RETURN n.id AS StartNode, type(r) AS Relationship, m.id AS EndNode, labels(n) AS StartLabels, labels(m) AS EndLabels\n",
    "            LIMIT 10\n",
    "        \"\"\")\n",
    "        print(\"Sample Relationships:\")\n",
    "        for record in sample_result:\n",
    "            print(f\"{record['StartNode']} [{record['StartLabels']}] -[{record['Relationship']}]-> {record['EndNode']} [{record['EndLabels']}]\")\n",
    "\n",
    "# Fetch data\n",
    "fetch_data()\n",
    "\n",
    "# Close the driver connection\n",
    "driver.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "25f4d76d-dcbb-4388-a0bc-53258b6693d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n"
     ]
    }
   ],
   "source": [
    "print(len(filtered_df['RELATION'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8a794325-90b3-4569-a256-bda57c75aecf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting neo4j\n",
      "  Using cached neo4j-5.27.0-py3-none-any.whl.metadata (5.9 kB)\n",
      "Requirement already satisfied: pytz in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from neo4j) (2024.2)\n",
      "Using cached neo4j-5.27.0-py3-none-any.whl (301 kB)\n",
      "Installing collected packages: neo4j\n",
      "Successfully installed neo4j-5.27.0\n"
     ]
    }
   ],
   "source": [
    "!pip install neo4j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35a6bf4d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d568d621",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d580d661",
   "metadata": {},
   "source": [
    "QUESTION2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66aefb89-ede6-4b25-9cfd-ac7308511f12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting sentence-transformers\n",
      "  Downloading sentence_transformers-3.2.1-py3-none-any.whl.metadata (10 kB)\n",
      "Collecting transformers<5.0.0,>=4.41.0 (from sentence-transformers)\n",
      "  Downloading transformers-4.46.3-py3-none-any.whl.metadata (44 kB)\n",
      "Collecting tqdm (from sentence-transformers)\n",
      "  Downloading tqdm-4.67.1-py3-none-any.whl.metadata (57 kB)\n",
      "Collecting torch>=1.11.0 (from sentence-transformers)\n",
      "  Downloading torch-2.4.1-cp38-cp38-win_amd64.whl.metadata (27 kB)\n",
      "Collecting scikit-learn (from sentence-transformers)\n",
      "  Downloading scikit_learn-1.3.2-cp38-cp38-win_amd64.whl.metadata (11 kB)\n",
      "Collecting scipy (from sentence-transformers)\n",
      "  Downloading scipy-1.10.1-cp38-cp38-win_amd64.whl.metadata (58 kB)\n",
      "Collecting huggingface-hub>=0.20.0 (from sentence-transformers)\n",
      "  Downloading huggingface_hub-0.26.5-py3-none-any.whl.metadata (13 kB)\n",
      "Collecting Pillow (from sentence-transformers)\n",
      "  Downloading pillow-10.4.0-cp38-cp38-win_amd64.whl.metadata (9.3 kB)\n",
      "Collecting filelock (from huggingface-hub>=0.20.0->sentence-transformers)\n",
      "  Downloading filelock-3.16.1-py3-none-any.whl.metadata (2.9 kB)\n",
      "Collecting fsspec>=2023.5.0 (from huggingface-hub>=0.20.0->sentence-transformers)\n",
      "  Downloading fsspec-2024.10.0-py3-none-any.whl.metadata (11 kB)\n",
      "Requirement already satisfied: packaging>=20.9 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (24.2)\n",
      "Collecting pyyaml>=5.1 (from huggingface-hub>=0.20.0->sentence-transformers)\n",
      "  Downloading PyYAML-6.0.2-cp38-cp38-win_amd64.whl.metadata (2.1 kB)\n",
      "Collecting requests (from huggingface-hub>=0.20.0->sentence-transformers)\n",
      "  Using cached requests-2.32.3-py3-none-any.whl.metadata (4.6 kB)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (4.12.2)\n",
      "Collecting sympy (from torch>=1.11.0->sentence-transformers)\n",
      "  Downloading sympy-1.13.3-py3-none-any.whl.metadata (12 kB)\n",
      "Collecting networkx (from torch>=1.11.0->sentence-transformers)\n",
      "  Downloading networkx-3.1-py3-none-any.whl.metadata (5.3 kB)\n",
      "Collecting jinja2 (from torch>=1.11.0->sentence-transformers)\n",
      "  Using cached jinja2-3.1.4-py3-none-any.whl.metadata (2.6 kB)\n",
      "Requirement already satisfied: colorama in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from tqdm->sentence-transformers) (0.4.6)\n",
      "Requirement already satisfied: numpy>=1.17 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (1.24.4)\n",
      "Collecting regex!=2019.12.17 (from transformers<5.0.0,>=4.41.0->sentence-transformers)\n",
      "  Downloading regex-2024.11.6-cp38-cp38-win_amd64.whl.metadata (41 kB)\n",
      "Collecting tokenizers<0.21,>=0.20 (from transformers<5.0.0,>=4.41.0->sentence-transformers)\n",
      "  Downloading tokenizers-0.20.3-cp38-none-win_amd64.whl.metadata (6.9 kB)\n",
      "Collecting safetensors>=0.4.1 (from transformers<5.0.0,>=4.41.0->sentence-transformers)\n",
      "  Downloading safetensors-0.4.5-cp38-none-win_amd64.whl.metadata (3.9 kB)\n",
      "Collecting joblib>=1.1.1 (from scikit-learn->sentence-transformers)\n",
      "  Using cached joblib-1.4.2-py3-none-any.whl.metadata (5.4 kB)\n",
      "Collecting threadpoolctl>=2.0.0 (from scikit-learn->sentence-transformers)\n",
      "  Using cached threadpoolctl-3.5.0-py3-none-any.whl.metadata (13 kB)\n",
      "Collecting MarkupSafe>=2.0 (from jinja2->torch>=1.11.0->sentence-transformers)\n",
      "  Downloading MarkupSafe-2.1.5-cp38-cp38-win_amd64.whl.metadata (3.1 kB)\n",
      "Collecting charset-normalizer<4,>=2 (from requests->huggingface-hub>=0.20.0->sentence-transformers)\n",
      "  Downloading charset_normalizer-3.4.0-cp38-cp38-win_amd64.whl.metadata (34 kB)\n",
      "Collecting idna<4,>=2.5 (from requests->huggingface-hub>=0.20.0->sentence-transformers)\n",
      "  Using cached idna-3.10-py3-none-any.whl.metadata (10 kB)\n",
      "Collecting urllib3<3,>=1.21.1 (from requests->huggingface-hub>=0.20.0->sentence-transformers)\n",
      "  Using cached urllib3-2.2.3-py3-none-any.whl.metadata (6.5 kB)\n",
      "Collecting certifi>=2017.4.17 (from requests->huggingface-hub>=0.20.0->sentence-transformers)\n",
      "  Using cached certifi-2024.8.30-py3-none-any.whl.metadata (2.2 kB)\n",
      "Collecting mpmath<1.4,>=1.1.0 (from sympy->torch>=1.11.0->sentence-transformers)\n",
      "  Using cached mpmath-1.3.0-py3-none-any.whl.metadata (8.6 kB)\n",
      "Downloading sentence_transformers-3.2.1-py3-none-any.whl (255 kB)\n",
      "Downloading huggingface_hub-0.26.5-py3-none-any.whl (447 kB)\n",
      "Downloading torch-2.4.1-cp38-cp38-win_amd64.whl (199.4 MB)\n",
      "   ---------------------------------------- 0.0/199.4 MB ? eta -:--:--\n",
      "    --------------------------------------- 2.6/199.4 MB 12.6 MB/s eta 0:00:16\n",
      "   - -------------------------------------- 5.2/199.4 MB 12.8 MB/s eta 0:00:16\n",
      "   - -------------------------------------- 7.6/199.4 MB 12.4 MB/s eta 0:00:16\n",
      "   -- ------------------------------------- 10.5/199.4 MB 12.6 MB/s eta 0:00:16\n",
      "   -- ------------------------------------- 13.1/199.4 MB 12.8 MB/s eta 0:00:15\n",
      "   --- ------------------------------------ 16.3/199.4 MB 13.1 MB/s eta 0:00:14\n",
      "   --- ------------------------------------ 19.7/199.4 MB 13.5 MB/s eta 0:00:14\n",
      "   ---- ----------------------------------- 22.5/199.4 MB 13.7 MB/s eta 0:00:13\n",
      "   ----- ---------------------------------- 26.0/199.4 MB 13.9 MB/s eta 0:00:13\n",
      "   ----- ---------------------------------- 29.1/199.4 MB 14.1 MB/s eta 0:00:13\n",
      "   ------ --------------------------------- 31.7/199.4 MB 14.0 MB/s eta 0:00:12\n",
      "   ------ --------------------------------- 33.8/199.4 MB 13.6 MB/s eta 0:00:13\n",
      "   ------- -------------------------------- 35.1/199.4 MB 13.0 MB/s eta 0:00:13\n",
      "   ------- -------------------------------- 35.9/199.4 MB 12.5 MB/s eta 0:00:14\n",
      "   ------- -------------------------------- 37.0/199.4 MB 11.7 MB/s eta 0:00:14\n",
      "   ------- -------------------------------- 37.5/199.4 MB 11.3 MB/s eta 0:00:15\n",
      "   ------- -------------------------------- 38.0/199.4 MB 10.9 MB/s eta 0:00:15\n",
      "   ------- -------------------------------- 39.1/199.4 MB 10.5 MB/s eta 0:00:16\n",
      "   ------- -------------------------------- 39.8/199.4 MB 10.2 MB/s eta 0:00:16\n",
      "   -------- ------------------------------- 40.9/199.4 MB 9.9 MB/s eta 0:00:17\n",
      "   -------- ------------------------------- 41.9/199.4 MB 9.7 MB/s eta 0:00:17\n",
      "   -------- ------------------------------- 43.3/199.4 MB 9.4 MB/s eta 0:00:17\n",
      "   -------- ------------------------------- 44.3/199.4 MB 9.2 MB/s eta 0:00:17\n",
      "   --------- ------------------------------ 45.4/199.4 MB 9.1 MB/s eta 0:00:17\n",
      "   --------- ------------------------------ 46.4/199.4 MB 8.9 MB/s eta 0:00:18\n",
      "   --------- ------------------------------ 47.4/199.4 MB 8.8 MB/s eta 0:00:18\n",
      "   --------- ------------------------------ 48.8/199.4 MB 8.6 MB/s eta 0:00:18\n",
      "   --------- ------------------------------ 49.8/199.4 MB 8.5 MB/s eta 0:00:18\n",
      "   ---------- ----------------------------- 51.1/199.4 MB 8.4 MB/s eta 0:00:18\n",
      "   ---------- ----------------------------- 52.4/199.4 MB 8.3 MB/s eta 0:00:18\n",
      "   ---------- ----------------------------- 53.7/199.4 MB 8.3 MB/s eta 0:00:18\n",
      "   ----------- ---------------------------- 55.3/199.4 MB 8.2 MB/s eta 0:00:18\n",
      "   ----------- ---------------------------- 56.6/199.4 MB 8.2 MB/s eta 0:00:18\n",
      "   ----------- ---------------------------- 57.9/199.4 MB 8.1 MB/s eta 0:00:18\n",
      "   ----------- ---------------------------- 59.5/199.4 MB 8.1 MB/s eta 0:00:18\n",
      "   ------------ --------------------------- 60.8/199.4 MB 8.1 MB/s eta 0:00:18\n",
      "   ------------ --------------------------- 62.4/199.4 MB 8.0 MB/s eta 0:00:18\n",
      "   ------------ --------------------------- 63.7/199.4 MB 8.0 MB/s eta 0:00:17\n",
      "   ------------- -------------------------- 65.0/199.4 MB 8.0 MB/s eta 0:00:17\n",
      "   ------------- -------------------------- 66.3/199.4 MB 7.9 MB/s eta 0:00:17\n",
      "   ------------- -------------------------- 67.6/199.4 MB 7.9 MB/s eta 0:00:17\n",
      "   ------------- -------------------------- 68.7/199.4 MB 7.8 MB/s eta 0:00:17\n",
      "   ------------- -------------------------- 69.7/199.4 MB 7.8 MB/s eta 0:00:17\n",
      "   -------------- ------------------------- 71.0/199.4 MB 7.7 MB/s eta 0:00:17\n",
      "   -------------- ------------------------- 72.1/199.4 MB 7.7 MB/s eta 0:00:17\n",
      "   -------------- ------------------------- 73.1/199.4 MB 7.6 MB/s eta 0:00:17\n",
      "   -------------- ------------------------- 73.7/199.4 MB 7.5 MB/s eta 0:00:17\n",
      "   -------------- ------------------------- 74.2/199.4 MB 7.4 MB/s eta 0:00:17\n",
      "   -------------- ------------------------- 74.7/199.4 MB 7.3 MB/s eta 0:00:18\n",
      "   --------------- ------------------------ 75.0/199.4 MB 7.3 MB/s eta 0:00:18\n",
      "   --------------- ------------------------ 75.5/199.4 MB 7.1 MB/s eta 0:00:18\n",
      "   --------------- ------------------------ 76.3/199.4 MB 7.0 MB/s eta 0:00:18\n",
      "   --------------- ------------------------ 76.8/199.4 MB 7.0 MB/s eta 0:00:18\n",
      "   --------------- ------------------------ 77.6/199.4 MB 6.9 MB/s eta 0:00:18\n",
      "   --------------- ------------------------ 78.1/199.4 MB 6.8 MB/s eta 0:00:18\n",
      "   --------------- ------------------------ 78.9/199.4 MB 6.8 MB/s eta 0:00:18\n",
      "   --------------- ------------------------ 79.7/199.4 MB 6.7 MB/s eta 0:00:18\n",
      "   ---------------- ----------------------- 80.5/199.4 MB 6.7 MB/s eta 0:00:18\n",
      "   ---------------- ----------------------- 81.3/199.4 MB 6.6 MB/s eta 0:00:18\n",
      "   ---------------- ----------------------- 81.8/199.4 MB 6.6 MB/s eta 0:00:18\n",
      "   ---------------- ----------------------- 82.3/199.4 MB 6.5 MB/s eta 0:00:19\n",
      "   ---------------- ----------------------- 82.8/199.4 MB 6.4 MB/s eta 0:00:19\n",
      "   ---------------- ----------------------- 83.4/199.4 MB 6.3 MB/s eta 0:00:19\n",
      "   ---------------- ----------------------- 83.9/199.4 MB 6.3 MB/s eta 0:00:19\n",
      "   ---------------- ----------------------- 84.4/199.4 MB 6.2 MB/s eta 0:00:19\n",
      "   ----------------- ---------------------- 84.9/199.4 MB 6.2 MB/s eta 0:00:19\n",
      "   ----------------- ---------------------- 85.7/199.4 MB 6.1 MB/s eta 0:00:19\n",
      "   ----------------- ---------------------- 86.2/199.4 MB 6.1 MB/s eta 0:00:19\n",
      "   ----------------- ---------------------- 87.0/199.4 MB 6.0 MB/s eta 0:00:19\n",
      "   ----------------- ---------------------- 87.6/199.4 MB 6.0 MB/s eta 0:00:19\n",
      "   ----------------- ---------------------- 88.3/199.4 MB 6.0 MB/s eta 0:00:19\n",
      "   ----------------- ---------------------- 89.1/199.4 MB 5.9 MB/s eta 0:00:19\n",
      "   ------------------ --------------------- 90.2/199.4 MB 5.9 MB/s eta 0:00:19\n",
      "   ------------------ --------------------- 91.0/199.4 MB 5.9 MB/s eta 0:00:19\n",
      "   ------------------ --------------------- 91.8/199.4 MB 5.8 MB/s eta 0:00:19\n",
      "   ------------------ --------------------- 92.5/199.4 MB 5.8 MB/s eta 0:00:19\n",
      "   ------------------ --------------------- 93.6/199.4 MB 5.8 MB/s eta 0:00:19\n",
      "   ------------------ --------------------- 94.6/199.4 MB 5.8 MB/s eta 0:00:19\n",
      "   ------------------- -------------------- 95.7/199.4 MB 5.8 MB/s eta 0:00:18\n",
      "   ------------------- -------------------- 96.5/199.4 MB 5.8 MB/s eta 0:00:18\n",
      "   ------------------- -------------------- 97.3/199.4 MB 5.7 MB/s eta 0:00:18\n",
      "   ------------------- -------------------- 97.8/199.4 MB 5.7 MB/s eta 0:00:18\n",
      "   ------------------- -------------------- 98.6/199.4 MB 5.7 MB/s eta 0:00:18\n",
      "   ------------------- -------------------- 99.4/199.4 MB 5.6 MB/s eta 0:00:18\n",
      "   -------------------- ------------------- 100.1/199.4 MB 5.6 MB/s eta 0:00:18\n",
      "   -------------------- ------------------- 101.2/199.4 MB 5.6 MB/s eta 0:00:18\n",
      "   -------------------- ------------------- 101.2/199.4 MB 5.6 MB/s eta 0:00:18\n",
      "   -------------------- ------------------- 101.4/199.4 MB 5.5 MB/s eta 0:00:18\n",
      "   -------------------- ------------------- 102.5/199.4 MB 5.5 MB/s eta 0:00:18\n",
      "   -------------------- ------------------- 103.5/199.4 MB 5.5 MB/s eta 0:00:18\n",
      "   -------------------- ------------------- 104.3/199.4 MB 5.5 MB/s eta 0:00:18\n",
      "   --------------------- ------------------ 105.1/199.4 MB 5.5 MB/s eta 0:00:18\n",
      "   --------------------- ------------------ 105.6/199.4 MB 5.4 MB/s eta 0:00:18\n",
      "   --------------------- ------------------ 106.4/199.4 MB 5.4 MB/s eta 0:00:18\n",
      "   --------------------- ------------------ 107.0/199.4 MB 5.4 MB/s eta 0:00:18\n",
      "   --------------------- ------------------ 107.5/199.4 MB 5.4 MB/s eta 0:00:18\n",
      "   --------------------- ------------------ 108.3/199.4 MB 5.3 MB/s eta 0:00:18\n",
      "   --------------------- ------------------ 108.8/199.4 MB 5.3 MB/s eta 0:00:18\n",
      "   --------------------- ------------------ 109.6/199.4 MB 5.3 MB/s eta 0:00:17\n",
      "   ---------------------- ----------------- 110.4/199.4 MB 5.3 MB/s eta 0:00:17\n",
      "   ---------------------- ----------------- 111.1/199.4 MB 5.2 MB/s eta 0:00:17\n",
      "   ---------------------- ----------------- 111.9/199.4 MB 5.2 MB/s eta 0:00:17\n",
      "   ---------------------- ----------------- 112.7/199.4 MB 5.2 MB/s eta 0:00:17\n",
      "   ---------------------- ----------------- 113.5/199.4 MB 5.2 MB/s eta 0:00:17\n",
      "   ---------------------- ----------------- 114.0/199.4 MB 5.2 MB/s eta 0:00:17\n",
      "   ----------------------- ---------------- 114.8/199.4 MB 5.2 MB/s eta 0:00:17\n",
      "   ----------------------- ---------------- 115.3/199.4 MB 5.2 MB/s eta 0:00:17\n",
      "   ----------------------- ---------------- 116.1/199.4 MB 5.1 MB/s eta 0:00:17\n",
      "   ----------------------- ---------------- 116.7/199.4 MB 5.1 MB/s eta 0:00:17\n",
      "   ----------------------- ---------------- 117.2/199.4 MB 5.1 MB/s eta 0:00:17\n",
      "   ----------------------- ---------------- 117.7/199.4 MB 5.1 MB/s eta 0:00:17\n",
      "   ----------------------- ---------------- 118.2/199.4 MB 5.0 MB/s eta 0:00:17\n",
      "   ----------------------- ---------------- 118.8/199.4 MB 5.0 MB/s eta 0:00:17\n",
      "   ----------------------- ---------------- 119.5/199.4 MB 5.0 MB/s eta 0:00:16\n",
      "   ------------------------ --------------- 120.1/199.4 MB 5.0 MB/s eta 0:00:16\n",
      "   ------------------------ --------------- 120.8/199.4 MB 5.0 MB/s eta 0:00:16\n",
      "   ------------------------ --------------- 121.4/199.4 MB 5.0 MB/s eta 0:00:16\n",
      "   ------------------------ --------------- 122.2/199.4 MB 5.0 MB/s eta 0:00:16\n",
      "   ------------------------ --------------- 122.9/199.4 MB 4.9 MB/s eta 0:00:16\n",
      "   ------------------------ --------------- 123.7/199.4 MB 4.9 MB/s eta 0:00:16\n",
      "   ------------------------- -------------- 124.8/199.4 MB 4.9 MB/s eta 0:00:16\n",
      "   ------------------------- -------------- 125.6/199.4 MB 4.9 MB/s eta 0:00:16\n",
      "   ------------------------- -------------- 126.6/199.4 MB 4.9 MB/s eta 0:00:15\n",
      "   ------------------------- -------------- 127.4/199.4 MB 4.9 MB/s eta 0:00:15\n",
      "   ------------------------- -------------- 128.2/199.4 MB 4.9 MB/s eta 0:00:15\n",
      "   ------------------------- -------------- 128.7/199.4 MB 4.9 MB/s eta 0:00:15\n",
      "   ------------------------- -------------- 129.5/199.4 MB 4.9 MB/s eta 0:00:15\n",
      "   -------------------------- ------------- 130.3/199.4 MB 4.9 MB/s eta 0:00:15\n",
      "   -------------------------- ------------- 131.1/199.4 MB 4.9 MB/s eta 0:00:15\n",
      "   -------------------------- ------------- 131.9/199.4 MB 4.9 MB/s eta 0:00:14\n",
      "   -------------------------- ------------- 132.9/199.4 MB 4.9 MB/s eta 0:00:14\n",
      "   -------------------------- ------------- 133.7/199.4 MB 4.8 MB/s eta 0:00:14\n",
      "   -------------------------- ------------- 134.5/199.4 MB 4.8 MB/s eta 0:00:14\n",
      "   --------------------------- ------------ 135.5/199.4 MB 4.8 MB/s eta 0:00:14\n",
      "   --------------------------- ------------ 136.6/199.4 MB 4.8 MB/s eta 0:00:13\n",
      "   --------------------------- ------------ 137.6/199.4 MB 4.8 MB/s eta 0:00:13\n",
      "   --------------------------- ------------ 138.7/199.4 MB 4.8 MB/s eta 0:00:13\n",
      "   ---------------------------- ----------- 139.7/199.4 MB 4.8 MB/s eta 0:00:13\n",
      "   ---------------------------- ----------- 140.8/199.4 MB 4.8 MB/s eta 0:00:13\n",
      "   ---------------------------- ----------- 141.6/199.4 MB 4.8 MB/s eta 0:00:12\n",
      "   ---------------------------- ----------- 142.3/199.4 MB 4.8 MB/s eta 0:00:12\n",
      "   ---------------------------- ----------- 143.1/199.4 MB 4.8 MB/s eta 0:00:12\n",
      "   ---------------------------- ----------- 144.2/199.4 MB 4.8 MB/s eta 0:00:12\n",
      "   ----------------------------- ---------- 144.7/199.4 MB 4.8 MB/s eta 0:00:12\n",
      "   ----------------------------- ---------- 145.5/199.4 MB 4.7 MB/s eta 0:00:12\n",
      "   ----------------------------- ---------- 146.3/199.4 MB 4.7 MB/s eta 0:00:12\n",
      "   ----------------------------- ---------- 146.8/199.4 MB 4.6 MB/s eta 0:00:12\n",
      "   ----------------------------- ---------- 147.6/199.4 MB 4.5 MB/s eta 0:00:12\n",
      "   ----------------------------- ---------- 148.1/199.4 MB 4.5 MB/s eta 0:00:12\n",
      "   ----------------------------- ---------- 148.9/199.4 MB 4.4 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 149.7/199.4 MB 4.3 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 150.7/199.4 MB 4.2 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 151.3/199.4 MB 4.2 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 152.0/199.4 MB 4.1 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 152.6/199.4 MB 4.0 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 153.1/199.4 MB 4.0 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 153.6/199.4 MB 3.9 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 153.9/199.4 MB 3.9 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 154.4/199.4 MB 3.9 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 154.7/199.4 MB 3.9 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 154.9/199.4 MB 3.9 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 155.2/199.4 MB 3.9 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 155.7/199.4 MB 3.8 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 156.0/199.4 MB 3.8 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 156.5/199.4 MB 3.8 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 156.8/199.4 MB 3.8 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 157.3/199.4 MB 3.8 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 157.5/199.4 MB 3.7 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 158.1/199.4 MB 3.7 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 158.3/199.4 MB 3.7 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 158.6/199.4 MB 3.7 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 158.9/199.4 MB 3.6 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 159.4/199.4 MB 3.6 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 159.6/199.4 MB 3.6 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 159.9/199.4 MB 3.5 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 160.4/199.4 MB 3.5 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 161.0/199.4 MB 3.5 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 161.5/199.4 MB 3.4 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 162.0/199.4 MB 3.4 MB/s eta 0:00:11\n",
      "   -------------------------------- ------- 162.5/199.4 MB 3.4 MB/s eta 0:00:11\n",
      "   -------------------------------- ------- 163.1/199.4 MB 3.4 MB/s eta 0:00:11\n",
      "   -------------------------------- ------- 163.6/199.4 MB 3.3 MB/s eta 0:00:11\n",
      "   -------------------------------- ------- 164.1/199.4 MB 3.3 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 164.6/199.4 MB 3.3 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 165.4/199.4 MB 3.2 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 165.9/199.4 MB 3.2 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 166.5/199.4 MB 3.2 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 167.0/199.4 MB 3.2 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 167.2/199.4 MB 3.2 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 167.5/199.4 MB 3.1 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 167.8/199.4 MB 3.1 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 168.0/199.4 MB 3.1 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 168.3/199.4 MB 3.1 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 168.6/199.4 MB 3.1 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 168.8/199.4 MB 3.1 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 169.1/199.4 MB 3.1 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 169.1/199.4 MB 3.1 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 169.3/199.4 MB 3.1 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 169.6/199.4 MB 3.1 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 169.9/199.4 MB 3.0 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 170.1/199.4 MB 3.0 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 170.4/199.4 MB 3.0 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 170.7/199.4 MB 3.0 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 170.9/199.4 MB 3.0 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 171.2/199.4 MB 3.0 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 171.4/199.4 MB 3.0 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 171.7/199.4 MB 2.9 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 172.0/199.4 MB 2.9 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 172.2/199.4 MB 2.9 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 172.5/199.4 MB 2.9 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 172.8/199.4 MB 2.9 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 173.3/199.4 MB 2.9 MB/s eta 0:00:10\n",
      "   ---------------------------------- ----- 173.5/199.4 MB 2.9 MB/s eta 0:00:09\n",
      "   ---------------------------------- ----- 173.8/199.4 MB 2.9 MB/s eta 0:00:09\n",
      "   ---------------------------------- ----- 174.1/199.4 MB 2.9 MB/s eta 0:00:09\n",
      "   ----------------------------------- ---- 174.6/199.4 MB 2.8 MB/s eta 0:00:09\n",
      "   ----------------------------------- ---- 174.9/199.4 MB 2.8 MB/s eta 0:00:09\n",
      "   ----------------------------------- ---- 175.4/199.4 MB 2.8 MB/s eta 0:00:09\n",
      "   ----------------------------------- ---- 175.6/199.4 MB 2.8 MB/s eta 0:00:09\n",
      "   ----------------------------------- ---- 175.9/199.4 MB 2.8 MB/s eta 0:00:09\n",
      "   ----------------------------------- ---- 176.4/199.4 MB 2.8 MB/s eta 0:00:09\n",
      "   ----------------------------------- ---- 176.7/199.4 MB 2.7 MB/s eta 0:00:09\n",
      "   ----------------------------------- ---- 176.9/199.4 MB 2.7 MB/s eta 0:00:09\n",
      "   ----------------------------------- ---- 177.5/199.4 MB 2.7 MB/s eta 0:00:09\n",
      "   ----------------------------------- ---- 177.7/199.4 MB 2.7 MB/s eta 0:00:09\n",
      "   ----------------------------------- ---- 178.0/199.4 MB 2.7 MB/s eta 0:00:08\n",
      "   ----------------------------------- ---- 178.3/199.4 MB 2.7 MB/s eta 0:00:08\n",
      "   ----------------------------------- ---- 178.8/199.4 MB 2.7 MB/s eta 0:00:08\n",
      "   ----------------------------------- ---- 179.3/199.4 MB 2.6 MB/s eta 0:00:08\n",
      "   ------------------------------------ --- 179.6/199.4 MB 2.6 MB/s eta 0:00:08\n",
      "   ------------------------------------ --- 180.1/199.4 MB 2.6 MB/s eta 0:00:08\n",
      "   ------------------------------------ --- 180.6/199.4 MB 2.6 MB/s eta 0:00:08\n",
      "   ------------------------------------ --- 181.1/199.4 MB 2.6 MB/s eta 0:00:07\n",
      "   ------------------------------------ --- 181.7/199.4 MB 2.6 MB/s eta 0:00:07\n",
      "   ------------------------------------ --- 182.2/199.4 MB 2.6 MB/s eta 0:00:07\n",
      "   ------------------------------------ --- 183.0/199.4 MB 2.6 MB/s eta 0:00:07\n",
      "   ------------------------------------ --- 183.8/199.4 MB 2.6 MB/s eta 0:00:07\n",
      "   ------------------------------------ --- 184.3/199.4 MB 2.6 MB/s eta 0:00:06\n",
      "   ------------------------------------- -- 185.1/199.4 MB 2.6 MB/s eta 0:00:06\n",
      "   ------------------------------------- -- 185.9/199.4 MB 2.6 MB/s eta 0:00:06\n",
      "   ------------------------------------- -- 186.9/199.4 MB 2.6 MB/s eta 0:00:05\n",
      "   ------------------------------------- -- 187.7/199.4 MB 2.6 MB/s eta 0:00:05\n",
      "   ------------------------------------- -- 188.5/199.4 MB 2.6 MB/s eta 0:00:05\n",
      "   ------------------------------------- -- 189.3/199.4 MB 2.6 MB/s eta 0:00:04\n",
      "   -------------------------------------- - 190.3/199.4 MB 2.6 MB/s eta 0:00:04\n",
      "   -------------------------------------- - 191.1/199.4 MB 2.7 MB/s eta 0:00:04\n",
      "   -------------------------------------- - 192.2/199.4 MB 2.7 MB/s eta 0:00:03\n",
      "   -------------------------------------- - 192.9/199.4 MB 2.7 MB/s eta 0:00:03\n",
      "   -------------------------------------- - 194.0/199.4 MB 2.7 MB/s eta 0:00:03\n",
      "   ---------------------------------------  194.8/199.4 MB 2.7 MB/s eta 0:00:02\n",
      "   ---------------------------------------  195.6/199.4 MB 2.7 MB/s eta 0:00:02\n",
      "   ---------------------------------------  196.3/199.4 MB 2.7 MB/s eta 0:00:02\n",
      "   ---------------------------------------  197.1/199.4 MB 2.7 MB/s eta 0:00:01\n",
      "   ---------------------------------------  197.9/199.4 MB 2.7 MB/s eta 0:00:01\n",
      "   ---------------------------------------  199.0/199.4 MB 2.7 MB/s eta 0:00:01\n",
      "   ---------------------------------------  199.2/199.4 MB 2.7 MB/s eta 0:00:01\n",
      "   ---------------------------------------  199.2/199.4 MB 2.7 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 199.4/199.4 MB 2.7 MB/s eta 0:00:00\n",
      "Downloading tqdm-4.67.1-py3-none-any.whl (78 kB)\n",
      "Downloading transformers-4.46.3-py3-none-any.whl (10.0 MB)\n",
      "   ---------------------------------------- 0.0/10.0 MB ? eta -:--:--\n",
      "   --- ------------------------------------ 0.8/10.0 MB 4.2 MB/s eta 0:00:03\n",
      "   ------- -------------------------------- 1.8/10.0 MB 4.6 MB/s eta 0:00:02\n",
      "   ---------- ----------------------------- 2.6/10.0 MB 4.6 MB/s eta 0:00:02\n",
      "   -------------- ------------------------- 3.7/10.0 MB 4.6 MB/s eta 0:00:02\n",
      "   ------------------ --------------------- 4.7/10.0 MB 4.6 MB/s eta 0:00:02\n",
      "   --------------------- ------------------ 5.5/10.0 MB 4.5 MB/s eta 0:00:01\n",
      "   ------------------------- -------------- 6.3/10.0 MB 4.4 MB/s eta 0:00:01\n",
      "   ---------------------------- ----------- 7.1/10.0 MB 4.4 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 7.9/10.0 MB 4.3 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 8.9/10.0 MB 4.3 MB/s eta 0:00:01\n",
      "   ---------------------------------------  10.0/10.0 MB 4.3 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 10.0/10.0 MB 4.3 MB/s eta 0:00:00\n",
      "Downloading pillow-10.4.0-cp38-cp38-win_amd64.whl (2.6 MB)\n",
      "   ---------------------------------------- 0.0/2.6 MB ? eta -:--:--\n",
      "   ------------ --------------------------- 0.8/2.6 MB 4.8 MB/s eta 0:00:01\n",
      "   ---------------------------- ----------- 1.8/2.6 MB 4.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.6/2.6 MB 4.6 MB/s eta 0:00:00\n",
      "Downloading scikit_learn-1.3.2-cp38-cp38-win_amd64.whl (9.3 MB)\n",
      "   ---------------------------------------- 0.0/9.3 MB ? eta -:--:--\n",
      "   ---- ----------------------------------- 1.0/9.3 MB 5.0 MB/s eta 0:00:02\n",
      "   ------- -------------------------------- 1.8/9.3 MB 4.8 MB/s eta 0:00:02\n",
      "   ----------- ---------------------------- 2.6/9.3 MB 4.6 MB/s eta 0:00:02\n",
      "   -------------- ------------------------- 3.4/9.3 MB 4.2 MB/s eta 0:00:02\n",
      "   ---------------- ----------------------- 3.9/9.3 MB 4.0 MB/s eta 0:00:02\n",
      "   ------------------- -------------------- 4.5/9.3 MB 3.8 MB/s eta 0:00:02\n",
      "   ---------------------- ----------------- 5.2/9.3 MB 3.6 MB/s eta 0:00:02\n",
      "   ------------------------ --------------- 5.8/9.3 MB 3.6 MB/s eta 0:00:01\n",
      "   --------------------------- ------------ 6.3/9.3 MB 3.5 MB/s eta 0:00:01\n",
      "   ----------------------------- ---------- 6.8/9.3 MB 3.3 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 7.6/9.3 MB 3.3 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 8.1/9.3 MB 3.3 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 8.9/9.3 MB 3.3 MB/s eta 0:00:01\n",
      "   ---------------------------------------  9.2/9.3 MB 3.3 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 9.3/9.3 MB 3.1 MB/s eta 0:00:00\n",
      "Downloading scipy-1.10.1-cp38-cp38-win_amd64.whl (42.2 MB)\n",
      "   ---------------------------------------- 0.0/42.2 MB ? eta -:--:--\n",
      "    --------------------------------------- 0.8/42.2 MB 3.7 MB/s eta 0:00:12\n",
      "   - -------------------------------------- 1.6/42.2 MB 4.0 MB/s eta 0:00:11\n",
      "   -- ------------------------------------- 2.4/42.2 MB 3.9 MB/s eta 0:00:11\n",
      "   -- ------------------------------------- 3.1/42.2 MB 4.1 MB/s eta 0:00:10\n",
      "   --- ------------------------------------ 3.9/42.2 MB 3.9 MB/s eta 0:00:10\n",
      "   ---- ----------------------------------- 5.0/42.2 MB 4.1 MB/s eta 0:00:10\n",
      "   ----- ---------------------------------- 5.8/42.2 MB 4.1 MB/s eta 0:00:09\n",
      "   ------ --------------------------------- 6.8/42.2 MB 4.2 MB/s eta 0:00:09\n",
      "   ------- -------------------------------- 7.6/42.2 MB 4.1 MB/s eta 0:00:09\n",
      "   ------- -------------------------------- 8.4/42.2 MB 4.0 MB/s eta 0:00:09\n",
      "   -------- ------------------------------- 9.2/42.2 MB 4.0 MB/s eta 0:00:09\n",
      "   --------- ------------------------------ 10.0/42.2 MB 4.0 MB/s eta 0:00:09\n",
      "   ---------- ----------------------------- 10.7/42.2 MB 3.9 MB/s eta 0:00:09\n",
      "   ---------- ----------------------------- 11.3/42.2 MB 3.8 MB/s eta 0:00:09\n",
      "   ----------- ---------------------------- 11.8/42.2 MB 3.8 MB/s eta 0:00:08\n",
      "   ----------- ---------------------------- 12.6/42.2 MB 3.8 MB/s eta 0:00:08\n",
      "   ------------ --------------------------- 13.1/42.2 MB 3.7 MB/s eta 0:00:08\n",
      "   ------------ --------------------------- 13.6/42.2 MB 3.6 MB/s eta 0:00:08\n",
      "   ------------- -------------------------- 14.2/42.2 MB 3.6 MB/s eta 0:00:08\n",
      "   ------------- -------------------------- 14.7/42.2 MB 3.5 MB/s eta 0:00:08\n",
      "   -------------- ------------------------- 15.2/42.2 MB 3.5 MB/s eta 0:00:08\n",
      "   --------------- ------------------------ 16.0/42.2 MB 3.5 MB/s eta 0:00:08\n",
      "   --------------- ------------------------ 16.5/42.2 MB 3.5 MB/s eta 0:00:08\n",
      "   ---------------- ----------------------- 17.3/42.2 MB 3.5 MB/s eta 0:00:08\n",
      "   ----------------- ---------------------- 18.1/42.2 MB 3.5 MB/s eta 0:00:07\n",
      "   ----------------- ---------------------- 18.6/42.2 MB 3.5 MB/s eta 0:00:07\n",
      "   ------------------ --------------------- 19.4/42.2 MB 3.5 MB/s eta 0:00:07\n",
      "   ------------------- -------------------- 20.4/42.2 MB 3.5 MB/s eta 0:00:07\n",
      "   -------------------- ------------------- 21.2/42.2 MB 3.5 MB/s eta 0:00:06\n",
      "   -------------------- ------------------- 22.0/42.2 MB 3.5 MB/s eta 0:00:06\n",
      "   --------------------- ------------------ 23.1/42.2 MB 3.6 MB/s eta 0:00:06\n",
      "   ---------------------- ----------------- 23.9/42.2 MB 3.6 MB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 24.9/42.2 MB 3.6 MB/s eta 0:00:05\n",
      "   ------------------------ --------------- 26.0/42.2 MB 3.7 MB/s eta 0:00:05\n",
      "   ------------------------- -------------- 27.0/42.2 MB 3.7 MB/s eta 0:00:05\n",
      "   -------------------------- ------------- 27.8/42.2 MB 3.7 MB/s eta 0:00:04\n",
      "   --------------------------- ------------ 28.6/42.2 MB 3.7 MB/s eta 0:00:04\n",
      "   --------------------------- ------------ 29.4/42.2 MB 3.7 MB/s eta 0:00:04\n",
      "   ---------------------------- ----------- 30.1/42.2 MB 3.7 MB/s eta 0:00:04\n",
      "   ----------------------------- ---------- 30.7/42.2 MB 3.7 MB/s eta 0:00:04\n",
      "   ----------------------------- ---------- 31.5/42.2 MB 3.7 MB/s eta 0:00:03\n",
      "   ------------------------------ --------- 32.2/42.2 MB 3.7 MB/s eta 0:00:03\n",
      "   ------------------------------- -------- 32.8/42.2 MB 3.7 MB/s eta 0:00:03\n",
      "   ------------------------------- -------- 33.6/42.2 MB 3.7 MB/s eta 0:00:03\n",
      "   -------------------------------- ------- 34.3/42.2 MB 3.7 MB/s eta 0:00:03\n",
      "   --------------------------------- ------ 35.1/42.2 MB 3.7 MB/s eta 0:00:02\n",
      "   --------------------------------- ------ 35.7/42.2 MB 3.6 MB/s eta 0:00:02\n",
      "   ---------------------------------- ----- 36.4/42.2 MB 3.6 MB/s eta 0:00:02\n",
      "   ----------------------------------- ---- 37.2/42.2 MB 3.6 MB/s eta 0:00:02\n",
      "   ----------------------------------- ---- 37.7/42.2 MB 3.6 MB/s eta 0:00:02\n",
      "   ------------------------------------ --- 38.8/42.2 MB 3.6 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 39.6/42.2 MB 3.6 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 40.4/42.2 MB 3.6 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 41.2/42.2 MB 3.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  41.9/42.2 MB 3.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 42.2/42.2 MB 3.6 MB/s eta 0:00:00\n",
      "Downloading fsspec-2024.10.0-py3-none-any.whl (179 kB)\n",
      "Using cached joblib-1.4.2-py3-none-any.whl (301 kB)\n",
      "Downloading PyYAML-6.0.2-cp38-cp38-win_amd64.whl (162 kB)\n",
      "Downloading regex-2024.11.6-cp38-cp38-win_amd64.whl (274 kB)\n",
      "Downloading safetensors-0.4.5-cp38-none-win_amd64.whl (286 kB)\n",
      "Using cached threadpoolctl-3.5.0-py3-none-any.whl (18 kB)\n",
      "Downloading tokenizers-0.20.3-cp38-none-win_amd64.whl (2.4 MB)\n",
      "   ---------------------------------------- 0.0/2.4 MB ? eta -:--:--\n",
      "   -------- ------------------------------- 0.5/2.4 MB 2.8 MB/s eta 0:00:01\n",
      "   ----------------- ---------------------- 1.0/2.4 MB 2.6 MB/s eta 0:00:01\n",
      "   ---------------------- ----------------- 1.3/2.4 MB 2.8 MB/s eta 0:00:01\n",
      "   ------------------------------ --------- 1.8/2.4 MB 2.3 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.4/2.4 MB 2.4 MB/s eta 0:00:00\n",
      "Downloading filelock-3.16.1-py3-none-any.whl (16 kB)\n",
      "Using cached jinja2-3.1.4-py3-none-any.whl (133 kB)\n",
      "Downloading networkx-3.1-py3-none-any.whl (2.1 MB)\n",
      "   ---------------------------------------- 0.0/2.1 MB ? eta -:--:--\n",
      "   ---------- ----------------------------- 0.5/2.1 MB 2.4 MB/s eta 0:00:01\n",
      "   -------------------- ------------------- 1.0/2.1 MB 2.8 MB/s eta 0:00:01\n",
      "   ------------------------------ --------- 1.6/2.1 MB 2.8 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.1/2.1 MB 2.9 MB/s eta 0:00:00\n",
      "Using cached requests-2.32.3-py3-none-any.whl (64 kB)\n",
      "Downloading sympy-1.13.3-py3-none-any.whl (6.2 MB)\n",
      "   ---------------------------------------- 0.0/6.2 MB ? eta -:--:--\n",
      "   --- ------------------------------------ 0.5/6.2 MB 3.4 MB/s eta 0:00:02\n",
      "   -------- ------------------------------- 1.3/6.2 MB 3.5 MB/s eta 0:00:02\n",
      "   ------------- -------------------------- 2.1/6.2 MB 3.6 MB/s eta 0:00:02\n",
      "   ------------------ --------------------- 2.9/6.2 MB 3.6 MB/s eta 0:00:01\n",
      "   ----------------------- ---------------- 3.7/6.2 MB 3.6 MB/s eta 0:00:01\n",
      "   --------------------------- ------------ 4.2/6.2 MB 3.6 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 5.0/6.2 MB 3.4 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 5.5/6.2 MB 3.4 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 6.2/6.2 MB 3.3 MB/s eta 0:00:00\n",
      "Using cached certifi-2024.8.30-py3-none-any.whl (167 kB)\n",
      "Downloading charset_normalizer-3.4.0-cp38-cp38-win_amd64.whl (101 kB)\n",
      "Using cached idna-3.10-py3-none-any.whl (70 kB)\n",
      "Downloading MarkupSafe-2.1.5-cp38-cp38-win_amd64.whl (17 kB)\n",
      "Using cached mpmath-1.3.0-py3-none-any.whl (536 kB)\n",
      "Using cached urllib3-2.2.3-py3-none-any.whl (126 kB)\n",
      "Installing collected packages: mpmath, urllib3, tqdm, threadpoolctl, sympy, scipy, safetensors, regex, pyyaml, Pillow, networkx, MarkupSafe, joblib, idna, fsspec, filelock, charset-normalizer, certifi, scikit-learn, requests, jinja2, torch, huggingface-hub, tokenizers, transformers, sentence-transformers\n",
      "Successfully installed MarkupSafe-2.1.5 Pillow-10.4.0 certifi-2024.8.30 charset-normalizer-3.4.0 filelock-3.16.1 fsspec-2024.10.0 huggingface-hub-0.26.5 idna-3.10 jinja2-3.1.4 joblib-1.4.2 mpmath-1.3.0 networkx-3.1 pyyaml-6.0.2 regex-2024.11.6 requests-2.32.3 safetensors-0.4.5 scikit-learn-1.3.2 scipy-1.10.1 sentence-transformers-3.2.1 sympy-1.13.3 threadpoolctl-3.5.0 tokenizers-0.20.3 torch-2.4.1 tqdm-4.67.1 transformers-4.46.3 urllib3-2.2.3\n"
     ]
    }
   ],
   "source": [
    "!pip install sentence-transformers # downlaoding libraries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf9f9f5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Yasser\\AppData\\Local\\Temp\\ipykernel_16992\\2473166140.py:16: DeprecationWarning: `id` is deprecated, use `element_id` instead\n",
      "  nodes = [{\"id\": record[\"n\"].id, \"data\": dict(record[\"n\"])} for record in result]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodes data saved to 'nodes.json'\n"
     ]
    }
   ],
   "source": [
    "#SAVING THE NODES AS JSON\n",
    "import json\n",
    "from neo4j import GraphDatabase\n",
    "\n",
    "# Setup Neo4j connection\n",
    "\n",
    "\n",
    "# Connect to Neo4j\n",
    "driver = GraphDatabase.driver(url, auth=(username, password))\n",
    "\n",
    "# Function to fetch all nodes\n",
    "def fetch_nodes():\n",
    "    with driver.session() as session:\n",
    "        result = session.run(\"MATCH (n) RETURN n LIMIT 53212 \")  # Fetching first 100 nodes\n",
    "        nodes = [{\"id\": record[\"n\"].id, \"data\": dict(record[\"n\"])} for record in result]\n",
    "    return nodes\n",
    "\n",
    "# Fetch nodes\n",
    "nodes = fetch_nodes()\n",
    "\n",
    "# Save nodes to JSON file\n",
    "with open(\"nodes.json\", \"w\") as file:\n",
    "    json.dump(nodes, file, indent=4)\n",
    "\n",
    "print(f\"Nodes data saved to 'nodes.json'\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ef81f9f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#converting the nodes to embeddings \n",
    "from sentence_transformers import SentenceTransformer\n",
    "import pickle\n",
    "\n",
    "# Initialize the model\n",
    "model = SentenceTransformer('all-MiniLM-L6-v2')  # \n",
    "\n",
    "def embed_nodes(nodes):\n",
    "    # Safely extract descriptions and convert to strings\n",
    "    node_descriptions = []\n",
    "    for node in nodes:\n",
    "        description = node.get('data', '')  # Get the 'data' field or default to empty string\n",
    "        if isinstance(description, str):  # Check if it's already a string\n",
    "            node_descriptions.append(description)\n",
    "        elif isinstance(description, (int, float)):  # Convert int/float to string\n",
    "            node_descriptions.append(str(description))\n",
    "        else:\n",
    "            node_descriptions.append('')  # Handle any other non-string values\n",
    "\n",
    "    # Generate embeddings\n",
    "    embeddings = model.encode(node_descriptions)\n",
    "\n",
    "    # Save embeddings to a file\n",
    "    with open('embeddings.pkl', 'wb') as f:\n",
    "        pickle.dump(embeddings, f)\n",
    "\n",
    "    return embeddings\n",
    "\n",
    "embeddings = embed_nodes(nodes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "299d45bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 10000,\n",
       " 'data': {'NumRotatableBonds': 9,\n",
       "  'DRUG_INDICATION_ASSOCIATION': nan,\n",
       "  'C_SMILES': 'C#CCOc1ccc(Nc2nc(Nc3ccc(C)c(S(=O)(=O)NC(=O)CC)c3)ncc2F)cc1',\n",
       "  'LogP': 3.6382200000000022,\n",
       "  'CLASS': nan,\n",
       "  'DRUG_SIDEEFFECT_ASSOCIATION': nan,\n",
       "  'DRUGINCHKE': 'IGLNXKVGKIFNBQ-UHFFFAOYSA-N',\n",
       "  'DRUGCOMP': nan,\n",
       "  'COMPCLAS': nan,\n",
       "  'DRUGINCH': ['InChI=1S/C23H22FN5O4S/c1-4-12-33-18-10-8-16(9-11-18)26-22-19(24)14-25-23(28-22)27-17-7-6-15(3)20(13-17)34(31',\n",
       "   '32)29-21(30)5-2/h1',\n",
       "   '6-11',\n",
       "   '13-14H',\n",
       "   '5',\n",
       "   '12H2',\n",
       "   '2-3H3',\n",
       "   '(H',\n",
       "   '29',\n",
       "   '30)(H2',\n",
       "   '25',\n",
       "   '26',\n",
       "   '27',\n",
       "   '28)'],\n",
       "  'SUBSTITUENT': nan,\n",
       "  'PUBMED_ARTICLE': '18360272',\n",
       "  'DRUG_ATC_CODE': nan,\n",
       "  'PRODUCT': nan,\n",
       "  'PRODUCT_STAGE': 'investigational',\n",
       "  'KINGDOM': nan,\n",
       "  'DRUGBANK_NAME': 'R-348',\n",
       "  'id': 'DB06321',\n",
       "  'TYPE': 'DRUG',\n",
       "  'MolecularWeight': 483.52500000000015,\n",
       "  'SUBCLASS': nan,\n",
       "  'SUPERCLASS': nan,\n",
       "  'TPSA': 122.30999999999999,\n",
       "  'NumHDonors': 3,\n",
       "  'NAME': 'R_348',\n",
       "  'SYNONYM': '5_Fluoro_N__4_methyl_3_propionylaminosulfonylphenyl__N___4__prop_2_ynyloxy_phenyl__2_4_pyrimidinediamine',\n",
       "  'LipinskiViolationCount': 0,\n",
       "  'SOURCE': 'BIOKG',\n",
       "  'NumHAcceptors': 8,\n",
       "  'DRUGINKE': nan,\n",
       "  'DIRECT_PARENT': nan,\n",
       "  'ALTERNATIVE_PARENT': nan}}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes[10000] # just exploring the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a0545eb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(53210, 384)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "embeddings.shape # checking the emebedding shape "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "062d2768-cf13-4c34-b11b-3c0777243af6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting chromadb\n",
      "  Using cached chromadb-0.5.23-py3-none-any.whl.metadata (6.8 kB)\n",
      "Collecting build>=1.0.3 (from chromadb)\n",
      "  Using cached build-1.2.2.post1-py3-none-any.whl.metadata (6.5 kB)\n",
      "Collecting pydantic>=1.9 (from chromadb)\n",
      "  Downloading pydantic-2.10.3-py3-none-any.whl.metadata (172 kB)\n",
      "Collecting chroma-hnswlib==0.7.6 (from chromadb)\n",
      "  Downloading chroma_hnswlib-0.7.6-cp38-cp38-win_amd64.whl.metadata (262 bytes)\n",
      "Collecting fastapi>=0.95.2 (from chromadb)\n",
      "  Using cached fastapi-0.115.6-py3-none-any.whl.metadata (27 kB)\n",
      "Collecting uvicorn>=0.18.3 (from uvicorn[standard]>=0.18.3->chromadb)\n",
      "  Using cached uvicorn-0.32.1-py3-none-any.whl.metadata (6.6 kB)\n",
      "Requirement already satisfied: numpy>=1.22.5 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from chromadb) (1.24.4)\n",
      "Collecting posthog>=2.4.0 (from chromadb)\n",
      "  Using cached posthog-3.7.4-py2.py3-none-any.whl.metadata (2.0 kB)\n",
      "Requirement already satisfied: typing_extensions>=4.5.0 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from chromadb) (4.12.2)\n",
      "Requirement already satisfied: onnxruntime>=1.14.1 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from chromadb) (1.19.2)\n",
      "Collecting opentelemetry-api>=1.2.0 (from chromadb)\n",
      "  Using cached opentelemetry_api-1.29.0-py3-none-any.whl.metadata (1.4 kB)\n",
      "Collecting opentelemetry-exporter-otlp-proto-grpc>=1.2.0 (from chromadb)\n",
      "  Using cached opentelemetry_exporter_otlp_proto_grpc-1.29.0-py3-none-any.whl.metadata (2.2 kB)\n",
      "Collecting opentelemetry-instrumentation-fastapi>=0.41b0 (from chromadb)\n",
      "  Using cached opentelemetry_instrumentation_fastapi-0.50b0-py3-none-any.whl.metadata (2.1 kB)\n",
      "Collecting opentelemetry-sdk>=1.2.0 (from chromadb)\n",
      "  Using cached opentelemetry_sdk-1.29.0-py3-none-any.whl.metadata (1.5 kB)\n",
      "Requirement already satisfied: tokenizers<=0.20.3,>=0.13.2 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from chromadb) (0.20.3)\n",
      "Collecting pypika>=0.48.9 (from chromadb)\n",
      "  Using cached PyPika-0.48.9.tar.gz (67 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Requirement already satisfied: tqdm>=4.65.0 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from chromadb) (4.67.1)\n",
      "Collecting overrides>=7.3.1 (from chromadb)\n",
      "  Using cached overrides-7.7.0-py3-none-any.whl.metadata (5.8 kB)\n",
      "Collecting importlib-resources (from chromadb)\n",
      "  Using cached importlib_resources-6.4.5-py3-none-any.whl.metadata (4.0 kB)\n",
      "Collecting graphlib_backport>=1.0.3 (from chromadb)\n",
      "  Downloading graphlib_backport-1.1.0-py3-none-any.whl.metadata (4.4 kB)\n",
      "Requirement already satisfied: grpcio>=1.58.0 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from chromadb) (1.67.1)\n",
      "Collecting bcrypt>=4.0.1 (from chromadb)\n",
      "  Downloading bcrypt-4.2.1-cp37-abi3-win_amd64.whl.metadata (10 kB)\n",
      "Collecting typer>=0.9.0 (from chromadb)\n",
      "  Downloading typer-0.15.1-py3-none-any.whl.metadata (15 kB)\n",
      "Collecting kubernetes>=28.1.0 (from chromadb)\n",
      "  Using cached kubernetes-31.0.0-py2.py3-none-any.whl.metadata (1.5 kB)\n",
      "Collecting tenacity>=8.2.3 (from chromadb)\n",
      "  Using cached tenacity-9.0.0-py3-none-any.whl.metadata (1.2 kB)\n",
      "Requirement already satisfied: PyYAML>=6.0.0 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from chromadb) (6.0.2)\n",
      "Collecting mmh3>=4.0.1 (from chromadb)\n",
      "  Downloading mmh3-5.0.1-cp38-cp38-win_amd64.whl.metadata (14 kB)\n",
      "Collecting orjson>=3.9.12 (from chromadb)\n",
      "  Downloading orjson-3.10.12-cp38-none-win_amd64.whl.metadata (42 kB)\n",
      "Collecting httpx>=0.27.0 (from chromadb)\n",
      "  Using cached httpx-0.28.1-py3-none-any.whl.metadata (7.1 kB)\n",
      "Collecting rich>=10.11.0 (from chromadb)\n",
      "  Downloading rich-13.9.4-py3-none-any.whl.metadata (18 kB)\n",
      "Requirement already satisfied: packaging>=19.1 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from build>=1.0.3->chromadb) (24.2)\n",
      "Collecting pyproject_hooks (from build>=1.0.3->chromadb)\n",
      "  Using cached pyproject_hooks-1.2.0-py3-none-any.whl.metadata (1.3 kB)\n",
      "Requirement already satisfied: colorama in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from build>=1.0.3->chromadb) (0.4.6)\n",
      "Requirement already satisfied: importlib-metadata>=4.6 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from build>=1.0.3->chromadb) (8.5.0)\n",
      "Collecting tomli>=1.1.0 (from build>=1.0.3->chromadb)\n",
      "  Downloading tomli-2.2.1-py3-none-any.whl.metadata (10 kB)\n",
      "Collecting starlette<0.42.0,>=0.40.0 (from fastapi>=0.95.2->chromadb)\n",
      "  Using cached starlette-0.41.3-py3-none-any.whl.metadata (6.0 kB)\n",
      "Collecting anyio (from httpx>=0.27.0->chromadb)\n",
      "  Downloading anyio-4.5.2-py3-none-any.whl.metadata (4.7 kB)\n",
      "Requirement already satisfied: certifi in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from httpx>=0.27.0->chromadb) (2024.8.30)\n",
      "Collecting httpcore==1.* (from httpx>=0.27.0->chromadb)\n",
      "  Downloading httpcore-1.0.7-py3-none-any.whl.metadata (21 kB)\n",
      "Requirement already satisfied: idna in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from httpx>=0.27.0->chromadb) (3.10)\n",
      "Collecting h11<0.15,>=0.13 (from httpcore==1.*->httpx>=0.27.0->chromadb)\n",
      "  Using cached h11-0.14.0-py3-none-any.whl.metadata (8.2 kB)\n",
      "Requirement already satisfied: six>=1.9.0 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from kubernetes>=28.1.0->chromadb) (1.17.0)\n",
      "Requirement already satisfied: python-dateutil>=2.5.3 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from kubernetes>=28.1.0->chromadb) (2.9.0.post0)\n",
      "Collecting google-auth>=1.0.1 (from kubernetes>=28.1.0->chromadb)\n",
      "  Using cached google_auth-2.37.0-py2.py3-none-any.whl.metadata (4.8 kB)\n",
      "Collecting websocket-client!=0.40.0,!=0.41.*,!=0.42.*,>=0.32.0 (from kubernetes>=28.1.0->chromadb)\n",
      "  Using cached websocket_client-1.8.0-py3-none-any.whl.metadata (8.0 kB)\n",
      "Requirement already satisfied: requests in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from kubernetes>=28.1.0->chromadb) (2.32.3)\n",
      "Collecting requests-oauthlib (from kubernetes>=28.1.0->chromadb)\n",
      "  Using cached requests_oauthlib-2.0.0-py2.py3-none-any.whl.metadata (11 kB)\n",
      "Collecting oauthlib>=3.2.2 (from kubernetes>=28.1.0->chromadb)\n",
      "  Using cached oauthlib-3.2.2-py3-none-any.whl.metadata (7.5 kB)\n",
      "Requirement already satisfied: urllib3>=1.24.2 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from kubernetes>=28.1.0->chromadb) (2.2.3)\n",
      "Collecting durationpy>=0.7 (from kubernetes>=28.1.0->chromadb)\n",
      "  Using cached durationpy-0.9-py3-none-any.whl.metadata (338 bytes)\n",
      "Requirement already satisfied: coloredlogs in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from onnxruntime>=1.14.1->chromadb) (15.0.1)\n",
      "Requirement already satisfied: flatbuffers in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from onnxruntime>=1.14.1->chromadb) (24.3.25)\n",
      "Requirement already satisfied: protobuf in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from onnxruntime>=1.14.1->chromadb) (5.29.1)\n",
      "Requirement already satisfied: sympy in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from onnxruntime>=1.14.1->chromadb) (1.13.3)\n",
      "Collecting deprecated>=1.2.6 (from opentelemetry-api>=1.2.0->chromadb)\n",
      "  Using cached Deprecated-1.2.15-py2.py3-none-any.whl.metadata (5.5 kB)\n",
      "Collecting googleapis-common-protos~=1.52 (from opentelemetry-exporter-otlp-proto-grpc>=1.2.0->chromadb)\n",
      "  Using cached googleapis_common_protos-1.66.0-py2.py3-none-any.whl.metadata (1.5 kB)\n",
      "Collecting opentelemetry-exporter-otlp-proto-common==1.29.0 (from opentelemetry-exporter-otlp-proto-grpc>=1.2.0->chromadb)\n",
      "  Using cached opentelemetry_exporter_otlp_proto_common-1.29.0-py3-none-any.whl.metadata (1.8 kB)\n",
      "Collecting opentelemetry-proto==1.29.0 (from opentelemetry-exporter-otlp-proto-grpc>=1.2.0->chromadb)\n",
      "  Using cached opentelemetry_proto-1.29.0-py3-none-any.whl.metadata (2.3 kB)\n",
      "Collecting opentelemetry-instrumentation-asgi==0.50b0 (from opentelemetry-instrumentation-fastapi>=0.41b0->chromadb)\n",
      "  Using cached opentelemetry_instrumentation_asgi-0.50b0-py3-none-any.whl.metadata (1.9 kB)\n",
      "Collecting opentelemetry-instrumentation==0.50b0 (from opentelemetry-instrumentation-fastapi>=0.41b0->chromadb)\n",
      "  Using cached opentelemetry_instrumentation-0.50b0-py3-none-any.whl.metadata (6.1 kB)\n",
      "Collecting opentelemetry-semantic-conventions==0.50b0 (from opentelemetry-instrumentation-fastapi>=0.41b0->chromadb)\n",
      "  Using cached opentelemetry_semantic_conventions-0.50b0-py3-none-any.whl.metadata (2.3 kB)\n",
      "Collecting opentelemetry-util-http==0.50b0 (from opentelemetry-instrumentation-fastapi>=0.41b0->chromadb)\n",
      "  Using cached opentelemetry_util_http-0.50b0-py3-none-any.whl.metadata (2.5 kB)\n",
      "Collecting wrapt<2.0.0,>=1.0.0 (from opentelemetry-instrumentation==0.50b0->opentelemetry-instrumentation-fastapi>=0.41b0->chromadb)\n",
      "  Downloading wrapt-1.17.0-cp38-cp38-win_amd64.whl.metadata (6.5 kB)\n",
      "Collecting asgiref~=3.0 (from opentelemetry-instrumentation-asgi==0.50b0->opentelemetry-instrumentation-fastapi>=0.41b0->chromadb)\n",
      "  Using cached asgiref-3.8.1-py3-none-any.whl.metadata (9.3 kB)\n",
      "Collecting monotonic>=1.5 (from posthog>=2.4.0->chromadb)\n",
      "  Using cached monotonic-1.6-py2.py3-none-any.whl.metadata (1.5 kB)\n",
      "Collecting backoff>=1.10.0 (from posthog>=2.4.0->chromadb)\n",
      "  Using cached backoff-2.2.1-py3-none-any.whl.metadata (14 kB)\n",
      "Collecting annotated-types>=0.6.0 (from pydantic>=1.9->chromadb)\n",
      "  Using cached annotated_types-0.7.0-py3-none-any.whl.metadata (15 kB)\n",
      "Collecting pydantic-core==2.27.1 (from pydantic>=1.9->chromadb)\n",
      "  Downloading pydantic_core-2.27.1-cp38-none-win_amd64.whl.metadata (6.7 kB)\n",
      "Collecting markdown-it-py>=2.2.0 (from rich>=10.11.0->chromadb)\n",
      "  Using cached markdown_it_py-3.0.0-py3-none-any.whl.metadata (6.9 kB)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from rich>=10.11.0->chromadb) (2.18.0)\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.16.4 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from tokenizers<=0.20.3,>=0.13.2->chromadb) (0.26.5)\n",
      "Collecting click>=8.0.0 (from typer>=0.9.0->chromadb)\n",
      "  Using cached click-8.1.7-py3-none-any.whl.metadata (3.0 kB)\n",
      "Collecting shellingham>=1.3.0 (from typer>=0.9.0->chromadb)\n",
      "  Using cached shellingham-1.5.4-py2.py3-none-any.whl.metadata (3.5 kB)\n",
      "Collecting httptools>=0.6.3 (from uvicorn[standard]>=0.18.3->chromadb)\n",
      "  Downloading httptools-0.6.4-cp38-cp38-win_amd64.whl.metadata (3.7 kB)\n",
      "Requirement already satisfied: python-dotenv>=0.13 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from uvicorn[standard]>=0.18.3->chromadb) (1.0.1)\n",
      "Collecting watchfiles>=0.13 (from uvicorn[standard]>=0.18.3->chromadb)\n",
      "  Downloading watchfiles-0.24.0-cp38-none-win_amd64.whl.metadata (5.0 kB)\n",
      "Collecting websockets>=10.4 (from uvicorn[standard]>=0.18.3->chromadb)\n",
      "  Downloading websockets-13.1-cp38-cp38-win_amd64.whl.metadata (7.0 kB)\n",
      "Requirement already satisfied: zipp>=3.1.0 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from importlib-resources->chromadb) (3.20.2)\n",
      "Collecting cachetools<6.0,>=2.0.0 (from google-auth>=1.0.1->kubernetes>=28.1.0->chromadb)\n",
      "  Downloading cachetools-5.5.0-py3-none-any.whl.metadata (5.3 kB)\n",
      "Collecting pyasn1-modules>=0.2.1 (from google-auth>=1.0.1->kubernetes>=28.1.0->chromadb)\n",
      "  Downloading pyasn1_modules-0.4.1-py3-none-any.whl.metadata (3.5 kB)\n",
      "Collecting rsa<5,>=3.1.4 (from google-auth>=1.0.1->kubernetes>=28.1.0->chromadb)\n",
      "  Using cached rsa-4.9-py3-none-any.whl.metadata (4.2 kB)\n",
      "Requirement already satisfied: filelock in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from huggingface-hub<1.0,>=0.16.4->tokenizers<=0.20.3,>=0.13.2->chromadb) (3.16.1)\n",
      "Requirement already satisfied: fsspec>=2023.5.0 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from huggingface-hub<1.0,>=0.16.4->tokenizers<=0.20.3,>=0.13.2->chromadb) (2024.10.0)\n",
      "Collecting mdurl~=0.1 (from markdown-it-py>=2.2.0->rich>=10.11.0->chromadb)\n",
      "  Using cached mdurl-0.1.2-py3-none-any.whl.metadata (1.6 kB)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from requests->kubernetes>=28.1.0->chromadb) (3.4.0)\n",
      "Collecting sniffio>=1.1 (from anyio->httpx>=0.27.0->chromadb)\n",
      "  Using cached sniffio-1.3.1-py3-none-any.whl.metadata (3.9 kB)\n",
      "Collecting exceptiongroup>=1.0.2 (from anyio->httpx>=0.27.0->chromadb)\n",
      "  Downloading exceptiongroup-1.2.2-py3-none-any.whl.metadata (6.6 kB)\n",
      "Requirement already satisfied: humanfriendly>=9.1 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from coloredlogs->onnxruntime>=1.14.1->chromadb) (10.0)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from sympy->onnxruntime>=1.14.1->chromadb) (1.3.0)\n",
      "Requirement already satisfied: pyreadline3 in c:\\users\\yasser\\anaconda3\\envs\\milvus\\lib\\site-packages (from humanfriendly>=9.1->coloredlogs->onnxruntime>=1.14.1->chromadb) (3.5.4)\n",
      "Collecting pyasn1<0.7.0,>=0.4.6 (from pyasn1-modules>=0.2.1->google-auth>=1.0.1->kubernetes>=28.1.0->chromadb)\n",
      "  Downloading pyasn1-0.6.1-py3-none-any.whl.metadata (8.4 kB)\n",
      "Using cached chromadb-0.5.23-py3-none-any.whl (628 kB)\n",
      "Downloading chroma_hnswlib-0.7.6-cp38-cp38-win_amd64.whl (150 kB)\n",
      "Downloading bcrypt-4.2.1-cp37-abi3-win_amd64.whl (152 kB)\n",
      "Using cached build-1.2.2.post1-py3-none-any.whl (22 kB)\n",
      "Using cached fastapi-0.115.6-py3-none-any.whl (94 kB)\n",
      "Downloading graphlib_backport-1.1.0-py3-none-any.whl (7.1 kB)\n",
      "Using cached httpx-0.28.1-py3-none-any.whl (73 kB)\n",
      "Downloading httpcore-1.0.7-py3-none-any.whl (78 kB)\n",
      "Using cached kubernetes-31.0.0-py2.py3-none-any.whl (1.9 MB)\n",
      "Downloading mmh3-5.0.1-cp38-cp38-win_amd64.whl (39 kB)\n",
      "Using cached opentelemetry_api-1.29.0-py3-none-any.whl (64 kB)\n",
      "Using cached opentelemetry_exporter_otlp_proto_grpc-1.29.0-py3-none-any.whl (18 kB)\n",
      "Using cached opentelemetry_exporter_otlp_proto_common-1.29.0-py3-none-any.whl (18 kB)\n",
      "Using cached opentelemetry_proto-1.29.0-py3-none-any.whl (55 kB)\n",
      "Using cached opentelemetry_instrumentation_fastapi-0.50b0-py3-none-any.whl (12 kB)\n",
      "Using cached opentelemetry_instrumentation-0.50b0-py3-none-any.whl (30 kB)\n",
      "Using cached opentelemetry_instrumentation_asgi-0.50b0-py3-none-any.whl (16 kB)\n",
      "Using cached opentelemetry_semantic_conventions-0.50b0-py3-none-any.whl (166 kB)\n",
      "Using cached opentelemetry_util_http-0.50b0-py3-none-any.whl (6.9 kB)\n",
      "Using cached opentelemetry_sdk-1.29.0-py3-none-any.whl (118 kB)\n",
      "Downloading orjson-3.10.12-cp38-none-win_amd64.whl (134 kB)\n",
      "Using cached overrides-7.7.0-py3-none-any.whl (17 kB)\n",
      "Using cached posthog-3.7.4-py2.py3-none-any.whl (54 kB)\n",
      "Downloading pydantic-2.10.3-py3-none-any.whl (456 kB)\n",
      "Downloading pydantic_core-2.27.1-cp38-none-win_amd64.whl (2.0 MB)\n",
      "   ---------------------------------------- 0.0/2.0 MB ? eta -:--:--\n",
      "   ------------------------------- -------- 1.6/2.0 MB 7.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.0/2.0 MB 7.8 MB/s eta 0:00:00\n",
      "Downloading rich-13.9.4-py3-none-any.whl (242 kB)\n",
      "Using cached tenacity-9.0.0-py3-none-any.whl (28 kB)\n",
      "Downloading typer-0.15.1-py3-none-any.whl (44 kB)\n",
      "Using cached uvicorn-0.32.1-py3-none-any.whl (63 kB)\n",
      "Using cached importlib_resources-6.4.5-py3-none-any.whl (36 kB)\n",
      "Using cached annotated_types-0.7.0-py3-none-any.whl (13 kB)\n",
      "Using cached backoff-2.2.1-py3-none-any.whl (15 kB)\n",
      "Using cached click-8.1.7-py3-none-any.whl (97 kB)\n",
      "Using cached Deprecated-1.2.15-py2.py3-none-any.whl (9.9 kB)\n",
      "Using cached durationpy-0.9-py3-none-any.whl (3.5 kB)\n",
      "Using cached google_auth-2.37.0-py2.py3-none-any.whl (209 kB)\n",
      "Using cached googleapis_common_protos-1.66.0-py2.py3-none-any.whl (221 kB)\n",
      "Using cached h11-0.14.0-py3-none-any.whl (58 kB)\n",
      "Downloading httptools-0.6.4-cp38-cp38-win_amd64.whl (89 kB)\n",
      "Using cached markdown_it_py-3.0.0-py3-none-any.whl (87 kB)\n",
      "Using cached monotonic-1.6-py2.py3-none-any.whl (8.2 kB)\n",
      "Using cached oauthlib-3.2.2-py3-none-any.whl (151 kB)\n",
      "Using cached shellingham-1.5.4-py2.py3-none-any.whl (9.8 kB)\n",
      "Using cached starlette-0.41.3-py3-none-any.whl (73 kB)\n",
      "Downloading anyio-4.5.2-py3-none-any.whl (89 kB)\n",
      "Downloading tomli-2.2.1-py3-none-any.whl (14 kB)\n",
      "Downloading watchfiles-0.24.0-cp38-none-win_amd64.whl (277 kB)\n",
      "Using cached websocket_client-1.8.0-py3-none-any.whl (58 kB)\n",
      "Downloading websockets-13.1-cp38-cp38-win_amd64.whl (159 kB)\n",
      "Using cached pyproject_hooks-1.2.0-py3-none-any.whl (10 kB)\n",
      "Using cached requests_oauthlib-2.0.0-py2.py3-none-any.whl (24 kB)\n",
      "Using cached asgiref-3.8.1-py3-none-any.whl (23 kB)\n",
      "Downloading cachetools-5.5.0-py3-none-any.whl (9.5 kB)\n",
      "Downloading exceptiongroup-1.2.2-py3-none-any.whl (16 kB)\n",
      "Using cached mdurl-0.1.2-py3-none-any.whl (10.0 kB)\n",
      "Downloading pyasn1_modules-0.4.1-py3-none-any.whl (181 kB)\n",
      "Using cached rsa-4.9-py3-none-any.whl (34 kB)\n",
      "Using cached sniffio-1.3.1-py3-none-any.whl (10 kB)\n",
      "Downloading wrapt-1.17.0-cp38-cp38-win_amd64.whl (38 kB)\n",
      "Downloading pyasn1-0.6.1-py3-none-any.whl (83 kB)\n",
      "Building wheels for collected packages: pypika\n",
      "  Building wheel for pypika (pyproject.toml): started\n",
      "  Building wheel for pypika (pyproject.toml): finished with status 'done'\n",
      "  Created wheel for pypika: filename=PyPika-0.48.9-py2.py3-none-any.whl size=53835 sha256=1e9ecb4fbf493b19501e4f61132c3bd1509f9f6ec4244bdd394bfc8b22aa53f5\n",
      "  Stored in directory: c:\\users\\yasser\\appdata\\local\\pip\\cache\\wheels\\54\\4a\\f8\\2803c6041841502d0d21fb2a62d401d14716dfeb2261a3a70b\n",
      "Successfully built pypika\n",
      "Installing collected packages: pypika, monotonic, durationpy, wrapt, websockets, websocket-client, tomli, tenacity, sniffio, shellingham, pyproject_hooks, pydantic-core, pyasn1, overrides, orjson, opentelemetry-util-http, opentelemetry-proto, oauthlib, mmh3, mdurl, importlib-resources, httptools, h11, graphlib_backport, googleapis-common-protos, exceptiongroup, click, chroma-hnswlib, cachetools, bcrypt, backoff, asgiref, annotated-types, uvicorn, rsa, requests-oauthlib, pydantic, pyasn1-modules, posthog, opentelemetry-exporter-otlp-proto-common, markdown-it-py, httpcore, deprecated, build, anyio, watchfiles, starlette, rich, opentelemetry-api, httpx, google-auth, typer, opentelemetry-semantic-conventions, kubernetes, fastapi, opentelemetry-sdk, opentelemetry-instrumentation, opentelemetry-instrumentation-asgi, opentelemetry-exporter-otlp-proto-grpc, opentelemetry-instrumentation-fastapi, chromadb\n",
      "Successfully installed annotated-types-0.7.0 anyio-4.5.2 asgiref-3.8.1 backoff-2.2.1 bcrypt-4.2.1 build-1.2.2.post1 cachetools-5.5.0 chroma-hnswlib-0.7.6 chromadb-0.5.23 click-8.1.7 deprecated-1.2.15 durationpy-0.9 exceptiongroup-1.2.2 fastapi-0.115.6 google-auth-2.37.0 googleapis-common-protos-1.66.0 graphlib_backport-1.1.0 h11-0.14.0 httpcore-1.0.7 httptools-0.6.4 httpx-0.28.1 importlib-resources-6.4.5 kubernetes-31.0.0 markdown-it-py-3.0.0 mdurl-0.1.2 mmh3-5.0.1 monotonic-1.6 oauthlib-3.2.2 opentelemetry-api-1.29.0 opentelemetry-exporter-otlp-proto-common-1.29.0 opentelemetry-exporter-otlp-proto-grpc-1.29.0 opentelemetry-instrumentation-0.50b0 opentelemetry-instrumentation-asgi-0.50b0 opentelemetry-instrumentation-fastapi-0.50b0 opentelemetry-proto-1.29.0 opentelemetry-sdk-1.29.0 opentelemetry-semantic-conventions-0.50b0 opentelemetry-util-http-0.50b0 orjson-3.10.12 overrides-7.7.0 posthog-3.7.4 pyasn1-0.6.1 pyasn1-modules-0.4.1 pydantic-2.10.3 pydantic-core-2.27.1 pypika-0.48.9 pyproject_hooks-1.2.0 requests-oauthlib-2.0.0 rich-13.9.4 rsa-4.9 shellingham-1.5.4 sniffio-1.3.1 starlette-0.41.3 tenacity-9.0.0 tomli-2.2.1 typer-0.15.1 uvicorn-0.32.1 watchfiles-0.24.0 websocket-client-1.8.0 websockets-13.1 wrapt-1.17.0\n"
     ]
    }
   ],
   "source": [
    "!pip install chromadb #installing the vector database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "1dbdf9d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Embeddings added successfully\n"
     ]
    }
   ],
   "source": [
    "#ADDING THE EMBEDDINGS to the vector database\n",
    "import chromadb \n",
    "\n",
    "client = chromadb.Client()\n",
    "\n",
    "# Create or load a collection\n",
    "collection = client.create_collection(\"neo4j_nodes\")  # Create a collection for the embeddings\n",
    "def store_embeddings_in_batches(embeddings, nodes, batch_size=10):\n",
    "    for i in range(0, len(nodes), batch_size):\n",
    "        batch_nodes = nodes[i:i + batch_size]\n",
    "        batch_embeddings = embeddings[i:i + batch_size]\n",
    "        for node, embedding in zip(batch_nodes, batch_embeddings):\n",
    "            document = json.dumps(node['data']) if isinstance(node['data'], dict) else str(node['data'])\n",
    "            collection.add(\n",
    "                ids=[node['data']['id']],\n",
    "                documents=[document],\n",
    "                embeddings=[embedding]\n",
    "            )\n",
    "store_embeddings_in_batches(embeddings, nodes)\n",
    "print(\"Embeddings added successfully\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebc9caf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#FUNCTION TO CREATE THE EMBEDDING FOR THE QUERY \n",
    "def query_collection_with_embedding(collection, query_text, model, n_results=2):\n",
    "    \"\"\"\n",
    "    Function to query a collection using a generated embedding for the provided query text.\n",
    "    \n",
    "    Parameters:\n",
    "        collection: The collection object to query.\n",
    "        query_text (str): The query text to generate an embedding for.\n",
    "        model: The model used for generating embeddings.\n",
    "        n_results (int): The number of results to retrieve (default is 2).\n",
    "    \n",
    "    Returns:\n",
    "        list: A list of query results.\n",
    "    \"\"\"\n",
    "    def generate_query_embedding(query_text):\n",
    "        \"\"\"Generates embedding for a given query text.\"\"\"\n",
    "        return model.encode([query_text])[0]\n",
    "    \n",
    "    # Perform the query with the generated embedding\n",
    "    results = collection.query(\n",
    "        query_embeddings=generate_query_embedding(query_text),\n",
    "        n_results=n_results\n",
    "    )\n",
    "    \n",
    "    return results\n",
    "\n",
    "# Example usage:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b6f3b370",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ids': [['D012513', 'D012804']], 'embeddings': None, 'documents': [['{\"DISEASE_SUPERGRP\": [\"C04.619\", \" C04.557.450.795.830\", \" C04\", \" C04.557.450.795\", \" C04.557\", \" C04.557.450\", \" C04.619.857\"], \"id\": \"D012513\", \"TYPE\": \"DISEASE\", \"NAME\": [\"Sarcoma\", \" Experimental\"]}', '{\"DISEASE_SUPERGRP\": [\"C14.280.067.093.249\", \" C23.550.073.425\", \" C23.550\", \" C14.280.067.093\", \" C14.280.123.500\", \" C14.280.067.558.536\", \" C23.550.073.425.440\", \" C14.280.123\", \" C23.550.073.093.249\", \" C14.280.123.500.536\", \" C23\", \" C14.280.067\", \" C23.550.073\", \" C14.280\", \" C14.280.067\", \" C23.550.073\", \" C14\"], \"id\": \"D012804\", \"TYPE\": \"DISEASE\", \"NAME\": \"Sick Sinus Syndrome\"}']], 'uris': None, 'data': None, 'metadatas': [[None, None]], 'distances': [[1.7967413663864136, 1.7967413663864136]], 'included': [\"<IncludeEnum.distances: 'distances'>\", \"<IncludeEnum.documents: 'documents'>\", \"<IncludeEnum.metadatas: 'metadatas'>\"]}\n"
     ]
    }
   ],
   "source": [
    "#RUNNING THE QUERIES IN VECTOR DATABASE\n",
    "query_text = \"Drugs for covid\"\n",
    "results = query_collection_with_embedding(collection, query_text, model)\n",
    "print(results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "bbcf4df7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Yasser\\AppData\\Local\\Temp\\ipykernel_3948\\15230747.py:30: DeprecationWarning: Using a driver after it has been closed is deprecated. Future versions of the driver will raise an error.\n",
      "  with self.driver.session() as session:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4311' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C04.619', ' C04.557.450.795.830', ' C04', ' C04.557.450.795', ' C04.557', ' C04.557.450', ' C04.619.857'], 'id': 'D012513', 'TYPE': 'DISEASE', 'NAME': ['Sarcoma', ' Experimental']}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4372' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C14.280.067.093.249', ' C23.550.073.425', ' C23.550', ' C14.280.067.093', ' C14.280.123.500', ' C14.280.067.558.536', ' C23.550.073.425.440', ' C14.280.123', ' C23.550.073.093.249', ' C14.280.123.500.536', ' C23', ' C14.280.067.558', ' C23.550.073.093', ' C14.280', ' C14.280.067', ' C23.550.073', ' C14'], 'id': 'D012804', 'TYPE': 'DISEASE', 'NAME': 'Sick Sinus Syndrome'}>>\n"
     ]
    }
   ],
   "source": [
    "#FETCHING THE NODES FROM GRAPH BASED ON THE ids\n",
    "node_id = ['D012513','D012804']  # Replace with the actual node ID\n",
    "for i in node_id:\n",
    "    node = neo4j_manager.fetch_node_by_id(i)\n",
    "    print(node)\n",
    "neo4j_manager.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10b295ae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "62cf17e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"ids\": [[\"D011469\", \"D012078\", \"D012513\", \"D012804\"]], \"embeddings\": null, \"documents\": [[\"{\\\"DISEASE_SUPERGRP\\\": [\\\"C12\\\", \\\" C12.294\\\", \\\" C12.294.565\\\"], \\\"id\\\": \\\"D011469\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": \\\"Prostatic Diseases\\\"}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C14.907.137.727\\\", \\\" C14.907\\\", \\\" C14\\\", \\\" C12.777.419\\\", \\\" C14.907.137\\\", \\\" C13.351.968\\\", \\\" C12\\\", \\\" C13\\\", \\\" C13.351.968.419.775\\\", \\\" C12.777\\\", \\\" C13.351.968.419\\\", \\\" C12.777.419.775\\\", \\\" C13.351\\\"], \\\"id\\\": \\\"D012078\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": \\\"Renal Artery Obstruction\\\"}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C04.619\\\", \\\" C04.557.450.795.830\\\", \\\" C04\\\", \\\" C04.557.450.795\\\", \\\" C04.557\\\", \\\" C04.557.450\\\", \\\" C04.619.857\\\"], \\\"id\\\": \\\"D012513\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": [\\\"Sarcoma\\\", \\\" Experimental\\\"]}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C14.280.067.093.249\\\", \\\" C23.550.073.425\\\", \\\" C23.550\\\", \\\" C14.280.067.093\\\", \\\" C14.280.123.500\\\", \\\" C14.280.067.558.536\\\", \\\" C23.550.073.425.440\\\", \\\" C14.280.123\\\", \\\" C23.550.073.093.249\\\", \\\" C14.280.123.500.536\\\", \\\" C23\\\", \\\" C14.280.067.558\\\", \\\" C23.550.073.093\\\", \\\" C14.280\\\", \\\" C14.280.067\\\", \\\" C23.550.073\\\", \\\" C14\\\"], \\\"id\\\": \\\"D012804\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": \\\"Sick Sinus Syndrome\\\"}\"]], \"uris\": null, \"data\": null, \"metadatas\": [[null, null, null, null]], \"distances\": [[2.050719738006592, 2.050719738006592, 2.050719738006592, 2.050719738006592]], \"included\": [\"<IncludeEnum.distances: 'distances'>\", \"<IncludeEnum.documents: 'documents'>\", \"<IncludeEnum.metadatas: 'metadatas'>\"]}\n"
     ]
    }
   ],
   "source": [
    "#RUNNING THE QUERIES IN VECTOR DATABASE\n",
    "query_text = \"Proteins and drugs associated with Covid and Tuberculosis \"\n",
    "results = query_collection_with_embedding(collection, query_text, model,n_results=4)\n",
    "print(results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "21962287",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Yasser\\AppData\\Local\\Temp\\ipykernel_3948\\15230747.py:30: DeprecationWarning: Using a driver after it has been closed is deprecated. Future versions of the driver will raise an error.\n",
      "  with self.driver.session() as session:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4173' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C12', ' C12.294', ' C12.294.565'], 'id': 'D011469', 'TYPE': 'DISEASE', 'NAME': 'Prostatic Diseases'}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4212' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C14.907.137.727', ' C14.907', ' C14', ' C12.777.419', ' C14.907.137', ' C13.351.968', ' C12', ' C13', ' C13.351.968.419.775', ' C12.777', ' C13.351.968.419', ' C12.777.419.775', ' C13.351'], 'id': 'D012078', 'TYPE': 'DISEASE', 'NAME': 'Renal Artery Obstruction'}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4311' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C04.619', ' C04.557.450.795.830', ' C04', ' C04.557.450.795', ' C04.557', ' C04.557.450', ' C04.619.857'], 'id': 'D012513', 'TYPE': 'DISEASE', 'NAME': ['Sarcoma', ' Experimental']}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4372' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C14.280.067.093.249', ' C23.550.073.425', ' C23.550', ' C14.280.067.093', ' C14.280.123.500', ' C14.280.067.558.536', ' C23.550.073.425.440', ' C14.280.123', ' C23.550.073.093.249', ' C14.280.123.500.536', ' C23', ' C14.280.067.558', ' C23.550.073.093', ' C14.280', ' C14.280.067', ' C23.550.073', ' C14'], 'id': 'D012804', 'TYPE': 'DISEASE', 'NAME': 'Sick Sinus Syndrome'}>>\n"
     ]
    }
   ],
   "source": [
    "#FETCHING THE NODES FROM GRAPH BASED ON THE ids\n",
    "node_id = [\"D011469\", \"D012078\", \"D012513\", \"D012804\"] \n",
    "for i in node_id:\n",
    "    node = neo4j_manager.fetch_node_by_id(i)\n",
    "    print(node)\n",
    "neo4j_manager.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "375129d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"ids\": [[\"D011271\", \"D011469\", \"D011535\", \"D012078\", \"D011694\", \"D012030\", \"D012513\", \"D012913\", \"D012804\", \"D012817\"]], \"embeddings\": null, \"documents\": [[\"{\\\"DISEASE_SUPERGRP\\\": [\\\"C13.703\\\", \\\" C13\\\", \\\" C13.703.733\\\"], \\\"id\\\": \\\"D011271\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": [\\\"Pregnancy\\\", \\\" Ectopic\\\"]}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C12\\\", \\\" C12.294\\\", \\\" C12.294.565\\\"], \\\"id\\\": \\\"D011469\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": \\\"Prostatic Diseases\\\"}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C16.131.077\\\", \\\" C16.131\\\", \\\" C16.131.077.745\\\", \\\" C16\\\"], \\\"id\\\": \\\"D011535\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": \\\"Prune Belly Syndrome\\\"}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C14.907.137.727\\\", \\\" C14.907\\\", \\\" C14\\\", \\\" C12.777.419\\\", \\\" C14.907.137\\\", \\\" C13.351.968\\\", \\\" C12\\\", \\\" C13\\\", \\\" C13.351.968.419.775\\\", \\\" C12.777\\\", \\\" C13.351.968.419\\\", \\\" C12.777.419.775\\\", \\\" C13.351\\\"], \\\"id\\\": \\\"D012078\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": \\\"Renal Artery Obstruction\\\"}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C15.378\\\", \\\" C23.888\\\", \\\" C14.907\\\", \\\" C23.888.885.687\\\", \\\" C23.550\\\", \\\" C15\\\", \\\" C23\\\", \\\" C15.378.100.802\\\", \\\" C14\\\", \\\" C23.550.414\\\", \\\" C23.888.885\\\", \\\" C23.550.414.950\\\", \\\" C15.378.100.802.250\\\", \\\" C23.550.414.950.250\\\", \\\" C14.907.454.550\\\", \\\" C15.378.463\\\", \\\" C14.907.454\\\", \\\" C15.378.463.515.550\\\", \\\" C15.378.100\\\", \\\" C15.378.463.515\\\", \\\" C23.888.885.687.250\\\"], \\\"id\\\": \\\"D011694\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": [\\\"Purpura\\\", \\\" Hyperglobulinemic\\\"]}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C11.744\\\", \\\" C11\\\"], \\\"id\\\": \\\"D012030\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": \\\"Refractive Errors\\\"}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C04.619\\\", \\\" C04.557.450.795.830\\\", \\\" C04\\\", \\\" C04.557.450.795\\\", \\\" C04.557\\\", \\\" C04.557.450\\\", \\\" C04.619.857\\\"], \\\"id\\\": \\\"D012513\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": [\\\"Sarcoma\\\", \\\" Experimental\\\"]}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C23.888.852.779\\\", \\\" C23.888.852.779.850\\\", \\\" C23.888\\\", \\\" C23\\\", \\\" C23.888.852\\\"], \\\"id\\\": \\\"D012913\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": \\\"Snoring\\\"}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C14.280.067.093.249\\\", \\\" C23.550.073.425\\\", \\\" C23.550\\\", \\\" C14.280.067.093\\\", \\\" C14.280.123.500\\\", \\\" C14.280.067.558.536\\\", \\\" C23.550.073.425.440\\\", \\\" C14.280.123\\\", \\\" C23.550.073.093.249\\\", \\\" C14.280.123.500.536\\\", \\\" C23\\\", \\\" C14.280.067.558\\\", \\\" C23.550.073.093\\\", \\\" C14.280\\\", \\\" C14.280.067\\\", \\\" C23.550.073\\\", \\\" C14\\\"], \\\"id\\\": \\\"D012804\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": \\\"Sick Sinus Syndrome\\\"}\", \"{\\\"DISEASE_SUPERGRP\\\": [\\\"C23.888\\\", \\\" C23.888.821\\\", \\\" C23\\\"], \\\"id\\\": \\\"D012817\\\", \\\"TYPE\\\": \\\"DISEASE\\\", \\\"NAME\\\": [\\\"Signs and Symptoms\\\", \\\" Digestive\\\"]}\"]], \"uris\": null, \"data\": null, \"metadatas\": [[null, null, null, null, null, null, null, null, null, null]], \"distances\": [[1.895555019378662, 1.895555019378662, 1.895555019378662, 1.895555019378662, 1.895555019378662, 1.895555019378662, 1.895555019378662, 1.895555019378662, 1.895555019378662, 1.895555019378662]], \"included\": [\"<IncludeEnum.distances: 'distances'>\", \"<IncludeEnum.documents: 'documents'>\", \"<IncludeEnum.metadatas: 'metadatas'>\"]}\n"
     ]
    }
   ],
   "source": [
    "#RUNNING THE QUERIES IN VECTOR DATABASE\n",
    "query_text = \"Give Pubmed papers for the drugs for covid\"\n",
    "results = query_collection_with_embedding(collection, query_text, model,n_results=10)\n",
    "print(results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "58b15f51",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Yasser\\AppData\\Local\\Temp\\ipykernel_3948\\15230747.py:30: DeprecationWarning: Using a driver after it has been closed is deprecated. Future versions of the driver will raise an error.\n",
      "  with self.driver.session() as session:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4164' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C13.703', ' C13', ' C13.703.733'], 'id': 'D011271', 'TYPE': 'DISEASE', 'NAME': ['Pregnancy', ' Ectopic']}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4173' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C12', ' C12.294', ' C12.294.565'], 'id': 'D011469', 'TYPE': 'DISEASE', 'NAME': 'Prostatic Diseases'}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4177' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C16.131.077', ' C16.131', ' C16.131.077.745', ' C16'], 'id': 'D011535', 'TYPE': 'DISEASE', 'NAME': 'Prune Belly Syndrome'}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4212' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C14.907.137.727', ' C14.907', ' C14', ' C12.777.419', ' C14.907.137', ' C13.351.968', ' C12', ' C13', ' C13.351.968.419.775', ' C12.777', ' C13.351.968.419', ' C12.777.419.775', ' C13.351'], 'id': 'D012078', 'TYPE': 'DISEASE', 'NAME': 'Renal Artery Obstruction'}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4222' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C15.378', ' C23.888', ' C14.907', ' C23.888.885.687', ' C23.550', ' C15', ' C23', ' C15.378.100.802', ' C14', ' C23.550.414', ' C23.888.885', ' C23.550.414.950', ' C15.378.100.802.250', ' C23.550.414.950.250', ' C14.907.454.550', ' C15.378.463', ' C14.907.454', ' C15.378.463.515.550', ' C15.378.100', ' C15.378.463.515', ' C23.888.885.687.250'], 'id': 'D011694', 'TYPE': 'DISEASE', 'NAME': ['Purpura', ' Hyperglobulinemic']}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4228' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C11.744', ' C11'], 'id': 'D012030', 'TYPE': 'DISEASE', 'NAME': 'Refractive Errors'}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4311' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C04.619', ' C04.557.450.795.830', ' C04', ' C04.557.450.795', ' C04.557', ' C04.557.450', ' C04.619.857'], 'id': 'D012513', 'TYPE': 'DISEASE', 'NAME': ['Sarcoma', ' Experimental']}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4363' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C23.888.852.779', ' C23.888.852.779.850', ' C23.888', ' C23', ' C23.888.852'], 'id': 'D012913', 'TYPE': 'DISEASE', 'NAME': 'Snoring'}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4372' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C14.280.067.093.249', ' C23.550.073.425', ' C23.550', ' C14.280.067.093', ' C14.280.123.500', ' C14.280.067.558.536', ' C23.550.073.425.440', ' C14.280.123', ' C23.550.073.093.249', ' C14.280.123.500.536', ' C23', ' C14.280.067.558', ' C23.550.073.093', ' C14.280', ' C14.280.067', ' C23.550.073', ' C14'], 'id': 'D012804', 'TYPE': 'DISEASE', 'NAME': 'Sick Sinus Syndrome'}>>\n",
      "<Record n=<Node element_id='4:989e726f-cbf2-40b1-9a31-ae15018dc948:4373' labels=frozenset({'Disease'}) properties={'DISEASE_SUPERGRP': ['C23.888', ' C23.888.821', ' C23'], 'id': 'D012817', 'TYPE': 'DISEASE', 'NAME': ['Signs and Symptoms', ' Digestive']}>>\n"
     ]
    }
   ],
   "source": [
    "#FETCHING THE NODES FROM GRAPH BASED ON THE ids\n",
    "node_id = [\"D011271\", \"D011469\", \"D011535\", \"D012078\", \"D011694\", \"D012030\", \"D012513\", \"D012913\", \"D012804\", \"D012817\"] \n",
    "for i in node_id:\n",
    "    node = neo4j_manager.fetch_node_by_id(i)\n",
    "    print(node)\n",
    "neo4j_manager.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95eaa698",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d758719",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e4471072",
   "metadata": {},
   "source": [
    "QUESTION 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8fef274f",
   "metadata": {},
   "outputs": [],
   "source": [
    "groq_api_key=\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62f8c81b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6997fded",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Yasser\\AppData\\Local\\Temp\\ipykernel_24248\\832041514.py:17: LangChainDeprecationWarning: The class `Neo4jGraph` was deprecated in LangChain 0.3.8 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-neo4j package and should be used instead. To use it run `pip install -U :class:`~langchain-neo4j` and import as `from :class:`~langchain_neo4j import Neo4jGraph``.\n",
      "  graph=Neo4jGraph(\n"
     ]
    }
   ],
   "source": [
    "# INSTALLING THE NECESARY LIBRARIES\n",
    "# !pip install --upgrade --quiet  langchain langchain-community langchain-groq neo4j\n",
    "# !pip install rapidfuzz\n",
    "# !pip install spacy\n",
    "# !python -m spacy download en_core_web_md\n",
    "# !pip install --upgrade --quiet langchain_experimental\n",
    "\n",
    "\n",
    "## Graphdb configuration\n",
    "\n",
    "import os\n",
    "\n",
    "os.environ[\"url\"]=url # setting the keys in the environemnt\n",
    "os.environ[\"username\"]=username\n",
    "os.environ[\"password\"]=password\n",
    "os.environ['groq_api_key']=groq_api_key\n",
    "from langchain_community.graphs import Neo4jGraph\n",
    "graph=Neo4jGraph(\n",
    "    url=url,\n",
    "    username=username,\n",
    "    password=password,\n",
    ")\n",
    "\n",
    "from langchain_groq import ChatGroq\n",
    "\n",
    "llm=ChatGroq(groq_api_key=groq_api_key,model_name=\"llama3-70b-8192\") # calling the LLM MODEL\n",
    "             ##\"llama3-70b-8192\")\n",
    "             ##\"llama-3.1-70b-versatile\")\n",
    "             ##\"llama3-8b-8192\")\n",
    "             ##\"Gemma2-9b-It\")\n",
    "             ##\"llama-3.2-3b-preview\")\n",
    "             ##\"llama-3.1-8b-instant\")\n",
    "\n",
    "#SETTING UP THE MODEL USING THE LANGCHAIN  AND LLTRANSFORMER AND GRAPHSCYPHER \n",
    "from langchain_experimental.graph_transformers import LLMGraphTransformer\n",
    "llm_transformer=LLMGraphTransformer(llm=llm)\n",
    "from langchain.chains import GraphCypherQAChain\n",
    "chain=GraphCypherQAChain.from_llm(llm=llm,graph=graph,verbose=True,allow_dangerous_requests=True,return_intermediate_steps=True,) \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5121dbed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: spacy in c:\\users\\yasser\\anaconda3\\lib\\site-packages (3.7.5)Note: you may need to restart the kernel to use updated packages.\n",
      "\n",
      "Requirement already satisfied: spacy-legacy<3.1.0,>=3.0.11 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (3.0.12)\n",
      "Requirement already satisfied: spacy-loggers<2.0.0,>=1.0.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (1.0.5)\n",
      "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (1.0.10)\n",
      "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (2.0.8)\n",
      "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (3.0.9)\n",
      "Requirement already satisfied: thinc<8.3.0,>=8.2.2 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (8.2.5)\n",
      "Requirement already satisfied: wasabi<1.2.0,>=0.9.1 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (1.1.3)\n",
      "Requirement already satisfied: srsly<3.0.0,>=2.4.3 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (2.4.8)\n",
      "Requirement already satisfied: catalogue<2.1.0,>=2.0.6 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (2.0.10)\n",
      "Requirement already satisfied: weasel<0.5.0,>=0.1.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (0.4.1)\n",
      "Requirement already satisfied: typer<1.0.0,>=0.3.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (0.12.3)\n",
      "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (4.66.4)\n",
      "Requirement already satisfied: requests<3.0.0,>=2.13.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (2.32.3)\n",
      "Requirement already satisfied: pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (2.9.2)\n",
      "Requirement already satisfied: jinja2 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (3.1.3)\n",
      "Requirement already satisfied: setuptools in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (75.6.0)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (24.1)\n",
      "Requirement already satisfied: langcodes<4.0.0,>=3.2.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (3.4.0)\n",
      "Requirement already satisfied: numpy>=1.19.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from spacy) (1.26.4)\n",
      "Requirement already satisfied: language-data>=1.2 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from langcodes<4.0.0,>=3.2.0->spacy) (1.2.0)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.23.4 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy) (2.23.4)\n",
      "Requirement already satisfied: typing-extensions>=4.6.1 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy) (4.9.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from requests<3.0.0,>=2.13.0->spacy) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from requests<3.0.0,>=2.13.0->spacy) (3.4)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from requests<3.0.0,>=2.13.0->spacy) (2.0.7)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from requests<3.0.0,>=2.13.0->spacy) (2024.7.4)\n",
      "Requirement already satisfied: blis<0.8.0,>=0.7.8 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from thinc<8.3.0,>=8.2.2->spacy) (0.7.11)\n",
      "Requirement already satisfied: confection<1.0.0,>=0.0.1 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from thinc<8.3.0,>=8.2.2->spacy) (0.1.5)\n",
      "Requirement already satisfied: colorama in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from tqdm<5.0.0,>=4.38.0->spacy) (0.4.6)\n",
      "Requirement already satisfied: click>=8.0.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from typer<1.0.0,>=0.3.0->spacy) (8.1.7)\n",
      "Requirement already satisfied: shellingham>=1.3.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from typer<1.0.0,>=0.3.0->spacy) (1.5.4)\n",
      "Requirement already satisfied: rich>=10.11.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from typer<1.0.0,>=0.3.0->spacy) (13.3.5)\n",
      "Requirement already satisfied: cloudpathlib<1.0.0,>=0.7.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from weasel<0.5.0,>=0.1.0->spacy) (0.18.1)\n",
      "Requirement already satisfied: smart-open<8.0.0,>=5.2.1 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from weasel<0.5.0,>=0.1.0->spacy) (5.2.1)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from jinja2->spacy) (2.1.3)\n",
      "Requirement already satisfied: marisa-trie>=0.7.7 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from language-data>=1.2->langcodes<4.0.0,>=3.2.0->spacy) (1.2.0)\n",
      "Requirement already satisfied: markdown-it-py<3.0.0,>=2.2.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from rich>=10.11.0->typer<1.0.0,>=0.3.0->spacy) (2.2.0)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from rich>=10.11.0->typer<1.0.0,>=0.3.0->spacy) (2.15.1)\n",
      "Requirement already satisfied: mdurl~=0.1 in c:\\users\\yasser\\anaconda3\\lib\\site-packages (from markdown-it-py<3.0.0,>=2.2.0->rich>=10.11.0->typer<1.0.0,>=0.3.0->spacy) (0.1.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install spacy #INSTALLING LIBRARY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b3a16e43",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "104e7410",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9243b4bf",
   "metadata": {},
   "source": [
    "In the code we give the query to the LLM, LLM FIRST AUGMENT THE USER QUERY USING BIOMEDICAL BASED NLP MODEL AND THE FUZZY LOGIC, LIKE if user enters the tuberculs  it correct it to tuberculosis.Moreover if the syntax is not captured it runs the model max three times.\n",
    "It also performs the erroe handling.AND IF THE GRPAH DOES NOT GET THE SYSNTAX IT PRINTS LIKE I DONT KNOW OR SOMETHING LIKE THAT.\n",
    "IT IS DESIGNED AS THE CONTINUOS CHATBOT\n",
    "\n",
    "Inconsistent Handling of Variations in Input\n",
    "When entering terms like \"Breast cancer\" instead of \"Breast_Cancer,\" the query failed to work correctly due to mismatched term formatting.\n",
    "Context Retrieval Issue\n",
    "Although the Cypher query was generated successfully and returned the correct results from the graph, the LLM sometimes failed to properly capture or utilize the response as context.\n",
    "Inconsistent Query Generation\n",
    "For identical user queries, the LLM occasionally failed to generate the correct Cypher query, leading to unsuccessful data retrieval.\n",
    "Issues with Reverse Relationship Handling\n",
    "The system struggled to properly interpret or capture reverse relationships, resulting in incomplete or incorrect responses.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Inconsistent Handling of Input Variations\n",
    "Problem:\n",
    "Queries like \"Breast cancer\" failed to match \"Breast_Cancer\" due to formatting inconsistencies.\n",
    "Solution:\n",
    "Before passing the user query to the system, we enhanced it by:\n",
    "Correcting spelling errors.\n",
    "Standardizing any term related to breast cancer as \"Breast_Cancer\" to ensure accurate matching.\n",
    "Context Retrieval Issue from Cypher Query\n",
    "Problem:\n",
    "Even though the Cypher query generated the correct response from the graph, the LLM sometimes failed to capture it properly as context.\n",
    "Solution:\n",
    "When Cypher returns a valid result, we pass that context back to the LLM to convert it into a natural language response for the user.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Inconsistent Handling of Input Variations\n",
    "Problem:\n",
    "Queries like \"Breast cancer\" failed to match \"Breast_Cancer\" due to formatting inconsistencies.\n",
    "Solution:\n",
    "Before passing the user query to the system, we enhanced it by:\n",
    "Correcting spelling errors.\n",
    "Standardizing any term related to breast cancer as \"Breast_Cancer\" to ensure accurate matching.\n",
    "Context Retrieval Issue from Cypher Query\n",
    "Problem:\n",
    "Even though the Cypher query generated the correct response from the graph, the LLM sometimes failed to capture it properly as context.\n",
    "Solution:\n",
    "When Cypher returns a valid result, we pass that context back to the LLM to convert it into a natural language response for the user.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "26c1719a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONTIINUOS CHATBOT\n",
    "import spacy\n",
    "from rapidfuzz import fuzz, process  # For fuzzy matching\n",
    "\n",
    "class ContinuousChatbot:\n",
    "    def __init__(self, chain, nlp_model=\"en_core_web_md\"):\n",
    "        self.chain = chain\n",
    "        self.previous_context = \"\"\n",
    "        self.nlp = spacy.load(nlp_model)  # Load a robust NLP model\n",
    "\n",
    "        # Standardized disease terms and their variations\n",
    "        self.standard_diseases = {\n",
    "            \"Tuberculosis\": [\"tb\", \"tuberculoses\", \"tubercular\", \"Tuberculosis\", \"tuberculosis\"],\n",
    "            \"Ovarian Cancer\": [\"ovarian cancer\", \"ovarian carcinoma\", \"ovarian neoplasm\", \"Ovarian cancer\", \"ovarian cancer\"],\n",
    "            \"Breast Cancer\": [\"breast cancer\", \"mammary carcinoma\", \"breast neoplasm\", \"Breast cancer\", \"breast carcinoma\"],\n",
    "            \"Diabetes Mellitus\": [\"diabetes\", \"diabetes mellitus\", \"type 1 diabetes\", \"type 2 diabetes\", \"diabetes type 2\", \"diabetes mellitus\", \"DM\"],\n",
    "            \"Asthma\": [\"asthma\", \"bronchial asthma\", \"reactive airway disease\", \"asthma\", \"bronchial asthma\"],\n",
    "            \"HIV/AIDS\": [\"HIV\", \"aids\", \"human immunodeficiency virus\", \"acquired immunodeficiency syndrome\", \"HIV/AIDS\"],\n",
    "            \"Hypertension\": [\"hypertension\", \"high blood pressure\", \"HBP\", \"hypertensive disorder\", \"hypertension\"],\n",
    "            \"Parkinson's Disease\": [\"parkinson's disease\", \"parkinsonism\", \"PD\", \"parkinson disease\", \"Parkinson's disease\"],\n",
    "            \"Alzheimer's Disease\": [\"alzheimer's disease\", \"alzheimer disease\", \"AD\", \"alzheimers\", \"Alzheimer's disease\"],\n",
    "            \"Influenza\": [\"influenza\", \"flu\", \"seasonal flu\", \"swine flu\", \"H1N1\", \"influenza virus\"],\n",
    "            \"Malaria\": [\"malaria\", \"plasmodium infection\", \"malaria fever\", \"malaria\"],\n",
    "            \"Cystic Fibrosis\": [\"cystic fibrosis\", \"CF\", \"cystic fibrosis disorder\", \"cystic fibrosis disease\"],\n",
    "            \"Sickle Cell Anemia\": [\"sickle cell anemia\", \"sickle cell disease\", \"sickle cell disorder\", \"SCD\", \"sickle cell\"],\n",
    "            \"Tuberculous Meningitis\": [\"tuberculous meningitis\", \"TB meningitis\", \"meningitis tuberculosa\"],\n",
    "            \"Chronic Obstructive Pulmonary Disease\": [\"COPD\", \"chronic obstructive pulmonary disease\", \"emphysema\", \"chronic bronchitis\"],\n",
    "            \"Multiple Sclerosis\": [\"multiple sclerosis\", \"MS\", \"sclerosis multiple\", \"multiple sclerosis disease\"],\n",
    "            \"Epilepsy\": [\"epilepsy\", \"seizure disorder\", \"convulsive disorder\", \"epileptic seizure\", \"epilepsy\"],\n",
    "            \"Leukemia\": [\"leukemia\", \"blood cancer\", \"leukemia\", \"leukemic\", \"leukemia cancer\"],\n",
    "            \"Lung Cancer\": [\"lung cancer\", \"pulmonary carcinoma\", \"lung carcinoma\", \"lung neoplasm\", \"non-small cell lung cancer\"],\n",
    "            \"Renal Failure\": [\"renal failure\", \"kidney failure\", \"chronic kidney disease\", \"CKD\", \"renal dysfunction\"],\n",
    "            \"Peptic Ulcer Disease\": [\"peptic ulcer\", \"gastric ulcer\", \"stomach ulcer\", \"peptic ulcer disease\", \"gastric ulcer disease\"],\n",
    "            \"Anemia\": [\"anemia\", \"iron deficiency anemia\", \"aplastic anemia\", \"sickle cell anemia\", \"hemolytic anemia\"],\n",
    "            \"Psoriasis\": [\"psoriasis\", \"psoriatic skin disease\", \"skin psoriasis\", \"psoriatic arthritis\"],\n",
    "        }\n",
    "\n",
    "\n",
    "        # Define fallback phrases for semantic similarity checking\n",
    "        self.fallback_phrases = [\n",
    "            \"I don't know the answer\",\n",
    "            \"I'm not sure\",\n",
    "            \"I don't have enough information\",\n",
    "            \"I'm sorry, I can't help with that\"\n",
    "        ]\n",
    "        # Convert fallback phrases to spaCy Doc objects for similarity comparison\n",
    "        self.fallback_docs = [self.nlp(phrase) for phrase in self.fallback_phrases]\n",
    "\n",
    "    # def format_context1(self, context):\n",
    "    #     if isinstance(context, list):\n",
    "    #         formatted_context = [\n",
    "    #             item['p.name'] for item in context if isinstance(item, dict) and 'p.name' in item and item['p.name'] is not None\n",
    "    #         ]\n",
    "    #         if formatted_context:\n",
    "    #             return \", \".join(formatted_context)\n",
    "    #         else:\n",
    "    #             return \"\"\n",
    "    #     return \"No additional context available.\"\n",
    "\n",
    "    def format_context(self, context, user_query):\n",
    "        instructions = (\n",
    "            \"You are extracting the name of diseases, drugs, genes or proteins, exposures, molecular function, \"\n",
    "            \"pathways, phenotypes, cellular components, biological processes entities from the text. \"\n",
    "            \"You are extracting multi-word entities (like the name of pathways or gene names) from the text. \"\n",
    "            \"Ensure to capture complete and meaningful phrases, and avoid partial entity identification. \"\n",
    "            \"Do semantic searching to understand context and relationships.\"\n",
    "        )\n",
    "        formatted_context_query = f\"{instructions} {user_query} {context}\"\n",
    "        response = self.chain.invoke({\"query\": formatted_context_query})\n",
    "        return response.get(\"result\", \"No relevant context available.\")\n",
    "\n",
    "\n",
    "    def format_context2(self, context):\n",
    "        print(\"context:\", context)\n",
    "\n",
    "        # Define the instruction to make the model extract important terms\n",
    "        formatted_context_query = f\"List out the important terms: {context}\"\n",
    "\n",
    "        for attempt in range(3):  # Retry up to 3 times if fallback is detected\n",
    "            print(f\"Attempt {attempt + 1} for format_context2...\")\n",
    "\n",
    "            try:\n",
    "                # Pass the formatted context to the chain for processing\n",
    "                response = self.chain.invoke({\"query\": formatted_context_query})\n",
    "\n",
    "                # Extract and clean the result text\n",
    "                result_text = response.get(\"result\", \"\").strip()\n",
    "\n",
    "                # Check if the result is semantically close to a fallback message\n",
    "                if not result_text or self.is_fallback_message(result_text):\n",
    "                    print(f\"Response deemed fallback on attempt {attempt + 1}: '{result_text}'\")\n",
    "                    continue  # Retry if it's a fallback message\n",
    "\n",
    "                # If a valid response is found, return it\n",
    "                return result_text\n",
    "\n",
    "            except Exception as e:\n",
    "                print(f\"Attempt {attempt + 1} failed with error: {e}\")\n",
    "\n",
    "        # If all attempts fail, return a default message\n",
    "        return \"No relevant context available after three attempts.\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    def nlp_process(self, user_query):\n",
    "        \"\"\"Apply NLP to standardize disease terms in the query.\"\"\"\n",
    "        doc = self.nlp(user_query)\n",
    "\n",
    "        # Replace terms with the closest standardized disease names\n",
    "        processed_tokens = [\n",
    "            self.find_closest_standard_term(token.text) for token in doc\n",
    "        ]\n",
    "\n",
    "        return \" \".join(processed_tokens)\n",
    "\n",
    "    def find_closest_standard_term(self, token_text):\n",
    "        \"\"\"Find the closest matching disease name using fuzzy matching.\"\"\"\n",
    "        token_text = token_text\n",
    "\n",
    "        # Check all disease names and their variations\n",
    "        for standard_name, variations in self.standard_diseases.items():\n",
    "            variations = [standard_name] + variations\n",
    "\n",
    "            # Use fuzzy matching to find the closest match\n",
    "            result = process.extractOne(token_text, variations, scorer=fuzz.partial_ratio)\n",
    "\n",
    "            if result:\n",
    "                match, score, *_ = result  # Unpack additional values safely\n",
    "                if score >= 80:  # Threshold for accepting a close match\n",
    "                    return standard_name  # Return the standardized name\n",
    "\n",
    "        return token_text  # Return the original token if no close match is found\n",
    "\n",
    "    def is_fallback_message(self, result_text):\n",
    "\n",
    "        \"\"\"Check if the result text is a fallback message by comparing semantic similarity.\"\"\"\n",
    "        result_doc = self.nlp(result_text)\n",
    "\n",
    "        # Compare result_doc with each fallback_doc, return True if any similarity is above the threshold\n",
    "        for fallback_doc in self.fallback_docs:\n",
    "            if result_doc.similarity(fallback_doc) >= 0.75:  # Threshold can be adjusted\n",
    "                return True\n",
    "        return False\n",
    "\n",
    "\n",
    "\n",
    "    def ask(self, user_query):\n",
    "        enhanced_query = self.nlp_process(user_query)\n",
    "        #full_query = f\"{self.previous_context}\\n{enhanced_query}\"\n",
    "        full_query = f\"{enhanced_query}\"\n",
    "\n",
    "        for attempt in range(3):  # Retry up to 3 times if context is empty\n",
    "            print(f\"Attempt {attempt + 1}...\")\n",
    "            try:\n",
    "                # Pass the enhanced query to the chain for processing\n",
    "                response = self.chain.invoke({\"query\": user_query})\n",
    "\n",
    "                result_text = response[\"result\"].strip()\n",
    "\n",
    "                # If the result is a fallback message, process the intermediate context\n",
    "                if self.is_fallback_message(result_text):\n",
    "                    intermediate_steps = response.get(\"intermediate_steps\", [])\n",
    "                    context_data_list = [step[\"context\"] for step in intermediate_steps if \"context\" in step]\n",
    "                    print(\"context_data_list\",context_data_list)\n",
    "\n",
    "                    if context_data_list:\n",
    "                        # Pass context to the LLM for natural language formatting\n",
    "                        all_contexts = [self.format_context2(data) for data in context_data_list]\n",
    "                        combined_readable_context = \"\\n\".join(all_contexts)\n",
    "                        print(\"combined_readable_context\",combined_readable_context)\n",
    "                        response[\"result\"] = f\"\\n{combined_readable_context}\"\n",
    "\n",
    "                # If the result is non-empty, append it to the previous context\n",
    "                if response[\"result\"].strip():\n",
    "                    self.previous_context += f\"\\n{response['result']}\"\n",
    "                    response['result']=self.format_context2(response['result'])\n",
    "                    return response\n",
    "\n",
    "\n",
    "\n",
    "            except Exception as e:\n",
    "                print(f\"Attempt {attempt + 1} failed with error: {e}\")\n",
    "\n",
    "        # If all attempts fail or no context is found, return a fallback response\n",
    "        return {\"result\": \"I don't know the answer\", \"intermediate_steps\": []}\n",
    "\n",
    "    def chat(self):\n",
    "        \"\"\"Start the chatbot conversation.\"\"\"\n",
    "        print(\"Welcome to the Medical Knowledge Chatbot! \")\n",
    "        while True:\n",
    "            user_input = input(\"You: \")\n",
    "            print(\"You: \",user_input)\n",
    "            if user_input.lower() == 'exit':\n",
    "                print(\"Goodbye!\")\n",
    "                break\n",
    "\n",
    "            response = self.ask(user_input)\n",
    "            print(\"CHATBOT:\" ,self.format_context2(response['result']))\n",
    "            #print(f\"\\nChatbot: {response['result']}\")\n",
    "\n",
    "            # intermediate_steps = response.get(\"intermediate_steps\", [])\n",
    "            # context_data_list = [\n",
    "            #     step[\"context\"] for step in intermediate_steps if \"context\" in step\n",
    "            # ]\n",
    "\n",
    "            # if context_data_list:\n",
    "            #     print(\"\\n[Context Extracted]:\")\n",
    "            #     for context_data in context_data_list:\n",
    "            #         print(self.format_context(context_data))\n",
    "            # else:\n",
    "            #     print(\"[No Relevant Context Available]\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4462484a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to the Medical Knowledge Chatbot! \n",
      "You:  give me the Drugs and associated Proteins\n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug)-[:DRUG_TARGET|DRUG_ENZYME|DRUG_CARRIER|DRUG_TRANSPORTER]->(p:Protein) \n",
      "RETURN d, p\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['12626323', ' 24370975', ' 659632'], 'DRUG_ATC_CODE': ['ATC:H04AA', ' ATC:H04AA01', ' ATC:H04A', ' ATC:H', ' ATC:H04'], 'PRODUCT': ['Baqsimi', ' Glucag0n', ' GlucaGen', ' GlucaGen_HypoKit', ' Glucagon', ' Glucagon_Ampoule_666_1mg', ' Gvoke_HypoPen_0_5_mg_Auto_Injector', ' Gvoke_HypoPen_1_mg_Auto_Injector', ' Gvoke_PFS_0_5_mg_Pre_filled_Syringe', ' Gvoke_PFS_1_mg_Pre_filled_Syringe'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00040', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Glucagon', 'SYNONYM': ['Glucagon', ' Glucagon__recombinant_dna_origin_', ' Glucagon_recombinant', ' Glucagon__human', ' Glucagon__porcine', ' Glucagone'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR017981', ' IPR001879'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0008528', ' GO:0004930', ' GO:0004967', ' GO:0017046'], 'SHORT_NAME': ['GLP-2 receptor', ' GLP-2-R', ' GLP-2R'], 'NAME': 'GLP2R', 'FAMILY': ['IPR039125', ' IPR000832'], 'RELATED_KEYWORD': ['Cell membrane', ' Disulfide bond', ' G-protein coupled receptor', ' Glycoprotein', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Transducer', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Skin_2__Epidermal_cells', ' Oral_mucosa__Squamous_epithelial_cells', ' Cervix', '_uterine', ' Skin_1__Keratinocytes', ' Esophagus', ' Gallbladder__Glandular_cells', ' Skin_2', ' Skin_1', ' Vagina__Squamous_epithelial_cells', ' Adrenal_gland', ' Skin_1__Langerhans', ' Gallbladder', ' Vagina', ' Adrenal_gland__Glandular_cells', ' Cervix', '_uterine__Squamous_epithelial_cells', ' Esophagus__Squamous_epithelial_cells', ' Oral_mucosa', ' Duodenum__Glandular_cells', ' Small_intestine', ' Rectum', ' Appendix__Glandular_cells', ' Colon__Glandular_cells', ' Rectum__Glandular_cells', ' Duodenum', ' Small_intestine__Glandular_cells', ' Appendix', ' Colon'], 'RELATED_PUBMED_ID': ['pubmed:9990065', ' pubmed:15489334'], 'FULL_NAME': 'Glucagon-like peptide 2 receptor', 'CONSERVED_SITE': 'IPR017983', 'GO_CC': ['GO:0016021', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036445', 'id': 'O95838', 'GO_MF': ['GO:0007188', ' GO:0007166', ' GO:0007186', ' GO:0008284'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['12626323', ' 24370975', ' 659632'], 'DRUG_ATC_CODE': ['ATC:H04AA', ' ATC:H04AA01', ' ATC:H04A', ' ATC:H', ' ATC:H04'], 'PRODUCT': ['Baqsimi', ' Glucag0n', ' GlucaGen', ' GlucaGen_HypoKit', ' Glucagon', ' Glucagon_Ampoule_666_1mg', ' Gvoke_HypoPen_0_5_mg_Auto_Injector', ' Gvoke_HypoPen_1_mg_Auto_Injector', ' Gvoke_PFS_0_5_mg_Pre_filled_Syringe', ' Gvoke_PFS_1_mg_Pre_filled_Syringe'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00040', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Glucagon', 'SYNONYM': ['Glucagon', ' Glucagon__recombinant_dna_origin_', ' Glucagon_recombinant', ' Glucagon__human', ' Glucagon__porcine', ' Glucagone'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR017981', ' IPR001879'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0008528', ' GO:0004967', ' GO:0044508', ' GO:0017046', ' GO:0004888'], 'SHORT_NAME': ['GLP-1 receptor', ' GLP-1-R', ' GLP-1R'], 'NAME': 'GLP1R', 'FAMILY': ['IPR003290', ' IPR003292', ' IPR000832'], 'RELATED_KEYWORD': ['3D-structure', ' ADP-ribosylation', ' Cell membrane', ' Disulfide bond', ' G-protein coupled receptor', ' Glycoprotein', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Signal', ' Transducer', ' Transmembrane', ' Transmembrane helix'], 'OTHER_ID': 'Q99669', 'RELATED_PUBMED_ID': ['pubmed:8405712', ' pubmed:8404634', ' pubmed:8216285', ' pubmed:7517895', ' pubmed:7843404', ' pubmed:14574404', ' pubmed:15489334', ' pubmed:9213353', ' pubmed:20869417', ' pubmed:21901419', ' pubmed:22412906', ' pubmed:18287102', ' pubmed:19861722', ' pubmed:26308095', ' pubmed:27196125', ' pubmed:28514449'], 'FULL_NAME': 'Glucagon-like peptide 1 receptor', 'CONSERVED_SITE': 'IPR017983', 'GO_CC': ['GO:0005623', ' GO:0016021', ' GO:0005887', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036445', 'id': 'P43220', 'GO_MF': ['GO:0007190', ' GO:0007189', ' GO:0019933', ' GO:0007166', ' GO:0007186', ' GO:0007611', ' GO:0045777', ' GO:0007204', ' GO:0008016', ' GO:0050796', ' GO:1990911'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['12626323', ' 24370975', ' 659632'], 'DRUG_ATC_CODE': ['ATC:H04AA', ' ATC:H04AA01', ' ATC:H04A', ' ATC:H', ' ATC:H04'], 'PRODUCT': ['Baqsimi', ' Glucag0n', ' GlucaGen', ' GlucaGen_HypoKit', ' Glucagon', ' Glucagon_Ampoule_666_1mg', ' Gvoke_HypoPen_0_5_mg_Auto_Injector', ' Gvoke_HypoPen_1_mg_Auto_Injector', ' Gvoke_PFS_0_5_mg_Pre_filled_Syringe', ' Gvoke_PFS_1_mg_Pre_filled_Syringe'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00040', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Glucagon', 'SYNONYM': ['Glucagon', ' Glucagon__recombinant_dna_origin_', ' Glucagon_recombinant', ' Glucagon__human', ' Glucagon__porcine', ' Glucagone'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR017981', ' IPR001879'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0008528', ' GO:0004967', ' GO:0005085', ' GO:0017046'], 'SHORT_NAME': 'GL-R', 'NAME': 'GLR', 'FAMILY': ['IPR003290', ' IPR003291', ' IPR000832'], 'RELATED_KEYWORD': ['3D-structure', ' Cell membrane', ' Diabetes mellitus', ' Disulfide bond', ' G-protein coupled receptor', ' Glycoprotein', ' Membrane', ' Phosphoprotein', ' Polymorphism', ' Receptor', ' Reference proteome', ' Signal', ' Transducer', ' Transmembrane', ' Transmembrane helix'], 'RELATED_PUBMED_ID': ['pubmed:7507321', ' pubmed:8144028', ' pubmed:15489334', ' pubmed:8020989', ' pubmed:9287038', ' pubmed:24275569', ' pubmed:16221670', ' pubmed:22908259', ' pubmed:23863937', ' pubmed:27111510', ' pubmed:28514451', ' pubmed:7589886', ' pubmed:19657311'], 'FULL_NAME': 'Glucagon receptor', 'CONSERVED_SITE': 'IPR017983', 'GO_CC': ['GO:0005768', ' GO:0016021', ' GO:0005887', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036445', 'id': 'P47871', 'GO_MF': ['GO:0007189', ' GO:0007188', ' GO:0007166', ' GO:0071377', ' GO:0009267', ' GO:0006887', ' GO:0007186', ' GO:0006091', ' GO:0042593', ' GO:0009755', ' GO:0010628', ' GO:0008217', ' GO:0070873', ' GO:0007584', ' GO:0042594'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['21456630', ' 12010086', ' 15078138', ' 25998211', ' 27789332'], 'DRUG_ATC_CODE': ['ATC:L', ' ATC:L03AA', ' ATC:L03', ' ATC:L03A', ' ATC:L03AA13'], 'PRODUCT': ['Fulphila', ' Lapelga', ' Neulasta', ' Neupopeg', ' Udenyca', ' Ziextenzo'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00019', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Pegfilgrastim', 'SYNONYM': ['Granulocyte_colony_stimulating_factor_pegfilgrastim', ' peg_filgrastim', ' pegfilgrastim_apgf', ' pegfilgrastim_bmez', ' pegfilgrastim_cbqv', ' pegfilgrastim_jmdb'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR003961', ' IPR010457'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019955', ' GO:0004896', ' GO:0051916', ' GO:0038023'], 'SHORT_NAME': ['G-CSF receptor', ' G-CSF-R'], 'NAME': 'CSF3R', 'RELATED_KEYWORD': ['3D-structure', ' Alternative splicing', ' Cell adhesion', ' Cell membrane', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Immunoglobulin domain', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Repeat', ' Secreted', ' Signal', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Bone_marrow', ' Lung__Macrophages', ' Lung', ' Bone_marrow__Hematopoietic_cells', ' Skin_1__Fibroblasts', ' Skin_1__Melanocytes', ' Skin_1', ' Heart_muscle', ' Testis', ' Testis__Leydig_cells', ' Cerebral_cortex', ' Cerebral_cortex__Neuronal_cells', ' Heart_muscle__Myocytes', ' Placenta', ' Placenta__Trophoblastic_cells'], 'RELATED_PUBMED_ID': ['pubmed:2147944', ' pubmed:1701053', ' pubmed:1530796', ' pubmed:15489334', ' pubmed:8554326', ' pubmed:1717255', ' pubmed:7514305', ' pubmed:17127321', ' pubmed:19120359', ' pubmed:9187659', ' pubmed:9368043', ' pubmed:16492764', ' pubmed:10449521', ' pubmed:19620628', ' pubmed:24753537', ' pubmed:26324699'], 'FULL_NAME': 'Granulocyte colony-stimulating factor receptor', 'CONSERVED_SITE': 'IPR003529', 'GO_CC': ['GO:0009897', ' GO:0005576', ' GO:0005887', ' GO:0005886', ' GO:0043235'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036116', ' IPR036179', ' IPR013783'], 'id': 'Q99062', 'GO_MF': ['GO:0097186', ' GO:0007155', ' GO:0019221', ' GO:0006952', ' GO:0030593', ' GO:0045637', ' GO:0007165'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['21456630', ' 12010086', ' 15078138', ' 25998211', ' 27789332'], 'DRUG_ATC_CODE': ['ATC:L', ' ATC:L03AA', ' ATC:L03', ' ATC:L03A', ' ATC:L03AA13'], 'PRODUCT': ['Fulphila', ' Lapelga', ' Neulasta', ' Neupopeg', ' Udenyca', ' Ziextenzo'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00019', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Pegfilgrastim', 'SYNONYM': ['Granulocyte_colony_stimulating_factor_pegfilgrastim', ' peg_filgrastim', ' pegfilgrastim_apgf', ' pegfilgrastim_bmez', ' pegfilgrastim_cbqv', ' pegfilgrastim_jmdb'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': 'IPR001254', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019955', ' GO:0004175', ' GO:0008201', ' GO:0008233', ' GO:0002020', ' GO:0004252', ' GO:0003714'], 'ACTIVE_SITE': ['IPR018114', ' IPR033116'], 'SHORT_NAME': 'HLE', 'NAME': 'ELNE', 'FAMILY': 'IPR001314', 'RELATED_KEYWORD': ['3D-structure', ' Cytoplasmic vesicle', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Hydrolase', ' Polymorphism', ' Protease', ' Reference proteome', ' Serine protease', ' Signal', ' Zymogen'], 'PROTEIN_EXPRESSED_IN': ['Spleen__Cells_in_red_pulp', ' Spleen', ' Bone_marrow__Hematopoietic_cells', ' Bone_marrow'], 'OTHER_ID': ['Q6B0D9', ' Q6LDP5'], 'RELATED_PUBMED_ID': ['pubmed:3479752', ' pubmed:2902087', ' pubmed:2775493', ' pubmed:2322278', ' pubmed:15489334', ' pubmed:2822677', ' pubmed:3550808', ' pubmed:2501794', ' pubmed:7897245', ' pubmed:3422232', ' pubmed:2462434', ' pubmed:1859409', ' pubmed:10947984', ' pubmed:15140022', ' pubmed:14673143', ' pubmed:19159218', ' pubmed:21269460', ' pubmed:25944712', ' pubmed:2911584', ' pubmed:3391280', ' pubmed:3640709', ' pubmed:10581030', ' pubmed:11001877', ' pubmed:11675333', ' pubmed:12091371', ' pubmed:14962902', ' pubmed:17436313', ' pubmed:18946670', ' pubmed:19036076', ' pubmed:19415009', ' pubmed:20220065', ' pubmed:19927291', ' pubmed:20803142', ' pubmed:21425445', ' pubmed:23463630'], 'FULL_NAME': ['Neutrophil elastase', ' Bone marrow serine protease', ' Elastase-2', ' Human leukocyte elastase', ' Medullasin', ' PMN elastase'], 'GO_CC': ['GO:0035578', ' GO:0005623', ' GO:0009986', ' GO:0062023', ' GO:0005737', ' GO:0070062', ' GO:0005576', ' GO:0005615', ' GO:0045335', ' GO:0030141', ' GO:0035580', ' GO:0017053'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR009003', 'id': 'P08246', 'GO_MF': ['GO:0002438', ' GO:0019730', ' GO:0002812', ' GO:0006874', ' GO:0042742', ' GO:0022617', ' GO:0002523', ' GO:0045079', ' GO:0050922', ' GO:0050728', ' GO:0045415', ' GO:0000122', ' GO:0043312', ' GO:0070947', ' GO:0070945', ' GO:0006909', ' GO:0050778', ' GO:0045416', ' GO:1903238', ' GO:0043406', ' GO:0048661', ' GO:0030163', ' GO:0006508', ' GO:0032496', ' GO:0009411', ' GO:0001878'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['17187516', ' 17454642'], 'DRUG_ATC_CODE': ['ATC:L01XX', ' ATC:L01XX29', ' ATC:L', ' ATC:L01X', ' ATC:L01'], 'PRODUCT': 'Ontak', 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00004', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Denileukin_diftitox', 'SYNONYM': ['Denileukin', ' Interleukin_2_diptheria_toxin_fusion_protein'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': 'IPR000436', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019976', ' GO:0004911'], 'SHORT_NAME': ['IL-2 receptor subunit alpha', ' IL-2-RA', ' IL-2R subunit alpha', ' IL2-RA'], 'NAME': 'IL2RA', 'FAMILY': 'IPR015486', 'RELATED_KEYWORD': ['3D-structure', ' Diabetes mellitus', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Immunity', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Repeat', ' Signal', ' Sushi', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Appendix', ' Bone_marrow', ' Appendix__Lymphoid_tissue', ' Bone_marrow__Hematopoietic_cells', ' Lymph_node', ' Lymph_node__Germinal_center_cells', ' Tonsil__Germinal_center_cells', ' Spleen', ' Spleen__Cells_in_red_pulp', ' Tonsil__Non-germinal_center_cells', ' Lymph_node__Non-germinal_center_cells', ' Tonsil'], 'RELATED_PUBMED_ID': ['pubmed:6090949', ' pubmed:6090948', ' pubmed:2999698', ' pubmed:2996141', ' pubmed:15164054', ' pubmed:3030566', ' pubmed:3134887', ' pubmed:7529123', ' pubmed:16293754', ' pubmed:16477002', ' pubmed:17676041', ' pubmed:23416241', ' pubmed:24116927'], 'FULL_NAME': ['Interleukin-2 receptor subunit alpha', ' TAC antigen', ' p55'], 'GO_CC': ['GO:0009897', ' GO:0005893', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR035976', 'id': 'P01589', 'GO_MF': ['GO:0006924', ' GO:0006915', ' GO:0007166', ' GO:0019221', ' GO:0006955', ' GO:0006954', ' GO:0002437', ' GO:0038110', ' GO:0000165', ' GO:0050687', ' GO:0050777', ' GO:0050728', ' GO:0042130', ' GO:0007219', ' GO:0042104', ' GO:0045582', ' GO:0045589', ' GO:0046013', ' GO:0002664'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['17187516', ' 17454642'], 'DRUG_ATC_CODE': ['ATC:L01XX', ' ATC:L01XX29', ' ATC:L', ' ATC:L01X', ' ATC:L01'], 'PRODUCT': 'Ontak', 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00004', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Denileukin_diftitox', 'SYNONYM': ['Denileukin', ' Interleukin_2_diptheria_toxin_fusion_protein'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR003961', ' IPR040951'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0042010', ' GO:0019976', ' GO:0004911'], 'SHORT_NAME': ['IL-2 receptor subunit beta', ' IL-2R subunit beta', ' IL-2RB', ' p75'], 'NAME': 'IL2RB', 'RELATED_KEYWORD': ['3D-structure', ' Cell membrane', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Host-virus interaction', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Signal', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Spleen', ' Lymph_node__Non-germinal_center_cells', ' Tonsil__Non-germinal_center_cells', ' Spleen__Cells_in_white_pulp', ' Tonsil', ' Spleen__Cells_in_red_pulp', ' Bone_marrow', ' Lymph_node', ' Bone_marrow__Hematopoietic_cells'], 'RELATED_PUBMED_ID': ['pubmed:2785715', ' pubmed:15461802', ' pubmed:14702039', ' pubmed:10591208', ' pubmed:15489334', ' pubmed:8648694', ' pubmed:12200137', ' pubmed:15123770', ' pubmed:7529123', ' pubmed:16293754', ' pubmed:16477002', ' pubmed:31040184', ' pubmed:31040185'], 'FULL_NAME': ['Interleukin-2 receptor subunit beta', ' High affinity IL-2 receptor subunit beta', ' Interleukin-15 receptor subunit beta {ECO:0000312|HGNC:HGNC:6009}', ' p70-75'], 'CONSERVED_SITE': 'IPR003531', 'GO_CC': ['GO:0009986', ' GO:0005829', ' GO:0005768', ' GO:0009897', ' GO:0005887', ' GO:0005893', ' GO:0016020', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036116', ' IPR013783'], 'id': 'P14784', 'GO_MF': ['GO:0019221', ' GO:0035723', ' GO:0038110', ' GO:0000165', ' GO:0043066', ' GO:0050766', ' GO:0065003', ' GO:0007165', ' GO:0016032'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['17187516', ' 17454642'], 'DRUG_ATC_CODE': ['ATC:L01XX', ' ATC:L01XX29', ' ATC:L', ' ATC:L01X', ' ATC:L01'], 'PRODUCT': 'Ontak', 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00004', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Denileukin_diftitox', 'SYNONYM': ['Denileukin', ' Interleukin_2_diptheria_toxin_fusion_protein'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR003961', ' IPR015321'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019955', ' GO:0004896', ' GO:0042010', ' GO:0019976'], 'SHORT_NAME': ['IL-2 receptor subunit gamma', ' IL-2R subunit gamma', ' IL-2RG'], 'NAME': 'IL2RG', 'RELATED_KEYWORD': ['3D-structure', ' Alternative splicing', ' Cell membrane', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Host-virus interaction', ' Membrane', ' Phosphoprotein', ' Polymorphism', ' Receptor', ' Reference proteome', ' SCID', ' Signal', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Bone_marrow__Hematopoietic_cells', ' Tonsil', ' Bone_marrow', ' Tonsil__Squamous_epithelial_cells', ' Spleen__Cells_in_red_pulp', ' Spleen'], 'RELATED_PUBMED_ID': ['pubmed:1631559', ' pubmed:8514792', ' pubmed:8401490', ' pubmed:15772651', ' pubmed:15489334', ' pubmed:8266076', ' pubmed:8266078', ' pubmed:8266077', ' pubmed:8648694', ' pubmed:12200137', ' pubmed:15123770', ' pubmed:19690332', ' pubmed:7529123', ' pubmed:16293754', ' pubmed:16477002', ' pubmed:8299698', ' pubmed:8088810', ' pubmed:8027558', ' pubmed:7937790', ' pubmed:7668284', ' pubmed:7557965', ' pubmed:7860773', ' pubmed:7883965', ' pubmed:8900089', ' pubmed:9150740', ' pubmed:9049783', ' pubmed:9399950'], 'FULL_NAME': ['Cytokine receptor common subunit gamma', ' Interleukin-2 receptor subunit gamma', ' gammaC', ' p64'], 'CONSERVED_SITE': 'IPR003531', 'GO_CC': ['GO:0009986', ' GO:0005829', ' GO:0005768', ' GO:0009897', ' GO:0005887', ' GO:0016020', ' GO:0005886', ' GO:0043235'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036116', ' IPR013783'], 'id': 'P31785', 'GO_MF': ['GO:0019221', ' GO:0006955', ' GO:0035723', ' GO:0038110', ' GO:0038114', ' GO:0035771', ' GO:0038111', ' GO:0038113', ' GO:0000165', ' GO:0050766', ' GO:0007165', ' GO:0016032'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['10480573', ' 10601294', ' 10628369', ' 11408594', ' 11431346', ' 11752352', ' 15269308', ' 16336752'], 'DRUG_ATC_CODE': ['ATC:L01', ' ATC:L', ' ATC:L01XC06', ' ATC:L01XC', ' ATC:L01X'], 'PRODUCT': 'Erbitux', 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00002', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Cetuximab', 'SYNONYM': nan, 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR007110', ' IPR003599'], 'SPECIES': 'HUMAN', 'GO_BP': 'GO:0019864', 'SHORT_NAME': ['Fc-gamma RIII', ' Fc-gamma RIIIb', ' FcRIII', ' FcRIIIb'], 'NAME': 'FCG3B', 'RELATED_KEYWORD': ['3D-structure', ' Cell membrane', ' Disulfide bond', ' Glycoprotein', ' GPI-anchor', ' IgG-binding protein', ' Immunoglobulin domain', ' Lipoprotein', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Repeat', ' Secreted', ' Signal'], 'RELATED_PUBMED_ID': ['pubmed:2526846', ' pubmed:2967436', ' pubmed:2521732', ' pubmed:2525780', ' pubmed:15245367', ' pubmed:16710414', ' pubmed:7836402', ' pubmed:10917521', ' pubmed:11021536', ' pubmed:9028335'], 'FULL_NAME': ['Low affinity immunoglobulin gamma Fc region receptor III-B', ' Fc-gamma RIII-beta', ' FcR-10', ' IgG Fc receptor III-1'], 'GO_CC': ['GO:0031225', ' GO:0070062', ' GO:0005576', ' GO:0005886', ' GO:0030667'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036179', ' IPR013783'], 'id': 'O75015', 'GO_MF': ['GO:0006955', ' GO:0043312'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['10480573', ' 10601294', ' 10628369', ' 11408594', ' 11431346', ' 11752352', ' 15269308', ' 16336752'], 'DRUG_ATC_CODE': ['ATC:L01', ' ATC:L', ' ATC:L01XC06', ' ATC:L01XC', ' ATC:L01X'], 'PRODUCT': 'Erbitux', 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00002', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Cetuximab', 'SYNONYM': nan, 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR006211', ' IPR032778', ' IPR000719', ' IPR000494', ' IPR001245', ' IPR020635'], 'REPEAT': 'IPR006212', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0051015', ' GO:0005524', ' GO:0051117', ' GO:0045296', ' GO:0005516', ' GO:0003682', ' GO:0003690', ' GO:0019899', ' GO:0048408', ' GO:0005006', ' GO:0042802', ' GO:0005178', ' GO:0004709', ' GO:0019901', ' GO:0019903', ' GO:0004713', ' GO:0004714', ' GO:0004888', ' GO:0031625', ' GO:0001618'], 'ACTIVE_SITE': 'IPR008266', 'NAME': 'EGFR', 'FAMILY': 'IPR016245', 'RELATED_KEYWORD': ['3D-structure', ' Alternative splicing', ' ATP-binding', ' Cell membrane', ' Developmental protein', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Endoplasmic reticulum', ' Endosome', ' Glycoprotein', ' Golgi apparatus', ' Host cell receptor for virus entry', ' Host-virus interaction', ' Isopeptide bond', ' Kinase', ' Lipoprotein', ' Membrane', ' Methylation', ' Nucleotide-binding', ' Nucleus', ' Palmitate', ' Phosphoprotein', ' Polymorphism', ' Proto-oncogene', ' Receptor', ' Reference proteome', ' Repeat', ' Secreted', ' Signal', ' Transferase', ' Transmembrane', ' Transmembrane helix', ' Tyrosine-protein kinase', ' Ubl conjugation'], 'PROTEIN_EXPRESSED_IN': ['Skin_1__Melanocytes', ' Tonsil', ' Vagina__Squamous_epithelial_cells', ' Skeletal_muscle', ' Urinary_bladder__Urothelial_cells', ' Urinary_bladder', ' Skin_1', ' Skeletal_muscle__Myocytes', ' Testis__Leydig_cells', ' Tonsil__Squamous_epithelial_cells', ' Liver__Hepatocytes', ' Vagina', ' Liver', ' Testis', ' Placenta__Trophoblastic_cells', ' Placenta'], 'OTHER_ID': ['O00732', ' P06268', ' Q14225', ' Q68GS5', ' Q92795', ' Q9BZS2', ' Q9GZX1', ' Q9H2C9', ' Q9H3C9', ' Q9UMD7', ' Q9UMD8', ' Q9UMG5'], 'RELATED_PUBMED_ID': ['pubmed:6328312', ' pubmed:7654368', ' pubmed:8918811', ' pubmed:9103388', ' pubmed:11161793', ' pubmed:6326261', ' pubmed:6330563', ' pubmed:6093780', ' pubmed:3329716', ' pubmed:1988448', ' pubmed:2991899', ' pubmed:6324343', ' pubmed:3138233', ' pubmed:15340161', ' pubmed:2985580', ' pubmed:16543144', ' pubmed:9556602', ' pubmed:6325948', ' pubmed:3039909', ' pubmed:2790960', ' pubmed:2543678', ' pubmed:7679104', ' pubmed:8650580', ' pubmed:9852145', ' pubmed:10523301', ' pubmed:8144591', ' pubmed:7657591', ' pubmed:8962717', ' pubmed:9419975', ' pubmed:9488479', ' pubmed:10228163', ' pubmed:10026169', ' pubmed:10731668', ' pubmed:11602604', ' pubmed:12731890', ' pubmed:12873986', ' pubmed:15282549', ' pubmed:10805725', ' pubmed:11116146', ' pubmed:11483589', ' pubmed:15590694', ' pubmed:15611079', ' pubmed:16083266', ' pubmed:16140940', ' pubmed:17081983', ' pubmed:17115032', ' pubmed:16420529', ' pubmed:17671655', ' pubmed:17182860', ' pubmed:18602463', ' pubmed:19172738', ' pubmed:18220336', ' pubmed:18691976', ' pubmed:18669648', ' pubmed:19413330', ' pubmed:19836242', ' pubmed:19509291', ' pubmed:19159218', ' pubmed:19369195', ' pubmed:19749156', ' pubmed:19718021', ' pubmed:20674546', ' pubmed:20551055', ' pubmed:20462955', ' pubmed:20068231', ' pubmed:21269460', ' pubmed:21487020', ' pubmed:21258366', ' pubmed:21516087', ' pubmed:21518868', ' pubmed:22298428', ' pubmed:22179831', ' pubmed:23418353', ' pubmed:23186163', ' pubmed:24275569', ' pubmed:25187647', ' pubmed:23589287', ' pubmed:25311788', ' pubmed:23912460', ' pubmed:27153536', ' pubmed:26988023', ' pubmed:28479384', ' pubmed:12297049', ' pubmed:12297050', ' pubmed:12620237', ' pubmed:15374980', ' pubmed:15837620', ' pubmed:15840573', ' pubmed:17349580', ' pubmed:18046415', ' pubmed:18227510', ' pubmed:19563760', ' pubmed:19560417', ' pubmed:20471394', ' pubmed:20837704', ' pubmed:22888118', ' pubmed:15623594', ' pubmed:15118125', ' pubmed:16533793', ' pubmed:16205628', ' pubmed:16672372', ' pubmed:17344846', ' pubmed:24691054'], 'BINDING_SITE': 'IPR017441', 'FULL_NAME': ['Epidermal growth factor receptor', ' Proto-oncogene c-ErbB-1', ' Receptor tyrosine-protein kinase erbB-1'], 'GO_CC': ['GO:0016324', ' GO:0009925', ' GO:0016323', ' GO:0030054', ' GO:0009986', ' GO:0030665', ' GO:0005737', ' GO:0031901', ' GO:0030139', ' GO:0005789', ' GO:0005768', ' GO:0010008', ' GO:0005615', ' GO:0005925', ' GO:0000139', ' GO:0005887', ' GO:0016020', ' GO:0045121', ' GO:0097489', ' GO:0031965', ' GO:0005634', ' GO:0048471', ' GO:0005886', ' GO:0032991', ' GO:0043235', ' GO:0070435', ' GO:0045202'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR009030', ' IPR011009', ' IPR036941'], 'id': 'P00533', 'GO_MF': ['GO:0043006', ' GO:0007202', ' GO:0048143', ' GO:0030154', ' GO:0007166', ' GO:0098609', ' GO:0071230', ' GO:0071276', ' GO:0071549', ' GO:0071364', ' GO:0071392', ' GO:0071260', ' GO:0034614', ' GO:0021795', ' GO:0007623', ' GO:0048546', ' GO:0016101', ' GO:0001892', ' GO:0007173', ' GO:0038128', ' GO:0061029', ' GO:0001942', ' GO:0042743', ' GO:0007611', ' GO:0097421', ' GO:0030324', ' GO:0010960', ' GO:0000165', ' GO:0061024', ' GO:0007494', ' GO:0060571', ' GO:0007275', ' GO:0043066', ' GO:1905208', ' GO:0042059', ' GO:1901185', ' GO:0045930', ' GO:0045746', ' GO:0042177', ' GO:0048812', ' GO:0001503', ' GO:0042698', ' GO:0038083', ' GO:0018108', ' GO:0097755', ' GO:0045780', ' GO:0090263', ' GO:0030307', ' GO:0030335', ' GO:0008284', ' GO:0045737', ' GO:0045739', ' GO:0045740', ' GO:0050679', ' GO:0070374', ' GO:0048146', ' GO:1900087', ' GO:0050729', ' GO:0033674', ' GO:0043406', ' GO:1901224', ' GO:0010750', ' GO:0033138', ' GO:0042327', ' GO:1903800', ' GO:1902722', ' GO:0051897', ' GO:1900020', ' GO:1903078', ' GO:0001934', ' GO:0048661', ' GO:0032930', ' GO:0051968', ' GO:0045944', ' GO:0045893', ' GO:0045907', ' GO:0046777', ' GO:0051205', ' GO:2000145', ' GO:0070372', ' GO:0046328', ' GO:0050999', ' GO:0050730', ' GO:0014066', ' GO:0006357', ' GO:0051592', ' GO:0033590', ' GO:0033594', ' GO:0006970', ' GO:0070141', ' GO:0007435', ' GO:0007165', ' GO:0043586', ' GO:0006412', ' GO:0007169', ' GO:0042060'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}]\u001b[0m\n",
      "Attempt 1 failed with error: Error code: 413 - {'error': {'message': 'Request too large for model `llama3-70b-8192` in organization `org_01j9yv18bffqksnp3w7k6gzqk0` on tokens per minute (TPM): Limit 6000, Requested 8215, please reduce your message size and try again. Visit https://console.groq.com/docs/rate-limits for more information.', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}\n",
      "Attempt 2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug)-[:DRUG_TARGET|DRUG_ENZYME|DRUG_CARRIER|DRUG_TRANSPORTER]->(p:Protein) RETURN d, p\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['12626323', ' 24370975', ' 659632'], 'DRUG_ATC_CODE': ['ATC:H04AA', ' ATC:H04AA01', ' ATC:H04A', ' ATC:H', ' ATC:H04'], 'PRODUCT': ['Baqsimi', ' Glucag0n', ' GlucaGen', ' GlucaGen_HypoKit', ' Glucagon', ' Glucagon_Ampoule_666_1mg', ' Gvoke_HypoPen_0_5_mg_Auto_Injector', ' Gvoke_HypoPen_1_mg_Auto_Injector', ' Gvoke_PFS_0_5_mg_Pre_filled_Syringe', ' Gvoke_PFS_1_mg_Pre_filled_Syringe'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00040', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Glucagon', 'SYNONYM': ['Glucagon', ' Glucagon__recombinant_dna_origin_', ' Glucagon_recombinant', ' Glucagon__human', ' Glucagon__porcine', ' Glucagone'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR017981', ' IPR001879'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0008528', ' GO:0004930', ' GO:0004967', ' GO:0017046'], 'SHORT_NAME': ['GLP-2 receptor', ' GLP-2-R', ' GLP-2R'], 'NAME': 'GLP2R', 'FAMILY': ['IPR039125', ' IPR000832'], 'RELATED_KEYWORD': ['Cell membrane', ' Disulfide bond', ' G-protein coupled receptor', ' Glycoprotein', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Transducer', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Skin_2__Epidermal_cells', ' Oral_mucosa__Squamous_epithelial_cells', ' Cervix', '_uterine', ' Skin_1__Keratinocytes', ' Esophagus', ' Gallbladder__Glandular_cells', ' Skin_2', ' Skin_1', ' Vagina__Squamous_epithelial_cells', ' Adrenal_gland', ' Skin_1__Langerhans', ' Gallbladder', ' Vagina', ' Adrenal_gland__Glandular_cells', ' Cervix', '_uterine__Squamous_epithelial_cells', ' Esophagus__Squamous_epithelial_cells', ' Oral_mucosa', ' Duodenum__Glandular_cells', ' Small_intestine', ' Rectum', ' Appendix__Glandular_cells', ' Colon__Glandular_cells', ' Rectum__Glandular_cells', ' Duodenum', ' Small_intestine__Glandular_cells', ' Appendix', ' Colon'], 'RELATED_PUBMED_ID': ['pubmed:9990065', ' pubmed:15489334'], 'FULL_NAME': 'Glucagon-like peptide 2 receptor', 'CONSERVED_SITE': 'IPR017983', 'GO_CC': ['GO:0016021', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036445', 'id': 'O95838', 'GO_MF': ['GO:0007188', ' GO:0007166', ' GO:0007186', ' GO:0008284'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['12626323', ' 24370975', ' 659632'], 'DRUG_ATC_CODE': ['ATC:H04AA', ' ATC:H04AA01', ' ATC:H04A', ' ATC:H', ' ATC:H04'], 'PRODUCT': ['Baqsimi', ' Glucag0n', ' GlucaGen', ' GlucaGen_HypoKit', ' Glucagon', ' Glucagon_Ampoule_666_1mg', ' Gvoke_HypoPen_0_5_mg_Auto_Injector', ' Gvoke_HypoPen_1_mg_Auto_Injector', ' Gvoke_PFS_0_5_mg_Pre_filled_Syringe', ' Gvoke_PFS_1_mg_Pre_filled_Syringe'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00040', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Glucagon', 'SYNONYM': ['Glucagon', ' Glucagon__recombinant_dna_origin_', ' Glucagon_recombinant', ' Glucagon__human', ' Glucagon__porcine', ' Glucagone'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR017981', ' IPR001879'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0008528', ' GO:0004967', ' GO:0044508', ' GO:0017046', ' GO:0004888'], 'SHORT_NAME': ['GLP-1 receptor', ' GLP-1-R', ' GLP-1R'], 'NAME': 'GLP1R', 'FAMILY': ['IPR003290', ' IPR003292', ' IPR000832'], 'RELATED_KEYWORD': ['3D-structure', ' ADP-ribosylation', ' Cell membrane', ' Disulfide bond', ' G-protein coupled receptor', ' Glycoprotein', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Signal', ' Transducer', ' Transmembrane', ' Transmembrane helix'], 'OTHER_ID': 'Q99669', 'RELATED_PUBMED_ID': ['pubmed:8405712', ' pubmed:8404634', ' pubmed:8216285', ' pubmed:7517895', ' pubmed:7843404', ' pubmed:14574404', ' pubmed:15489334', ' pubmed:9213353', ' pubmed:20869417', ' pubmed:21901419', ' pubmed:22412906', ' pubmed:18287102', ' pubmed:19861722', ' pubmed:26308095', ' pubmed:27196125', ' pubmed:28514449'], 'FULL_NAME': 'Glucagon-like peptide 1 receptor', 'CONSERVED_SITE': 'IPR017983', 'GO_CC': ['GO:0005623', ' GO:0016021', ' GO:0005887', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036445', 'id': 'P43220', 'GO_MF': ['GO:0007190', ' GO:0007189', ' GO:0019933', ' GO:0007166', ' GO:0007186', ' GO:0007611', ' GO:0045777', ' GO:0007204', ' GO:0008016', ' GO:0050796', ' GO:1990911'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['12626323', ' 24370975', ' 659632'], 'DRUG_ATC_CODE': ['ATC:H04AA', ' ATC:H04AA01', ' ATC:H04A', ' ATC:H', ' ATC:H04'], 'PRODUCT': ['Baqsimi', ' Glucag0n', ' GlucaGen', ' GlucaGen_HypoKit', ' Glucagon', ' Glucagon_Ampoule_666_1mg', ' Gvoke_HypoPen_0_5_mg_Auto_Injector', ' Gvoke_HypoPen_1_mg_Auto_Injector', ' Gvoke_PFS_0_5_mg_Pre_filled_Syringe', ' Gvoke_PFS_1_mg_Pre_filled_Syringe'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00040', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Glucagon', 'SYNONYM': ['Glucagon', ' Glucagon__recombinant_dna_origin_', ' Glucagon_recombinant', ' Glucagon__human', ' Glucagon__porcine', ' Glucagone'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR017981', ' IPR001879'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0008528', ' GO:0004967', ' GO:0005085', ' GO:0017046'], 'SHORT_NAME': 'GL-R', 'NAME': 'GLR', 'FAMILY': ['IPR003290', ' IPR003291', ' IPR000832'], 'RELATED_KEYWORD': ['3D-structure', ' Cell membrane', ' Diabetes mellitus', ' Disulfide bond', ' G-protein coupled receptor', ' Glycoprotein', ' Membrane', ' Phosphoprotein', ' Polymorphism', ' Receptor', ' Reference proteome', ' Signal', ' Transducer', ' Transmembrane', ' Transmembrane helix'], 'RELATED_PUBMED_ID': ['pubmed:7507321', ' pubmed:8144028', ' pubmed:15489334', ' pubmed:8020989', ' pubmed:9287038', ' pubmed:24275569', ' pubmed:16221670', ' pubmed:22908259', ' pubmed:23863937', ' pubmed:27111510', ' pubmed:28514451', ' pubmed:7589886', ' pubmed:19657311'], 'FULL_NAME': 'Glucagon receptor', 'CONSERVED_SITE': 'IPR017983', 'GO_CC': ['GO:0005768', ' GO:0016021', ' GO:0005887', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036445', 'id': 'P47871', 'GO_MF': ['GO:0007189', ' GO:0007188', ' GO:0007166', ' GO:0071377', ' GO:0009267', ' GO:0006887', ' GO:0007186', ' GO:0006091', ' GO:0042593', ' GO:0009755', ' GO:0010628', ' GO:0008217', ' GO:0070873', ' GO:0007584', ' GO:0042594'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['21456630', ' 12010086', ' 15078138', ' 25998211', ' 27789332'], 'DRUG_ATC_CODE': ['ATC:L', ' ATC:L03AA', ' ATC:L03', ' ATC:L03A', ' ATC:L03AA13'], 'PRODUCT': ['Fulphila', ' Lapelga', ' Neulasta', ' Neupopeg', ' Udenyca', ' Ziextenzo'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00019', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Pegfilgrastim', 'SYNONYM': ['Granulocyte_colony_stimulating_factor_pegfilgrastim', ' peg_filgrastim', ' pegfilgrastim_apgf', ' pegfilgrastim_bmez', ' pegfilgrastim_cbqv', ' pegfilgrastim_jmdb'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR003961', ' IPR010457'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019955', ' GO:0004896', ' GO:0051916', ' GO:0038023'], 'SHORT_NAME': ['G-CSF receptor', ' G-CSF-R'], 'NAME': 'CSF3R', 'RELATED_KEYWORD': ['3D-structure', ' Alternative splicing', ' Cell adhesion', ' Cell membrane', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Immunoglobulin domain', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Repeat', ' Secreted', ' Signal', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Bone_marrow', ' Lung__Macrophages', ' Lung', ' Bone_marrow__Hematopoietic_cells', ' Skin_1__Fibroblasts', ' Skin_1__Melanocytes', ' Skin_1', ' Heart_muscle', ' Testis', ' Testis__Leydig_cells', ' Cerebral_cortex', ' Cerebral_cortex__Neuronal_cells', ' Heart_muscle__Myocytes', ' Placenta', ' Placenta__Trophoblastic_cells'], 'RELATED_PUBMED_ID': ['pubmed:2147944', ' pubmed:1701053', ' pubmed:1530796', ' pubmed:15489334', ' pubmed:8554326', ' pubmed:1717255', ' pubmed:7514305', ' pubmed:17127321', ' pubmed:19120359', ' pubmed:9187659', ' pubmed:9368043', ' pubmed:16492764', ' pubmed:10449521', ' pubmed:19620628', ' pubmed:24753537', ' pubmed:26324699'], 'FULL_NAME': 'Granulocyte colony-stimulating factor receptor', 'CONSERVED_SITE': 'IPR003529', 'GO_CC': ['GO:0009897', ' GO:0005576', ' GO:0005887', ' GO:0005886', ' GO:0043235'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036116', ' IPR036179', ' IPR013783'], 'id': 'Q99062', 'GO_MF': ['GO:0097186', ' GO:0007155', ' GO:0019221', ' GO:0006952', ' GO:0030593', ' GO:0045637', ' GO:0007165'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['21456630', ' 12010086', ' 15078138', ' 25998211', ' 27789332'], 'DRUG_ATC_CODE': ['ATC:L', ' ATC:L03AA', ' ATC:L03', ' ATC:L03A', ' ATC:L03AA13'], 'PRODUCT': ['Fulphila', ' Lapelga', ' Neulasta', ' Neupopeg', ' Udenyca', ' Ziextenzo'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00019', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Pegfilgrastim', 'SYNONYM': ['Granulocyte_colony_stimulating_factor_pegfilgrastim', ' peg_filgrastim', ' pegfilgrastim_apgf', ' pegfilgrastim_bmez', ' pegfilgrastim_cbqv', ' pegfilgrastim_jmdb'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': 'IPR001254', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019955', ' GO:0004175', ' GO:0008201', ' GO:0008233', ' GO:0002020', ' GO:0004252', ' GO:0003714'], 'ACTIVE_SITE': ['IPR018114', ' IPR033116'], 'SHORT_NAME': 'HLE', 'NAME': 'ELNE', 'FAMILY': 'IPR001314', 'RELATED_KEYWORD': ['3D-structure', ' Cytoplasmic vesicle', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Hydrolase', ' Polymorphism', ' Protease', ' Reference proteome', ' Serine protease', ' Signal', ' Zymogen'], 'PROTEIN_EXPRESSED_IN': ['Spleen__Cells_in_red_pulp', ' Spleen', ' Bone_marrow__Hematopoietic_cells', ' Bone_marrow'], 'OTHER_ID': ['Q6B0D9', ' Q6LDP5'], 'RELATED_PUBMED_ID': ['pubmed:3479752', ' pubmed:2902087', ' pubmed:2775493', ' pubmed:2322278', ' pubmed:15489334', ' pubmed:2822677', ' pubmed:3550808', ' pubmed:2501794', ' pubmed:7897245', ' pubmed:3422232', ' pubmed:2462434', ' pubmed:1859409', ' pubmed:10947984', ' pubmed:15140022', ' pubmed:14673143', ' pubmed:19159218', ' pubmed:21269460', ' pubmed:25944712', ' pubmed:2911584', ' pubmed:3391280', ' pubmed:3640709', ' pubmed:10581030', ' pubmed:11001877', ' pubmed:11675333', ' pubmed:12091371', ' pubmed:14962902', ' pubmed:17436313', ' pubmed:18946670', ' pubmed:19036076', ' pubmed:19415009', ' pubmed:20220065', ' pubmed:19927291', ' pubmed:20803142', ' pubmed:21425445', ' pubmed:23463630'], 'FULL_NAME': ['Neutrophil elastase', ' Bone marrow serine protease', ' Elastase-2', ' Human leukocyte elastase', ' Medullasin', ' PMN elastase'], 'GO_CC': ['GO:0035578', ' GO:0005623', ' GO:0009986', ' GO:0062023', ' GO:0005737', ' GO:0070062', ' GO:0005576', ' GO:0005615', ' GO:0045335', ' GO:0030141', ' GO:0035580', ' GO:0017053'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR009003', 'id': 'P08246', 'GO_MF': ['GO:0002438', ' GO:0019730', ' GO:0002812', ' GO:0006874', ' GO:0042742', ' GO:0022617', ' GO:0002523', ' GO:0045079', ' GO:0050922', ' GO:0050728', ' GO:0045415', ' GO:0000122', ' GO:0043312', ' GO:0070947', ' GO:0070945', ' GO:0006909', ' GO:0050778', ' GO:0045416', ' GO:1903238', ' GO:0043406', ' GO:0048661', ' GO:0030163', ' GO:0006508', ' GO:0032496', ' GO:0009411', ' GO:0001878'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['17187516', ' 17454642'], 'DRUG_ATC_CODE': ['ATC:L01XX', ' ATC:L01XX29', ' ATC:L', ' ATC:L01X', ' ATC:L01'], 'PRODUCT': 'Ontak', 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00004', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Denileukin_diftitox', 'SYNONYM': ['Denileukin', ' Interleukin_2_diptheria_toxin_fusion_protein'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': 'IPR000436', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019976', ' GO:0004911'], 'SHORT_NAME': ['IL-2 receptor subunit alpha', ' IL-2-RA', ' IL-2R subunit alpha', ' IL2-RA'], 'NAME': 'IL2RA', 'FAMILY': 'IPR015486', 'RELATED_KEYWORD': ['3D-structure', ' Diabetes mellitus', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Immunity', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Repeat', ' Signal', ' Sushi', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Appendix', ' Bone_marrow', ' Appendix__Lymphoid_tissue', ' Bone_marrow__Hematopoietic_cells', ' Lymph_node', ' Lymph_node__Germinal_center_cells', ' Tonsil__Germinal_center_cells', ' Spleen', ' Spleen__Cells_in_red_pulp', ' Tonsil__Non-germinal_center_cells', ' Lymph_node__Non-germinal_center_cells', ' Tonsil'], 'RELATED_PUBMED_ID': ['pubmed:6090949', ' pubmed:6090948', ' pubmed:2999698', ' pubmed:2996141', ' pubmed:15164054', ' pubmed:3030566', ' pubmed:3134887', ' pubmed:7529123', ' pubmed:16293754', ' pubmed:16477002', ' pubmed:17676041', ' pubmed:23416241', ' pubmed:24116927'], 'FULL_NAME': ['Interleukin-2 receptor subunit alpha', ' TAC antigen', ' p55'], 'GO_CC': ['GO:0009897', ' GO:0005893', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR035976', 'id': 'P01589', 'GO_MF': ['GO:0006924', ' GO:0006915', ' GO:0007166', ' GO:0019221', ' GO:0006955', ' GO:0006954', ' GO:0002437', ' GO:0038110', ' GO:0000165', ' GO:0050687', ' GO:0050777', ' GO:0050728', ' GO:0042130', ' GO:0007219', ' GO:0042104', ' GO:0045582', ' GO:0045589', ' GO:0046013', ' GO:0002664'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['17187516', ' 17454642'], 'DRUG_ATC_CODE': ['ATC:L01XX', ' ATC:L01XX29', ' ATC:L', ' ATC:L01X', ' ATC:L01'], 'PRODUCT': 'Ontak', 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00004', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Denileukin_diftitox', 'SYNONYM': ['Denileukin', ' Interleukin_2_diptheria_toxin_fusion_protein'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR003961', ' IPR040951'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0042010', ' GO:0019976', ' GO:0004911'], 'SHORT_NAME': ['IL-2 receptor subunit beta', ' IL-2R subunit beta', ' IL-2RB', ' p75'], 'NAME': 'IL2RB', 'RELATED_KEYWORD': ['3D-structure', ' Cell membrane', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Host-virus interaction', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Signal', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Spleen', ' Lymph_node__Non-germinal_center_cells', ' Tonsil__Non-germinal_center_cells', ' Spleen__Cells_in_white_pulp', ' Tonsil', ' Spleen__Cells_in_red_pulp', ' Bone_marrow', ' Lymph_node', ' Bone_marrow__Hematopoietic_cells'], 'RELATED_PUBMED_ID': ['pubmed:2785715', ' pubmed:15461802', ' pubmed:14702039', ' pubmed:10591208', ' pubmed:15489334', ' pubmed:8648694', ' pubmed:12200137', ' pubmed:15123770', ' pubmed:7529123', ' pubmed:16293754', ' pubmed:16477002', ' pubmed:31040184', ' pubmed:31040185'], 'FULL_NAME': ['Interleukin-2 receptor subunit beta', ' High affinity IL-2 receptor subunit beta', ' Interleukin-15 receptor subunit beta {ECO:0000312|HGNC:HGNC:6009}', ' p70-75'], 'CONSERVED_SITE': 'IPR003531', 'GO_CC': ['GO:0009986', ' GO:0005829', ' GO:0005768', ' GO:0009897', ' GO:0005887', ' GO:0005893', ' GO:0016020', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036116', ' IPR013783'], 'id': 'P14784', 'GO_MF': ['GO:0019221', ' GO:0035723', ' GO:0038110', ' GO:0000165', ' GO:0043066', ' GO:0050766', ' GO:0065003', ' GO:0007165', ' GO:0016032'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['17187516', ' 17454642'], 'DRUG_ATC_CODE': ['ATC:L01XX', ' ATC:L01XX29', ' ATC:L', ' ATC:L01X', ' ATC:L01'], 'PRODUCT': 'Ontak', 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00004', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Denileukin_diftitox', 'SYNONYM': ['Denileukin', ' Interleukin_2_diptheria_toxin_fusion_protein'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR003961', ' IPR015321'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019955', ' GO:0004896', ' GO:0042010', ' GO:0019976'], 'SHORT_NAME': ['IL-2 receptor subunit gamma', ' IL-2R subunit gamma', ' IL-2RG'], 'NAME': 'IL2RG', 'RELATED_KEYWORD': ['3D-structure', ' Alternative splicing', ' Cell membrane', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Host-virus interaction', ' Membrane', ' Phosphoprotein', ' Polymorphism', ' Receptor', ' Reference proteome', ' SCID', ' Signal', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Bone_marrow__Hematopoietic_cells', ' Tonsil', ' Bone_marrow', ' Tonsil__Squamous_epithelial_cells', ' Spleen__Cells_in_red_pulp', ' Spleen'], 'RELATED_PUBMED_ID': ['pubmed:1631559', ' pubmed:8514792', ' pubmed:8401490', ' pubmed:15772651', ' pubmed:15489334', ' pubmed:8266076', ' pubmed:8266078', ' pubmed:8266077', ' pubmed:8648694', ' pubmed:12200137', ' pubmed:15123770', ' pubmed:19690332', ' pubmed:7529123', ' pubmed:16293754', ' pubmed:16477002', ' pubmed:8299698', ' pubmed:8088810', ' pubmed:8027558', ' pubmed:7937790', ' pubmed:7668284', ' pubmed:7557965', ' pubmed:7860773', ' pubmed:7883965', ' pubmed:8900089', ' pubmed:9150740', ' pubmed:9049783', ' pubmed:9399950'], 'FULL_NAME': ['Cytokine receptor common subunit gamma', ' Interleukin-2 receptor subunit gamma', ' gammaC', ' p64'], 'CONSERVED_SITE': 'IPR003531', 'GO_CC': ['GO:0009986', ' GO:0005829', ' GO:0005768', ' GO:0009897', ' GO:0005887', ' GO:0016020', ' GO:0005886', ' GO:0043235'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036116', ' IPR013783'], 'id': 'P31785', 'GO_MF': ['GO:0019221', ' GO:0006955', ' GO:0035723', ' GO:0038110', ' GO:0038114', ' GO:0035771', ' GO:0038111', ' GO:0038113', ' GO:0000165', ' GO:0050766', ' GO:0007165', ' GO:0016032'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['10480573', ' 10601294', ' 10628369', ' 11408594', ' 11431346', ' 11752352', ' 15269308', ' 16336752'], 'DRUG_ATC_CODE': ['ATC:L01', ' ATC:L', ' ATC:L01XC06', ' ATC:L01XC', ' ATC:L01X'], 'PRODUCT': 'Erbitux', 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00002', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Cetuximab', 'SYNONYM': nan, 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR007110', ' IPR003599'], 'SPECIES': 'HUMAN', 'GO_BP': 'GO:0019864', 'SHORT_NAME': ['Fc-gamma RIII', ' Fc-gamma RIIIb', ' FcRIII', ' FcRIIIb'], 'NAME': 'FCG3B', 'RELATED_KEYWORD': ['3D-structure', ' Cell membrane', ' Disulfide bond', ' Glycoprotein', ' GPI-anchor', ' IgG-binding protein', ' Immunoglobulin domain', ' Lipoprotein', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Repeat', ' Secreted', ' Signal'], 'RELATED_PUBMED_ID': ['pubmed:2526846', ' pubmed:2967436', ' pubmed:2521732', ' pubmed:2525780', ' pubmed:15245367', ' pubmed:16710414', ' pubmed:7836402', ' pubmed:10917521', ' pubmed:11021536', ' pubmed:9028335'], 'FULL_NAME': ['Low affinity immunoglobulin gamma Fc region receptor III-B', ' Fc-gamma RIII-beta', ' FcR-10', ' IgG Fc receptor III-1'], 'GO_CC': ['GO:0031225', ' GO:0070062', ' GO:0005576', ' GO:0005886', ' GO:0030667'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036179', ' IPR013783'], 'id': 'O75015', 'GO_MF': ['GO:0006955', ' GO:0043312'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['10480573', ' 10601294', ' 10628369', ' 11408594', ' 11431346', ' 11752352', ' 15269308', ' 16336752'], 'DRUG_ATC_CODE': ['ATC:L01', ' ATC:L', ' ATC:L01XC06', ' ATC:L01XC', ' ATC:L01X'], 'PRODUCT': 'Erbitux', 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00002', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Cetuximab', 'SYNONYM': nan, 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR006211', ' IPR032778', ' IPR000719', ' IPR000494', ' IPR001245', ' IPR020635'], 'REPEAT': 'IPR006212', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0051015', ' GO:0005524', ' GO:0051117', ' GO:0045296', ' GO:0005516', ' GO:0003682', ' GO:0003690', ' GO:0019899', ' GO:0048408', ' GO:0005006', ' GO:0042802', ' GO:0005178', ' GO:0004709', ' GO:0019901', ' GO:0019903', ' GO:0004713', ' GO:0004714', ' GO:0004888', ' GO:0031625', ' GO:0001618'], 'ACTIVE_SITE': 'IPR008266', 'NAME': 'EGFR', 'FAMILY': 'IPR016245', 'RELATED_KEYWORD': ['3D-structure', ' Alternative splicing', ' ATP-binding', ' Cell membrane', ' Developmental protein', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Endoplasmic reticulum', ' Endosome', ' Glycoprotein', ' Golgi apparatus', ' Host cell receptor for virus entry', ' Host-virus interaction', ' Isopeptide bond', ' Kinase', ' Lipoprotein', ' Membrane', ' Methylation', ' Nucleotide-binding', ' Nucleus', ' Palmitate', ' Phosphoprotein', ' Polymorphism', ' Proto-oncogene', ' Receptor', ' Reference proteome', ' Repeat', ' Secreted', ' Signal', ' Transferase', ' Transmembrane', ' Transmembrane helix', ' Tyrosine-protein kinase', ' Ubl conjugation'], 'PROTEIN_EXPRESSED_IN': ['Skin_1__Melanocytes', ' Tonsil', ' Vagina__Squamous_epithelial_cells', ' Skeletal_muscle', ' Urinary_bladder__Urothelial_cells', ' Urinary_bladder', ' Skin_1', ' Skeletal_muscle__Myocytes', ' Testis__Leydig_cells', ' Tonsil__Squamous_epithelial_cells', ' Liver__Hepatocytes', ' Vagina', ' Liver', ' Testis', ' Placenta__Trophoblastic_cells', ' Placenta'], 'OTHER_ID': ['O00732', ' P06268', ' Q14225', ' Q68GS5', ' Q92795', ' Q9BZS2', ' Q9GZX1', ' Q9H2C9', ' Q9H3C9', ' Q9UMD7', ' Q9UMD8', ' Q9UMG5'], 'RELATED_PUBMED_ID': ['pubmed:6328312', ' pubmed:7654368', ' pubmed:8918811', ' pubmed:9103388', ' pubmed:11161793', ' pubmed:6326261', ' pubmed:6330563', ' pubmed:6093780', ' pubmed:3329716', ' pubmed:1988448', ' pubmed:2991899', ' pubmed:6324343', ' pubmed:3138233', ' pubmed:15340161', ' pubmed:2985580', ' pubmed:16543144', ' pubmed:9556602', ' pubmed:6325948', ' pubmed:3039909', ' pubmed:2790960', ' pubmed:2543678', ' pubmed:7679104', ' pubmed:8650580', ' pubmed:9852145', ' pubmed:10523301', ' pubmed:8144591', ' pubmed:7657591', ' pubmed:8962717', ' pubmed:9419975', ' pubmed:9488479', ' pubmed:10228163', ' pubmed:10026169', ' pubmed:10731668', ' pubmed:11602604', ' pubmed:12731890', ' pubmed:12873986', ' pubmed:15282549', ' pubmed:10805725', ' pubmed:11116146', ' pubmed:11483589', ' pubmed:15590694', ' pubmed:15611079', ' pubmed:16083266', ' pubmed:16140940', ' pubmed:17081983', ' pubmed:17115032', ' pubmed:16420529', ' pubmed:17671655', ' pubmed:17182860', ' pubmed:18602463', ' pubmed:19172738', ' pubmed:18220336', ' pubmed:18691976', ' pubmed:18669648', ' pubmed:19413330', ' pubmed:19836242', ' pubmed:19509291', ' pubmed:19159218', ' pubmed:19369195', ' pubmed:19749156', ' pubmed:19718021', ' pubmed:20674546', ' pubmed:20551055', ' pubmed:20462955', ' pubmed:20068231', ' pubmed:21269460', ' pubmed:21487020', ' pubmed:21258366', ' pubmed:21516087', ' pubmed:21518868', ' pubmed:22298428', ' pubmed:22179831', ' pubmed:23418353', ' pubmed:23186163', ' pubmed:24275569', ' pubmed:25187647', ' pubmed:23589287', ' pubmed:25311788', ' pubmed:23912460', ' pubmed:27153536', ' pubmed:26988023', ' pubmed:28479384', ' pubmed:12297049', ' pubmed:12297050', ' pubmed:12620237', ' pubmed:15374980', ' pubmed:15837620', ' pubmed:15840573', ' pubmed:17349580', ' pubmed:18046415', ' pubmed:18227510', ' pubmed:19563760', ' pubmed:19560417', ' pubmed:20471394', ' pubmed:20837704', ' pubmed:22888118', ' pubmed:15623594', ' pubmed:15118125', ' pubmed:16533793', ' pubmed:16205628', ' pubmed:16672372', ' pubmed:17344846', ' pubmed:24691054'], 'BINDING_SITE': 'IPR017441', 'FULL_NAME': ['Epidermal growth factor receptor', ' Proto-oncogene c-ErbB-1', ' Receptor tyrosine-protein kinase erbB-1'], 'GO_CC': ['GO:0016324', ' GO:0009925', ' GO:0016323', ' GO:0030054', ' GO:0009986', ' GO:0030665', ' GO:0005737', ' GO:0031901', ' GO:0030139', ' GO:0005789', ' GO:0005768', ' GO:0010008', ' GO:0005615', ' GO:0005925', ' GO:0000139', ' GO:0005887', ' GO:0016020', ' GO:0045121', ' GO:0097489', ' GO:0031965', ' GO:0005634', ' GO:0048471', ' GO:0005886', ' GO:0032991', ' GO:0043235', ' GO:0070435', ' GO:0045202'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR009030', ' IPR011009', ' IPR036941'], 'id': 'P00533', 'GO_MF': ['GO:0043006', ' GO:0007202', ' GO:0048143', ' GO:0030154', ' GO:0007166', ' GO:0098609', ' GO:0071230', ' GO:0071276', ' GO:0071549', ' GO:0071364', ' GO:0071392', ' GO:0071260', ' GO:0034614', ' GO:0021795', ' GO:0007623', ' GO:0048546', ' GO:0016101', ' GO:0001892', ' GO:0007173', ' GO:0038128', ' GO:0061029', ' GO:0001942', ' GO:0042743', ' GO:0007611', ' GO:0097421', ' GO:0030324', ' GO:0010960', ' GO:0000165', ' GO:0061024', ' GO:0007494', ' GO:0060571', ' GO:0007275', ' GO:0043066', ' GO:1905208', ' GO:0042059', ' GO:1901185', ' GO:0045930', ' GO:0045746', ' GO:0042177', ' GO:0048812', ' GO:0001503', ' GO:0042698', ' GO:0038083', ' GO:0018108', ' GO:0097755', ' GO:0045780', ' GO:0090263', ' GO:0030307', ' GO:0030335', ' GO:0008284', ' GO:0045737', ' GO:0045739', ' GO:0045740', ' GO:0050679', ' GO:0070374', ' GO:0048146', ' GO:1900087', ' GO:0050729', ' GO:0033674', ' GO:0043406', ' GO:1901224', ' GO:0010750', ' GO:0033138', ' GO:0042327', ' GO:1903800', ' GO:1902722', ' GO:0051897', ' GO:1900020', ' GO:1903078', ' GO:0001934', ' GO:0048661', ' GO:0032930', ' GO:0051968', ' GO:0045944', ' GO:0045893', ' GO:0045907', ' GO:0046777', ' GO:0051205', ' GO:2000145', ' GO:0070372', ' GO:0046328', ' GO:0050999', ' GO:0050730', ' GO:0014066', ' GO:0006357', ' GO:0051592', ' GO:0033590', ' GO:0033594', ' GO:0006970', ' GO:0070141', ' GO:0007435', ' GO:0007165', ' GO:0043586', ' GO:0006412', ' GO:0007169', ' GO:0042060'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}]\u001b[0m\n",
      "Attempt 2 failed with error: Error code: 413 - {'error': {'message': 'Request too large for model `llama3-70b-8192` in organization `org_01j9yv18bffqksnp3w7k6gzqk0` on tokens per minute (TPM): Limit 6000, Requested 8215, please reduce your message size and try again. Visit https://console.groq.com/docs/rate-limits for more information.', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}\n",
      "Attempt 3...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug)-[:DRUG_TARGET|DRUG_ENZYME|DRUG_CARRIER|DRUG_TRANSPORTER]->(p:Protein) RETURN d, p\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['12626323', ' 24370975', ' 659632'], 'DRUG_ATC_CODE': ['ATC:H04AA', ' ATC:H04AA01', ' ATC:H04A', ' ATC:H', ' ATC:H04'], 'PRODUCT': ['Baqsimi', ' Glucag0n', ' GlucaGen', ' GlucaGen_HypoKit', ' Glucagon', ' Glucagon_Ampoule_666_1mg', ' Gvoke_HypoPen_0_5_mg_Auto_Injector', ' Gvoke_HypoPen_1_mg_Auto_Injector', ' Gvoke_PFS_0_5_mg_Pre_filled_Syringe', ' Gvoke_PFS_1_mg_Pre_filled_Syringe'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00040', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Glucagon', 'SYNONYM': ['Glucagon', ' Glucagon__recombinant_dna_origin_', ' Glucagon_recombinant', ' Glucagon__human', ' Glucagon__porcine', ' Glucagone'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR017981', ' IPR001879'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0008528', ' GO:0004930', ' GO:0004967', ' GO:0017046'], 'SHORT_NAME': ['GLP-2 receptor', ' GLP-2-R', ' GLP-2R'], 'NAME': 'GLP2R', 'FAMILY': ['IPR039125', ' IPR000832'], 'RELATED_KEYWORD': ['Cell membrane', ' Disulfide bond', ' G-protein coupled receptor', ' Glycoprotein', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Transducer', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Skin_2__Epidermal_cells', ' Oral_mucosa__Squamous_epithelial_cells', ' Cervix', '_uterine', ' Skin_1__Keratinocytes', ' Esophagus', ' Gallbladder__Glandular_cells', ' Skin_2', ' Skin_1', ' Vagina__Squamous_epithelial_cells', ' Adrenal_gland', ' Skin_1__Langerhans', ' Gallbladder', ' Vagina', ' Adrenal_gland__Glandular_cells', ' Cervix', '_uterine__Squamous_epithelial_cells', ' Esophagus__Squamous_epithelial_cells', ' Oral_mucosa', ' Duodenum__Glandular_cells', ' Small_intestine', ' Rectum', ' Appendix__Glandular_cells', ' Colon__Glandular_cells', ' Rectum__Glandular_cells', ' Duodenum', ' Small_intestine__Glandular_cells', ' Appendix', ' Colon'], 'RELATED_PUBMED_ID': ['pubmed:9990065', ' pubmed:15489334'], 'FULL_NAME': 'Glucagon-like peptide 2 receptor', 'CONSERVED_SITE': 'IPR017983', 'GO_CC': ['GO:0016021', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036445', 'id': 'O95838', 'GO_MF': ['GO:0007188', ' GO:0007166', ' GO:0007186', ' GO:0008284'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['12626323', ' 24370975', ' 659632'], 'DRUG_ATC_CODE': ['ATC:H04AA', ' ATC:H04AA01', ' ATC:H04A', ' ATC:H', ' ATC:H04'], 'PRODUCT': ['Baqsimi', ' Glucag0n', ' GlucaGen', ' GlucaGen_HypoKit', ' Glucagon', ' Glucagon_Ampoule_666_1mg', ' Gvoke_HypoPen_0_5_mg_Auto_Injector', ' Gvoke_HypoPen_1_mg_Auto_Injector', ' Gvoke_PFS_0_5_mg_Pre_filled_Syringe', ' Gvoke_PFS_1_mg_Pre_filled_Syringe'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00040', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Glucagon', 'SYNONYM': ['Glucagon', ' Glucagon__recombinant_dna_origin_', ' Glucagon_recombinant', ' Glucagon__human', ' Glucagon__porcine', ' Glucagone'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR017981', ' IPR001879'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0008528', ' GO:0004967', ' GO:0044508', ' GO:0017046', ' GO:0004888'], 'SHORT_NAME': ['GLP-1 receptor', ' GLP-1-R', ' GLP-1R'], 'NAME': 'GLP1R', 'FAMILY': ['IPR003290', ' IPR003292', ' IPR000832'], 'RELATED_KEYWORD': ['3D-structure', ' ADP-ribosylation', ' Cell membrane', ' Disulfide bond', ' G-protein coupled receptor', ' Glycoprotein', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Signal', ' Transducer', ' Transmembrane', ' Transmembrane helix'], 'OTHER_ID': 'Q99669', 'RELATED_PUBMED_ID': ['pubmed:8405712', ' pubmed:8404634', ' pubmed:8216285', ' pubmed:7517895', ' pubmed:7843404', ' pubmed:14574404', ' pubmed:15489334', ' pubmed:9213353', ' pubmed:20869417', ' pubmed:21901419', ' pubmed:22412906', ' pubmed:18287102', ' pubmed:19861722', ' pubmed:26308095', ' pubmed:27196125', ' pubmed:28514449'], 'FULL_NAME': 'Glucagon-like peptide 1 receptor', 'CONSERVED_SITE': 'IPR017983', 'GO_CC': ['GO:0005623', ' GO:0016021', ' GO:0005887', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036445', 'id': 'P43220', 'GO_MF': ['GO:0007190', ' GO:0007189', ' GO:0019933', ' GO:0007166', ' GO:0007186', ' GO:0007611', ' GO:0045777', ' GO:0007204', ' GO:0008016', ' GO:0050796', ' GO:1990911'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['12626323', ' 24370975', ' 659632'], 'DRUG_ATC_CODE': ['ATC:H04AA', ' ATC:H04AA01', ' ATC:H04A', ' ATC:H', ' ATC:H04'], 'PRODUCT': ['Baqsimi', ' Glucag0n', ' GlucaGen', ' GlucaGen_HypoKit', ' Glucagon', ' Glucagon_Ampoule_666_1mg', ' Gvoke_HypoPen_0_5_mg_Auto_Injector', ' Gvoke_HypoPen_1_mg_Auto_Injector', ' Gvoke_PFS_0_5_mg_Pre_filled_Syringe', ' Gvoke_PFS_1_mg_Pre_filled_Syringe'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00040', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Glucagon', 'SYNONYM': ['Glucagon', ' Glucagon__recombinant_dna_origin_', ' Glucagon_recombinant', ' Glucagon__human', ' Glucagon__porcine', ' Glucagone'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR017981', ' IPR001879'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0008528', ' GO:0004967', ' GO:0005085', ' GO:0017046'], 'SHORT_NAME': 'GL-R', 'NAME': 'GLR', 'FAMILY': ['IPR003290', ' IPR003291', ' IPR000832'], 'RELATED_KEYWORD': ['3D-structure', ' Cell membrane', ' Diabetes mellitus', ' Disulfide bond', ' G-protein coupled receptor', ' Glycoprotein', ' Membrane', ' Phosphoprotein', ' Polymorphism', ' Receptor', ' Reference proteome', ' Signal', ' Transducer', ' Transmembrane', ' Transmembrane helix'], 'RELATED_PUBMED_ID': ['pubmed:7507321', ' pubmed:8144028', ' pubmed:15489334', ' pubmed:8020989', ' pubmed:9287038', ' pubmed:24275569', ' pubmed:16221670', ' pubmed:22908259', ' pubmed:23863937', ' pubmed:27111510', ' pubmed:28514451', ' pubmed:7589886', ' pubmed:19657311'], 'FULL_NAME': 'Glucagon receptor', 'CONSERVED_SITE': 'IPR017983', 'GO_CC': ['GO:0005768', ' GO:0016021', ' GO:0005887', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036445', 'id': 'P47871', 'GO_MF': ['GO:0007189', ' GO:0007188', ' GO:0007166', ' GO:0071377', ' GO:0009267', ' GO:0006887', ' GO:0007186', ' GO:0006091', ' GO:0042593', ' GO:0009755', ' GO:0010628', ' GO:0008217', ' GO:0070873', ' GO:0007584', ' GO:0042594'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['21456630', ' 12010086', ' 15078138', ' 25998211', ' 27789332'], 'DRUG_ATC_CODE': ['ATC:L', ' ATC:L03AA', ' ATC:L03', ' ATC:L03A', ' ATC:L03AA13'], 'PRODUCT': ['Fulphila', ' Lapelga', ' Neulasta', ' Neupopeg', ' Udenyca', ' Ziextenzo'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00019', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Pegfilgrastim', 'SYNONYM': ['Granulocyte_colony_stimulating_factor_pegfilgrastim', ' peg_filgrastim', ' pegfilgrastim_apgf', ' pegfilgrastim_bmez', ' pegfilgrastim_cbqv', ' pegfilgrastim_jmdb'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR003961', ' IPR010457'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019955', ' GO:0004896', ' GO:0051916', ' GO:0038023'], 'SHORT_NAME': ['G-CSF receptor', ' G-CSF-R'], 'NAME': 'CSF3R', 'RELATED_KEYWORD': ['3D-structure', ' Alternative splicing', ' Cell adhesion', ' Cell membrane', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Immunoglobulin domain', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Repeat', ' Secreted', ' Signal', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Bone_marrow', ' Lung__Macrophages', ' Lung', ' Bone_marrow__Hematopoietic_cells', ' Skin_1__Fibroblasts', ' Skin_1__Melanocytes', ' Skin_1', ' Heart_muscle', ' Testis', ' Testis__Leydig_cells', ' Cerebral_cortex', ' Cerebral_cortex__Neuronal_cells', ' Heart_muscle__Myocytes', ' Placenta', ' Placenta__Trophoblastic_cells'], 'RELATED_PUBMED_ID': ['pubmed:2147944', ' pubmed:1701053', ' pubmed:1530796', ' pubmed:15489334', ' pubmed:8554326', ' pubmed:1717255', ' pubmed:7514305', ' pubmed:17127321', ' pubmed:19120359', ' pubmed:9187659', ' pubmed:9368043', ' pubmed:16492764', ' pubmed:10449521', ' pubmed:19620628', ' pubmed:24753537', ' pubmed:26324699'], 'FULL_NAME': 'Granulocyte colony-stimulating factor receptor', 'CONSERVED_SITE': 'IPR003529', 'GO_CC': ['GO:0009897', ' GO:0005576', ' GO:0005887', ' GO:0005886', ' GO:0043235'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036116', ' IPR036179', ' IPR013783'], 'id': 'Q99062', 'GO_MF': ['GO:0097186', ' GO:0007155', ' GO:0019221', ' GO:0006952', ' GO:0030593', ' GO:0045637', ' GO:0007165'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['21456630', ' 12010086', ' 15078138', ' 25998211', ' 27789332'], 'DRUG_ATC_CODE': ['ATC:L', ' ATC:L03AA', ' ATC:L03', ' ATC:L03A', ' ATC:L03AA13'], 'PRODUCT': ['Fulphila', ' Lapelga', ' Neulasta', ' Neupopeg', ' Udenyca', ' Ziextenzo'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00019', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Pegfilgrastim', 'SYNONYM': ['Granulocyte_colony_stimulating_factor_pegfilgrastim', ' peg_filgrastim', ' pegfilgrastim_apgf', ' pegfilgrastim_bmez', ' pegfilgrastim_cbqv', ' pegfilgrastim_jmdb'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': 'IPR001254', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019955', ' GO:0004175', ' GO:0008201', ' GO:0008233', ' GO:0002020', ' GO:0004252', ' GO:0003714'], 'ACTIVE_SITE': ['IPR018114', ' IPR033116'], 'SHORT_NAME': 'HLE', 'NAME': 'ELNE', 'FAMILY': 'IPR001314', 'RELATED_KEYWORD': ['3D-structure', ' Cytoplasmic vesicle', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Hydrolase', ' Polymorphism', ' Protease', ' Reference proteome', ' Serine protease', ' Signal', ' Zymogen'], 'PROTEIN_EXPRESSED_IN': ['Spleen__Cells_in_red_pulp', ' Spleen', ' Bone_marrow__Hematopoietic_cells', ' Bone_marrow'], 'OTHER_ID': ['Q6B0D9', ' Q6LDP5'], 'RELATED_PUBMED_ID': ['pubmed:3479752', ' pubmed:2902087', ' pubmed:2775493', ' pubmed:2322278', ' pubmed:15489334', ' pubmed:2822677', ' pubmed:3550808', ' pubmed:2501794', ' pubmed:7897245', ' pubmed:3422232', ' pubmed:2462434', ' pubmed:1859409', ' pubmed:10947984', ' pubmed:15140022', ' pubmed:14673143', ' pubmed:19159218', ' pubmed:21269460', ' pubmed:25944712', ' pubmed:2911584', ' pubmed:3391280', ' pubmed:3640709', ' pubmed:10581030', ' pubmed:11001877', ' pubmed:11675333', ' pubmed:12091371', ' pubmed:14962902', ' pubmed:17436313', ' pubmed:18946670', ' pubmed:19036076', ' pubmed:19415009', ' pubmed:20220065', ' pubmed:19927291', ' pubmed:20803142', ' pubmed:21425445', ' pubmed:23463630'], 'FULL_NAME': ['Neutrophil elastase', ' Bone marrow serine protease', ' Elastase-2', ' Human leukocyte elastase', ' Medullasin', ' PMN elastase'], 'GO_CC': ['GO:0035578', ' GO:0005623', ' GO:0009986', ' GO:0062023', ' GO:0005737', ' GO:0070062', ' GO:0005576', ' GO:0005615', ' GO:0045335', ' GO:0030141', ' GO:0035580', ' GO:0017053'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR009003', 'id': 'P08246', 'GO_MF': ['GO:0002438', ' GO:0019730', ' GO:0002812', ' GO:0006874', ' GO:0042742', ' GO:0022617', ' GO:0002523', ' GO:0045079', ' GO:0050922', ' GO:0050728', ' GO:0045415', ' GO:0000122', ' GO:0043312', ' GO:0070947', ' GO:0070945', ' GO:0006909', ' GO:0050778', ' GO:0045416', ' GO:1903238', ' GO:0043406', ' GO:0048661', ' GO:0030163', ' GO:0006508', ' GO:0032496', ' GO:0009411', ' GO:0001878'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['17187516', ' 17454642'], 'DRUG_ATC_CODE': ['ATC:L01XX', ' ATC:L01XX29', ' ATC:L', ' ATC:L01X', ' ATC:L01'], 'PRODUCT': 'Ontak', 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00004', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Denileukin_diftitox', 'SYNONYM': ['Denileukin', ' Interleukin_2_diptheria_toxin_fusion_protein'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': 'IPR000436', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019976', ' GO:0004911'], 'SHORT_NAME': ['IL-2 receptor subunit alpha', ' IL-2-RA', ' IL-2R subunit alpha', ' IL2-RA'], 'NAME': 'IL2RA', 'FAMILY': 'IPR015486', 'RELATED_KEYWORD': ['3D-structure', ' Diabetes mellitus', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Immunity', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Repeat', ' Signal', ' Sushi', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Appendix', ' Bone_marrow', ' Appendix__Lymphoid_tissue', ' Bone_marrow__Hematopoietic_cells', ' Lymph_node', ' Lymph_node__Germinal_center_cells', ' Tonsil__Germinal_center_cells', ' Spleen', ' Spleen__Cells_in_red_pulp', ' Tonsil__Non-germinal_center_cells', ' Lymph_node__Non-germinal_center_cells', ' Tonsil'], 'RELATED_PUBMED_ID': ['pubmed:6090949', ' pubmed:6090948', ' pubmed:2999698', ' pubmed:2996141', ' pubmed:15164054', ' pubmed:3030566', ' pubmed:3134887', ' pubmed:7529123', ' pubmed:16293754', ' pubmed:16477002', ' pubmed:17676041', ' pubmed:23416241', ' pubmed:24116927'], 'FULL_NAME': ['Interleukin-2 receptor subunit alpha', ' TAC antigen', ' p55'], 'GO_CC': ['GO:0009897', ' GO:0005893', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR035976', 'id': 'P01589', 'GO_MF': ['GO:0006924', ' GO:0006915', ' GO:0007166', ' GO:0019221', ' GO:0006955', ' GO:0006954', ' GO:0002437', ' GO:0038110', ' GO:0000165', ' GO:0050687', ' GO:0050777', ' GO:0050728', ' GO:0042130', ' GO:0007219', ' GO:0042104', ' GO:0045582', ' GO:0045589', ' GO:0046013', ' GO:0002664'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['17187516', ' 17454642'], 'DRUG_ATC_CODE': ['ATC:L01XX', ' ATC:L01XX29', ' ATC:L', ' ATC:L01X', ' ATC:L01'], 'PRODUCT': 'Ontak', 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00004', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Denileukin_diftitox', 'SYNONYM': ['Denileukin', ' Interleukin_2_diptheria_toxin_fusion_protein'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR003961', ' IPR040951'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0042010', ' GO:0019976', ' GO:0004911'], 'SHORT_NAME': ['IL-2 receptor subunit beta', ' IL-2R subunit beta', ' IL-2RB', ' p75'], 'NAME': 'IL2RB', 'RELATED_KEYWORD': ['3D-structure', ' Cell membrane', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Host-virus interaction', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Signal', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Spleen', ' Lymph_node__Non-germinal_center_cells', ' Tonsil__Non-germinal_center_cells', ' Spleen__Cells_in_white_pulp', ' Tonsil', ' Spleen__Cells_in_red_pulp', ' Bone_marrow', ' Lymph_node', ' Bone_marrow__Hematopoietic_cells'], 'RELATED_PUBMED_ID': ['pubmed:2785715', ' pubmed:15461802', ' pubmed:14702039', ' pubmed:10591208', ' pubmed:15489334', ' pubmed:8648694', ' pubmed:12200137', ' pubmed:15123770', ' pubmed:7529123', ' pubmed:16293754', ' pubmed:16477002', ' pubmed:31040184', ' pubmed:31040185'], 'FULL_NAME': ['Interleukin-2 receptor subunit beta', ' High affinity IL-2 receptor subunit beta', ' Interleukin-15 receptor subunit beta {ECO:0000312|HGNC:HGNC:6009}', ' p70-75'], 'CONSERVED_SITE': 'IPR003531', 'GO_CC': ['GO:0009986', ' GO:0005829', ' GO:0005768', ' GO:0009897', ' GO:0005887', ' GO:0005893', ' GO:0016020', ' GO:0005886'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036116', ' IPR013783'], 'id': 'P14784', 'GO_MF': ['GO:0019221', ' GO:0035723', ' GO:0038110', ' GO:0000165', ' GO:0043066', ' GO:0050766', ' GO:0065003', ' GO:0007165', ' GO:0016032'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['17187516', ' 17454642'], 'DRUG_ATC_CODE': ['ATC:L01XX', ' ATC:L01XX29', ' ATC:L', ' ATC:L01X', ' ATC:L01'], 'PRODUCT': 'Ontak', 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00004', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Denileukin_diftitox', 'SYNONYM': ['Denileukin', ' Interleukin_2_diptheria_toxin_fusion_protein'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR003961', ' IPR015321'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019955', ' GO:0004896', ' GO:0042010', ' GO:0019976'], 'SHORT_NAME': ['IL-2 receptor subunit gamma', ' IL-2R subunit gamma', ' IL-2RG'], 'NAME': 'IL2RG', 'RELATED_KEYWORD': ['3D-structure', ' Alternative splicing', ' Cell membrane', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Host-virus interaction', ' Membrane', ' Phosphoprotein', ' Polymorphism', ' Receptor', ' Reference proteome', ' SCID', ' Signal', ' Transmembrane', ' Transmembrane helix'], 'PROTEIN_EXPRESSED_IN': ['Bone_marrow__Hematopoietic_cells', ' Tonsil', ' Bone_marrow', ' Tonsil__Squamous_epithelial_cells', ' Spleen__Cells_in_red_pulp', ' Spleen'], 'RELATED_PUBMED_ID': ['pubmed:1631559', ' pubmed:8514792', ' pubmed:8401490', ' pubmed:15772651', ' pubmed:15489334', ' pubmed:8266076', ' pubmed:8266078', ' pubmed:8266077', ' pubmed:8648694', ' pubmed:12200137', ' pubmed:15123770', ' pubmed:19690332', ' pubmed:7529123', ' pubmed:16293754', ' pubmed:16477002', ' pubmed:8299698', ' pubmed:8088810', ' pubmed:8027558', ' pubmed:7937790', ' pubmed:7668284', ' pubmed:7557965', ' pubmed:7860773', ' pubmed:7883965', ' pubmed:8900089', ' pubmed:9150740', ' pubmed:9049783', ' pubmed:9399950'], 'FULL_NAME': ['Cytokine receptor common subunit gamma', ' Interleukin-2 receptor subunit gamma', ' gammaC', ' p64'], 'CONSERVED_SITE': 'IPR003531', 'GO_CC': ['GO:0009986', ' GO:0005829', ' GO:0005768', ' GO:0009897', ' GO:0005887', ' GO:0016020', ' GO:0005886', ' GO:0043235'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036116', ' IPR013783'], 'id': 'P31785', 'GO_MF': ['GO:0019221', ' GO:0006955', ' GO:0035723', ' GO:0038110', ' GO:0038114', ' GO:0035771', ' GO:0038111', ' GO:0038113', ' GO:0000165', ' GO:0050766', ' GO:0007165', ' GO:0016032'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['10480573', ' 10601294', ' 10628369', ' 11408594', ' 11431346', ' 11752352', ' 15269308', ' 16336752'], 'DRUG_ATC_CODE': ['ATC:L01', ' ATC:L', ' ATC:L01XC06', ' ATC:L01XC', ' ATC:L01X'], 'PRODUCT': 'Erbitux', 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00002', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Cetuximab', 'SYNONYM': nan, 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR007110', ' IPR003599'], 'SPECIES': 'HUMAN', 'GO_BP': 'GO:0019864', 'SHORT_NAME': ['Fc-gamma RIII', ' Fc-gamma RIIIb', ' FcRIII', ' FcRIIIb'], 'NAME': 'FCG3B', 'RELATED_KEYWORD': ['3D-structure', ' Cell membrane', ' Disulfide bond', ' Glycoprotein', ' GPI-anchor', ' IgG-binding protein', ' Immunoglobulin domain', ' Lipoprotein', ' Membrane', ' Polymorphism', ' Receptor', ' Reference proteome', ' Repeat', ' Secreted', ' Signal'], 'RELATED_PUBMED_ID': ['pubmed:2526846', ' pubmed:2967436', ' pubmed:2521732', ' pubmed:2525780', ' pubmed:15245367', ' pubmed:16710414', ' pubmed:7836402', ' pubmed:10917521', ' pubmed:11021536', ' pubmed:9028335'], 'FULL_NAME': ['Low affinity immunoglobulin gamma Fc region receptor III-B', ' Fc-gamma RIII-beta', ' FcR-10', ' IgG Fc receptor III-1'], 'GO_CC': ['GO:0031225', ' GO:0070062', ' GO:0005576', ' GO:0005886', ' GO:0030667'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036179', ' IPR013783'], 'id': 'O75015', 'GO_MF': ['GO:0006955', ' GO:0043312'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['10480573', ' 10601294', ' 10628369', ' 11408594', ' 11431346', ' 11752352', ' 15269308', ' 16336752'], 'DRUG_ATC_CODE': ['ATC:L01', ' ATC:L', ' ATC:L01XC06', ' ATC:L01XC', ' ATC:L01X'], 'PRODUCT': 'Erbitux', 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00002', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Cetuximab', 'SYNONYM': nan, 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'p': {'DOMAIN': ['IPR006211', ' IPR032778', ' IPR000719', ' IPR000494', ' IPR001245', ' IPR020635'], 'REPEAT': 'IPR006212', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0051015', ' GO:0005524', ' GO:0051117', ' GO:0045296', ' GO:0005516', ' GO:0003682', ' GO:0003690', ' GO:0019899', ' GO:0048408', ' GO:0005006', ' GO:0042802', ' GO:0005178', ' GO:0004709', ' GO:0019901', ' GO:0019903', ' GO:0004713', ' GO:0004714', ' GO:0004888', ' GO:0031625', ' GO:0001618'], 'ACTIVE_SITE': 'IPR008266', 'NAME': 'EGFR', 'FAMILY': 'IPR016245', 'RELATED_KEYWORD': ['3D-structure', ' Alternative splicing', ' ATP-binding', ' Cell membrane', ' Developmental protein', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Endoplasmic reticulum', ' Endosome', ' Glycoprotein', ' Golgi apparatus', ' Host cell receptor for virus entry', ' Host-virus interaction', ' Isopeptide bond', ' Kinase', ' Lipoprotein', ' Membrane', ' Methylation', ' Nucleotide-binding', ' Nucleus', ' Palmitate', ' Phosphoprotein', ' Polymorphism', ' Proto-oncogene', ' Receptor', ' Reference proteome', ' Repeat', ' Secreted', ' Signal', ' Transferase', ' Transmembrane', ' Transmembrane helix', ' Tyrosine-protein kinase', ' Ubl conjugation'], 'PROTEIN_EXPRESSED_IN': ['Skin_1__Melanocytes', ' Tonsil', ' Vagina__Squamous_epithelial_cells', ' Skeletal_muscle', ' Urinary_bladder__Urothelial_cells', ' Urinary_bladder', ' Skin_1', ' Skeletal_muscle__Myocytes', ' Testis__Leydig_cells', ' Tonsil__Squamous_epithelial_cells', ' Liver__Hepatocytes', ' Vagina', ' Liver', ' Testis', ' Placenta__Trophoblastic_cells', ' Placenta'], 'OTHER_ID': ['O00732', ' P06268', ' Q14225', ' Q68GS5', ' Q92795', ' Q9BZS2', ' Q9GZX1', ' Q9H2C9', ' Q9H3C9', ' Q9UMD7', ' Q9UMD8', ' Q9UMG5'], 'RELATED_PUBMED_ID': ['pubmed:6328312', ' pubmed:7654368', ' pubmed:8918811', ' pubmed:9103388', ' pubmed:11161793', ' pubmed:6326261', ' pubmed:6330563', ' pubmed:6093780', ' pubmed:3329716', ' pubmed:1988448', ' pubmed:2991899', ' pubmed:6324343', ' pubmed:3138233', ' pubmed:15340161', ' pubmed:2985580', ' pubmed:16543144', ' pubmed:9556602', ' pubmed:6325948', ' pubmed:3039909', ' pubmed:2790960', ' pubmed:2543678', ' pubmed:7679104', ' pubmed:8650580', ' pubmed:9852145', ' pubmed:10523301', ' pubmed:8144591', ' pubmed:7657591', ' pubmed:8962717', ' pubmed:9419975', ' pubmed:9488479', ' pubmed:10228163', ' pubmed:10026169', ' pubmed:10731668', ' pubmed:11602604', ' pubmed:12731890', ' pubmed:12873986', ' pubmed:15282549', ' pubmed:10805725', ' pubmed:11116146', ' pubmed:11483589', ' pubmed:15590694', ' pubmed:15611079', ' pubmed:16083266', ' pubmed:16140940', ' pubmed:17081983', ' pubmed:17115032', ' pubmed:16420529', ' pubmed:17671655', ' pubmed:17182860', ' pubmed:18602463', ' pubmed:19172738', ' pubmed:18220336', ' pubmed:18691976', ' pubmed:18669648', ' pubmed:19413330', ' pubmed:19836242', ' pubmed:19509291', ' pubmed:19159218', ' pubmed:19369195', ' pubmed:19749156', ' pubmed:19718021', ' pubmed:20674546', ' pubmed:20551055', ' pubmed:20462955', ' pubmed:20068231', ' pubmed:21269460', ' pubmed:21487020', ' pubmed:21258366', ' pubmed:21516087', ' pubmed:21518868', ' pubmed:22298428', ' pubmed:22179831', ' pubmed:23418353', ' pubmed:23186163', ' pubmed:24275569', ' pubmed:25187647', ' pubmed:23589287', ' pubmed:25311788', ' pubmed:23912460', ' pubmed:27153536', ' pubmed:26988023', ' pubmed:28479384', ' pubmed:12297049', ' pubmed:12297050', ' pubmed:12620237', ' pubmed:15374980', ' pubmed:15837620', ' pubmed:15840573', ' pubmed:17349580', ' pubmed:18046415', ' pubmed:18227510', ' pubmed:19563760', ' pubmed:19560417', ' pubmed:20471394', ' pubmed:20837704', ' pubmed:22888118', ' pubmed:15623594', ' pubmed:15118125', ' pubmed:16533793', ' pubmed:16205628', ' pubmed:16672372', ' pubmed:17344846', ' pubmed:24691054'], 'BINDING_SITE': 'IPR017441', 'FULL_NAME': ['Epidermal growth factor receptor', ' Proto-oncogene c-ErbB-1', ' Receptor tyrosine-protein kinase erbB-1'], 'GO_CC': ['GO:0016324', ' GO:0009925', ' GO:0016323', ' GO:0030054', ' GO:0009986', ' GO:0030665', ' GO:0005737', ' GO:0031901', ' GO:0030139', ' GO:0005789', ' GO:0005768', ' GO:0010008', ' GO:0005615', ' GO:0005925', ' GO:0000139', ' GO:0005887', ' GO:0016020', ' GO:0045121', ' GO:0097489', ' GO:0031965', ' GO:0005634', ' GO:0048471', ' GO:0005886', ' GO:0032991', ' GO:0043235', ' GO:0070435', ' GO:0045202'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR009030', ' IPR011009', ' IPR036941'], 'id': 'P00533', 'GO_MF': ['GO:0043006', ' GO:0007202', ' GO:0048143', ' GO:0030154', ' GO:0007166', ' GO:0098609', ' GO:0071230', ' GO:0071276', ' GO:0071549', ' GO:0071364', ' GO:0071392', ' GO:0071260', ' GO:0034614', ' GO:0021795', ' GO:0007623', ' GO:0048546', ' GO:0016101', ' GO:0001892', ' GO:0007173', ' GO:0038128', ' GO:0061029', ' GO:0001942', ' GO:0042743', ' GO:0007611', ' GO:0097421', ' GO:0030324', ' GO:0010960', ' GO:0000165', ' GO:0061024', ' GO:0007494', ' GO:0060571', ' GO:0007275', ' GO:0043066', ' GO:1905208', ' GO:0042059', ' GO:1901185', ' GO:0045930', ' GO:0045746', ' GO:0042177', ' GO:0048812', ' GO:0001503', ' GO:0042698', ' GO:0038083', ' GO:0018108', ' GO:0097755', ' GO:0045780', ' GO:0090263', ' GO:0030307', ' GO:0030335', ' GO:0008284', ' GO:0045737', ' GO:0045739', ' GO:0045740', ' GO:0050679', ' GO:0070374', ' GO:0048146', ' GO:1900087', ' GO:0050729', ' GO:0033674', ' GO:0043406', ' GO:1901224', ' GO:0010750', ' GO:0033138', ' GO:0042327', ' GO:1903800', ' GO:1902722', ' GO:0051897', ' GO:1900020', ' GO:1903078', ' GO:0001934', ' GO:0048661', ' GO:0032930', ' GO:0051968', ' GO:0045944', ' GO:0045893', ' GO:0045907', ' GO:0046777', ' GO:0051205', ' GO:2000145', ' GO:0070372', ' GO:0046328', ' GO:0050999', ' GO:0050730', ' GO:0014066', ' GO:0006357', ' GO:0051592', ' GO:0033590', ' GO:0033594', ' GO:0006970', ' GO:0070141', ' GO:0007435', ' GO:0007165', ' GO:0043586', ' GO:0006412', ' GO:0007169', ' GO:0042060'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}]\u001b[0m\n",
      "Attempt 3 failed with error: Error code: 413 - {'error': {'message': 'Request too large for model `llama3-70b-8192` in organization `org_01j9yv18bffqksnp3w7k6gzqk0` on tokens per minute (TPM): Limit 6000, Requested 8215, please reduce your message size and try again. Visit https://console.groq.com/docs/rate-limits for more information.', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}\n",
      "context: I don't know the answer\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease)-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug)-[:DRUG_TARGET]-(p:Protein) \n",
      "RETURN d.NAME, dr.NAME, p.NAME\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d.NAME': 'Albuminuria', 'dr.NAME': 'Glucagon', 'p.NAME': 'GLP2R'}, {'d.NAME': ['Arrhythmias', ' Cardiac'], 'dr.NAME': 'Glucagon', 'p.NAME': 'GLP2R'}, {'d.NAME': 'Bradycardia', 'dr.NAME': 'Glucagon', 'p.NAME': 'GLP2R'}, {'d.NAME': 'Diabetic Nephropathies', 'dr.NAME': 'Glucagon', 'p.NAME': 'GLP2R'}, {'d.NAME': 'Fibrosis', 'dr.NAME': 'Glucagon', 'p.NAME': 'GLP2R'}, {'d.NAME': 'Heart Block', 'dr.NAME': 'Glucagon', 'p.NAME': 'GLP2R'}, {'d.NAME': 'Heart Diseases', 'dr.NAME': 'Glucagon', 'p.NAME': 'GLP2R'}, {'d.NAME': 'Hyperglycemia', 'dr.NAME': 'Glucagon', 'p.NAME': 'GLP2R'}, {'d.NAME': 'Hyperinsulinism', 'dr.NAME': 'Glucagon', 'p.NAME': 'GLP2R'}, {'d.NAME': 'Hypotension', 'dr.NAME': 'Glucagon', 'p.NAME': 'GLP2R'}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: Here are the important terms: Albuminuria, Arrhythmias, Cardiac, Bradycardia, Diabetic Nephropathies, Fibrosis, Heart Block, Heart Diseases, Hyperglycemia, Hyperinsulinism, Hypotension, Glucagon, GLP2R.\n",
      "You:  give pubmed papers for drugs for tuberculosis\n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease {NAME: \"Tuberculosis\"})-[:DRUG_DISEASE_ASSOCIATION]-(drug:Drug) RETURN drug.PUBMED_ARTICLE;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'drug.PUBMED_ARTICLE': ['17224143', ' 16157873', ' 3541289']}, {'drug.PUBMED_ARTICLE': nan}, {'drug.PUBMED_ARTICLE': nan}, {'drug.PUBMED_ARTICLE': nan}, {'drug.PUBMED_ARTICLE': ['6109855', ' 6356538', ' 6386028', ' 12569078', ' 11914348', ' 17101678', ' 17485499', ' 21835980']}, {'drug.PUBMED_ARTICLE': nan}, {'drug.PUBMED_ARTICLE': ['30285357', ' 15634032', ' 23213332', ' 29411351', ' 27279294', ' 24347992', ' 29568666', ' 29273683', ' 31808984', ' 5442584', ' 31643719']}, {'drug.PUBMED_ARTICLE': ['10319086', ' 29257114']}, {'drug.PUBMED_ARTICLE': ['30285357', ' 15634032', ' 2199128', ' 25800201', ' 7310640', ' 378499', ' 23300763']}, {'drug.PUBMED_ARTICLE': nan}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: 17224143, 16157873, 3541289, 6109855, 6356538, 6386028, 12569078, 11914348, 17101678, 17485499, 21835980, 30285357, 15634032, 23213332, 29411351, 27279294, 24347992, 29568666, 29273683, 31808984, 5442584, 31643719, 10319086, 29257114, 2199128, 25800201, 7310640, 378499, 23300763.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease)<-[:DRUG_DISEASE_ASSOCIATION]-(drug:Drug)<-[:DRUG_TARGET]-(p:Protein) \n",
      "WHERE drug.PUBMED_ARTICLE IN [\"17224143\", \"16157873\", \"3541289\", \"6109855\", \"6356538\", \"6386028\", \"12569078\", \"11914348\", \"17101678\", \"17485499\", \"21835980\", \"30285357\", \"15634032\", \"23213332\", \"29411351\", \"27279294\", \"24347992\", \"29568666\", \"29273683\", \"31808984\", \"5442584\", \"31643719\", \"10319086\", \"29257114\", \"2199128\", \"25800201\", \"7310640\", \"378499\", \"23300763\"]\n",
      "RETURN p.NAME, d.NAME, drug.NAME;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Response deemed fallback on attempt 1: 'I don't know the answer.'\n",
      "Attempt 2 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease)<-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug)-[:PUBMED_ARTICLE]->(p) \n",
      "WHERE p = 17224143 OR p = 16157873 OR p = 3541289 OR p = 6109855 OR p = 6356538 OR p = 6386028 OR p = 12569078 OR p = 11914348 OR p = 17101678 OR p = 17485499 OR p = 21835980 OR p = 30285357 OR p = 15634032 OR p = 23213332 OR p = 29411351 OR p = 27279294 OR p = 24347992 OR p = 29568666 OR p = 29273683 OR p = 31808984 OR p = 5442584 OR p = 31643719 OR p = 10319086 OR p = 29257114 OR p = 2199128 OR p = 25800201 OR p = 7310640 OR p = 378499 OR p = 23300763 \n",
      "RETURN d, dr;\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Received notification from DBMS server: {severity: WARNING} {code: Neo.ClientNotification.Statement.UnknownRelationshipTypeWarning} {category: UNRECOGNIZED} {title: The provided relationship type is not in the database.} {description: One of the relationship types in your query is not available in the database, make sure you didn't misspell it or that the label is available when you run this statement in your application (the missing relationship type is: PUBMED_ARTICLE)} {position: line: 1, column: 60, offset: 59} for query: 'MATCH (d:Disease)<-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug)-[:PUBMED_ARTICLE]->(p) \\nWHERE p = 17224143 OR p = 16157873 OR p = 3541289 OR p = 6109855 OR p = 6356538 OR p = 6386028 OR p = 12569078 OR p = 11914348 OR p = 17101678 OR p = 17485499 OR p = 21835980 OR p = 30285357 OR p = 15634032 OR p = 23213332 OR p = 29411351 OR p = 27279294 OR p = 24347992 OR p = 29568666 OR p = 29273683 OR p = 31808984 OR p = 5442584 OR p = 31643719 OR p = 10319086 OR p = 29257114 OR p = 2199128 OR p = 25800201 OR p = 7310640 OR p = 378499 OR p = 23300763 \\nRETURN d, dr;'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Response deemed fallback on attempt 2: 'I don't know the answer.'\n",
      "Attempt 3 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease) WHERE d.id IN [\"17224143\", \"16157873\", \"3541289\", \"6109855\", \"6356538\", \"6386028\", \"12569078\", \"11914348\", \"17101678\", \"17485499\", \"21835980\", \"30285357\", \"15634032\", \"23213332\", \"29411351\", \"27279294\", \"24347992\", \"29568666\", \"29273683\", \"31808984\", \"5442584\", \"31643719\", \"10319086\", \"29257114\", \"2199128\", \"25800201\", \"7310640\", \"378499\", \"23300763\"] RETURN d;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Response deemed fallback on attempt 3: 'I don't know the answer.'\n",
      "context: No relevant context available after three attempts.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease)<-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug) \n",
      "RETURN d.NAME, dr.NAME, dr.TYPE, dr.DRUG_INDICATION_ASSOCIATION, dr.LOGP, dr.MolecularWeight, dr.TPSA, dr.NumHDonors, dr.NumHAcceptors, dr.LipinskiViolationCount, dr.SYNONYM, dr.CLASS, dr.SUBCLASS, dr.SUPERCLASS\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Received notification from DBMS server: {severity: WARNING} {code: Neo.ClientNotification.Statement.UnknownPropertyKeyWarning} {category: UNRECOGNIZED} {title: The provided property key is not in the database} {description: One of the property names in your query is not available in the database, make sure you didn't misspell it or that the label is available when you run this statement in your application (the missing property name is: LOGP)} {position: line: 2, column: 69, offset: 126} for query: 'MATCH (d:Disease)<-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug) \\nRETURN d.NAME, dr.NAME, dr.TYPE, dr.DRUG_INDICATION_ASSOCIATION, dr.LOGP, dr.MolecularWeight, dr.TPSA, dr.NumHDonors, dr.NumHAcceptors, dr.LipinskiViolationCount, dr.SYNONYM, dr.CLASS, dr.SUBCLASS, dr.SUPERCLASS'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d.NAME': 'COVID-19', 'dr.NAME': 'Choline', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.LOGP': None, 'dr.MolecularWeight': 104.17299999999999, 'dr.TPSA': 20.23, 'dr.NumHDonors': 1, 'dr.NumHAcceptors': 1, 'dr.LipinskiViolationCount': 0, 'dr.SYNONYM': ['Bilineurine', ' Choline_ion', ' N_trimethylethanolamine', ' N_N_N_trimethylethanol_ammonium', ' trimethylethanolamine', ' Vitamin_J'], 'dr.CLASS': 'Organonitrogen_compounds', 'dr.SUBCLASS': 'Quaternary_ammonium_salts', 'dr.SUPERCLASS': 'Organic_nitrogen_compounds'}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Azithromycin', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0032285', ' C0023895', ' C0600260', ' C0702166', ' C3495832', ' C0031350', ' C0151281', ' C0271429', ' C0008677', ' C0021051', ' C0026918', ' C0085605', ' C0035243', ' C0024117', ' C0339959', ' C1328428', ' C0035436', ' C0009450', ' C0040425', ' C0000833', ' C0242172', ' C0231218', ' C0497209', ' C0010674', ' C0687133', ' C0009763', ' C0264344', ' C0015663', ' C0043167', ' C1262006', ' C0006277', ' C1412002', ' C0600031', ' C0149512', ' C0036689', ' C0205721', ' C1565489', ' C0004610', ' C0009768', ' C0032302', ' C0035078', ' C0007947', ' C0041976', ' C0041912', ' C0085129', ' C0149725', ' C0086565', ' C0024198', ' C0007860', ' C0039128', ' C0853847', ' C0036916', ' C0040592', ' C1328337', ' C0852859', ' C0018081', ' C0026916'], 'dr.LOGP': None, 'dr.MolecularWeight': 748.9960000000001, 'dr.TPSA': 180.08, 'dr.NumHDonors': 5, 'dr.NumHAcceptors': 14, 'dr.LipinskiViolationCount': 2, 'dr.SYNONYM': 'Azithromycin', 'dr.CLASS': 'Organooxygen_compounds', 'dr.SUBCLASS': 'Carbohydrates_and_carbohydrate_conjugates', 'dr.SUPERCLASS': 'Organic_oxygen_compounds'}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Tenofovir_disoproxil', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0524909', ' C0009450', ' C0023895', ' C0392390', ' C0239571', ' C0151763'], 'dr.LOGP': None, 'dr.MolecularWeight': 519.4480000000002, 'dr.TPSA': 185.43999999999997, 'dr.NumHDonors': 1, 'dr.NumHAcceptors': 15, 'dr.LipinskiViolationCount': 2, 'dr.SYNONYM': ['Bis_POC_PMPA', ' Tenofovir_bis_isopropyloxycarbonyloxymethyl__ester'], 'dr.CLASS': 'Imidazopyrimidines', 'dr.SUBCLASS': 'Purines_and_purine_derivatives', 'dr.SUPERCLASS': 'Organoheterocyclic_compounds'}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Mefloquine', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0858320', ' C0858321', ' C0024530', ' C0858319', ' C0009450', ' C0858318', ' C0853847'], 'dr.LOGP': None, 'dr.MolecularWeight': 378.31600000000014, 'dr.TPSA': 45.150000000000006, 'dr.NumHDonors': 2, 'dr.NumHAcceptors': 3, 'dr.LipinskiViolationCount': 0, 'dr.SYNONYM': ['__R__S___2_8_bis_trifluoromethyl_quinolin_4_yl___2_piperidyl_methanol', ' alpha_2_Piperidinyl_2_8_bis_trifluoromethyl__4_quinolinemethanol', ' Mefloquine'], 'dr.CLASS': 'Quinolines_and_derivatives', 'dr.SUBCLASS': '4_quinolinemethanols', 'dr.SUPERCLASS': 'Organoheterocyclic_compounds'}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Ritonavir', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0019069', ' C0004364', ' C0011849', ' C0001175'], 'dr.LOGP': None, 'dr.MolecularWeight': 720.9620000000001, 'dr.TPSA': 145.77999999999997, 'dr.NumHDonors': 4, 'dr.NumHAcceptors': 9, 'dr.LipinskiViolationCount': 2, 'dr.SYNONYM': nan, 'dr.CLASS': 'Carboxylic_acids_and_derivatives', 'dr.SUBCLASS': 'Amino_acids__peptides__and_analogues', 'dr.SUPERCLASS': 'Organic_acids_and_derivatives'}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Acetazolamide', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0017605', ' C0018802', ' C0238284', ' C0017601', ' C0014553', ' C0013604', ' C0036572', ' C0014544'], 'dr.LOGP': None, 'dr.MolecularWeight': 222.251, 'dr.TPSA': 115.03999999999999, 'dr.NumHDonors': 2, 'dr.NumHAcceptors': 6, 'dr.LipinskiViolationCount': 0, 'dr.SYNONYM': ['2_acetylamino_1_3_4_thiadiazole_5_sulfonamide', ' 5_acetamido_1_3_4_thiadiazole_2_sulfonamide', ' 5_acetylamino_1_3_4_thiadiazole_2_sulfonamide', ' Acetazolamide', ' N__5__aminosulfonyl__1_3_4_thiadiazol_2_yl_acetamide', ' N__5__aminosulfonyl__1_3_5_thiadiazol_2_yl_acetamide'], 'dr.CLASS': 'Azoles', 'dr.SUBCLASS': 'Thiadiazoles', 'dr.SUPERCLASS': 'Organoheterocyclic_compounds'}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Ribavirin', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0019189', ' C0014412', ' C0019158', ' C0021368', ' C0009450', ' C0023895', ' C0013378', ' C0085129', ' C0035243', ' C0043096', ' C0311468', ' C0023890', ' C0524910', ' C0151905', ' C0021400', ' C0035235', ' C0239571', ' C0152021', ' C0019196', ' C0085605', ' C0034072', ' C0151763', ' C0024117', ' C0242656', ' C0042769', ' C0149725', ' C0016059'], 'dr.LOGP': None, 'dr.MolecularWeight': 244.20700000000002, 'dr.TPSA': 143.72000000000003, 'dr.NumHDonors': 4, 'dr.NumHAcceptors': 8, 'dr.LipinskiViolationCount': 0, 'dr.SYNONYM': ['1_beta_D_Ribofuranosyl_1_2_4_triazole_3_carboxamide', ' 1_beta_D_Ribofuranosyl_1H_1_2_4_triazole_3_carboxamide', ' RBV', ' Tribavirin'], 'dr.CLASS': 'Triazole_ribonucleosides_and_ribonucleotides', 'dr.SUBCLASS': nan, 'dr.SUPERCLASS': 'Nucleosides__nucleotides__and_analogues'}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Ivermectin', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0038463', ' C0016085', ' C0028259', ' C0009450', ' C0029001', ' C0013884', ' C0042961', ' C0281981', ' C0033774', ' C1504343'], 'dr.LOGP': None, 'dr.MolecularWeight': 1736.1849999999997, 'dr.TPSA': 340.12, 'dr.NumHDonors': 6, 'dr.NumHAcceptors': 28, 'dr.LipinskiViolationCount': 4, 'dr.SYNONYM': 'Ivermectin', 'dr.CLASS': 'Macrolides_and_analogues', 'dr.SUBCLASS': 'Milbemycins', 'dr.SUPERCLASS': 'Phenylpropanoids_and_polyketides'}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Toremifene', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0027651', ' C0278488', ' C0153690'], 'dr.LOGP': None, 'dr.MolecularWeight': 405.969, 'dr.TPSA': 12.47, 'dr.NumHDonors': 0, 'dr.NumHAcceptors': 2, 'dr.LipinskiViolationCount': 1, 'dr.SYNONYM': 'Toremifene', 'dr.CLASS': 'Stilbenes', 'dr.SUBCLASS': nan, 'dr.SUPERCLASS': 'Phenylpropanoids_and_polyketides'}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Chloroquine', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0858318', ' C0002438', ' C0858319', ' C0009450', ' C0858320', ' C0858321'], 'dr.LOGP': None, 'dr.MolecularWeight': 319.88000000000005, 'dr.TPSA': 28.16, 'dr.NumHDonors': 1, 'dr.NumHAcceptors': 3, 'dr.LipinskiViolationCount': 0, 'dr.SYNONYM': ['Chloraquine', ' Chloroquina', ' Chloroquinium', ' N4__7_chloro_4_quinolinyl__N1_N1_diethyl_1_4_pentanediamine'], 'dr.CLASS': 'Quinolines_and_derivatives', 'dr.SUBCLASS': 'Aminoquinolines_and_derivatives', 'dr.SUPERCLASS': 'Organoheterocyclic_compounds'}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: Choline, Azithromycin, Tenofovir_disoproxil, Mefloquine, Ritonavir, Acetazolamide, Ribavirin, Ivermectin, Toremifene, Chloroquine.\n",
      "You:  Give me the exit\n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mThere is no question to generate a Cypher statement. Please provide a specific query request, and I'll be happy to help.\u001b[0m\n",
      "Attempt 1 failed with error: {code: Neo.ClientError.Statement.SyntaxError} {message: Invalid input 'There': expected 'FOREACH', 'ALTER', 'ORDER BY', 'CALL', 'USING PERIODIC COMMIT', 'CREATE', 'LOAD CSV', 'START DATABASE', 'STOP DATABASE', 'DEALLOCATE', 'DELETE', 'DENY', 'DETACH', 'DROP', 'DRYRUN', 'FINISH', 'GRANT', 'INSERT', 'LIMIT', 'MATCH', 'MERGE', 'NODETACH', 'OFFSET', 'OPTIONAL', 'REALLOCATE', 'REMOVE', 'RENAME', 'RETURN', 'REVOKE', 'ENABLE SERVER', 'SET', 'SHOW', 'SKIP', 'TERMINATE', 'UNWIND', 'USE' or 'WITH' (line 1, column 1 (offset: 0))\n",
      "\"There is no question to generate a Cypher statement. Please provide a specific query request, and I'll be happy to help.\"\n",
      " ^}\n",
      "Attempt 2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mNone\u001b[0m\n",
      "Attempt 2 failed with error: {code: Neo.ClientError.Statement.SyntaxError} {message: Invalid input 'None': expected 'FOREACH', 'ALTER', 'ORDER BY', 'CALL', 'USING PERIODIC COMMIT', 'CREATE', 'LOAD CSV', 'START DATABASE', 'STOP DATABASE', 'DEALLOCATE', 'DELETE', 'DENY', 'DETACH', 'DROP', 'DRYRUN', 'FINISH', 'GRANT', 'INSERT', 'LIMIT', 'MATCH', 'MERGE', 'NODETACH', 'OFFSET', 'OPTIONAL', 'REALLOCATE', 'REMOVE', 'RENAME', 'RETURN', 'REVOKE', 'ENABLE SERVER', 'SET', 'SHOW', 'SKIP', 'TERMINATE', 'UNWIND', 'USE' or 'WITH' (line 1, column 1 (offset: 0))\n",
      "\"None\"\n",
      " ^}\n",
      "Attempt 3...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mNo Cypher statement can be generated based on the provided schema and the question \"Give me the exit\". The question does not make sense in the context of querying a graph database.\n",
      "\n",
      "If you meant to ask for a Cypher statement to exit or terminate a query, there is no such statement in Cypher. Cypher is used to query and manipulate graph data, not to exit or terminate a query.\n",
      "\n",
      "If you have a specific question or query in mind, please rephrase it, and I'll be happy to help you with a Cypher statement.\u001b[0m\n",
      "Attempt 3 failed with error: {code: Neo.ClientError.Statement.SyntaxError} {message: Invalid input 'No': expected 'FOREACH', 'ALTER', 'ORDER BY', 'CALL', 'USING PERIODIC COMMIT', 'CREATE', 'LOAD CSV', 'START DATABASE', 'STOP DATABASE', 'DEALLOCATE', 'DELETE', 'DENY', 'DETACH', 'DROP', 'DRYRUN', 'FINISH', 'GRANT', 'INSERT', 'LIMIT', 'MATCH', 'MERGE', 'NODETACH', 'OFFSET', 'OPTIONAL', 'REALLOCATE', 'REMOVE', 'RENAME', 'RETURN', 'REVOKE', 'ENABLE SERVER', 'SET', 'SHOW', 'SKIP', 'TERMINATE', 'UNWIND', 'USE' or 'WITH' (line 1, column 1 (offset: 0))\n",
      "\"No Cypher statement can be generated based on the provided schema and the question \"Give me the exit\". The question does not make sense in the context of querying a graph database.\"\n",
      " ^}\n",
      "context: I don't know the answer\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease)-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug) \n",
      "RETURN d.NAME, dr.NAME, dr.DRUG_INDICATION_ASSOCIATION, dr.DRUGINCHKE, dr.DRUGINCH, dr.DRUG_ATC_CODE, dr.MolecularWeight, dr.TPSA, dr.NumHDonors, dr.NumHAcceptors, dr.LipinskiViolationCount \n",
      "ORDER BY d.DISEASE_SUPERGRP DESC;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d.NAME': 'COVID-19', 'dr.NAME': 'Choline', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.DRUGINCHKE': 'OEYIOHPDSNJKLS-UHFFFAOYSA-N', 'dr.DRUGINCH': ['InChI=1S/C5H14NO/c1-6(2', '3)4-5-7/h7H', '4-5H2', '1-3H3/q+1'], 'dr.DRUG_ATC_CODE': nan, 'dr.MolecularWeight': 104.17299999999999, 'dr.TPSA': 20.23, 'dr.NumHDonors': 1, 'dr.NumHAcceptors': 1, 'dr.LipinskiViolationCount': 0}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Azithromycin', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0032285', ' C0023895', ' C0600260', ' C0702166', ' C3495832', ' C0031350', ' C0151281', ' C0271429', ' C0008677', ' C0021051', ' C0026918', ' C0085605', ' C0035243', ' C0024117', ' C0339959', ' C1328428', ' C0035436', ' C0009450', ' C0040425', ' C0000833', ' C0242172', ' C0231218', ' C0497209', ' C0010674', ' C0687133', ' C0009763', ' C0264344', ' C0015663', ' C0043167', ' C1262006', ' C0006277', ' C1412002', ' C0600031', ' C0149512', ' C0036689', ' C0205721', ' C1565489', ' C0004610', ' C0009768', ' C0032302', ' C0035078', ' C0007947', ' C0041976', ' C0041912', ' C0085129', ' C0149725', ' C0086565', ' C0024198', ' C0007860', ' C0039128', ' C0853847', ' C0036916', ' C0040592', ' C1328337', ' C0852859', ' C0018081', ' C0026916'], 'dr.DRUGINCHKE': 'MQTOSJVFKKJCRP-BICOPXKESA-N', 'dr.DRUGINCH': ['InChI=1S/C38H72N2O12/c1-15-27-38(10', '46)31(42)24(6)40(13)19-20(2)17-36(8', '45)33(52-35-29(41)26(39(11)12)16-21(3)48-35)22(4)30(23(5)34(44)50-27)51-28-18-37(9', '47-14)32(43)25(7)49-28/h20-33', '35', '41-43', '45-46H', '15-19H2', '1-14H3/t20-', '21-', '22+', '23-', '24-', '25+', '26+', '27-', '28+', '29-', '30+', '31-', '32+', '33-', '35+', '36-', '37-', '38-/m1/s1'], 'dr.DRUG_ATC_CODE': ['ATC:J01', ' ATC:J01FA10', ' ATC:S01AA26', ' ATC:J01R', ' ATC:S', ' ATC:J', ' ATC:S01', ' ATC:J01RA', ' ATC:J01RA07', ' ATC:J01FA', ' ATC:J01F', ' ATC:S01AA', ' ATC:S01A'], 'dr.MolecularWeight': 748.9960000000001, 'dr.TPSA': 180.08, 'dr.NumHDonors': 5, 'dr.NumHAcceptors': 14, 'dr.LipinskiViolationCount': 2}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Tenofovir_disoproxil', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0524909', ' C0009450', ' C0023895', ' C0392390', ' C0239571', ' C0151763'], 'dr.DRUGINCHKE': 'JFVZFKDSXNQEJW-CQSZACIVSA-N', 'dr.DRUGINCH': ['InChI=1S/C19H30N5O10P/c1-12(2)33-18(25)28-9-31-35(27', '32-10-29-19(26)34-13(3)4)11-30-14(5)6-24-8-23-15-16(20)21-7-22-17(15)24/h7-8', '12-14H', '6', '9-11H2', '1-5H3', '(H2', '20', '21', '22)/t14-/m1/s1'], 'dr.DRUG_ATC_CODE': ['ATC:J05AR11', ' ATC:J05AR24', ' ATC:J05AF', ' ATC:J05AR08', ' ATC:J05', ' ATC:J05AF07', ' ATC:J05AR12', ' ATC:J05A', ' ATC:J05AR', ' ATC:J05AR09', ' ATC:J', ' ATC:J05AR06', ' ATC:J05AR03'], 'dr.MolecularWeight': 519.4480000000002, 'dr.TPSA': 185.43999999999997, 'dr.NumHDonors': 1, 'dr.NumHAcceptors': 15, 'dr.LipinskiViolationCount': 2}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Mefloquine', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0858320', ' C0858321', ' C0024530', ' C0858319', ' C0009450', ' C0858318', ' C0853847'], 'dr.DRUGINCHKE': 'XEEQGYMUWCZPDN-UHFFFAOYSA-N', 'dr.DRUGINCH': ['InChI=1S/C17H16F6N2O/c18-16(19', '20)11-5-3-4-9-10(15(26)12-6-1-2-7-24-12)8-13(17(21', '22)23)25-14(9)11/h3-5', '8', '12', '15', '24', '26H', '1-2', '6-7H2'], 'dr.DRUG_ATC_CODE': ['ATC:P01BC02', ' ATC:P01BC', ' ATC:P', ' ATC:P01BF', ' ATC:P01B', ' ATC:P01', ' ATC:P01BF02'], 'dr.MolecularWeight': 378.31600000000014, 'dr.TPSA': 45.150000000000006, 'dr.NumHDonors': 2, 'dr.NumHAcceptors': 3, 'dr.LipinskiViolationCount': 0}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Ritonavir', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0019069', ' C0004364', ' C0011849', ' C0001175'], 'dr.DRUGINCHKE': 'NCDNCNXCDXHOMX-XGKFQTDJSA-N', 'dr.DRUGINCH': ['InChI=1S/C37H48N6O5S2/c1-24(2)33(42-36(46)43(5)20-29-22-49-35(40-29)25(3)4)34(45)39-28(16-26-12-8-6-9-13-26)18-32(44)31(17-27-14-10-7-11-15-27)41-37(47)48-21-30-19-38-23-50-30/h6-15', '19', '22-25', '28', '31-33', '44H', '16-18', '20-21H2', '1-5H3', '(H', '39', '45)(H', '41', '47)(H', '42', '46)/t28-', '31-', '32-', '33-/m0/s1'], 'dr.DRUG_ATC_CODE': ['ATC:J05AR23', ' ATC:J05AE03', ' ATC:J', ' ATC:J05A', ' ATC:J05AE', ' ATC:J05AP52', ' ATC:J05AR10', ' ATC:J05AR', ' ATC:J05', ' ATC:J05AP53', ' ATC:J05AP'], 'dr.MolecularWeight': 720.9620000000001, 'dr.TPSA': 145.77999999999997, 'dr.NumHDonors': 4, 'dr.NumHAcceptors': 9, 'dr.LipinskiViolationCount': 2}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Acetazolamide', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0017605', ' C0018802', ' C0238284', ' C0017601', ' C0014553', ' C0013604', ' C0036572', ' C0014544'], 'dr.DRUGINCHKE': 'BZKPWHYZMXOIDC-UHFFFAOYSA-N', 'dr.DRUGINCH': ['InChI=1S/C4H6N4O3S2/c1-2(9)6-3-7-8-4(12-3)13(5', '10)11/h1H3', '(H2', '5', '10', '11)(H', '6', '7', '9)'], 'dr.DRUG_ATC_CODE': ['ATC:G01', ' ATC:S01EC', ' ATC:G01A', ' ATC:S01EC01', ' ATC:S01E', ' ATC:G01AE', ' ATC:G', ' ATC:G01AE10', ' ATC:S01', ' ATC:S'], 'dr.MolecularWeight': 222.251, 'dr.TPSA': 115.03999999999999, 'dr.NumHDonors': 2, 'dr.NumHAcceptors': 6, 'dr.LipinskiViolationCount': 0}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Ribavirin', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0019189', ' C0014412', ' C0019158', ' C0021368', ' C0009450', ' C0023895', ' C0013378', ' C0085129', ' C0035243', ' C0043096', ' C0311468', ' C0023890', ' C0524910', ' C0151905', ' C0021400', ' C0035235', ' C0239571', ' C0152021', ' C0019196', ' C0085605', ' C0034072', ' C0151763', ' C0024117', ' C0242656', ' C0042769', ' C0149725', ' C0016059'], 'dr.DRUGINCHKE': 'IWUCXVSUMQZMFG-AFCXAGJDSA-N', 'dr.DRUGINCH': ['InChI=1S/C8H12N4O5/c9-6(16)7-10-2-12(11-7)8-5(15)4(14)3(1-13)17-8/h2-5', '8', '13-15H', '1H2', '(H2', '9', '16)/t3-', '4-', '5-', '8-/m1/s1'], 'dr.DRUG_ATC_CODE': ['ATC:J', ' ATC:J05', ' ATC:J05AP01', ' ATC:J05A', ' ATC:J05AP'], 'dr.MolecularWeight': 244.20700000000002, 'dr.TPSA': 143.72000000000003, 'dr.NumHDonors': 4, 'dr.NumHAcceptors': 8, 'dr.LipinskiViolationCount': 0}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Ivermectin', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0038463', ' C0016085', ' C0028259', ' C0009450', ' C0029001', ' C0013884', ' C0042961', ' C0281981', ' C0033774', ' C1504343'], 'dr.DRUGINCHKE': 'SPBDXSGPUHCETR-JFUDTMANSA-N', 'dr.DRUGINCH': ['InChI=1S/C48H74O14.C47H72O14/c1-11-25(2)43-28(5)17-18-47(62-43)23-34-20-33(61-47)16-15-27(4)42(26(3)13-12-14-32-24-55-45-40(49)29(6)19-35(46(51)58-34)48(32', '45)52)59-39-22-37(54-10)44(31(8)57-39)60-38-21-36(53-9)41(50)30(7)56-38;1-24(2)41-27(5)16-17-46(61-41)22-33-19-32(60-46)15-14-26(4)42(25(3)12-11-13-31-23-54-44-39(48)28(6)18-34(45(50)57-33)47(31', '44)51)58-38-21-36(53-10)43(30(8)56-38)59-37-20-35(52-9)40(49)29(7)55-37/h12-15', '19', '25-26', '28', '30-31', '33-45', '49-50', '52H', '11', '16-18', '20-24H2', '1-10H3;11-14', '18', '24-25', '27', '29-30', '32-44', '48-49', '51H', '15-17', '19-23H2', '1-10H3/b13-12+', '27-15+', '32-14+;12-11+', '26-14+', '31-13+/t25-', '26-', '28-', '30-', '31-', '33+', '34-', '35-', '36-', '37-', '38-', '39-', '40+', '41-', '42-', '43+', '44-', '45+', '47+', '48+;25-', '27-', '29-', '30-', '32+', '33-', '34-', '35-', '36-', '37-', '38-', '39+', '40-', '41+', '42-', '43-', '44+', '46+', '47+/m00/s1'], 'dr.DRUG_ATC_CODE': ['ATC:D', ' ATC:P02', ' ATC:P02CF01', ' ATC:D11', ' ATC:P', ' ATC:P02CF', ' ATC:D11AX22', ' ATC:P02C', ' ATC:D11A', ' ATC:D11AX'], 'dr.MolecularWeight': 1736.1849999999997, 'dr.TPSA': 340.12, 'dr.NumHDonors': 6, 'dr.NumHAcceptors': 28, 'dr.LipinskiViolationCount': 4}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Toremifene', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0027651', ' C0278488', ' C0153690'], 'dr.DRUGINCHKE': 'XFCLJVABOIYOMF-QPLCGJKRSA-N', 'dr.DRUGINCH': ['InChI=1S/C26H28ClNO/c1-28(2)19-20-29-24-15-13-23(14-16-24)26(22-11-7-4-8-12-22)25(17-18-27)21-9-5-3-6-10-21/h3-16H', '17-20H2', '1-2H3/b26-25-'], 'dr.DRUG_ATC_CODE': ['ATC:L02', ' ATC:L', ' ATC:L02BA', ' ATC:L02BA02', ' ATC:L02B'], 'dr.MolecularWeight': 405.969, 'dr.TPSA': 12.47, 'dr.NumHDonors': 0, 'dr.NumHAcceptors': 2, 'dr.LipinskiViolationCount': 1}, {'d.NAME': 'COVID-19', 'dr.NAME': 'Chloroquine', 'dr.DRUG_INDICATION_ASSOCIATION': ['C0858318', ' C0002438', ' C0858319', ' C0009450', ' C0858320', ' C0858321'], 'dr.DRUGINCHKE': 'WHTVZRBIWZFKQO-UHFFFAOYSA-N', 'dr.DRUGINCH': ['InChI=1S/C18H26ClN3/c1-4-22(5-2)12-6-7-14(3)21-17-10-11-20-18-13-15(19)8-9-16(17)18/h8-11', '13-14H', '4-7', '12H2', '1-3H3', '(H', '20', '21)'], 'dr.DRUG_ATC_CODE': ['ATC:P01BA01', ' ATC:P01B', ' ATC:P01', ' ATC:P', ' ATC:P01BA'], 'dr.MolecularWeight': 319.88000000000005, 'dr.TPSA': 28.16, 'dr.NumHDonors': 1, 'dr.NumHAcceptors': 3, 'dr.LipinskiViolationCount': 0}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Response deemed fallback on attempt 1: 'I don't know the answer.'\n",
      "Attempt 2 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease)-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug) \n",
      "RETURN d.NAME, dr.NAME ORDER BY d.NAME;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d.NAME': ['ACTH Deficiency', ' Isolated'], 'dr.NAME': 'Hydrocortisone'}, {'d.NAME': ['ACTH Deficiency', ' Isolated'], 'dr.NAME': 'Dexamethasone'}, {'d.NAME': ['Abdomen', ' Acute'], 'dr.NAME': 'Enalapril'}, {'d.NAME': ['Abdomen', ' Acute'], 'dr.NAME': 'Isoniazid'}, {'d.NAME': ['Abdomen', ' Acute'], 'dr.NAME': 'Cocaine'}, {'d.NAME': ['Abdomen', ' Acute'], 'dr.NAME': 'Metoclopramide'}, {'d.NAME': ['Abnormalities', ' Drug-Induced'], 'dr.NAME': 'Vitamin_A'}, {'d.NAME': ['Abnormalities', ' Drug-Induced'], 'dr.NAME': 'Folic_acid'}, {'d.NAME': ['Abnormalities', ' Drug-Induced'], 'dr.NAME': 'Nicotine'}, {'d.NAME': ['Abnormalities', ' Drug-Induced'], 'dr.NAME': 'Caffeine'}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: Here are the important terms: ACTH Deficiency, Isolated, Abdomen, Acute, Abnormalities, Drug-Induced, Hydrocortisone, Dexamethasone, Enalapril, Isoniazid, Cocaine, Metoclopramide, Vitamin_A, Folic_acid, Nicotine, Caffeine.\n",
      "You:  exit\n",
      "Goodbye!\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Initialize the chatbot with the chain\n",
    "chatbot = ContinuousChatbot(chain)\n",
    "\n",
    "# Start the chat\n",
    "chatbot.chat()\n",
    " # THIS IS MULTIHOP QUERY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "983b2a43",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to the Medical Knowledge Chatbot! \n",
      "You:  Give me the Pubmed Articles for Drugs fro Covid 19\n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease {NAME: \"Covid 19\"})-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug) RETURN dr.PUBMED_ARTICLE;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context_data_list [[]]\n",
      "context: []\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mRETURN DISTINCT d.NAME, p.NAME\n",
      "FROM (:Drug)-[:DRUG_DISEASE_ASSOCIATION]->(:Disease)<-[:PROTEIN_DISEASE_ASSOCIATION]-(:Protein) \n",
      "MATCH (d:Drug)-[:DRUG_TARGET|DRUG_ENZYME|DRUG_CARRIER|DRUG_TRANSPORTER]->(p:Protein) \n",
      "RETURN d, p;\u001b[0m\n",
      "Attempt 1 failed with error: {code: Neo.ClientError.Statement.SyntaxError} {message: Invalid input 'FROM': expected an expression, 'FOREACH', ',', 'AS', 'ORDER BY', 'CALL', 'CREATE', 'LOAD CSV', 'DELETE', 'DETACH', 'FINISH', 'INSERT', 'LIMIT', 'MATCH', 'MERGE', 'NODETACH', 'OFFSET', 'OPTIONAL', 'REMOVE', 'RETURN', 'SET', 'SKIP', 'UNION', 'UNWIND', 'USE', 'WITH' or <EOF> (line 2, column 1 (offset: 31))\n",
      "\"FROM (:Drug)-[:DRUG_DISEASE_ASSOCIATION]->(:Disease)<-[:PROTEIN_DISEASE_ASSOCIATION]-(:Protein)\"\n",
      " ^}\n",
      "Attempt 2 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (n) RETURN n.NAME, n.TYPE, n.DRUG_INDICATION_ASSOCIATION, n.DRUG_SIDEEFFECT_ASSOCIATION, n.LOGP, n.TPSA, n.NumHDonors, n.NumHAcceptors, n.LipinskiViolationCount, n.DRUGBANK_NAME, n.MolecularWeight, n.SYNONYM, n.PRODUCT, n.KINGDOM, n.SOURCE, n.DIRECT_PARENT, n.ALTERNATIVE_PARENT, n.C_SMILES\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Received notification from DBMS server: {severity: WARNING} {code: Neo.ClientNotification.Statement.UnknownPropertyKeyWarning} {category: UNRECOGNIZED} {title: The provided property key is not in the database} {description: One of the property names in your query is not available in the database, make sure you didn't misspell it or that the label is available when you run this statement in your application (the missing property name is: LOGP)} {position: line: 1, column: 98, offset: 97} for query: 'MATCH (n) RETURN n.NAME, n.TYPE, n.DRUG_INDICATION_ASSOCIATION, n.DRUG_SIDEEFFECT_ASSOCIATION, n.LOGP, n.TPSA, n.NumHDonors, n.NumHAcceptors, n.LipinskiViolationCount, n.DRUGBANK_NAME, n.MolecularWeight, n.SYNONYM, n.PRODUCT, n.KINGDOM, n.SOURCE, n.DIRECT_PARENT, n.ALTERNATIVE_PARENT, n.C_SMILES'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'n.NAME': ['Ehlers-Danlos Syndrome', ' musculocontractural type 1'], 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.LOGP': None, 'n.TPSA': None, 'n.NumHDonors': None, 'n.NumHAcceptors': None, 'n.LipinskiViolationCount': None, 'n.DRUGBANK_NAME': None, 'n.MolecularWeight': None, 'n.SYNONYM': None, 'n.PRODUCT': None, 'n.KINGDOM': None, 'n.SOURCE': None, 'n.DIRECT_PARENT': None, 'n.ALTERNATIVE_PARENT': None, 'n.C_SMILES': None}, {'n.NAME': 'COVID-19', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.LOGP': None, 'n.TPSA': None, 'n.NumHDonors': None, 'n.NumHAcceptors': None, 'n.LipinskiViolationCount': None, 'n.DRUGBANK_NAME': None, 'n.MolecularWeight': None, 'n.SYNONYM': None, 'n.PRODUCT': None, 'n.KINGDOM': None, 'n.SOURCE': None, 'n.DIRECT_PARENT': None, 'n.ALTERNATIVE_PARENT': None, 'n.C_SMILES': None}, {'n.NAME': 'Diffuse alopecia', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.LOGP': None, 'n.TPSA': None, 'n.NumHDonors': None, 'n.NumHAcceptors': None, 'n.LipinskiViolationCount': None, 'n.DRUGBANK_NAME': None, 'n.MolecularWeight': None, 'n.SYNONYM': None, 'n.PRODUCT': None, 'n.KINGDOM': None, 'n.SOURCE': None, 'n.DIRECT_PARENT': None, 'n.ALTERNATIVE_PARENT': None, 'n.C_SMILES': None}, {'n.NAME': 'Amyotrophic lateral sclerosis 1', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.LOGP': None, 'n.TPSA': None, 'n.NumHDonors': None, 'n.NumHAcceptors': None, 'n.LipinskiViolationCount': None, 'n.DRUGBANK_NAME': None, 'n.MolecularWeight': None, 'n.SYNONYM': None, 'n.PRODUCT': None, 'n.KINGDOM': None, 'n.SOURCE': None, 'n.DIRECT_PARENT': None, 'n.ALTERNATIVE_PARENT': None, 'n.C_SMILES': None}, {'n.NAME': ['Rhizomelic chondrodysplasia punctata', ' type 1'], 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.LOGP': None, 'n.TPSA': None, 'n.NumHDonors': None, 'n.NumHAcceptors': None, 'n.LipinskiViolationCount': None, 'n.DRUGBANK_NAME': None, 'n.MolecularWeight': None, 'n.SYNONYM': None, 'n.PRODUCT': None, 'n.KINGDOM': None, 'n.SOURCE': None, 'n.DIRECT_PARENT': None, 'n.ALTERNATIVE_PARENT': None, 'n.C_SMILES': None}, {'n.NAME': 'Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.LOGP': None, 'n.TPSA': None, 'n.NumHDonors': None, 'n.NumHAcceptors': None, 'n.LipinskiViolationCount': None, 'n.DRUGBANK_NAME': None, 'n.MolecularWeight': None, 'n.SYNONYM': None, 'n.PRODUCT': None, 'n.KINGDOM': None, 'n.SOURCE': None, 'n.DIRECT_PARENT': None, 'n.ALTERNATIVE_PARENT': None, 'n.C_SMILES': None}, {'n.NAME': 'Idiopathic intracranial hypertension with papilledema', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.LOGP': None, 'n.TPSA': None, 'n.NumHDonors': None, 'n.NumHAcceptors': None, 'n.LipinskiViolationCount': None, 'n.DRUGBANK_NAME': None, 'n.MolecularWeight': None, 'n.SYNONYM': None, 'n.PRODUCT': None, 'n.KINGDOM': None, 'n.SOURCE': None, 'n.DIRECT_PARENT': None, 'n.ALTERNATIVE_PARENT': None, 'n.C_SMILES': None}, {'n.NAME': \"Hypopituitarism and septooptic 'dysplasia'\", 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.LOGP': None, 'n.TPSA': None, 'n.NumHDonors': None, 'n.NumHAcceptors': None, 'n.LipinskiViolationCount': None, 'n.DRUGBANK_NAME': None, 'n.MolecularWeight': None, 'n.SYNONYM': None, 'n.PRODUCT': None, 'n.KINGDOM': None, 'n.SOURCE': None, 'n.DIRECT_PARENT': None, 'n.ALTERNATIVE_PARENT': None, 'n.C_SMILES': None}, {'n.NAME': 'Lysosomal acid lipase deficiency', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.LOGP': None, 'n.TPSA': None, 'n.NumHDonors': None, 'n.NumHAcceptors': None, 'n.LipinskiViolationCount': None, 'n.DRUGBANK_NAME': None, 'n.MolecularWeight': None, 'n.SYNONYM': None, 'n.PRODUCT': None, 'n.KINGDOM': None, 'n.SOURCE': None, 'n.DIRECT_PARENT': None, 'n.ALTERNATIVE_PARENT': None, 'n.C_SMILES': None}, {'n.NAME': 'Esophageal atresia with or without tracheoesophageal fistula', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.LOGP': None, 'n.TPSA': None, 'n.NumHDonors': None, 'n.NumHAcceptors': None, 'n.LipinskiViolationCount': None, 'n.DRUGBANK_NAME': None, 'n.MolecularWeight': None, 'n.SYNONYM': None, 'n.PRODUCT': None, 'n.KINGDOM': None, 'n.SOURCE': None, 'n.DIRECT_PARENT': None, 'n.ALTERNATIVE_PARENT': None, 'n.C_SMILES': None}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "combined_readable_context Ehlers-Danlos Syndrome, musculocontractural type 1, COVID-19, Diffuse alopecia, Amyotrophic lateral sclerosis 1, Rhizomelic chondrodysplasia punctata, type 1, Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules, Idiopathic intracranial hypertension with papilledema, Hypopituitarism and septooptic dysplasia, Lysosomal acid lipase deficiency, Esophageal atresia with or without tracheoesophageal fistula.\n",
      "context: \n",
      "Ehlers-Danlos Syndrome, musculocontractural type 1, COVID-19, Diffuse alopecia, Amyotrophic lateral sclerosis 1, Rhizomelic chondrodysplasia punctata, type 1, Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules, Idiopathic intracranial hypertension with papilledema, Hypopituitarism and septooptic dysplasia, Lysosomal acid lipase deficiency, Esophageal atresia with or without tracheoesophageal fistula.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease) WHERE d.NAME IN [\"Ehlers-Danlos Syndrome, musculocontractural type 1\", \"COVID-19\", \"Diffuse alopecia\", \"Amyotrophic lateral sclerosis 1\", \"Rhizomelic chondrodysplasia punctata, type 1\", \"Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules\", \"Idiopathic intracranial hypertension with papilledema\", \"Hypopituitarism and septooptic dysplasia\", \"Lysosomal acid lipase deficiency\", \"Esophageal atresia with or without tracheoesophageal fistula\"] RETURN d;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'DISEASE_SUPERGRP': nan, 'id': 'C000657245', 'TYPE': 'SCR_DISEASE', 'NAME': 'COVID-19'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531609', 'TYPE': 'SCR_DISEASE', 'NAME': 'Diffuse alopecia'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531617', 'TYPE': 'SCR_DISEASE', 'NAME': 'Amyotrophic lateral sclerosis 1'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531755', 'TYPE': 'SCR_DISEASE', 'NAME': 'Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531795', 'TYPE': 'SCR_DISEASE', 'NAME': 'Idiopathic intracranial hypertension with papilledema'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531854', 'TYPE': 'SCR_DISEASE', 'NAME': 'Lysosomal acid lipase deficiency'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531835', 'TYPE': 'SCR_DISEASE', 'NAME': 'Esophageal atresia with or without tracheoesophageal fistula'}}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: The important terms that are relevant to the provided information are: COVID-19, Diffuse alopecia, Amyotrophic lateral sclerosis 1, Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules, Idiopathic intracranial hypertension with papilledema, Lysosomal acid lipase deficiency, and Esophageal atresia with or without tracheoesophageal fistula.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease) WHERE d.NAME IN ['COVID-19', 'Diffuse alopecia', 'Amyotrophic lateral sclerosis 1', 'Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules', 'Idiopathic intracranial hypertension with papilledema', 'Lysosomal acid lipase deficiency', 'Esophageal atresia with or without tracheoesophageal fistula'] RETURN d;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'DISEASE_SUPERGRP': nan, 'id': 'C000657245', 'TYPE': 'SCR_DISEASE', 'NAME': 'COVID-19'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531609', 'TYPE': 'SCR_DISEASE', 'NAME': 'Diffuse alopecia'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531617', 'TYPE': 'SCR_DISEASE', 'NAME': 'Amyotrophic lateral sclerosis 1'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531755', 'TYPE': 'SCR_DISEASE', 'NAME': 'Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531795', 'TYPE': 'SCR_DISEASE', 'NAME': 'Idiopathic intracranial hypertension with papilledema'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531854', 'TYPE': 'SCR_DISEASE', 'NAME': 'Lysosomal acid lipase deficiency'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531835', 'TYPE': 'SCR_DISEASE', 'NAME': 'Esophageal atresia with or without tracheoesophageal fistula'}}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: The important terms that are relevant to the provided information are: COVID-19, Diffuse alopecia, Amyotrophic lateral sclerosis 1, Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules, Idiopathic intracranial hypertension with papilledema, Lysosomal acid lipase deficiency, and Esophageal atresia with or without tracheoesophageal fistula.\n",
      "You:  exit\n",
      "Goodbye!\n"
     ]
    }
   ],
   "source": [
    "chatbot.chat() #just fetching the node metadata. OF THE NODE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "088a36d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to the Medical Knowledge Chatbot! \n",
      "You:  Give the drugs for the Breast Neoplasms\n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease {NAME: \"Breast Neoplasms\"})-[:DRUG_DISEASE_ASSOCIATION]-(drug:Drug) RETURN drug.id, drug.NAME;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'drug.id': 'DB00019', 'drug.NAME': 'Pegfilgrastim'}, {'drug.id': 'DB00014', 'drug.NAME': 'Goserelin'}, {'drug.id': 'DB00072', 'drug.NAME': 'Trastuzumab'}, {'drug.id': 'DB00126', 'drug.NAME': 'Ascorbic_acid'}, {'drug.id': 'DB00136', 'drug.NAME': 'Calcitriol'}, {'drug.id': 'DB00091', 'drug.NAME': 'Cyclosporine'}, {'drug.id': 'DB00158', 'drug.NAME': 'Folic_acid'}, {'drug.id': 'DB00140', 'drug.NAME': 'Riboflavin'}, {'drug.id': 'DB00184', 'drug.NAME': 'Nicotine'}, {'drug.id': 'DB00188', 'drug.NAME': 'Bortezomib'}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Trastuzumab is a drug for Breast Neoplasms.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug {NAME: \"Trastuzumab\"})-[:DRUG_DISEASE_ASSOCIATION]->(di:Disease {NAME: \"Breast Neoplasms\"}) RETURN d, di;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['11329054', ' 14528282', ' 14750129', ' 12908564', ' 14613027', ' 21632460', ' 22720269', ' 24906530', ' 23474221'], 'DRUG_ATC_CODE': ['ATC:L01X', ' ATC:L01XC', ' ATC:L', ' ATC:L01XC03', ' ATC:L01'], 'PRODUCT': ['Herceptin', ' Herceptin_Hylecta', ' Herceptin_Sc', ' Herzuma', ' Kanjinti', ' Ogivri', ' Ontruzant', ' Perjeta_herceptin', ' Trazimera', ' Trazimera_qyyp'], 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00072', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Trastuzumab', 'SYNONYM': ['RHUMAB_HER2', ' trastuzumab_anns', ' trastuzumab_dkst', ' trastuzumab_dttb', ' trastuzumab_pkrb', ' trastuzumab_qyyp'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'di': {'DISEASE_SUPERGRP': ['C17.800', ' C17.800.090', ' C17', ' C04.588', ' C17.800.090.500', ' C04', ' C04.588.180'], 'id': 'D001943', 'TYPE': 'DISEASE', 'NAME': 'Breast Neoplasms'}}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Here is a helpful answer:\n",
      "\n",
      "Trastuzumab is a drug for Breast Neoplasms.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease {NAME: \"Breast Neoplasms\"})<-[:DRUG_DISEASE_ASSOCIATION]-(drug:Drug {NAME: \"Trastuzumab\"}) RETURN drug, d\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'drug': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['11329054', ' 14528282', ' 14750129', ' 12908564', ' 14613027', ' 21632460', ' 22720269', ' 24906530', ' 23474221'], 'DRUG_ATC_CODE': ['ATC:L01X', ' ATC:L01XC', ' ATC:L', ' ATC:L01XC03', ' ATC:L01'], 'PRODUCT': ['Herceptin', ' Herceptin_Hylecta', ' Herceptin_Sc', ' Herzuma', ' Kanjinti', ' Ogivri', ' Ontruzant', ' Perjeta_herceptin', ' Trazimera', ' Trazimera_qyyp'], 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00072', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Trastuzumab', 'SYNONYM': ['RHUMAB_HER2', ' trastuzumab_anns', ' trastuzumab_dkst', ' trastuzumab_dttb', ' trastuzumab_pkrb', ' trastuzumab_qyyp'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'd': {'DISEASE_SUPERGRP': ['C17.800', ' C17.800.090', ' C17', ' C04.588', ' C17.800.090.500', ' C04', ' C04.588.180'], 'id': 'D001943', 'TYPE': 'DISEASE', 'NAME': 'Breast Neoplasms'}}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: Here is a helpful answer:\n",
      "\n",
      "Trastuzumab is a drug for Breast Neoplasms.\n",
      "You:  exit\n",
      "Goodbye!\n"
     ]
    }
   ],
   "source": [
    "chatbot.chat()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0d431bb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to the Medical Knowledge Chatbot! \n",
      "You:  Give the name of drug that binds to Interleukin-11 receptor subunit alpha\n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug)-[:DRUG_TARGET]->(p:Protein) WHERE p.FULL_NAME = \"Interleukin-11 receptor subunit alpha\" RETURN d.NAME;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d.NAME': 'Oprelvekin'}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Oprelvekin.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug {NAME: \"Oprelvekin\"}) RETURN d;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': nan, 'DRUG_ATC_CODE': ['ATC:L03A', ' ATC:L03AC', ' ATC:L03', ' ATC:L', ' ATC:L03AC02'], 'PRODUCT': 'Neumega', 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00038', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Oprelvekin', 'SYNONYM': ['IL_11', ' Interleukin_11', ' Oprelvekin', ' Recombinant_interleukin_11'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Here are the important terms related to Oprelvekin: IL_11, Interleukin_11, Recombinant_interleukin_11, Neumega.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease)-[:DRUG_DISEASE_ASSOCIATION]-(drug:Drug)-[:DRUG_TARGET|DRUG_ENZYME|DRUG_CARRIER|DRUG_TRANSPORTER]-(p:Protein) \n",
      "WHERE drug.NAME IN ['Oprelvekin', 'IL_11', 'Interleukin_11', 'Recombinant_interleukin_11', 'Neumega'] \n",
      "RETURN drug.NAME, p.NAME, d.NAME ORDER BY drug.NAME;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'drug.NAME': 'Oprelvekin', 'p.NAME': 'I11RA', 'd.NAME': 'Thrombocytopenia'}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: Here are the important terms related to Oprelvekin: I11RA, Thrombocytopenia.\n",
      "You:  Nma the properties of the drug histidine\n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug {NAME: \"histidine\"}) \n",
      "RETURN d{id, TYPE, NAME, NumRotatableBonds, LogP, CLASS, DRUG_INDICATION_ASSOCIATION, DRUG_SIDEEFFECT_ASSOCIATION, DRUGCOMP, COMPCLAS, SUBSTITUENT, PUBMED_ARTICLE, DRUG_ATC_CODE, PRODUCT, PRODUCT_STAGE, KINGDOM, DRUGBANK_NAME, MolecularWeight, SUBCLASS, SUPERCLASS, TPSA, NumHDonors, SYNONYM, LipinskiViolationCount, SOURCE, NumHAcceptors, DRUGINKE, DIRECT_PARENT, ALTERNATIVE_PARENT, C_SMILES, DRUGINCHKE, DRUGINCH}\u001b[0m\n",
      "Attempt 1 failed with error: {code: Neo.ClientError.Statement.SyntaxError} {message: Variable `id` not defined (line 2, column 10 (offset: 45))\n",
      "\"RETURN d{id, TYPE, NAME, NumRotatableBonds, LogP, CLASS, DRUG_INDICATION_ASSOCIATION, DRUG_SIDEEFFECT_ASSOCIATION, DRUGCOMP, COMPCLAS, SUBSTITUENT, PUBMED_ARTICLE, DRUG_ATC_CODE, PRODUCT, PRODUCT_STAGE, KINGDOM, DRUGBANK_NAME, MolecularWeight, SUBCLASS, SUPERCLASS, TPSA, NumHDonors, SYNONYM, LipinskiViolationCount, SOURCE, NumHAcceptors, DRUGINKE, DIRECT_PARENT, ALTERNATIVE_PARENT, C_SMILES, DRUGINCHKE, DRUGINCH}\"\n",
      "          ^}\n",
      "Attempt 2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug {NAME: \"histidine\"}) RETURN d.id, d.TYPE, d.NAME, d.NumRotatableBonds, d.DRUG_INDICATION_ASSOCIATION, d.LogP, d.CLASS, d.DRUG_SIDEEFFECT_ASSOCIATION, d.DRUGCOMP, d.COMPCLAS, d.SUBSTITUENT, d.PUBMED_ARTICLE, d.DRUG_ATC_CODE, d.PRODUCT, d.PRODUCT_STAGE, d.KINGDOM, d.DRUGBANK_NAME, d.MolecularWeight, d.SUBCLASS, d.SUPERCLASS, d.TPSA, d.NumHDonors, d.SYNONYM, d.LipinskiViolationCount, d.SOURCE, d.NumHAcceptors, d.DRUGINKE, d.DIRECT_PARENT, d.ALTERNATIVE_PARENT, d.C_SMILES, d.DRUGINCHKE, d.DRUGINCH;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context_data_list [[]]\n",
      "context: []\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mRETURN []\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'[]': []}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Response deemed fallback on attempt 1: 'I don't know the answer.'\n",
      "Attempt 2 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (n) RETURN n.NAME, n.TYPE, n.id, n.DRUG_INDICATION_ASSOCIATION, n.LogP, n.CLASS, n.DRUG_SIDEEFFECT_ASSOCIATION, n.DRUGCOMP, n.COMPCLAS, n.SUBSTITUENT, n.PUBMED_ARTICLE, n.DRUG_ATC_CODE, n.PRODUCT, n.PRODUCT_STAGE, n.KINGDOM, n.DRUGBANK_NAME, n.MolecularWeight, n.SUBCLASS, n.SUPERCLASS, n.TPSA, n.NumHDonors, n.SYNONYM, n.LipinskiViolationCount, n.SOURCE, n.NumHAcceptors, n.DRUGINKE, n.DIRECT_PARENT, n.ALTERNATIVE_PARENT, n.C_SMILES, n.DRUGINCHKE, n.DRUGINCH, n.DOMAIN, n.SPECIES, n.GO_BP, n.RELATED_KEYWORD, n.RELATED_PUBMED_ID, n.FULL_NAME, n.GO_CC, n.HOMOLOGOUS_SUPERFAMILY, n.GO_MF, n.ORGANISM_CLASS, n.FAMILY, n.CONSERVED_SITE, n.REPEAT, n.PTM, n.SHORT_NAME, n.OTHER_ID, n.PROTEIN_EXPRESSED_IN, n.ACTIVE_SITE, n.BINDING_SITE\u001b[0m\n",
      "Attempt 2 failed with error: Existing exports of data: object cannot be re-sized\n",
      "Attempt 3 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (n) RETURN n.NAME, n.id\u001b[0m\n",
      "Attempt 3 failed with error: Existing exports of data: object cannot be re-sized\n",
      "combined_readable_context No relevant context available after three attempts.\n",
      "context: \n",
      "No relevant context available after three attempts.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[#D202]  _: <CONNECTION> error: Failed to read from defunct connection IPv4Address(('localhost', 7687)) (ResolvedIPv4Address(('127.0.0.1', 7687))): ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None)\n",
      "Transaction failed and will be retried in 1.0697077528730017s (Failed to read from defunct connection IPv4Address(('localhost', 7687)) (ResolvedIPv4Address(('127.0.0.1', 7687))))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mHere is the Cypher statement to query the graph database:\n",
      "\n",
      "MATCH (n) RETURN n.NAME, labels(n), keys(n) ORDER BY n.NAME;\u001b[0m\n",
      "Attempt 1 failed with error: {code: Neo.ClientError.Statement.SyntaxError} {message: Invalid input 'Here': expected 'FOREACH', 'ALTER', 'ORDER BY', 'CALL', 'USING PERIODIC COMMIT', 'CREATE', 'LOAD CSV', 'START DATABASE', 'STOP DATABASE', 'DEALLOCATE', 'DELETE', 'DENY', 'DETACH', 'DROP', 'DRYRUN', 'FINISH', 'GRANT', 'INSERT', 'LIMIT', 'MATCH', 'MERGE', 'NODETACH', 'OFFSET', 'OPTIONAL', 'REALLOCATE', 'REMOVE', 'RENAME', 'RETURN', 'REVOKE', 'ENABLE SERVER', 'SET', 'SHOW', 'SKIP', 'TERMINATE', 'UNWIND', 'USE' or 'WITH' (line 1, column 1 (offset: 0))\n",
      "\"Here is the Cypher statement to query the graph database:\"\n",
      " ^}\n",
      "Attempt 2 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (n) RETURN n.NAME, n.TYPE, n.DRUG_INDICATION_ASSOCIATION, n.DRUG_SIDEEFFECT_ASSOCIATION, n.PUBMED_ARTICLE, n.PRODUCT, n.DRUG_ATC_CODE, n.SYNONYM, n.GO_BP, n.RELATED_KEYWORD, n.GO_CC, n.GO_MF, n.ORGANISM_CLASS, n.FAMILY, n.SHORT_NAME, n.OTHER_ID;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'n.NAME': ['Ehlers-Danlos Syndrome', ' musculocontractural type 1'], 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.PUBMED_ARTICLE': None, 'n.PRODUCT': None, 'n.DRUG_ATC_CODE': None, 'n.SYNONYM': None, 'n.GO_BP': None, 'n.RELATED_KEYWORD': None, 'n.GO_CC': None, 'n.GO_MF': None, 'n.ORGANISM_CLASS': None, 'n.FAMILY': None, 'n.SHORT_NAME': None, 'n.OTHER_ID': None}, {'n.NAME': 'COVID-19', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.PUBMED_ARTICLE': None, 'n.PRODUCT': None, 'n.DRUG_ATC_CODE': None, 'n.SYNONYM': None, 'n.GO_BP': None, 'n.RELATED_KEYWORD': None, 'n.GO_CC': None, 'n.GO_MF': None, 'n.ORGANISM_CLASS': None, 'n.FAMILY': None, 'n.SHORT_NAME': None, 'n.OTHER_ID': None}, {'n.NAME': 'Diffuse alopecia', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.PUBMED_ARTICLE': None, 'n.PRODUCT': None, 'n.DRUG_ATC_CODE': None, 'n.SYNONYM': None, 'n.GO_BP': None, 'n.RELATED_KEYWORD': None, 'n.GO_CC': None, 'n.GO_MF': None, 'n.ORGANISM_CLASS': None, 'n.FAMILY': None, 'n.SHORT_NAME': None, 'n.OTHER_ID': None}, {'n.NAME': 'Amyotrophic lateral sclerosis 1', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.PUBMED_ARTICLE': None, 'n.PRODUCT': None, 'n.DRUG_ATC_CODE': None, 'n.SYNONYM': None, 'n.GO_BP': None, 'n.RELATED_KEYWORD': None, 'n.GO_CC': None, 'n.GO_MF': None, 'n.ORGANISM_CLASS': None, 'n.FAMILY': None, 'n.SHORT_NAME': None, 'n.OTHER_ID': None}, {'n.NAME': ['Rhizomelic chondrodysplasia punctata', ' type 1'], 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.PUBMED_ARTICLE': None, 'n.PRODUCT': None, 'n.DRUG_ATC_CODE': None, 'n.SYNONYM': None, 'n.GO_BP': None, 'n.RELATED_KEYWORD': None, 'n.GO_CC': None, 'n.GO_MF': None, 'n.ORGANISM_CLASS': None, 'n.FAMILY': None, 'n.SHORT_NAME': None, 'n.OTHER_ID': None}, {'n.NAME': 'Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.PUBMED_ARTICLE': None, 'n.PRODUCT': None, 'n.DRUG_ATC_CODE': None, 'n.SYNONYM': None, 'n.GO_BP': None, 'n.RELATED_KEYWORD': None, 'n.GO_CC': None, 'n.GO_MF': None, 'n.ORGANISM_CLASS': None, 'n.FAMILY': None, 'n.SHORT_NAME': None, 'n.OTHER_ID': None}, {'n.NAME': 'Idiopathic intracranial hypertension with papilledema', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.PUBMED_ARTICLE': None, 'n.PRODUCT': None, 'n.DRUG_ATC_CODE': None, 'n.SYNONYM': None, 'n.GO_BP': None, 'n.RELATED_KEYWORD': None, 'n.GO_CC': None, 'n.GO_MF': None, 'n.ORGANISM_CLASS': None, 'n.FAMILY': None, 'n.SHORT_NAME': None, 'n.OTHER_ID': None}, {'n.NAME': \"Hypopituitarism and septooptic 'dysplasia'\", 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.PUBMED_ARTICLE': None, 'n.PRODUCT': None, 'n.DRUG_ATC_CODE': None, 'n.SYNONYM': None, 'n.GO_BP': None, 'n.RELATED_KEYWORD': None, 'n.GO_CC': None, 'n.GO_MF': None, 'n.ORGANISM_CLASS': None, 'n.FAMILY': None, 'n.SHORT_NAME': None, 'n.OTHER_ID': None}, {'n.NAME': 'Lysosomal acid lipase deficiency', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.PUBMED_ARTICLE': None, 'n.PRODUCT': None, 'n.DRUG_ATC_CODE': None, 'n.SYNONYM': None, 'n.GO_BP': None, 'n.RELATED_KEYWORD': None, 'n.GO_CC': None, 'n.GO_MF': None, 'n.ORGANISM_CLASS': None, 'n.FAMILY': None, 'n.SHORT_NAME': None, 'n.OTHER_ID': None}, {'n.NAME': 'Esophageal atresia with or without tracheoesophageal fistula', 'n.TYPE': 'SCR_DISEASE', 'n.DRUG_INDICATION_ASSOCIATION': None, 'n.DRUG_SIDEEFFECT_ASSOCIATION': None, 'n.PUBMED_ARTICLE': None, 'n.PRODUCT': None, 'n.DRUG_ATC_CODE': None, 'n.SYNONYM': None, 'n.GO_BP': None, 'n.RELATED_KEYWORD': None, 'n.GO_CC': None, 'n.GO_MF': None, 'n.ORGANISM_CLASS': None, 'n.FAMILY': None, 'n.SHORT_NAME': None, 'n.OTHER_ID': None}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Ehlers-Danlos Syndrome, musculocontractural type 1, COVID-19, Diffuse alopecia, Amyotrophic lateral sclerosis 1, Rhizomelic chondrodysplasia punctata, type 1, Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules, Idiopathic intracranial hypertension with papilledema, Hypopituitarism and septooptic 'dysplasia', Lysosomal acid lipase deficiency, Esophageal atresia with or without tracheoesophageal fistula.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease) WHERE d.NAME IN ['Ehlers-Danlos Syndrome, musculocontractural type 1', 'COVID-19', 'Diffuse alopecia', 'Amyotrophic lateral sclerosis 1', 'Rhizomelic chondrodysplasia punctata, type 1', 'Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules', 'Idiopathic intracranial hypertension with papilledema', 'Hypopituitarism and septooptic dysplasia', 'Lysosomal acid lipase deficiency', 'Esophageal atresia with or without tracheoesophageal fistula'] RETURN d;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'DISEASE_SUPERGRP': nan, 'id': 'C000657245', 'TYPE': 'SCR_DISEASE', 'NAME': 'COVID-19'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531609', 'TYPE': 'SCR_DISEASE', 'NAME': 'Diffuse alopecia'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531617', 'TYPE': 'SCR_DISEASE', 'NAME': 'Amyotrophic lateral sclerosis 1'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531755', 'TYPE': 'SCR_DISEASE', 'NAME': 'Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531795', 'TYPE': 'SCR_DISEASE', 'NAME': 'Idiopathic intracranial hypertension with papilledema'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531854', 'TYPE': 'SCR_DISEASE', 'NAME': 'Lysosomal acid lipase deficiency'}}, {'d': {'DISEASE_SUPERGRP': nan, 'id': 'C531835', 'TYPE': 'SCR_DISEASE', 'NAME': 'Esophageal atresia with or without tracheoesophageal fistula'}}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: The important terms are: COVID-19, Diffuse alopecia, Amyotrophic lateral sclerosis 1, Kidney disorder involving deposition of calcium and oxalate or phosphate in the renal tubules, Idiopathic intracranial hypertension with papilledema, Lysosomal acid lipase deficiency, Esophageal atresia with or without tracheoesophageal fistula.\n",
      "You:  exit\n",
      "Goodbye!\n"
     ]
    }
   ],
   "source": [
    "chatbot.chat()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6f5a41d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to the Medical Knowledge Chatbot! \n",
      "You:  Give all those proteins which are carrier or interact with the drug Histidine\n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug {NAME: \"Histidine\"})-[:DRUG_CARRIER|DRUG_TARGET]*->(p:Protein) RETURN p;\u001b[0m\n",
      "Attempt 1 failed with error: {code: Neo.ClientError.Statement.SyntaxError} {message: Invalid input '*': expected '-' (line 1, column 63 (offset: 62))\n",
      "\"MATCH (d:Drug {NAME: \"Histidine\"})-[:DRUG_CARRIER|DRUG_TARGET]*->(p:Protein) RETURN p;\"\n",
      "                                                               ^}\n",
      "Attempt 2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug {NAME: 'Histidine'})-[:DRUG_CARRIER|DRUG_TARGET|PPI]-(p:Protein) RETURN p;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'p': {'DOMAIN': ['IPR006195', ' IPR004154', ' IPR041715', ' IPR033656', ' IPR000738'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0005524', ' GO:0004821', ' GO:0042802', ' GO:0042803'], 'SHORT_NAME': 'HisRS', 'NAME': 'HARS1', 'FAMILY': ['IPR015807', ' IPR004516'], 'RELATED_KEYWORD': ['3D-structure', ' Acetylation', ' Alternative splicing', ' Aminoacyl-tRNA synthetase', ' ATP-binding', ' Charcot-Marie-Tooth disease', ' Cytoplasm', ' Deafness', ' Disease mutation', ' Ligase', ' Neurodegeneration', ' Neuropathy', ' Nucleotide-binding', ' Phosphoprotein', ' Polymorphism', ' Protein biosynthesis', ' Reference proteome', ' Retinitis pigmentosa', ' Usher syndrome'], 'PROTEIN_EXPRESSED_IN': ['Smooth_muscle__Smooth_muscle_cells', ' Cerebellum__Cells_in_granular_layer', ' Caudate__Glial_cells', ' Colon__Endothelial_cells', ' Hippocampus__Glial_cells', ' Cerebral_cortex__Neuropil', ' Smooth_muscle', ' Pancreas__Islets_of_Langerhans', ' Hippocampus__Neuronal_cells', ' Cerebral_cortex__Endothelial_cells', ' Hippocampus', ' Caudate__Neuronal_cells', ' Cerebral_cortex', ' Ovary__Ovarian_stroma_cells', ' Spleen__Cells_in_red_pulp', ' Endometrium_1__Cells_in_endometrial_stroma', ' Soft_tissue_2__Fibroblasts', ' Thyroid_gland', ' Liver__Bile_duct_cells', ' Placenta__Trophoblastic_cells', ' Cervix', '_uterine__Squamous_epithelial_cells', ' Thyroid_gland__Glandular_cells', ' Lung', ' Kidney', ' Cervix', '_uterine', ' Skin_1__Keratinocytes', ' Lung__Pneumocytes', ' Testis__Leydig_cells', ' Soft_tissue_2', ' Kidney__Cells_in_tubules', ' Skin_1', ' Kidney__Cells_in_glomeruli', ' Skin_1__Melanocytes', ' Cerebral_cortex__Neuronal_cells', ' Cervix', '_uterine__Glandular_cells', ' Cerebellum__Cells_in_molecular_layer', ' Adrenal_gland__Glandular_cells', ' Soft_tissue_2__Chondrocytes', ' Placenta', ' Endometrium_2__Cells_in_endometrial_stroma', ' Skin_1__Fibroblasts', ' Caudate', ' Adrenal_gland', ' Colon__Peripheral_nerve/ganglion', ' Skin_1__Langerhans', ' Liver__Hepatocytes', ' Lung__Macrophages', ' Placenta__Decidual_cells', ' Cerebral_cortex__Glial_cells', ' Liver', ' Lymph_node', ' Skin_2__Epidermal_cells', ' Rectum__Glandular_cells', ' Appendix__Lymphoid_tissue', ' Spleen__Cells_in_white_pulp', ' Endometrium_2', ' Endometrium_1', ' Bronchus__Respiratory_epithelial_cells', ' Cerebellum', ' Pancreas', ' Nasopharynx__Respiratory_epithelial_cells', ' Prostate__Glandular_cells', ' Breast', ' Rectum', ' Gallbladder', ' Bronchus', ' Tonsil__Germinal_center_cells', ' Lymph_node__Germinal_center_cells', ' Duodenum', ' Ovary__Follicle_cells', ' Vagina__Squamous_epithelial_cells', ' Tonsil__Non-germinal_center_cells', ' Testis__Cells_in_seminiferous_ducts', ' Oral_mucosa__Squamous_epithelial_cells', ' Appendix', ' Stomach_2__Glandular_cells', ' Lymph_node__Non-germinal_center_cells', ' Endometrium_1__Glandular_cells', ' Colon', ' Bone_marrow__Hematopoietic_cells', ' Fallopian_tube__Glandular_cells', ' Esophagus__Squamous_epithelial_cells', ' Parathyroid_gland', ' Gallbladder__Glandular_cells', ' Spleen', ' Endometrium_2__Glandular_cells', ' Parathyroid_gland__Glandular_cells', ' Salivary_gland', ' Salivary_gland__Glandular_cells', ' Tonsil__Squamous_epithelial_cells', ' Small_intestine__Glandular_cells', ' Prostate', ' Pancreas__Exocrine_glandular_cells', ' Seminal_vesicle__Glandular_cells', ' Esophagus', ' Breast__Glandular_cells', ' Epididymis', ' Seminal_vesicle', ' Vagina', ' Skin_2', ' Small_intestine', ' Stomach_1__Glandular_cells', ' Cerebellum__Purkinje_cells', ' Nasopharynx', ' Testis', ' Colon__Glandular_cells', ' Urinary_bladder', ' Stomach_1', ' Epididymis__Glandular_cells', ' Stomach_2', ' Oral_mucosa', ' Tonsil', ' Urinary_bladder__Urothelial_cells', ' Duodenum__Glandular_cells', ' Fallopian_tube', ' Appendix__Glandular_cells', ' Bone_marrow', ' Ovary'], 'OTHER_ID': ['B4DY73', ' D6REN6', ' J3KNE5'], 'RELATED_PUBMED_ID': ['pubmed:1549469', ' pubmed:3554142', ' pubmed:14702039', ' pubmed:15372022', ' pubmed:15489334', ' pubmed:8406012', ' pubmed:7755634', ' pubmed:19413330', ' pubmed:21269460', ' pubmed:22930593', ' pubmed:24275569', ' pubmed:26072516', ' pubmed:25151410', ' pubmed:22279524', ' pubmed:29235198'], 'FULL_NAME': ['Histidine--tRNA ligase', ' cytoplasmic', ' Histidyl-tRNA synthetase'], 'GO_CC': ['GO:0005737', ' GO:0005829', ' GO:0005739'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036621', ' IPR009068'], 'id': 'P12081', 'GO_MF': ['GO:0006427', ' GO:0032543', ' GO:0006412', ' GO:0006418'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'p': {'SPECIES': 'HUMAN', 'GO_BP': ['GO:0004398', ' GO:0030170'], 'SHORT_NAME': 'HDC', 'NAME': 'DCHS', 'FAMILY': ['IPR010977', ' IPR002129'], 'RELATED_KEYWORD': ['3D-structure', ' Alternative splicing', ' Catecholamine biosynthesis', ' Decarboxylase', ' Lyase', ' Polymorphism', ' Pyridoxal phosphate', ' Reference proteome'], 'PROTEIN_EXPRESSED_IN': ['Lung', ' Vagina__Squamous_epithelial_cells', ' Skin_2', ' Pancreas', ' Testis', ' Skin_2__Epidermal_cells', ' Lung__Macrophages', ' Soft_tissue_1', ' Cerebral_cortex__Neuronal_cells', ' Endometrium_1__Glandular_cells', ' Lymph_node', ' Soft_tissue_1__Chondrocytes', ' Skin_1__Keratinocytes', ' Stomach_1__Glandular_cells', ' Testis__Leydig_cells', ' Pancreas__Islets_of_Langerhans', ' Skin_1', ' Vagina', ' Endometrium_1', ' Stomach_1', ' Cerebral_cortex', ' Lymph_node__Non-germinal_center_cells', ' Epididymis', ' Epididymis__Glandular_cells'], 'OTHER_ID': 'B7ZM01', 'RELATED_PUBMED_ID': ['pubmed:2216786', ' pubmed:1768863', ' pubmed:1425659', ' pubmed:8288622', ' pubmed:16572171', ' pubmed:15489334', ' pubmed:22767596', ' pubmed:16959974'], 'BINDING_SITE': 'IPR021115', 'FULL_NAME': 'Histidine decarboxylase', 'GO_CC': 'GO:0005829', 'HOMOLOGOUS_SUPERFAMILY': ['IPR015424', ' IPR015422', ' IPR015421'], 'id': 'P19113', 'GO_MF': ['GO:0042423', ' GO:0001694', ' GO:0006548', ' GO:0006547'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'p': {'DOMAIN': 'IPR021922', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0016841', ' GO:0004397'], 'ACTIVE_SITE': 'IPR022313', 'SHORT_NAME': 'Histidase', 'NAME': 'HUTH', 'FAMILY': ['IPR001106', ' IPR005921'], 'RELATED_KEYWORD': ['Alternative splicing', ' Disease mutation', ' Histidine metabolism', ' Lyase', ' Phosphoprotein', ' Polymorphism', ' Reference proteome'], 'PROTEIN_EXPRESSED_IN': ['Hippocampus__Neuronal_cells', ' Cerebral_cortex__Neuropil', ' Endometrium_1__Cells_in_endometrial_stroma', ' Kidney__Cells_in_glomeruli', ' Endometrium_2__Cells_in_endometrial_stroma', ' Caudate__Neuronal_cells', ' Spleen__Cells_in_white_pulp', ' Hippocampus', ' Caudate', ' Lymph_node__Germinal_center_cells', ' Breast__Adipocytes', ' Tonsil__Squamous_epithelial_cells', ' Bronchus', ' Adrenal_gland', ' Cerebellum__Purkinje_cells', ' Seminal_vesicle__Glandular_cells', ' Fallopian_tube', ' Salivary_gland__Glandular_cells', ' Endometrium_2', ' Skeletal_muscle', ' Pancreas', ' Bronchus__Respiratory_epithelial_cells', ' Appendix__Lymphoid_tissue', ' Esophagus__Squamous_epithelial_cells', ' Seminal_vesicle', ' Appendix__Glandular_cells', ' Fallopian_tube__Glandular_cells', ' Urinary_bladder__Urothelial_cells', ' Nasopharynx__Respiratory_epithelial_cells', ' Lymph_node', ' Skeletal_muscle__Myocytes', ' Tonsil', ' Bone_marrow__Hematopoietic_cells', ' Parathyroid_gland', ' Small_intestine', ' Lymph_node__Non-germinal_center_cells', ' Tonsil__Germinal_center_cells', ' Oral_mucosa__Squamous_epithelial_cells', ' Stomach_2__Glandular_cells', ' Skin_1__Fibroblasts', ' Parathyroid_gland__Glandular_cells', ' Cervix', '_uterine__Glandular_cells', ' Testis', ' Rectum__Glandular_cells', ' Cerebellum', ' Appendix', ' Soft_tissue_2', ' Prostate__Glandular_cells', ' Skin_2__Epidermal_cells', ' Adrenal_gland__Glandular_cells', ' Esophagus', ' Colon', ' Breast', ' Gallbladder', ' Prostate', ' Epididymis__Glandular_cells', ' Small_intestine__Glandular_cells', ' Skin_2', ' Adipose_tissue__Adipocytes', ' Stomach_1__Glandular_cells', ' Stomach_1', ' Colon__Glandular_cells', ' Placenta__Trophoblastic_cells', ' Duodenum', ' Endometrium_1__Glandular_cells', ' Testis__Cells_in_seminiferous_ducts', ' Soft_tissue_1', ' Colon__Peripheral_nerve/ganglion', ' Heart_muscle__Myocytes', ' Breast__Myoepithelial_cells', ' Kidney', ' Soft_tissue_2__Peripheral_nerve', ' Endometrium_1', ' Thyroid_gland__Glandular_cells', ' Testis__Leydig_cells', ' Epididymis', ' Cerebral_cortex', ' Skin_1__Langerhans', ' Pancreas__Exocrine_glandular_cells', ' Vagina__Squamous_epithelial_cells', ' Soft_tissue_1__Fibroblasts', ' Oral_mucosa', ' Bone_marrow', ' Heart_muscle', ' Nasopharynx', ' Urinary_bladder', ' Salivary_gland', ' Endometrium_2__Glandular_cells', ' Spleen__Cells_in_red_pulp', ' Lung__Macrophages', ' Stomach_2', ' Placenta', ' Smooth_muscle', ' Cervix', '_uterine', ' Spleen', ' Vagina', ' Soft_tissue_2__Fibroblasts', ' Duodenum__Glandular_cells', ' Adipose_tissue', ' Breast__Glandular_cells', ' Cerebral_cortex__Neuronal_cells', ' Smooth_muscle__Smooth_muscle_cells', ' Pancreas__Islets_of_Langerhans', ' Gallbladder__Glandular_cells', ' Cerebellum__Cells_in_granular_layer', ' Lung', ' Thyroid_gland', ' Rectum', ' Lung__Pneumocytes', ' Colon__Endothelial_cells', ' Cervix', '_uterine__Squamous_epithelial_cells', ' Kidney__Cells_in_tubules', ' Liver', ' Liver__Hepatocytes', ' Skin_1__Keratinocytes', ' Skin_1', ' Skin_1__Melanocytes'], 'OTHER_ID': ['B4E0V8', ' F5GXF2', ' F5H1U5', ' Q4VB92', ' Q4VB93'], 'RELATED_PUBMED_ID': ['pubmed:7916645', ' pubmed:8530107', ' pubmed:14702039', ' pubmed:16541075', ' pubmed:15489334', ' pubmed:24275569', ' pubmed:15806399'], 'FULL_NAME': 'Histidine ammonia-lyase', 'GO_CC': 'GO:0005829', 'HOMOLOGOUS_SUPERFAMILY': ['IPR024083', ' IPR008948'], 'id': 'P42357', 'GO_MF': ['GO:0006548', ' GO:0019556', ' GO:0019557'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'p': {'RELATED_KEYWORD': ['Amino-acid transport', ' Antiport', ' Cell membrane', ' Disulfide bond', ' Glycoprotein', ' Ion transport', ' Membrane', ' Reference proteome', ' Sodium', ' Sodium transport', ' Symport', ' Transmembrane', ' Transmembrane helix', ' Transport'], 'DOMAIN': 'IPR013057', 'OTHER_ID': 'Q6IB34', 'RELATED_PUBMED_ID': ['pubmed:10823827', ' pubmed:14702039', ' pubmed:17974005', ' pubmed:16641997', ' pubmed:15489334', ' pubmed:11085536', ' pubmed:24275569'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0015171', ' GO:0015297', ' GO:0015180', ' GO:0015182', ' GO:0015186', ' GO:0005290', ' GO:0015293'], 'FULL_NAME': ['Sodium-coupled neutral amino acid transporter 3', ' N-system amino acid transporter 1', ' Na(+)-coupled neutral amino acid transporter 3', ' Solute carrier family 38 member 3', ' System N amino acid transporter 1'], 'GO_CC': ['GO:0016323', ' GO:0005887', ' GO:0005886'], 'id': 'Q99624', 'GO_MF': ['GO:0003333', ' GO:0006865', ' GO:0006867', ' GO:0007420', ' GO:0051365', ' GO:0007565', ' GO:0006868', ' GO:0015817', ' GO:0015808', ' GO:2000487', ' GO:0061402', ' GO:0006814', ' GO:0150104'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.'], 'NAME': 'S38A3'}}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Here are the proteins that are related to Histidine:\n",
      "\n",
      "HARS1 (Histidine--tRNA ligase), HDC (Histidine decarboxylase), HUTH (Histidine ammonia-lyase)\n",
      "\n",
      "These proteins are either carriers or interact with the drug Histidine.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (p:Protein) WHERE p.NAME IN [\"HARS1\", \"HDC\", \"HUTH\"] \n",
      "MATCH (d:Drug {NAME: \"Histidine\"}) \n",
      "MATCH (p)-[:DRUG_CARRIER|DRUG_TARGET|DRUG_ENZYME|PPI]-(d) \n",
      "RETURN p, d;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'p': {'DOMAIN': ['IPR006195', ' IPR004154', ' IPR041715', ' IPR033656', ' IPR000738'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0005524', ' GO:0004821', ' GO:0042802', ' GO:0042803'], 'SHORT_NAME': 'HisRS', 'NAME': 'HARS1', 'FAMILY': ['IPR015807', ' IPR004516'], 'RELATED_KEYWORD': ['3D-structure', ' Acetylation', ' Alternative splicing', ' Aminoacyl-tRNA synthetase', ' ATP-binding', ' Charcot-Marie-Tooth disease', ' Cytoplasm', ' Deafness', ' Disease mutation', ' Ligase', ' Neurodegeneration', ' Neuropathy', ' Nucleotide-binding', ' Phosphoprotein', ' Polymorphism', ' Protein biosynthesis', ' Reference proteome', ' Retinitis pigmentosa', ' Usher syndrome'], 'PROTEIN_EXPRESSED_IN': ['Smooth_muscle__Smooth_muscle_cells', ' Cerebellum__Cells_in_granular_layer', ' Caudate__Glial_cells', ' Colon__Endothelial_cells', ' Hippocampus__Glial_cells', ' Cerebral_cortex__Neuropil', ' Smooth_muscle', ' Pancreas__Islets_of_Langerhans', ' Hippocampus__Neuronal_cells', ' Cerebral_cortex__Endothelial_cells', ' Hippocampus', ' Caudate__Neuronal_cells', ' Cerebral_cortex', ' Ovary__Ovarian_stroma_cells', ' Spleen__Cells_in_red_pulp', ' Endometrium_1__Cells_in_endometrial_stroma', ' Soft_tissue_2__Fibroblasts', ' Thyroid_gland', ' Liver__Bile_duct_cells', ' Placenta__Trophoblastic_cells', ' Cervix', '_uterine__Squamous_epithelial_cells', ' Thyroid_gland__Glandular_cells', ' Lung', ' Kidney', ' Cervix', '_uterine', ' Skin_1__Keratinocytes', ' Lung__Pneumocytes', ' Testis__Leydig_cells', ' Soft_tissue_2', ' Kidney__Cells_in_tubules', ' Skin_1', ' Kidney__Cells_in_glomeruli', ' Skin_1__Melanocytes', ' Cerebral_cortex__Neuronal_cells', ' Cervix', '_uterine__Glandular_cells', ' Cerebellum__Cells_in_molecular_layer', ' Adrenal_gland__Glandular_cells', ' Soft_tissue_2__Chondrocytes', ' Placenta', ' Endometrium_2__Cells_in_endometrial_stroma', ' Skin_1__Fibroblasts', ' Caudate', ' Adrenal_gland', ' Colon__Peripheral_nerve/ganglion', ' Skin_1__Langerhans', ' Liver__Hepatocytes', ' Lung__Macrophages', ' Placenta__Decidual_cells', ' Cerebral_cortex__Glial_cells', ' Liver', ' Lymph_node', ' Skin_2__Epidermal_cells', ' Rectum__Glandular_cells', ' Appendix__Lymphoid_tissue', ' Spleen__Cells_in_white_pulp', ' Endometrium_2', ' Endometrium_1', ' Bronchus__Respiratory_epithelial_cells', ' Cerebellum', ' Pancreas', ' Nasopharynx__Respiratory_epithelial_cells', ' Prostate__Glandular_cells', ' Breast', ' Rectum', ' Gallbladder', ' Bronchus', ' Tonsil__Germinal_center_cells', ' Lymph_node__Germinal_center_cells', ' Duodenum', ' Ovary__Follicle_cells', ' Vagina__Squamous_epithelial_cells', ' Tonsil__Non-germinal_center_cells', ' Testis__Cells_in_seminiferous_ducts', ' Oral_mucosa__Squamous_epithelial_cells', ' Appendix', ' Stomach_2__Glandular_cells', ' Lymph_node__Non-germinal_center_cells', ' Endometrium_1__Glandular_cells', ' Colon', ' Bone_marrow__Hematopoietic_cells', ' Fallopian_tube__Glandular_cells', ' Esophagus__Squamous_epithelial_cells', ' Parathyroid_gland', ' Gallbladder__Glandular_cells', ' Spleen', ' Endometrium_2__Glandular_cells', ' Parathyroid_gland__Glandular_cells', ' Salivary_gland', ' Salivary_gland__Glandular_cells', ' Tonsil__Squamous_epithelial_cells', ' Small_intestine__Glandular_cells', ' Prostate', ' Pancreas__Exocrine_glandular_cells', ' Seminal_vesicle__Glandular_cells', ' Esophagus', ' Breast__Glandular_cells', ' Epididymis', ' Seminal_vesicle', ' Vagina', ' Skin_2', ' Small_intestine', ' Stomach_1__Glandular_cells', ' Cerebellum__Purkinje_cells', ' Nasopharynx', ' Testis', ' Colon__Glandular_cells', ' Urinary_bladder', ' Stomach_1', ' Epididymis__Glandular_cells', ' Stomach_2', ' Oral_mucosa', ' Tonsil', ' Urinary_bladder__Urothelial_cells', ' Duodenum__Glandular_cells', ' Fallopian_tube', ' Appendix__Glandular_cells', ' Bone_marrow', ' Ovary'], 'OTHER_ID': ['B4DY73', ' D6REN6', ' J3KNE5'], 'RELATED_PUBMED_ID': ['pubmed:1549469', ' pubmed:3554142', ' pubmed:14702039', ' pubmed:15372022', ' pubmed:15489334', ' pubmed:8406012', ' pubmed:7755634', ' pubmed:19413330', ' pubmed:21269460', ' pubmed:22930593', ' pubmed:24275569', ' pubmed:26072516', ' pubmed:25151410', ' pubmed:22279524', ' pubmed:29235198'], 'FULL_NAME': ['Histidine--tRNA ligase', ' cytoplasmic', ' Histidyl-tRNA synthetase'], 'GO_CC': ['GO:0005737', ' GO:0005829', ' GO:0005739'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036621', ' IPR009068'], 'id': 'P12081', 'GO_MF': ['GO:0006427', ' GO:0032543', ' GO:0006412', ' GO:0006418'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}, 'd': {'NumRotatableBonds': 3, 'DRUG_INDICATION_ASSOCIATION': nan, 'C_SMILES': 'N[C@@H](Cc1c[nH]cn1)C(=O)O', 'LogP': -0.6359000000000001, 'CLASS': 'Carboxylic_acids_and_derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGINCHKE': 'HNDVDQJCIGZPNO-YFKPBYRVSA-N', 'DRUGCOMP': nan, 'COMPCLAS': nan, 'DRUGINCH': ['InChI=1S/C6H9N3O2/c7-5(6(10)11)1-4-2-8-3-9-4/h2-3', '5H', '1', '7H2', '(H', '8', '9)(H', '10', '11)/t5-/m0/s1'], 'SUBSTITUENT': ['Alpha_amino_acid', ' Amine', ' Amino_acid', ' Aralkylamine', ' Aromatic_heteromonocyclic_compound', ' Azacycle', ' Azole', ' Carbonyl_group', ' Carboxylic_acid', ' Heteroaromatic_compound', ' Histidine_or_derivatives', ' Hydrocarbon_derivative', ' Imidazole', ' Imidazolyl_carboxylic_acid_derivative', ' L_alpha_amino_acid', ' Monocarboxylic_acid_or_derivatives', ' Organic_nitrogen_compound', ' Organic_oxide', ' Organic_oxygen_compound', ' Organoheterocyclic_compound', ' Organonitrogen_compound', ' Organooxygen_compound', ' Organopnictogen_compound', ' Primary_aliphatic_amine', ' Primary_amine'], 'PUBMED_ARTICLE': nan, 'DRUG_ATC_CODE': nan, 'PRODUCT': ['2_5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose', ' 2_5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose_Clinimix', ' 2_5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_10__Dextrose_Quickmix', ' 2_5_travasol_Amino_Acid_InJ_W_eleC_W_25_dex', ' 2_75__Travas__Amino_Acid_InJ_W_elecw_25_dex', ' 2_75__Travasol_Amino_Acid_Injection_With_Electrolytes_In_25__Dextrose_Quickmix', ' 2_75__Travasol_Amino_Acid_Injection_With_Electrolytes_In_5__Dextrose_Quickmix', ' 2_75_travasol_Amino_Acid_InJ_W_eleC_W_5_dex_', ' 20__Prosol', ' 4_25__Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_25__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_5__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrlytes_In_10__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_25__Dextrose', ' 4_25__Travasol_Amino_Acid_Without_Electrolytes_In_5__Dextrose_Quickmix', ' 4_25_trav__Amino_Acid_InJ_W_O_elect_5_dext_', ' 4_25_travasol_Amino_Acid_InJ_W_elecw_5_dex_', ' 5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose', ' 5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose', ' 5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_25__Dextrose', ' 8_5__Freamine_III', ' Aminosyn', ' Aminosyn_10__W_Electrolytes', ' Aminosyn_3_5__M', ' Aminosyn_5_', ' Aminosyn_7_', ' Aminosyn_8_5_', ' Aminosyn_8_5__Injection_With_Electrolytes', ' Aminosyn_HF', ' Aminosyn_II', ' Aminosyn_II_10__With_Electrolytes', ' Aminosyn_II_5__Inj', ' Aminosyn_II_7__Injection', ' Aminosyn_II_7__M_In_10__Dextrose_dual_Chamber_', ' Aminosyn_II_7__With_10__Dextrose', ' Aminosyn_II_7__With_50__Dextrose', ' Aminosyn_II_8_5__Injection', ' Aminosyn_II_8_5__M_In_20__Dextrose__dual_Chamber_', ' Aminosyn_II_8_5__With_50__Dextrose', ' Aminosyn_II_and_Dextrose', ' Aminosyn_II_in_Dextrose', ' Aminosyn_II_In_Dextrose_Injection', ' Aminosyn_II_with_Electrolytes', ' Aminosyn_II_with_Electrolytes_in_Dextrose_with_Calcium', ' Aminosyn_RF', ' Aminosyn_RF_Injection', ' Aminosyn_Sulfite_Free', ' Aminosyn_PF', ' Aminosyn_PF_10_', ' Aminosyn_PF_7_', ' CAREDO_Treating_Periodontitis_Toothpastes_100g', ' Clinimix', ' Clinimix_2_5__Travasol_Aa_Without_Electrolytes_In_10__Dextrose_Injec', ' Clinimix_E', ' Clinisol', ' FreAmine_HBC', ' FreAmine_III', ' HepatAmine', ' Hepatasol', ' Kabiven', ' NephrAmine', ' Nephramine_Liq_IV', ' Novamine', ' Nutrineal_Pd4', ' Olimel_3_3__E', ' Olimel_4_4_', ' Olimel_4_4__E', ' Olimel_5_7_', ' Olimel_5_7__E', ' Olimel_7_6_', ' Olimel_7_6__E', ' Perikabiven', ' Periolimel_2_5__E', ' Plenamine', ' Premasol___sulfite_free__Amino_Acid_', ' Primene', ' Primene_10_', ' Primene_10__liq_IV', ' ProcalAmine', ' ProSol', ' Quick_Mix_2_5__Travasol_Aa_With_Electrolytes_With_25__Dextrose_Injec', ' RenAmin', ' Renamin__amino_Acids__Injection', ' Smofkabiven', ' Smofkabiven_Electrolyte_Free', ' Smofkabiven_Peripheral', ' Synthamin', ' Travasol', ' Travasol_Amino_Acid_Inj_5_5_', ' Travasol_Amino_Acid_Inj_8_5_', ' Travasol_E', ' Travasol_Inj_Without_Electrolytes_5_5_', ' Travasol_Inj_Without_Electrolytes_8_5_', ' Trepoxen_250', ' Trepoxicam_7_5', ' TrophAmine', ' Vamin_18_Electrolyte_free', ' Vamin_N'], 'PRODUCT_STAGE': ['investigational', ' nutraceutical'], 'KINGDOM': 'Organic_compounds', 'DRUGBANK_NAME': 'Histidine', 'id': 'DB00117', 'TYPE': 'DRUG', 'MolecularWeight': 155.15699999999998, 'SUBCLASS': 'Amino_acids__peptides__and_analogues', 'SUPERCLASS': 'Organic_acids_and_derivatives', 'TPSA': 92.00000000000001, 'NumHDonors': 3, 'NAME': 'Histidine', 'SYNONYM': ['_S__4__2_Amino_2_carboxyethyl_imidazole', ' _S__a_Amino_1H_imidazole_4_propanoic_acid', ' _S__alpha_amino_1H_Imidazole_4_propanoic_acid', ' _S__alpha_Amino_1H_imidazole_4_propionic_acid', ' _S____amino_1H_Imidazole_4_propanoic_acid', ' HIS', ' Histidine', ' L_____histidine', ' L_Histidin', ' L_Histidine'], 'LipinskiViolationCount': 0, 'SOURCE': 'BIOKG', 'NumHAcceptors': 3, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Histidine_and_derivatives', 'ALTERNATIVE_PARENT': ['Amino_acids', ' Aralkylamines', ' Azacyclic_compounds', ' Carbonyl_compounds', ' Carboxylic_acids', ' Heteroaromatic_compounds', ' Hydrocarbon_derivatives', ' Imidazolyl_carboxylic_acids_and_derivatives', ' L_alpha_amino_acids', ' Monoalkylamines', ' Monocarboxylic_acids_and_derivatives', ' Organic_oxides', ' Organopnictogen_compounds']}}, {'p': {'DOMAIN': 'IPR021922', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0016841', ' GO:0004397'], 'ACTIVE_SITE': 'IPR022313', 'SHORT_NAME': 'Histidase', 'NAME': 'HUTH', 'FAMILY': ['IPR001106', ' IPR005921'], 'RELATED_KEYWORD': ['Alternative splicing', ' Disease mutation', ' Histidine metabolism', ' Lyase', ' Phosphoprotein', ' Polymorphism', ' Reference proteome'], 'PROTEIN_EXPRESSED_IN': ['Hippocampus__Neuronal_cells', ' Cerebral_cortex__Neuropil', ' Endometrium_1__Cells_in_endometrial_stroma', ' Kidney__Cells_in_glomeruli', ' Endometrium_2__Cells_in_endometrial_stroma', ' Caudate__Neuronal_cells', ' Spleen__Cells_in_white_pulp', ' Hippocampus', ' Caudate', ' Lymph_node__Germinal_center_cells', ' Breast__Adipocytes', ' Tonsil__Squamous_epithelial_cells', ' Bronchus', ' Adrenal_gland', ' Cerebellum__Purkinje_cells', ' Seminal_vesicle__Glandular_cells', ' Fallopian_tube', ' Salivary_gland__Glandular_cells', ' Endometrium_2', ' Skeletal_muscle', ' Pancreas', ' Bronchus__Respiratory_epithelial_cells', ' Appendix__Lymphoid_tissue', ' Esophagus__Squamous_epithelial_cells', ' Seminal_vesicle', ' Appendix__Glandular_cells', ' Fallopian_tube__Glandular_cells', ' Urinary_bladder__Urothelial_cells', ' Nasopharynx__Respiratory_epithelial_cells', ' Lymph_node', ' Skeletal_muscle__Myocytes', ' Tonsil', ' Bone_marrow__Hematopoietic_cells', ' Parathyroid_gland', ' Small_intestine', ' Lymph_node__Non-germinal_center_cells', ' Tonsil__Germinal_center_cells', ' Oral_mucosa__Squamous_epithelial_cells', ' Stomach_2__Glandular_cells', ' Skin_1__Fibroblasts', ' Parathyroid_gland__Glandular_cells', ' Cervix', '_uterine__Glandular_cells', ' Testis', ' Rectum__Glandular_cells', ' Cerebellum', ' Appendix', ' Soft_tissue_2', ' Prostate__Glandular_cells', ' Skin_2__Epidermal_cells', ' Adrenal_gland__Glandular_cells', ' Esophagus', ' Colon', ' Breast', ' Gallbladder', ' Prostate', ' Epididymis__Glandular_cells', ' Small_intestine__Glandular_cells', ' Skin_2', ' Adipose_tissue__Adipocytes', ' Stomach_1__Glandular_cells', ' Stomach_1', ' Colon__Glandular_cells', ' Placenta__Trophoblastic_cells', ' Duodenum', ' Endometrium_1__Glandular_cells', ' Testis__Cells_in_seminiferous_ducts', ' Soft_tissue_1', ' Colon__Peripheral_nerve/ganglion', ' Heart_muscle__Myocytes', ' Breast__Myoepithelial_cells', ' Kidney', ' Soft_tissue_2__Peripheral_nerve', ' Endometrium_1', ' Thyroid_gland__Glandular_cells', ' Testis__Leydig_cells', ' Epididymis', ' Cerebral_cortex', ' Skin_1__Langerhans', ' Pancreas__Exocrine_glandular_cells', ' Vagina__Squamous_epithelial_cells', ' Soft_tissue_1__Fibroblasts', ' Oral_mucosa', ' Bone_marrow', ' Heart_muscle', ' Nasopharynx', ' Urinary_bladder', ' Salivary_gland', ' Endometrium_2__Glandular_cells', ' Spleen__Cells_in_red_pulp', ' Lung__Macrophages', ' Stomach_2', ' Placenta', ' Smooth_muscle', ' Cervix', '_uterine', ' Spleen', ' Vagina', ' Soft_tissue_2__Fibroblasts', ' Duodenum__Glandular_cells', ' Adipose_tissue', ' Breast__Glandular_cells', ' Cerebral_cortex__Neuronal_cells', ' Smooth_muscle__Smooth_muscle_cells', ' Pancreas__Islets_of_Langerhans', ' Gallbladder__Glandular_cells', ' Cerebellum__Cells_in_granular_layer', ' Lung', ' Thyroid_gland', ' Rectum', ' Lung__Pneumocytes', ' Colon__Endothelial_cells', ' Cervix', '_uterine__Squamous_epithelial_cells', ' Kidney__Cells_in_tubules', ' Liver', ' Liver__Hepatocytes', ' Skin_1__Keratinocytes', ' Skin_1', ' Skin_1__Melanocytes'], 'OTHER_ID': ['B4E0V8', ' F5GXF2', ' F5H1U5', ' Q4VB92', ' Q4VB93'], 'RELATED_PUBMED_ID': ['pubmed:7916645', ' pubmed:8530107', ' pubmed:14702039', ' pubmed:16541075', ' pubmed:15489334', ' pubmed:24275569', ' pubmed:15806399'], 'FULL_NAME': 'Histidine ammonia-lyase', 'GO_CC': 'GO:0005829', 'HOMOLOGOUS_SUPERFAMILY': ['IPR024083', ' IPR008948'], 'id': 'P42357', 'GO_MF': ['GO:0006548', ' GO:0019556', ' GO:0019557'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}, 'd': {'NumRotatableBonds': 3, 'DRUG_INDICATION_ASSOCIATION': nan, 'C_SMILES': 'N[C@@H](Cc1c[nH]cn1)C(=O)O', 'LogP': -0.6359000000000001, 'CLASS': 'Carboxylic_acids_and_derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGINCHKE': 'HNDVDQJCIGZPNO-YFKPBYRVSA-N', 'DRUGCOMP': nan, 'COMPCLAS': nan, 'DRUGINCH': ['InChI=1S/C6H9N3O2/c7-5(6(10)11)1-4-2-8-3-9-4/h2-3', '5H', '1', '7H2', '(H', '8', '9)(H', '10', '11)/t5-/m0/s1'], 'SUBSTITUENT': ['Alpha_amino_acid', ' Amine', ' Amino_acid', ' Aralkylamine', ' Aromatic_heteromonocyclic_compound', ' Azacycle', ' Azole', ' Carbonyl_group', ' Carboxylic_acid', ' Heteroaromatic_compound', ' Histidine_or_derivatives', ' Hydrocarbon_derivative', ' Imidazole', ' Imidazolyl_carboxylic_acid_derivative', ' L_alpha_amino_acid', ' Monocarboxylic_acid_or_derivatives', ' Organic_nitrogen_compound', ' Organic_oxide', ' Organic_oxygen_compound', ' Organoheterocyclic_compound', ' Organonitrogen_compound', ' Organooxygen_compound', ' Organopnictogen_compound', ' Primary_aliphatic_amine', ' Primary_amine'], 'PUBMED_ARTICLE': nan, 'DRUG_ATC_CODE': nan, 'PRODUCT': ['2_5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose', ' 2_5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose_Clinimix', ' 2_5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_10__Dextrose_Quickmix', ' 2_5_travasol_Amino_Acid_InJ_W_eleC_W_25_dex', ' 2_75__Travas__Amino_Acid_InJ_W_elecw_25_dex', ' 2_75__Travasol_Amino_Acid_Injection_With_Electrolytes_In_25__Dextrose_Quickmix', ' 2_75__Travasol_Amino_Acid_Injection_With_Electrolytes_In_5__Dextrose_Quickmix', ' 2_75_travasol_Amino_Acid_InJ_W_eleC_W_5_dex_', ' 20__Prosol', ' 4_25__Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_25__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_5__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrlytes_In_10__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_25__Dextrose', ' 4_25__Travasol_Amino_Acid_Without_Electrolytes_In_5__Dextrose_Quickmix', ' 4_25_trav__Amino_Acid_InJ_W_O_elect_5_dext_', ' 4_25_travasol_Amino_Acid_InJ_W_elecw_5_dex_', ' 5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose', ' 5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose', ' 5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_25__Dextrose', ' 8_5__Freamine_III', ' Aminosyn', ' Aminosyn_10__W_Electrolytes', ' Aminosyn_3_5__M', ' Aminosyn_5_', ' Aminosyn_7_', ' Aminosyn_8_5_', ' Aminosyn_8_5__Injection_With_Electrolytes', ' Aminosyn_HF', ' Aminosyn_II', ' Aminosyn_II_10__With_Electrolytes', ' Aminosyn_II_5__Inj', ' Aminosyn_II_7__Injection', ' Aminosyn_II_7__M_In_10__Dextrose_dual_Chamber_', ' Aminosyn_II_7__With_10__Dextrose', ' Aminosyn_II_7__With_50__Dextrose', ' Aminosyn_II_8_5__Injection', ' Aminosyn_II_8_5__M_In_20__Dextrose__dual_Chamber_', ' Aminosyn_II_8_5__With_50__Dextrose', ' Aminosyn_II_and_Dextrose', ' Aminosyn_II_in_Dextrose', ' Aminosyn_II_In_Dextrose_Injection', ' Aminosyn_II_with_Electrolytes', ' Aminosyn_II_with_Electrolytes_in_Dextrose_with_Calcium', ' Aminosyn_RF', ' Aminosyn_RF_Injection', ' Aminosyn_Sulfite_Free', ' Aminosyn_PF', ' Aminosyn_PF_10_', ' Aminosyn_PF_7_', ' CAREDO_Treating_Periodontitis_Toothpastes_100g', ' Clinimix', ' Clinimix_2_5__Travasol_Aa_Without_Electrolytes_In_10__Dextrose_Injec', ' Clinimix_E', ' Clinisol', ' FreAmine_HBC', ' FreAmine_III', ' HepatAmine', ' Hepatasol', ' Kabiven', ' NephrAmine', ' Nephramine_Liq_IV', ' Novamine', ' Nutrineal_Pd4', ' Olimel_3_3__E', ' Olimel_4_4_', ' Olimel_4_4__E', ' Olimel_5_7_', ' Olimel_5_7__E', ' Olimel_7_6_', ' Olimel_7_6__E', ' Perikabiven', ' Periolimel_2_5__E', ' Plenamine', ' Premasol___sulfite_free__Amino_Acid_', ' Primene', ' Primene_10_', ' Primene_10__liq_IV', ' ProcalAmine', ' ProSol', ' Quick_Mix_2_5__Travasol_Aa_With_Electrolytes_With_25__Dextrose_Injec', ' RenAmin', ' Renamin__amino_Acids__Injection', ' Smofkabiven', ' Smofkabiven_Electrolyte_Free', ' Smofkabiven_Peripheral', ' Synthamin', ' Travasol', ' Travasol_Amino_Acid_Inj_5_5_', ' Travasol_Amino_Acid_Inj_8_5_', ' Travasol_E', ' Travasol_Inj_Without_Electrolytes_5_5_', ' Travasol_Inj_Without_Electrolytes_8_5_', ' Trepoxen_250', ' Trepoxicam_7_5', ' TrophAmine', ' Vamin_18_Electrolyte_free', ' Vamin_N'], 'PRODUCT_STAGE': ['investigational', ' nutraceutical'], 'KINGDOM': 'Organic_compounds', 'DRUGBANK_NAME': 'Histidine', 'id': 'DB00117', 'TYPE': 'DRUG', 'MolecularWeight': 155.15699999999998, 'SUBCLASS': 'Amino_acids__peptides__and_analogues', 'SUPERCLASS': 'Organic_acids_and_derivatives', 'TPSA': 92.00000000000001, 'NumHDonors': 3, 'NAME': 'Histidine', 'SYNONYM': ['_S__4__2_Amino_2_carboxyethyl_imidazole', ' _S__a_Amino_1H_imidazole_4_propanoic_acid', ' _S__alpha_amino_1H_Imidazole_4_propanoic_acid', ' _S__alpha_Amino_1H_imidazole_4_propionic_acid', ' _S____amino_1H_Imidazole_4_propanoic_acid', ' HIS', ' Histidine', ' L_____histidine', ' L_Histidin', ' L_Histidine'], 'LipinskiViolationCount': 0, 'SOURCE': 'BIOKG', 'NumHAcceptors': 3, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Histidine_and_derivatives', 'ALTERNATIVE_PARENT': ['Amino_acids', ' Aralkylamines', ' Azacyclic_compounds', ' Carbonyl_compounds', ' Carboxylic_acids', ' Heteroaromatic_compounds', ' Hydrocarbon_derivatives', ' Imidazolyl_carboxylic_acids_and_derivatives', ' L_alpha_amino_acids', ' Monoalkylamines', ' Monocarboxylic_acids_and_derivatives', ' Organic_oxides', ' Organopnictogen_compounds']}}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Here are the important terms related to Histidine:\n",
      "\n",
      "1. HARS1 (Histidine--tRNA ligase)\n",
      "2. HDC (Histidine decarboxylase)\n",
      "3. HUTH (Histidine ammonia-lyase)\n",
      "4. Histidine (the drug)\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug {NAME: \"Histidine\"})-[:DRUG_TARGET|DRUG_ENZYME|DRUG_CARRIER|DRUG_TRANSPORTER]->(p:Protein) \n",
      "WHERE p.NAME IN [\"HARS1\", \"HDC\", \"HUTH\"]\n",
      "RETURN d, p;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'NumRotatableBonds': 3, 'DRUG_INDICATION_ASSOCIATION': nan, 'C_SMILES': 'N[C@@H](Cc1c[nH]cn1)C(=O)O', 'LogP': -0.6359000000000001, 'CLASS': 'Carboxylic_acids_and_derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGINCHKE': 'HNDVDQJCIGZPNO-YFKPBYRVSA-N', 'DRUGCOMP': nan, 'COMPCLAS': nan, 'DRUGINCH': ['InChI=1S/C6H9N3O2/c7-5(6(10)11)1-4-2-8-3-9-4/h2-3', '5H', '1', '7H2', '(H', '8', '9)(H', '10', '11)/t5-/m0/s1'], 'SUBSTITUENT': ['Alpha_amino_acid', ' Amine', ' Amino_acid', ' Aralkylamine', ' Aromatic_heteromonocyclic_compound', ' Azacycle', ' Azole', ' Carbonyl_group', ' Carboxylic_acid', ' Heteroaromatic_compound', ' Histidine_or_derivatives', ' Hydrocarbon_derivative', ' Imidazole', ' Imidazolyl_carboxylic_acid_derivative', ' L_alpha_amino_acid', ' Monocarboxylic_acid_or_derivatives', ' Organic_nitrogen_compound', ' Organic_oxide', ' Organic_oxygen_compound', ' Organoheterocyclic_compound', ' Organonitrogen_compound', ' Organooxygen_compound', ' Organopnictogen_compound', ' Primary_aliphatic_amine', ' Primary_amine'], 'PUBMED_ARTICLE': nan, 'DRUG_ATC_CODE': nan, 'PRODUCT': ['2_5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose', ' 2_5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose_Clinimix', ' 2_5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_10__Dextrose_Quickmix', ' 2_5_travasol_Amino_Acid_InJ_W_eleC_W_25_dex', ' 2_75__Travas__Amino_Acid_InJ_W_elecw_25_dex', ' 2_75__Travasol_Amino_Acid_Injection_With_Electrolytes_In_25__Dextrose_Quickmix', ' 2_75__Travasol_Amino_Acid_Injection_With_Electrolytes_In_5__Dextrose_Quickmix', ' 2_75_travasol_Amino_Acid_InJ_W_eleC_W_5_dex_', ' 20__Prosol', ' 4_25__Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_25__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_5__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrlytes_In_10__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_25__Dextrose', ' 4_25__Travasol_Amino_Acid_Without_Electrolytes_In_5__Dextrose_Quickmix', ' 4_25_trav__Amino_Acid_InJ_W_O_elect_5_dext_', ' 4_25_travasol_Amino_Acid_InJ_W_elecw_5_dex_', ' 5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose', ' 5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose', ' 5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_25__Dextrose', ' 8_5__Freamine_III', ' Aminosyn', ' Aminosyn_10__W_Electrolytes', ' Aminosyn_3_5__M', ' Aminosyn_5_', ' Aminosyn_7_', ' Aminosyn_8_5_', ' Aminosyn_8_5__Injection_With_Electrolytes', ' Aminosyn_HF', ' Aminosyn_II', ' Aminosyn_II_10__With_Electrolytes', ' Aminosyn_II_5__Inj', ' Aminosyn_II_7__Injection', ' Aminosyn_II_7__M_In_10__Dextrose_dual_Chamber_', ' Aminosyn_II_7__With_10__Dextrose', ' Aminosyn_II_7__With_50__Dextrose', ' Aminosyn_II_8_5__Injection', ' Aminosyn_II_8_5__M_In_20__Dextrose__dual_Chamber_', ' Aminosyn_II_8_5__With_50__Dextrose', ' Aminosyn_II_and_Dextrose', ' Aminosyn_II_in_Dextrose', ' Aminosyn_II_In_Dextrose_Injection', ' Aminosyn_II_with_Electrolytes', ' Aminosyn_II_with_Electrolytes_in_Dextrose_with_Calcium', ' Aminosyn_RF', ' Aminosyn_RF_Injection', ' Aminosyn_Sulfite_Free', ' Aminosyn_PF', ' Aminosyn_PF_10_', ' Aminosyn_PF_7_', ' CAREDO_Treating_Periodontitis_Toothpastes_100g', ' Clinimix', ' Clinimix_2_5__Travasol_Aa_Without_Electrolytes_In_10__Dextrose_Injec', ' Clinimix_E', ' Clinisol', ' FreAmine_HBC', ' FreAmine_III', ' HepatAmine', ' Hepatasol', ' Kabiven', ' NephrAmine', ' Nephramine_Liq_IV', ' Novamine', ' Nutrineal_Pd4', ' Olimel_3_3__E', ' Olimel_4_4_', ' Olimel_4_4__E', ' Olimel_5_7_', ' Olimel_5_7__E', ' Olimel_7_6_', ' Olimel_7_6__E', ' Perikabiven', ' Periolimel_2_5__E', ' Plenamine', ' Premasol___sulfite_free__Amino_Acid_', ' Primene', ' Primene_10_', ' Primene_10__liq_IV', ' ProcalAmine', ' ProSol', ' Quick_Mix_2_5__Travasol_Aa_With_Electrolytes_With_25__Dextrose_Injec', ' RenAmin', ' Renamin__amino_Acids__Injection', ' Smofkabiven', ' Smofkabiven_Electrolyte_Free', ' Smofkabiven_Peripheral', ' Synthamin', ' Travasol', ' Travasol_Amino_Acid_Inj_5_5_', ' Travasol_Amino_Acid_Inj_8_5_', ' Travasol_E', ' Travasol_Inj_Without_Electrolytes_5_5_', ' Travasol_Inj_Without_Electrolytes_8_5_', ' Trepoxen_250', ' Trepoxicam_7_5', ' TrophAmine', ' Vamin_18_Electrolyte_free', ' Vamin_N'], 'PRODUCT_STAGE': ['investigational', ' nutraceutical'], 'KINGDOM': 'Organic_compounds', 'DRUGBANK_NAME': 'Histidine', 'id': 'DB00117', 'TYPE': 'DRUG', 'MolecularWeight': 155.15699999999998, 'SUBCLASS': 'Amino_acids__peptides__and_analogues', 'SUPERCLASS': 'Organic_acids_and_derivatives', 'TPSA': 92.00000000000001, 'NumHDonors': 3, 'NAME': 'Histidine', 'SYNONYM': ['_S__4__2_Amino_2_carboxyethyl_imidazole', ' _S__a_Amino_1H_imidazole_4_propanoic_acid', ' _S__alpha_amino_1H_Imidazole_4_propanoic_acid', ' _S__alpha_Amino_1H_imidazole_4_propionic_acid', ' _S____amino_1H_Imidazole_4_propanoic_acid', ' HIS', ' Histidine', ' L_____histidine', ' L_Histidin', ' L_Histidine'], 'LipinskiViolationCount': 0, 'SOURCE': 'BIOKG', 'NumHAcceptors': 3, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Histidine_and_derivatives', 'ALTERNATIVE_PARENT': ['Amino_acids', ' Aralkylamines', ' Azacyclic_compounds', ' Carbonyl_compounds', ' Carboxylic_acids', ' Heteroaromatic_compounds', ' Hydrocarbon_derivatives', ' Imidazolyl_carboxylic_acids_and_derivatives', ' L_alpha_amino_acids', ' Monoalkylamines', ' Monocarboxylic_acids_and_derivatives', ' Organic_oxides', ' Organopnictogen_compounds']}, 'p': {'DOMAIN': ['IPR006195', ' IPR004154', ' IPR041715', ' IPR033656', ' IPR000738'], 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0005524', ' GO:0004821', ' GO:0042802', ' GO:0042803'], 'SHORT_NAME': 'HisRS', 'NAME': 'HARS1', 'FAMILY': ['IPR015807', ' IPR004516'], 'RELATED_KEYWORD': ['3D-structure', ' Acetylation', ' Alternative splicing', ' Aminoacyl-tRNA synthetase', ' ATP-binding', ' Charcot-Marie-Tooth disease', ' Cytoplasm', ' Deafness', ' Disease mutation', ' Ligase', ' Neurodegeneration', ' Neuropathy', ' Nucleotide-binding', ' Phosphoprotein', ' Polymorphism', ' Protein biosynthesis', ' Reference proteome', ' Retinitis pigmentosa', ' Usher syndrome'], 'PROTEIN_EXPRESSED_IN': ['Smooth_muscle__Smooth_muscle_cells', ' Cerebellum__Cells_in_granular_layer', ' Caudate__Glial_cells', ' Colon__Endothelial_cells', ' Hippocampus__Glial_cells', ' Cerebral_cortex__Neuropil', ' Smooth_muscle', ' Pancreas__Islets_of_Langerhans', ' Hippocampus__Neuronal_cells', ' Cerebral_cortex__Endothelial_cells', ' Hippocampus', ' Caudate__Neuronal_cells', ' Cerebral_cortex', ' Ovary__Ovarian_stroma_cells', ' Spleen__Cells_in_red_pulp', ' Endometrium_1__Cells_in_endometrial_stroma', ' Soft_tissue_2__Fibroblasts', ' Thyroid_gland', ' Liver__Bile_duct_cells', ' Placenta__Trophoblastic_cells', ' Cervix', '_uterine__Squamous_epithelial_cells', ' Thyroid_gland__Glandular_cells', ' Lung', ' Kidney', ' Cervix', '_uterine', ' Skin_1__Keratinocytes', ' Lung__Pneumocytes', ' Testis__Leydig_cells', ' Soft_tissue_2', ' Kidney__Cells_in_tubules', ' Skin_1', ' Kidney__Cells_in_glomeruli', ' Skin_1__Melanocytes', ' Cerebral_cortex__Neuronal_cells', ' Cervix', '_uterine__Glandular_cells', ' Cerebellum__Cells_in_molecular_layer', ' Adrenal_gland__Glandular_cells', ' Soft_tissue_2__Chondrocytes', ' Placenta', ' Endometrium_2__Cells_in_endometrial_stroma', ' Skin_1__Fibroblasts', ' Caudate', ' Adrenal_gland', ' Colon__Peripheral_nerve/ganglion', ' Skin_1__Langerhans', ' Liver__Hepatocytes', ' Lung__Macrophages', ' Placenta__Decidual_cells', ' Cerebral_cortex__Glial_cells', ' Liver', ' Lymph_node', ' Skin_2__Epidermal_cells', ' Rectum__Glandular_cells', ' Appendix__Lymphoid_tissue', ' Spleen__Cells_in_white_pulp', ' Endometrium_2', ' Endometrium_1', ' Bronchus__Respiratory_epithelial_cells', ' Cerebellum', ' Pancreas', ' Nasopharynx__Respiratory_epithelial_cells', ' Prostate__Glandular_cells', ' Breast', ' Rectum', ' Gallbladder', ' Bronchus', ' Tonsil__Germinal_center_cells', ' Lymph_node__Germinal_center_cells', ' Duodenum', ' Ovary__Follicle_cells', ' Vagina__Squamous_epithelial_cells', ' Tonsil__Non-germinal_center_cells', ' Testis__Cells_in_seminiferous_ducts', ' Oral_mucosa__Squamous_epithelial_cells', ' Appendix', ' Stomach_2__Glandular_cells', ' Lymph_node__Non-germinal_center_cells', ' Endometrium_1__Glandular_cells', ' Colon', ' Bone_marrow__Hematopoietic_cells', ' Fallopian_tube__Glandular_cells', ' Esophagus__Squamous_epithelial_cells', ' Parathyroid_gland', ' Gallbladder__Glandular_cells', ' Spleen', ' Endometrium_2__Glandular_cells', ' Parathyroid_gland__Glandular_cells', ' Salivary_gland', ' Salivary_gland__Glandular_cells', ' Tonsil__Squamous_epithelial_cells', ' Small_intestine__Glandular_cells', ' Prostate', ' Pancreas__Exocrine_glandular_cells', ' Seminal_vesicle__Glandular_cells', ' Esophagus', ' Breast__Glandular_cells', ' Epididymis', ' Seminal_vesicle', ' Vagina', ' Skin_2', ' Small_intestine', ' Stomach_1__Glandular_cells', ' Cerebellum__Purkinje_cells', ' Nasopharynx', ' Testis', ' Colon__Glandular_cells', ' Urinary_bladder', ' Stomach_1', ' Epididymis__Glandular_cells', ' Stomach_2', ' Oral_mucosa', ' Tonsil', ' Urinary_bladder__Urothelial_cells', ' Duodenum__Glandular_cells', ' Fallopian_tube', ' Appendix__Glandular_cells', ' Bone_marrow', ' Ovary'], 'OTHER_ID': ['B4DY73', ' D6REN6', ' J3KNE5'], 'RELATED_PUBMED_ID': ['pubmed:1549469', ' pubmed:3554142', ' pubmed:14702039', ' pubmed:15372022', ' pubmed:15489334', ' pubmed:8406012', ' pubmed:7755634', ' pubmed:19413330', ' pubmed:21269460', ' pubmed:22930593', ' pubmed:24275569', ' pubmed:26072516', ' pubmed:25151410', ' pubmed:22279524', ' pubmed:29235198'], 'FULL_NAME': ['Histidine--tRNA ligase', ' cytoplasmic', ' Histidyl-tRNA synthetase'], 'GO_CC': ['GO:0005737', ' GO:0005829', ' GO:0005739'], 'HOMOLOGOUS_SUPERFAMILY': ['IPR036621', ' IPR009068'], 'id': 'P12081', 'GO_MF': ['GO:0006427', ' GO:0032543', ' GO:0006412', ' GO:0006418'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}, {'d': {'NumRotatableBonds': 3, 'DRUG_INDICATION_ASSOCIATION': nan, 'C_SMILES': 'N[C@@H](Cc1c[nH]cn1)C(=O)O', 'LogP': -0.6359000000000001, 'CLASS': 'Carboxylic_acids_and_derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGINCHKE': 'HNDVDQJCIGZPNO-YFKPBYRVSA-N', 'DRUGCOMP': nan, 'COMPCLAS': nan, 'DRUGINCH': ['InChI=1S/C6H9N3O2/c7-5(6(10)11)1-4-2-8-3-9-4/h2-3', '5H', '1', '7H2', '(H', '8', '9)(H', '10', '11)/t5-/m0/s1'], 'SUBSTITUENT': ['Alpha_amino_acid', ' Amine', ' Amino_acid', ' Aralkylamine', ' Aromatic_heteromonocyclic_compound', ' Azacycle', ' Azole', ' Carbonyl_group', ' Carboxylic_acid', ' Heteroaromatic_compound', ' Histidine_or_derivatives', ' Hydrocarbon_derivative', ' Imidazole', ' Imidazolyl_carboxylic_acid_derivative', ' L_alpha_amino_acid', ' Monocarboxylic_acid_or_derivatives', ' Organic_nitrogen_compound', ' Organic_oxide', ' Organic_oxygen_compound', ' Organoheterocyclic_compound', ' Organonitrogen_compound', ' Organooxygen_compound', ' Organopnictogen_compound', ' Primary_aliphatic_amine', ' Primary_amine'], 'PUBMED_ARTICLE': nan, 'DRUG_ATC_CODE': nan, 'PRODUCT': ['2_5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose', ' 2_5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose_Clinimix', ' 2_5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_10__Dextrose_Quickmix', ' 2_5_travasol_Amino_Acid_InJ_W_eleC_W_25_dex', ' 2_75__Travas__Amino_Acid_InJ_W_elecw_25_dex', ' 2_75__Travasol_Amino_Acid_Injection_With_Electrolytes_In_25__Dextrose_Quickmix', ' 2_75__Travasol_Amino_Acid_Injection_With_Electrolytes_In_5__Dextrose_Quickmix', ' 2_75_travasol_Amino_Acid_InJ_W_eleC_W_5_dex_', ' 20__Prosol', ' 4_25__Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_10__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_25__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_With_Electrolytes_In_5__Dextrose_Quickmix', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrlytes_In_10__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose', ' 4_25__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_25__Dextrose', ' 4_25__Travasol_Amino_Acid_Without_Electrolytes_In_5__Dextrose_Quickmix', ' 4_25_trav__Amino_Acid_InJ_W_O_elect_5_dext_', ' 4_25_travasol_Amino_Acid_InJ_W_elecw_5_dex_', ' 5__Travasol_Amino_Acid_Injection_With_Electrolytes_In_20__Dextrose', ' 5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_20__Dextrose', ' 5__Travasol_Amino_Acid_Injection_Without_Electrolytes_In_25__Dextrose', ' 8_5__Freamine_III', ' Aminosyn', ' Aminosyn_10__W_Electrolytes', ' Aminosyn_3_5__M', ' Aminosyn_5_', ' Aminosyn_7_', ' Aminosyn_8_5_', ' Aminosyn_8_5__Injection_With_Electrolytes', ' Aminosyn_HF', ' Aminosyn_II', ' Aminosyn_II_10__With_Electrolytes', ' Aminosyn_II_5__Inj', ' Aminosyn_II_7__Injection', ' Aminosyn_II_7__M_In_10__Dextrose_dual_Chamber_', ' Aminosyn_II_7__With_10__Dextrose', ' Aminosyn_II_7__With_50__Dextrose', ' Aminosyn_II_8_5__Injection', ' Aminosyn_II_8_5__M_In_20__Dextrose__dual_Chamber_', ' Aminosyn_II_8_5__With_50__Dextrose', ' Aminosyn_II_and_Dextrose', ' Aminosyn_II_in_Dextrose', ' Aminosyn_II_In_Dextrose_Injection', ' Aminosyn_II_with_Electrolytes', ' Aminosyn_II_with_Electrolytes_in_Dextrose_with_Calcium', ' Aminosyn_RF', ' Aminosyn_RF_Injection', ' Aminosyn_Sulfite_Free', ' Aminosyn_PF', ' Aminosyn_PF_10_', ' Aminosyn_PF_7_', ' CAREDO_Treating_Periodontitis_Toothpastes_100g', ' Clinimix', ' Clinimix_2_5__Travasol_Aa_Without_Electrolytes_In_10__Dextrose_Injec', ' Clinimix_E', ' Clinisol', ' FreAmine_HBC', ' FreAmine_III', ' HepatAmine', ' Hepatasol', ' Kabiven', ' NephrAmine', ' Nephramine_Liq_IV', ' Novamine', ' Nutrineal_Pd4', ' Olimel_3_3__E', ' Olimel_4_4_', ' Olimel_4_4__E', ' Olimel_5_7_', ' Olimel_5_7__E', ' Olimel_7_6_', ' Olimel_7_6__E', ' Perikabiven', ' Periolimel_2_5__E', ' Plenamine', ' Premasol___sulfite_free__Amino_Acid_', ' Primene', ' Primene_10_', ' Primene_10__liq_IV', ' ProcalAmine', ' ProSol', ' Quick_Mix_2_5__Travasol_Aa_With_Electrolytes_With_25__Dextrose_Injec', ' RenAmin', ' Renamin__amino_Acids__Injection', ' Smofkabiven', ' Smofkabiven_Electrolyte_Free', ' Smofkabiven_Peripheral', ' Synthamin', ' Travasol', ' Travasol_Amino_Acid_Inj_5_5_', ' Travasol_Amino_Acid_Inj_8_5_', ' Travasol_E', ' Travasol_Inj_Without_Electrolytes_5_5_', ' Travasol_Inj_Without_Electrolytes_8_5_', ' Trepoxen_250', ' Trepoxicam_7_5', ' TrophAmine', ' Vamin_18_Electrolyte_free', ' Vamin_N'], 'PRODUCT_STAGE': ['investigational', ' nutraceutical'], 'KINGDOM': 'Organic_compounds', 'DRUGBANK_NAME': 'Histidine', 'id': 'DB00117', 'TYPE': 'DRUG', 'MolecularWeight': 155.15699999999998, 'SUBCLASS': 'Amino_acids__peptides__and_analogues', 'SUPERCLASS': 'Organic_acids_and_derivatives', 'TPSA': 92.00000000000001, 'NumHDonors': 3, 'NAME': 'Histidine', 'SYNONYM': ['_S__4__2_Amino_2_carboxyethyl_imidazole', ' _S__a_Amino_1H_imidazole_4_propanoic_acid', ' _S__alpha_amino_1H_Imidazole_4_propanoic_acid', ' _S__alpha_Amino_1H_imidazole_4_propionic_acid', ' _S____amino_1H_Imidazole_4_propanoic_acid', ' HIS', ' Histidine', ' L_____histidine', ' L_Histidin', ' L_Histidine'], 'LipinskiViolationCount': 0, 'SOURCE': 'BIOKG', 'NumHAcceptors': 3, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Histidine_and_derivatives', 'ALTERNATIVE_PARENT': ['Amino_acids', ' Aralkylamines', ' Azacyclic_compounds', ' Carbonyl_compounds', ' Carboxylic_acids', ' Heteroaromatic_compounds', ' Hydrocarbon_derivatives', ' Imidazolyl_carboxylic_acids_and_derivatives', ' L_alpha_amino_acids', ' Monoalkylamines', ' Monocarboxylic_acids_and_derivatives', ' Organic_oxides', ' Organopnictogen_compounds']}, 'p': {'DOMAIN': 'IPR021922', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0016841', ' GO:0004397'], 'ACTIVE_SITE': 'IPR022313', 'SHORT_NAME': 'Histidase', 'NAME': 'HUTH', 'FAMILY': ['IPR001106', ' IPR005921'], 'RELATED_KEYWORD': ['Alternative splicing', ' Disease mutation', ' Histidine metabolism', ' Lyase', ' Phosphoprotein', ' Polymorphism', ' Reference proteome'], 'PROTEIN_EXPRESSED_IN': ['Hippocampus__Neuronal_cells', ' Cerebral_cortex__Neuropil', ' Endometrium_1__Cells_in_endometrial_stroma', ' Kidney__Cells_in_glomeruli', ' Endometrium_2__Cells_in_endometrial_stroma', ' Caudate__Neuronal_cells', ' Spleen__Cells_in_white_pulp', ' Hippocampus', ' Caudate', ' Lymph_node__Germinal_center_cells', ' Breast__Adipocytes', ' Tonsil__Squamous_epithelial_cells', ' Bronchus', ' Adrenal_gland', ' Cerebellum__Purkinje_cells', ' Seminal_vesicle__Glandular_cells', ' Fallopian_tube', ' Salivary_gland__Glandular_cells', ' Endometrium_2', ' Skeletal_muscle', ' Pancreas', ' Bronchus__Respiratory_epithelial_cells', ' Appendix__Lymphoid_tissue', ' Esophagus__Squamous_epithelial_cells', ' Seminal_vesicle', ' Appendix__Glandular_cells', ' Fallopian_tube__Glandular_cells', ' Urinary_bladder__Urothelial_cells', ' Nasopharynx__Respiratory_epithelial_cells', ' Lymph_node', ' Skeletal_muscle__Myocytes', ' Tonsil', ' Bone_marrow__Hematopoietic_cells', ' Parathyroid_gland', ' Small_intestine', ' Lymph_node__Non-germinal_center_cells', ' Tonsil__Germinal_center_cells', ' Oral_mucosa__Squamous_epithelial_cells', ' Stomach_2__Glandular_cells', ' Skin_1__Fibroblasts', ' Parathyroid_gland__Glandular_cells', ' Cervix', '_uterine__Glandular_cells', ' Testis', ' Rectum__Glandular_cells', ' Cerebellum', ' Appendix', ' Soft_tissue_2', ' Prostate__Glandular_cells', ' Skin_2__Epidermal_cells', ' Adrenal_gland__Glandular_cells', ' Esophagus', ' Colon', ' Breast', ' Gallbladder', ' Prostate', ' Epididymis__Glandular_cells', ' Small_intestine__Glandular_cells', ' Skin_2', ' Adipose_tissue__Adipocytes', ' Stomach_1__Glandular_cells', ' Stomach_1', ' Colon__Glandular_cells', ' Placenta__Trophoblastic_cells', ' Duodenum', ' Endometrium_1__Glandular_cells', ' Testis__Cells_in_seminiferous_ducts', ' Soft_tissue_1', ' Colon__Peripheral_nerve/ganglion', ' Heart_muscle__Myocytes', ' Breast__Myoepithelial_cells', ' Kidney', ' Soft_tissue_2__Peripheral_nerve', ' Endometrium_1', ' Thyroid_gland__Glandular_cells', ' Testis__Leydig_cells', ' Epididymis', ' Cerebral_cortex', ' Skin_1__Langerhans', ' Pancreas__Exocrine_glandular_cells', ' Vagina__Squamous_epithelial_cells', ' Soft_tissue_1__Fibroblasts', ' Oral_mucosa', ' Bone_marrow', ' Heart_muscle', ' Nasopharynx', ' Urinary_bladder', ' Salivary_gland', ' Endometrium_2__Glandular_cells', ' Spleen__Cells_in_red_pulp', ' Lung__Macrophages', ' Stomach_2', ' Placenta', ' Smooth_muscle', ' Cervix', '_uterine', ' Spleen', ' Vagina', ' Soft_tissue_2__Fibroblasts', ' Duodenum__Glandular_cells', ' Adipose_tissue', ' Breast__Glandular_cells', ' Cerebral_cortex__Neuronal_cells', ' Smooth_muscle__Smooth_muscle_cells', ' Pancreas__Islets_of_Langerhans', ' Gallbladder__Glandular_cells', ' Cerebellum__Cells_in_granular_layer', ' Lung', ' Thyroid_gland', ' Rectum', ' Lung__Pneumocytes', ' Colon__Endothelial_cells', ' Cervix', '_uterine__Squamous_epithelial_cells', ' Kidney__Cells_in_tubules', ' Liver', ' Liver__Hepatocytes', ' Skin_1__Keratinocytes', ' Skin_1', ' Skin_1__Melanocytes'], 'OTHER_ID': ['B4E0V8', ' F5GXF2', ' F5H1U5', ' Q4VB92', ' Q4VB93'], 'RELATED_PUBMED_ID': ['pubmed:7916645', ' pubmed:8530107', ' pubmed:14702039', ' pubmed:16541075', ' pubmed:15489334', ' pubmed:24275569', ' pubmed:15806399'], 'FULL_NAME': 'Histidine ammonia-lyase', 'GO_CC': 'GO:0005829', 'HOMOLOGOUS_SUPERFAMILY': ['IPR024083', ' IPR008948'], 'id': 'P42357', 'GO_MF': ['GO:0006548', ' GO:0019556', ' GO:0019557'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: Here are the important terms related to Histidine:\n",
      "\n",
      "1. HARS1 (Histidine--tRNA ligase)\n",
      "2. HDC (Histidine decarboxylase)\n",
      "3. HUTH (Histidine ammonia-lyase)\n",
      "4. Histidine (the drug)\n",
      "\n",
      "These terms are related to the biological functions and enzymes involved in histidine metabolism.\n",
      "You:  exit\n",
      "Goodbye!\n"
     ]
    }
   ],
   "source": [
    "chatbot.chat()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6561f6f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to the Medical Knowledge Chatbot! \n",
      "You:  Give the drugs for tuberculss\n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease {NAME: \"Tuberculosis\"})-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug) RETURN dr.NAME, dr.id;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'dr.NAME': 'Pyridoxine', 'dr.id': 'DB00165'}, {'dr.NAME': 'Aminosalicylic_acid', 'dr.id': 'DB00233'}, {'dr.NAME': 'Cycloserine', 'dr.id': 'DB00260'}, {'dr.NAME': 'Capreomycin', 'dr.id': 'DB00314'}, {'dr.NAME': 'Pyrazinamide', 'dr.id': 'DB00339'}, {'dr.NAME': 'Ethambutol', 'dr.id': 'DB00330'}, {'dr.NAME': 'Betamethasone', 'dr.id': 'DB00443'}, {'dr.NAME': 'Amikacin', 'dr.id': 'DB00479'}, {'dr.NAME': 'Prednisone', 'dr.id': 'DB00635'}, {'dr.NAME': 'Rifabutin', 'dr.id': 'DB00615'}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Pyridoxine, Aminosalicylic_acid, Cycloserine, Capreomycin, Pyrazinamide, Ethambutol, Amikacin, Rifabutin are drugs for tuberculosis.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease {NAME: \"Tuberculosis\"})-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug) \n",
      "WHERE dr.NAME IN [\"Pyridoxine\", \"Aminosalicylic_acid\", \"Cycloserine\", \"Capreomycin\", \"Pyrazinamide\", \"Ethambutol\", \"Amikacin\", \"Rifabutin\"] \n",
      "RETURN dr.NAME, dr.DRUGCOMP, dr.LogP, dr.MolecularWeight, dr.TPSA, dr.NumHDonors, dr.NumHAcceptors, dr.LipinskiViolationCount;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'dr.NAME': 'Pyridoxine', 'dr.DRUGCOMP': nan, 'dr.LogP': 0.08021999999999985, 'dr.MolecularWeight': 169.18, 'dr.TPSA': 73.58000000000001, 'dr.NumHDonors': 3, 'dr.NumHAcceptors': 4, 'dr.LipinskiViolationCount': 0}, {'dr.NAME': 'Aminosalicylic_acid', 'dr.DRUGCOMP': nan, 'dr.LogP': 0.6726000000000001, 'dr.MolecularWeight': 153.137, 'dr.TPSA': 83.55000000000001, 'dr.NumHDonors': 3, 'dr.NumHAcceptors': 3, 'dr.LipinskiViolationCount': 0}, {'dr.NAME': 'Cycloserine', 'dr.DRUGCOMP': nan, 'dr.LogP': -1.6249000000000007, 'dr.MolecularWeight': 102.09299999999999, 'dr.TPSA': 64.35, 'dr.NumHDonors': 2, 'dr.NumHAcceptors': 3, 'dr.LipinskiViolationCount': 0}, {'dr.NAME': 'Capreomycin', 'dr.DRUGCOMP': nan, 'dr.LogP': -15.441999999999899, 'dr.MolecularWeight': 1321.4349999999997, 'dr.TPSA': 736.6099999999998, 'dr.NumHDonors': 27, 'dr.NumHAcceptors': 27, 'dr.LipinskiViolationCount': 3}, {'dr.NAME': 'Pyrazinamide', 'dr.DRUGCOMP': nan, 'dr.LogP': -0.4245000000000001, 'dr.MolecularWeight': 123.11499999999998, 'dr.TPSA': 68.87, 'dr.NumHDonors': 1, 'dr.NumHAcceptors': 3, 'dr.LipinskiViolationCount': 0}, {'dr.NAME': 'Ethambutol', 'dr.DRUGCOMP': nan, 'dr.LogP': -0.292599999999999, 'dr.MolecularWeight': 204.314, 'dr.TPSA': 64.52, 'dr.NumHDonors': 4, 'dr.NumHAcceptors': 4, 'dr.LipinskiViolationCount': 0}, {'dr.NAME': 'Amikacin', 'dr.DRUGCOMP': nan, 'dr.LogP': -8.424200000000015, 'dr.MolecularWeight': 585.6080000000002, 'dr.TPSA': 331.93999999999994, 'dr.NumHDonors': 13, 'dr.NumHAcceptors': 17, 'dr.LipinskiViolationCount': 3}, {'dr.NAME': 'Rifabutin', 'dr.DRUGCOMP': nan, 'dr.LogP': 4.615720000000003, 'dr.MolecularWeight': 847.0190000000003, 'dr.TPSA': 205.54999999999995, 'dr.NumHDonors': 5, 'dr.NumHAcceptors': 14, 'dr.LipinskiViolationCount': 2}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: These are the names of drugs for tuberculosis: Pyridoxine, Aminosalicylic_acid, Cycloserine, Capreomycin, Pyrazinamide, Ethambutol, Amikacin, Rifabutin.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease {NAME: \"Tuberculosis\"})-[:DRUG_DISEASE_ASSOCIATION]-(drug:Drug) WHERE drug.NAME IN [\"Pyridoxine\", \"Aminosalicylic_acid\", \"Cycloserine\", \"Capreomycin\", \"Pyrazinamide\", \"Ethambutol\", \"Amikacin\", \"Rifabutin\"] RETURN drug.NAME AS DrugName;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'DrugName': 'Pyridoxine'}, {'DrugName': 'Aminosalicylic_acid'}, {'DrugName': 'Cycloserine'}, {'DrugName': 'Capreomycin'}, {'DrugName': 'Pyrazinamide'}, {'DrugName': 'Ethambutol'}, {'DrugName': 'Amikacin'}, {'DrugName': 'Rifabutin'}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: The important terms are Pyridoxine, Aminosalicylic_acid, Cycloserine, Capreomycin, Pyrazinamide, Ethambutol, Amikacin, Rifabutin.\n",
      "You:  exit\n",
      "Goodbye!\n"
     ]
    }
   ],
   "source": [
    "chatbot.chat()\n",
    "#ONE HOP QUERY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0e3df816",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to the Medical Knowledge Chatbot! \n",
      "You:  give me the proteins for histidn drug\n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug {NAME: \"Histidin\"})-[:DRUG_TARGET|DRUG_ENZYME|DRUG_CARRIER|DRUG_TRANSPORTER]->(p:Protein) RETURN p;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context_data_list [[]]\n",
      "context: []\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease)-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug) \n",
      "RETURN d.NAME, dr.NAME, dr.TYPE, dr.DRUG_INDICATION_ASSOCIATION, dr.LogP, dr.DRUGINKE, dr.TPSA, dr.NumHDonors, dr.NumHAcceptors, dr.LipinskiViolationCount, dr.MolecularWeight \n",
      "ORDER BY dr.LogP DESC;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d.NAME': 'Albuminuria', 'dr.NAME': 'Glucagon', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.LogP': nan, 'dr.DRUGINKE': nan, 'dr.TPSA': nan, 'dr.NumHDonors': nan, 'dr.NumHAcceptors': nan, 'dr.LipinskiViolationCount': nan, 'dr.MolecularWeight': nan}, {'d.NAME': ['Arrhythmias', ' Cardiac'], 'dr.NAME': 'Glucagon', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.LogP': nan, 'dr.DRUGINKE': nan, 'dr.TPSA': nan, 'dr.NumHDonors': nan, 'dr.NumHAcceptors': nan, 'dr.LipinskiViolationCount': nan, 'dr.MolecularWeight': nan}, {'d.NAME': 'Bradycardia', 'dr.NAME': 'Glucagon', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.LogP': nan, 'dr.DRUGINKE': nan, 'dr.TPSA': nan, 'dr.NumHDonors': nan, 'dr.NumHAcceptors': nan, 'dr.LipinskiViolationCount': nan, 'dr.MolecularWeight': nan}, {'d.NAME': 'Diabetic Nephropathies', 'dr.NAME': 'Glucagon', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.LogP': nan, 'dr.DRUGINKE': nan, 'dr.TPSA': nan, 'dr.NumHDonors': nan, 'dr.NumHAcceptors': nan, 'dr.LipinskiViolationCount': nan, 'dr.MolecularWeight': nan}, {'d.NAME': 'Fibrosis', 'dr.NAME': 'Glucagon', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.LogP': nan, 'dr.DRUGINKE': nan, 'dr.TPSA': nan, 'dr.NumHDonors': nan, 'dr.NumHAcceptors': nan, 'dr.LipinskiViolationCount': nan, 'dr.MolecularWeight': nan}, {'d.NAME': 'Heart Block', 'dr.NAME': 'Glucagon', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.LogP': nan, 'dr.DRUGINKE': nan, 'dr.TPSA': nan, 'dr.NumHDonors': nan, 'dr.NumHAcceptors': nan, 'dr.LipinskiViolationCount': nan, 'dr.MolecularWeight': nan}, {'d.NAME': 'Heart Diseases', 'dr.NAME': 'Glucagon', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.LogP': nan, 'dr.DRUGINKE': nan, 'dr.TPSA': nan, 'dr.NumHDonors': nan, 'dr.NumHAcceptors': nan, 'dr.LipinskiViolationCount': nan, 'dr.MolecularWeight': nan}, {'d.NAME': 'Hyperglycemia', 'dr.NAME': 'Glucagon', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.LogP': nan, 'dr.DRUGINKE': nan, 'dr.TPSA': nan, 'dr.NumHDonors': nan, 'dr.NumHAcceptors': nan, 'dr.LipinskiViolationCount': nan, 'dr.MolecularWeight': nan}, {'d.NAME': 'Hyperinsulinism', 'dr.NAME': 'Glucagon', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.LogP': nan, 'dr.DRUGINKE': nan, 'dr.TPSA': nan, 'dr.NumHDonors': nan, 'dr.NumHAcceptors': nan, 'dr.LipinskiViolationCount': nan, 'dr.MolecularWeight': nan}, {'d.NAME': 'Hypotension', 'dr.NAME': 'Glucagon', 'dr.TYPE': 'DRUG', 'dr.DRUG_INDICATION_ASSOCIATION': nan, 'dr.LogP': nan, 'dr.DRUGINKE': nan, 'dr.TPSA': nan, 'dr.NumHDonors': nan, 'dr.NumHAcceptors': nan, 'dr.LipinskiViolationCount': nan, 'dr.MolecularWeight': nan}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "combined_readable_context Here are the important terms: Albuminuria, Arrhythmias, Cardiac, Bradycardia, Diabetic Nephropathies, Fibrosis, Heart Block, Heart Diseases, Hyperglycemia, Hyperinsulinism, Hypotension, and Glucagon.\n",
      "context: \n",
      "Here are the important terms: Albuminuria, Arrhythmias, Cardiac, Bradycardia, Diabetic Nephropathies, Fibrosis, Heart Block, Heart Diseases, Hyperglycemia, Hyperinsulinism, Hypotension, and Glucagon.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease) WHERE d.NAME IN ['Albuminuria', 'Arrhythmias', 'Cardiac', 'Bradycardia', 'Diabetic Nephropathies', 'Fibrosis', 'Heart Block', 'Heart Diseases', 'Hyperglycemia', 'Hyperinsulinism', 'Hypotension', 'Glucagon'] RETURN d;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'DISEASE_SUPERGRP': ['C12.777.934', ' C12.777.934.734', ' C23.888.942.750.269', ' C13.351.968', ' C23', ' C23.888', ' C23.888.942.750', ' C23.888.942', ' C12.777', ' C13.351.968.934.734.269', ' C12.777.934.734.269', ' C13.351', ' C13', ' C13.351.968.934.734', ' C12', ' C13.351.968.934'], 'id': 'D000419', 'TYPE': 'DISEASE', 'NAME': 'Albuminuria'}}, {'d': {'DISEASE_SUPERGRP': ['C23.550.073', ' C23', ' C14', ' C23.550', ' C14.280.067.319', ' C14.280', ' C14.280.067', ' C23.550.073.300'], 'id': 'D001919', 'TYPE': 'DISEASE', 'NAME': 'Bradycardia'}}, {'d': {'DISEASE_SUPERGRP': ['C13.351.968', ' C13.351.968.419.192', ' C12.777', ' C19.246.099.875', ' C12.777.419.192', ' C13', ' C12.777.419', ' C19.246.099', ' C12', ' C13.351.968.419', ' C19.246', ' C13.351', ' C19'], 'id': 'D003928', 'TYPE': 'DISEASE', 'NAME': 'Diabetic Nephropathies'}}, {'d': {'DISEASE_SUPERGRP': ['C23', ' C23.550', ' C23.550.355'], 'id': 'D005355', 'TYPE': 'DISEASE', 'NAME': 'Fibrosis'}}, {'d': {'DISEASE_SUPERGRP': ['C14.280.067', ' C23', ' C23.550.073', ' C14.280.123.500', ' C14', ' C23.550', ' C23.550.073.425', ' C14.280', ' C14.280.067.558', ' C14.280.123'], 'id': 'D006327', 'TYPE': 'DISEASE', 'NAME': 'Heart Block'}}, {'d': {'DISEASE_SUPERGRP': ['C14', ' C14.280'], 'id': 'D006331', 'TYPE': 'DISEASE', 'NAME': 'Heart Diseases'}}, {'d': {'DISEASE_SUPERGRP': ['C18', ' C18.452', ' C18.452.394.952', ' C18.452.394'], 'id': 'D006943', 'TYPE': 'DISEASE', 'NAME': 'Hyperglycemia'}}, {'d': {'DISEASE_SUPERGRP': ['C18', ' C18.452.394', ' C18.452.394.968', ' C18.452'], 'id': 'D006946', 'TYPE': 'DISEASE', 'NAME': 'Hyperinsulinism'}}, {'d': {'DISEASE_SUPERGRP': ['C14.907.514', ' C14.907', ' C14'], 'id': 'D007022', 'TYPE': 'DISEASE', 'NAME': 'Hypotension'}}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Here are the important terms: Albuminuria, Bradycardia, Diabetic Nephropathies, Fibrosis, Heart Block, Heart Diseases, Hyperglycemia, Hyperinsulinism, Hypotension.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease) WHERE d.NAME IN ['Albuminuria', 'Bradycardia', 'Diabetic Nephropathies', 'Fibrosis', 'Heart Block', 'Heart Diseases', 'Hyperglycemia', 'Hyperinsulinism', 'Hypotension'] RETURN d;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'DISEASE_SUPERGRP': ['C12.777.934', ' C12.777.934.734', ' C23.888.942.750.269', ' C13.351.968', ' C23', ' C23.888', ' C23.888.942.750', ' C23.888.942', ' C12.777', ' C13.351.968.934.734.269', ' C12.777.934.734.269', ' C13.351', ' C13', ' C13.351.968.934.734', ' C12', ' C13.351.968.934'], 'id': 'D000419', 'TYPE': 'DISEASE', 'NAME': 'Albuminuria'}}, {'d': {'DISEASE_SUPERGRP': ['C23.550.073', ' C23', ' C14', ' C23.550', ' C14.280.067.319', ' C14.280', ' C14.280.067', ' C23.550.073.300'], 'id': 'D001919', 'TYPE': 'DISEASE', 'NAME': 'Bradycardia'}}, {'d': {'DISEASE_SUPERGRP': ['C13.351.968', ' C13.351.968.419.192', ' C12.777', ' C19.246.099.875', ' C12.777.419.192', ' C13', ' C12.777.419', ' C19.246.099', ' C12', ' C13.351.968.419', ' C19.246', ' C13.351', ' C19'], 'id': 'D003928', 'TYPE': 'DISEASE', 'NAME': 'Diabetic Nephropathies'}}, {'d': {'DISEASE_SUPERGRP': ['C23', ' C23.550', ' C23.550.355'], 'id': 'D005355', 'TYPE': 'DISEASE', 'NAME': 'Fibrosis'}}, {'d': {'DISEASE_SUPERGRP': ['C14.280.067', ' C23', ' C23.550.073', ' C14.280.123.500', ' C14', ' C23.550', ' C23.550.073.425', ' C14.280', ' C14.280.067.558', ' C14.280.123'], 'id': 'D006327', 'TYPE': 'DISEASE', 'NAME': 'Heart Block'}}, {'d': {'DISEASE_SUPERGRP': ['C14', ' C14.280'], 'id': 'D006331', 'TYPE': 'DISEASE', 'NAME': 'Heart Diseases'}}, {'d': {'DISEASE_SUPERGRP': ['C18', ' C18.452', ' C18.452.394.952', ' C18.452.394'], 'id': 'D006943', 'TYPE': 'DISEASE', 'NAME': 'Hyperglycemia'}}, {'d': {'DISEASE_SUPERGRP': ['C18', ' C18.452.394', ' C18.452.394.968', ' C18.452'], 'id': 'D006946', 'TYPE': 'DISEASE', 'NAME': 'Hyperinsulinism'}}, {'d': {'DISEASE_SUPERGRP': ['C14.907.514', ' C14.907', ' C14'], 'id': 'D007022', 'TYPE': 'DISEASE', 'NAME': 'Hypotension'}}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: Here are the important terms: Albuminuria, Bradycardia, Diabetic Nephropathies, Fibrosis, Heart Block, Heart Diseases, Hyperglycemia, Hyperinsulinism, Hypotension.\n",
      "You:  exit\n",
      "Goodbye!\n"
     ]
    }
   ],
   "source": [
    "chatbot.chat()\n",
    "# correcting the word  histidn drug to histidine drug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44aea08c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddbc01c5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8a02f013",
   "metadata": {},
   "source": [
    "SAME AS THE PREVIOUS JUST IT USES ALSO THE EMBEDDINGS OTHER THINGS ARE THE SAME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f3647f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import spacy\n",
    "from rapidfuzz import fuzz, process  # For fuzzy matching\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import numpy as np\n",
    "\n",
    "class ChatbotWithEmbeddings:\n",
    "    def __init__(self, chain, nlp_model=\"en_core_web_md\", embedding_model=None):\n",
    "        self.chain = chain\n",
    "        self.previous_context = \"\"\n",
    "        self.nlp = spacy.load(nlp_model)  # Load a robust NLP model\n",
    "        self.embedding_model = embedding_model  # Placeholder for any embedding model\n",
    "        self.standard_diseases = {\n",
    "            \"Tuberculosis\": [\"tb\", \"tuberculoses\", \"tubercular\", \"Tuberculosis\", \"tuberculosis\"],\n",
    "            \"Ovarian Cancer\": [\"ovarian cancer\", \"ovarian carcinoma\", \"ovarian neoplasm\", \"Ovarian cancer\", \"ovarian cancer\"],\n",
    "            # Add more diseases as needed\n",
    "        }\n",
    "\n",
    "        self.fallback_phrases = [\n",
    "            \"I don't know the answer\",\n",
    "            \"I'm not sure\",\n",
    "            \"I don't have enough information\",\n",
    "            \"I'm sorry, I can't help with that\"\n",
    "        ]\n",
    "        # Convert fallback phrases to spaCy Doc objects for similarity comparison\n",
    "        self.fallback_docs = [self.nlp(phrase) for phrase in self.fallback_phrases]\n",
    "\n",
    "    def nlp_process(self, user_query):\n",
    "        \"\"\"Apply NLP to process medical terms and map using embeddings and fuzzy logic.\"\"\"\n",
    "        doc = self.nlp(user_query)\n",
    "\n",
    "        # Process each token and map medical terms using embeddings and fuzzy matching\n",
    "        processed_tokens = [\n",
    "            self.find_closest_term_using_embeddings(token.text) for token in doc\n",
    "        ]\n",
    "\n",
    "        return \" \".join(processed_tokens)\n",
    "\n",
    "    def find_closest_term_using_embeddings(self, token_text):\n",
    "        \"\"\"Find the closest matching term using embeddings and fuzzy matching.\"\"\"\n",
    "        token_text = token_text.lower()\n",
    "\n",
    "        # Use embedding model to find the closest matching term if model exists\n",
    "        if self.embedding_model:\n",
    "            # Example placeholder code to use embeddings for semantic similarity\n",
    "            token_embedding = self.get_embedding(token_text)  # Get the embedding for the token\n",
    "            closest_term = self.find_closest_term_using_similarity(token_embedding)\n",
    "            if closest_term:\n",
    "                return closest_term\n",
    "\n",
    "        # Fallback to fuzzy matching if embedding-based matching fails or is not available\n",
    "        return self.find_closest_standard_term(token_text)\n",
    "\n",
    "    def find_closest_term_using_similarity(self, token_embedding):\n",
    "        \"\"\"Find closest matching term using cosine similarity on embeddings.\"\"\"\n",
    "        # Placeholder: Compare with known disease embeddings\n",
    "        diseases_embeddings = {term: self.get_embedding(term) for term in self.standard_diseases.keys()}\n",
    "        \n",
    "        similarities = {term: cosine_similarity([token_embedding], [embedding])[0][0] for term, embedding in diseases_embeddings.items()}\n",
    "        closest_term = max(similarities, key=similarities.get)\n",
    "        return closest_term if similarities[closest_term] >= 0.75 else None\n",
    "\n",
    "    def get_embedding(self, term):\n",
    "        \"\"\"Retrieve the embedding for a given term using the model.\"\"\"\n",
    "        # Placeholder for an actual embedding model\n",
    "        return np.random.rand(300)  # Random vector as a placeholder\n",
    "\n",
    "    def find_closest_standard_term(self, token_text):\n",
    "        \"\"\"Find the closest matching disease name using fuzzy matching.\"\"\"\n",
    "        for standard_name, variations in self.standard_diseases.items():\n",
    "            variations = [standard_name] + variations\n",
    "            result = process.extractOne(token_text, variations, scorer=fuzz.partial_ratio)\n",
    "\n",
    "            if result:\n",
    "                match, score, *_ = result  # Unpack additional values safely\n",
    "                if score >= 80:  # Threshold for accepting a close match\n",
    "                    return standard_name  # Return the standardized name\n",
    "\n",
    "        return token_text  # Return the original token if no close match is found\n",
    "\n",
    "    def is_fallback_message(self, result_text):\n",
    "        \"\"\"Check if the result text is a fallback message.\"\"\"\n",
    "        result_doc = self.nlp(result_text)\n",
    "        for fallback_doc in self.fallback_docs:\n",
    "            if result_doc.similarity(fallback_doc) >= 0.75:  # Similarity threshold\n",
    "                return True\n",
    "        return False\n",
    "\n",
    "    def ask(self, user_query):\n",
    "        \"\"\"Process user query and interact with the knowledge chain.\"\"\"\n",
    "        enhanced_query = self.nlp_process(user_query)\n",
    "        full_query = f\"{enhanced_query}\"\n",
    "\n",
    "        for attempt in range(3):  # Retry up to 3 times if context is empty\n",
    "            print(f\"Attempt {attempt + 1}...\")\n",
    "            try:\n",
    "                # Pass the enhanced query to the chain for processing\n",
    "                response = self.chain.invoke({\"query\": user_query})\n",
    "\n",
    "                result_text = response[\"result\"].strip()\n",
    "\n",
    "                # If the result is a fallback message, process the intermediate context\n",
    "                if self.is_fallback_message(result_text):\n",
    "                    intermediate_steps = response.get(\"intermediate_steps\", [])\n",
    "                    context_data_list = [step[\"context\"] for step in intermediate_steps if \"context\" in step]\n",
    "                    if context_data_list:\n",
    "                        # Pass context to the LLM for natural language formatting\n",
    "                        all_contexts = [self.format_context2(data) for data in context_data_list]\n",
    "                        combined_readable_context = \"\\n\".join(all_contexts)\n",
    "                        response[\"result\"] = f\"\\n{combined_readable_context}\"\n",
    "\n",
    "                # If the result is non-empty, append it to the previous context\n",
    "                if response[\"result\"].strip():\n",
    "                    self.previous_context += f\"\\n{response['result']}\"\n",
    "                    return response\n",
    "\n",
    "            except Exception as e:\n",
    "                print(f\"Attempt {attempt + 1} failed with error: {e}\")\n",
    "\n",
    "        # If all attempts fail or no context is found, return a fallback response\n",
    "        return {\"result\": \"I don't know the answer\", \"intermediate_steps\": []}\n",
    "\n",
    "    def chat(self):\n",
    "        \"\"\"Start the chatbot conversation.\"\"\"\n",
    "        print(\"Welcome to the Medical Knowledge Chatbot!\")\n",
    "        while True:\n",
    "            user_input = input(\"You: \")\n",
    "            if user_input.lower() == 'exit':\n",
    "                print(\"Goodbye!\")\n",
    "                break\n",
    "\n",
    "            response = self.ask(user_input)\n",
    "            print(f\"Chatbot: {response['result']}\")\n",
    "\n",
    "# Example usage with a placeholder for the knowledge chain\n",
    "class PlaceholderChain:\n",
    "    def invoke(self, query):\n",
    "        # Placeholder: return a mock response based on the query\n",
    "        return {\"result\": f\"Mock response for query: {query['query']}\"}\n",
    "\n",
    "# Instantiate the chatbot\n",
    "embedding_model = None  # Replace with an actual embedding model if available\n",
    "chain = PlaceholderChain()  # Replace with actual knowledge chain\n",
    "chatbot = ChatbotWithEmbeddings(chain, embedding_model=embedding_model)\n",
    "\n",
    "# Start the chat\n",
    "chatbot.chat()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9cfedce3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to the Medical Knowledge Chatbot! \n",
      "You:  Give the organism class for Matrix metallopeptidase-21 \n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (p:Protein {NAME: \"Matrix metallopeptidase-21\"}) RETURN p.ORGANISM_CLASS AS OrganismClass\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context_data_list [[]]\n",
      "context: []\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease)<-[:DRUG_DISEASE_ASSOCIATION]-(dg:Drug) RETURN d.NAME, dg.NAME ORDER BY d.NAME;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d.NAME': ['ACTH Deficiency', ' Isolated'], 'dg.NAME': 'Hydrocortisone'}, {'d.NAME': ['ACTH Deficiency', ' Isolated'], 'dg.NAME': 'Dexamethasone'}, {'d.NAME': ['Abdomen', ' Acute'], 'dg.NAME': 'Enalapril'}, {'d.NAME': ['Abdomen', ' Acute'], 'dg.NAME': 'Isoniazid'}, {'d.NAME': ['Abdomen', ' Acute'], 'dg.NAME': 'Cocaine'}, {'d.NAME': ['Abdomen', ' Acute'], 'dg.NAME': 'Metoclopramide'}, {'d.NAME': ['Abnormalities', ' Drug-Induced'], 'dg.NAME': 'Vitamin_A'}, {'d.NAME': ['Abnormalities', ' Drug-Induced'], 'dg.NAME': 'Folic_acid'}, {'d.NAME': ['Abnormalities', ' Drug-Induced'], 'dg.NAME': 'Nicotine'}, {'d.NAME': ['Abnormalities', ' Drug-Induced'], 'dg.NAME': 'Caffeine'}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "combined_readable_context Here are the important terms: ACTH Deficiency, Isolated, Hydrocortisone, Abdomen, Acute, Enalapril, Isoniazid, Cocaine, Metoclopramide, Abnormalities, Drug-Induced, Vitamin_A, Folic_acid, Nicotine, Caffeine.\n",
      "context: \n",
      "Here are the important terms: ACTH Deficiency, Isolated, Hydrocortisone, Abdomen, Acute, Enalapril, Isoniazid, Cocaine, Metoclopramide, Abnormalities, Drug-Induced, Vitamin_A, Folic_acid, Nicotine, Caffeine.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease {NAME: \"ACTH Deficiency, Isolated\"})-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug) \n",
      "WHERE dr.NAME IN [\"Hydrocortisone\", \"Enalapril\", \"Isoniazid\", \"Cocaine\", \"Metoclopramide\"] \n",
      "RETURN d, dr;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Here are the important terms: ACTH Deficiency, Isolated, Hydrocortisone, Abdomen, Acute, Enalapril, Isoniazid, Cocaine, Metoclopramide, Abnormalities, Drug-Induced, Vitamin_A, Folic_acid, Nicotine, Caffeine.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease {NAME: \"ACTH Deficiency, Isolated\"})-[:DRUG_DISEASE_ASSOCIATION]->(dr:Drug) \n",
      "WHERE dr.NAME IN [\"Hydrocortisone\", \"Enalapril\", \"Isoniazid\", \"Metoclopramide\"] \n",
      "RETURN d, dr \n",
      "\n",
      "UNION \n",
      "\n",
      "MATCH (p:Protein)-[:PROTEIN_DISEASE_ASSOCIATION]->(d:Disease {NAME: \"Abdomen, Acute\"}) \n",
      "MATCH (dr:Drug)-[:DRUG_TARGET|DRUG_ENZYME|DRUG_CARRIER|DRUG_TRANSPORTER]->(p) \n",
      "WHERE dr.NAME IN [\"Cocaine\", \"Nicotine\", \"Caffeine\"] \n",
      "RETURN p, d, dr \n",
      "\n",
      "UNION \n",
      "\n",
      "MATCH (d:Disease {NAME: \"Abnormalities, Drug-Induced\"}) \n",
      "MATCH (dr:Drug {NAME: \"Vitamin_A\"})-[:DRUG_DISEASE_ASSOCIATION]->(d) \n",
      "RETURN d, dr \n",
      "\n",
      "UNION \n",
      "\n",
      "MATCH (d:Disease {NAME: \"Abnormalities, Drug-Induced\"}) \n",
      "MATCH (dr:Drug {NAME: \"Folic_acid\"})-[:DRUG_DISEASE_ASSOCIATION]->(d) \n",
      "RETURN d, dr\u001b[0m\n",
      "Attempt 1 failed with error: {code: Neo.ClientError.Statement.SyntaxError} {message: All sub queries in an UNION must have the same return column names (line 5, column 1 (offset: 190))\n",
      "\"UNION\"\n",
      " ^}\n",
      "Attempt 2 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease {NAME: \"ACTH Deficiency, Isolated\"}) \n",
      "OPTIONAL MATCH (d)-[:DRUG_DISEASE_ASSOCIATION]-(dr:Drug) \n",
      "WHERE dr.NAME IN [\"Hydrocortisone\", \"Enalapril\", \"Isoniazid\", \"Metoclopramide\"] \n",
      "RETURN d, dr;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: Here are the important terms: ACTH Deficiency, Isolated, Hydrocortisone, Abdomen, Acute, Enalapril, Isoniazid, Cocaine, Metoclopramide, Abnormalities, Drug-Induced, Vitamin_A, Folic_acid, Nicotine, Caffeine.\n",
      "You:  exit\n",
      "Goodbye!\n"
     ]
    }
   ],
   "source": [
    "chatbot.chat()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e90e2c44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to the Medical Knowledge Chatbot! \n",
      "You:  Give me some drugs and their enxymes and diseases\n",
      "Attempt 1...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug)-[:DRUG_ENZYME]->(p:Protein)-[:PROTEIN_DISEASE_ASSOCIATION]->(di:Disease) \n",
      "RETURN d.NAME, p.NAME, di.NAME\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d.NAME': 'Pegfilgrastim', 'p.NAME': 'ELNE', 'di.NAME': 'Cyclic neutropenia'}, {'d.NAME': 'Pegfilgrastim', 'p.NAME': 'ELNE', 'di.NAME': ['Neutropenia', ' Severe Congenital', ' Autosomal Dominant 1']}, {'d.NAME': 'Pegfilgrastim', 'p.NAME': 'ELNE', 'di.NAME': 'Neutropenia'}, {'d.NAME': 'Insulin_human', 'p.NAME': 'CP1A2', 'di.NAME': 'Atrophy'}, {'d.NAME': 'Insulin_human', 'p.NAME': 'CP1A2', 'di.NAME': ['Carcinoma', ' Squamous Cell']}, {'d.NAME': 'Insulin_human', 'p.NAME': 'CP1A2', 'di.NAME': 'Cholestasis'}, {'d.NAME': 'Insulin_human', 'p.NAME': 'CP1A2', 'di.NAME': ['Diabetes Mellitus', ' Type 2']}, {'d.NAME': 'Insulin_human', 'p.NAME': 'CP1A2', 'di.NAME': 'Growth Disorders'}, {'d.NAME': 'Insulin_human', 'p.NAME': 'CP1A2', 'di.NAME': ['Carcinoma', ' Hepatocellular']}, {'d.NAME': 'Insulin_human', 'p.NAME': 'CP1A2', 'di.NAME': 'Hepatomegaly'}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Here are some drugs, their enzymes, and diseases: Pegfilgrastim is associated with enzyme ELNE and diseases Cyclic neutropenia, Neutropenia, and Severe Congenital Autosomal Dominant 1. Insulin_human is associated with enzyme CP1A2 and diseases Atrophy, Carcinoma Squamous Cell, Cholestasis, Diabetes Mellitus Type 2, Growth Disorders, Carcinoma Hepatocellular, and Hepatomegaly.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Drug {NAME: \"Pegfilgrastim\"})-[:DRUG_ENZYME]->(e:Protein {NAME: \"ELNE\"})-[:PROTEIN_DISEASE_ASSOCIATION]->(ds:Disease) \n",
      "WHERE ds.NAME IN [\"Cyclic neutropenia\", \"Neutropenia\", \"Severe Congenital Autosomal Dominant 1\"] \n",
      "RETURN d, e, ds \n",
      "\n",
      "UNION \n",
      "\n",
      "MATCH (d:Drug {NAME: \"Insulin_human\"})-[:DRUG_ENZYME]->(e:Protein {NAME: \"CP1A2\"})-[:PROTEIN_DISEASE_ASSOCIATION]->(ds:Disease) \n",
      "WHERE ds.NAME IN [\"Atrophy\", \"Carcinoma Squamous Cell\", \"Cholestasis\", \"Diabetes Mellitus Type 2\", \"Growth Disorders\", \"Carcinoma Hepatocellular\", \"Hepatomegaly\"] \n",
      "RETURN d, e, ds\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['21456630', ' 12010086', ' 15078138', ' 25998211', ' 27789332'], 'DRUG_ATC_CODE': ['ATC:L', ' ATC:L03AA', ' ATC:L03', ' ATC:L03A', ' ATC:L03AA13'], 'PRODUCT': ['Fulphila', ' Lapelga', ' Neulasta', ' Neupopeg', ' Udenyca', ' Ziextenzo'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00019', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Pegfilgrastim', 'SYNONYM': ['Granulocyte_colony_stimulating_factor_pegfilgrastim', ' peg_filgrastim', ' pegfilgrastim_apgf', ' pegfilgrastim_bmez', ' pegfilgrastim_cbqv', ' pegfilgrastim_jmdb'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'e': {'DOMAIN': 'IPR001254', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019955', ' GO:0004175', ' GO:0008201', ' GO:0008233', ' GO:0002020', ' GO:0004252', ' GO:0003714'], 'ACTIVE_SITE': ['IPR018114', ' IPR033116'], 'SHORT_NAME': 'HLE', 'NAME': 'ELNE', 'FAMILY': 'IPR001314', 'RELATED_KEYWORD': ['3D-structure', ' Cytoplasmic vesicle', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Hydrolase', ' Polymorphism', ' Protease', ' Reference proteome', ' Serine protease', ' Signal', ' Zymogen'], 'PROTEIN_EXPRESSED_IN': ['Spleen__Cells_in_red_pulp', ' Spleen', ' Bone_marrow__Hematopoietic_cells', ' Bone_marrow'], 'OTHER_ID': ['Q6B0D9', ' Q6LDP5'], 'RELATED_PUBMED_ID': ['pubmed:3479752', ' pubmed:2902087', ' pubmed:2775493', ' pubmed:2322278', ' pubmed:15489334', ' pubmed:2822677', ' pubmed:3550808', ' pubmed:2501794', ' pubmed:7897245', ' pubmed:3422232', ' pubmed:2462434', ' pubmed:1859409', ' pubmed:10947984', ' pubmed:15140022', ' pubmed:14673143', ' pubmed:19159218', ' pubmed:21269460', ' pubmed:25944712', ' pubmed:2911584', ' pubmed:3391280', ' pubmed:3640709', ' pubmed:10581030', ' pubmed:11001877', ' pubmed:11675333', ' pubmed:12091371', ' pubmed:14962902', ' pubmed:17436313', ' pubmed:18946670', ' pubmed:19036076', ' pubmed:19415009', ' pubmed:20220065', ' pubmed:19927291', ' pubmed:20803142', ' pubmed:21425445', ' pubmed:23463630'], 'FULL_NAME': ['Neutrophil elastase', ' Bone marrow serine protease', ' Elastase-2', ' Human leukocyte elastase', ' Medullasin', ' PMN elastase'], 'GO_CC': ['GO:0035578', ' GO:0005623', ' GO:0009986', ' GO:0062023', ' GO:0005737', ' GO:0070062', ' GO:0005576', ' GO:0005615', ' GO:0045335', ' GO:0030141', ' GO:0035580', ' GO:0017053'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR009003', 'id': 'P08246', 'GO_MF': ['GO:0002438', ' GO:0019730', ' GO:0002812', ' GO:0006874', ' GO:0042742', ' GO:0022617', ' GO:0002523', ' GO:0045079', ' GO:0050922', ' GO:0050728', ' GO:0045415', ' GO:0000122', ' GO:0043312', ' GO:0070947', ' GO:0070945', ' GO:0006909', ' GO:0050778', ' GO:0045416', ' GO:1903238', ' GO:0043406', ' GO:0048661', ' GO:0030163', ' GO:0006508', ' GO:0032496', ' GO:0009411', ' GO:0001878'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}, 'ds': {'DISEASE_SUPERGRP': nan, 'id': 'C536227', 'TYPE': 'SCR_DISEASE', 'NAME': 'Cyclic neutropenia'}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['21456630', ' 12010086', ' 15078138', ' 25998211', ' 27789332'], 'DRUG_ATC_CODE': ['ATC:L', ' ATC:L03AA', ' ATC:L03', ' ATC:L03A', ' ATC:L03AA13'], 'PRODUCT': ['Fulphila', ' Lapelga', ' Neulasta', ' Neupopeg', ' Udenyca', ' Ziextenzo'], 'PRODUCT_STAGE': 'approved', 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00019', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Pegfilgrastim', 'SYNONYM': ['Granulocyte_colony_stimulating_factor_pegfilgrastim', ' peg_filgrastim', ' pegfilgrastim_apgf', ' pegfilgrastim_bmez', ' pegfilgrastim_cbqv', ' pegfilgrastim_jmdb'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'e': {'DOMAIN': 'IPR001254', 'SPECIES': 'HUMAN', 'GO_BP': ['GO:0019955', ' GO:0004175', ' GO:0008201', ' GO:0008233', ' GO:0002020', ' GO:0004252', ' GO:0003714'], 'ACTIVE_SITE': ['IPR018114', ' IPR033116'], 'SHORT_NAME': 'HLE', 'NAME': 'ELNE', 'FAMILY': 'IPR001314', 'RELATED_KEYWORD': ['3D-structure', ' Cytoplasmic vesicle', ' Direct protein sequencing', ' Disease mutation', ' Disulfide bond', ' Glycoprotein', ' Hydrolase', ' Polymorphism', ' Protease', ' Reference proteome', ' Serine protease', ' Signal', ' Zymogen'], 'PROTEIN_EXPRESSED_IN': ['Spleen__Cells_in_red_pulp', ' Spleen', ' Bone_marrow__Hematopoietic_cells', ' Bone_marrow'], 'OTHER_ID': ['Q6B0D9', ' Q6LDP5'], 'RELATED_PUBMED_ID': ['pubmed:3479752', ' pubmed:2902087', ' pubmed:2775493', ' pubmed:2322278', ' pubmed:15489334', ' pubmed:2822677', ' pubmed:3550808', ' pubmed:2501794', ' pubmed:7897245', ' pubmed:3422232', ' pubmed:2462434', ' pubmed:1859409', ' pubmed:10947984', ' pubmed:15140022', ' pubmed:14673143', ' pubmed:19159218', ' pubmed:21269460', ' pubmed:25944712', ' pubmed:2911584', ' pubmed:3391280', ' pubmed:3640709', ' pubmed:10581030', ' pubmed:11001877', ' pubmed:11675333', ' pubmed:12091371', ' pubmed:14962902', ' pubmed:17436313', ' pubmed:18946670', ' pubmed:19036076', ' pubmed:19415009', ' pubmed:20220065', ' pubmed:19927291', ' pubmed:20803142', ' pubmed:21425445', ' pubmed:23463630'], 'FULL_NAME': ['Neutrophil elastase', ' Bone marrow serine protease', ' Elastase-2', ' Human leukocyte elastase', ' Medullasin', ' PMN elastase'], 'GO_CC': ['GO:0035578', ' GO:0005623', ' GO:0009986', ' GO:0062023', ' GO:0005737', ' GO:0070062', ' GO:0005576', ' GO:0005615', ' GO:0045335', ' GO:0030141', ' GO:0035580', ' GO:0017053'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR009003', 'id': 'P08246', 'GO_MF': ['GO:0002438', ' GO:0019730', ' GO:0002812', ' GO:0006874', ' GO:0042742', ' GO:0022617', ' GO:0002523', ' GO:0045079', ' GO:0050922', ' GO:0050728', ' GO:0045415', ' GO:0000122', ' GO:0043312', ' GO:0070947', ' GO:0070945', ' GO:0006909', ' GO:0050778', ' GO:0045416', ' GO:1903238', ' GO:0043406', ' GO:0048661', ' GO:0030163', ' GO:0006508', ' GO:0032496', ' GO:0009411', ' GO:0001878'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}, 'ds': {'DISEASE_SUPERGRP': ['C15.378.553.546', ' C15.378.553', ' C15.378.553.546.184.564', ' C15.378.553.546.184', ' C15.378', ' C15'], 'id': 'D009503', 'TYPE': 'DISEASE', 'NAME': 'Neutropenia'}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['23512415', ' 11118018', ' 10841002', ' 14578244', ' 18715209', ' 27574374'], 'DRUG_ATC_CODE': ['ATC:A10AE01', ' ATC:A10AD', ' ATC:A10AB01', ' ATC:A', ' ATC:A10AE', ' ATC:A10AC', ' ATC:A10AD01', ' ATC:A10', ' ATC:A10AF', ' ATC:A10AB', ' ATC:A10A', ' ATC:A10AF01', ' ATC:A10AC01'], 'PRODUCT': ['Actraphane_30', ' Actraphane_30_Flexpen', ' Actraphane_30_Innolet', ' Actraphane_30_Penfill', ' Actraphane_40_Penfill', ' Actraphane_50_Penfill', ' Actrapid', ' Actrapid_Flexpen', ' Actrapid_Innolet', ' Actrapid_Penfill', ' Afrezza', ' Entuzity_Kwikpen', ' Exubera', ' Humalog_70_30', ' Humulin_10_90_Cartridge', ' Humulin_10_90_Inj', ' Humulin_20_80_Cartridge', ' Humulin_20_80_Inj', ' Humulin_30_70__insulin_Human_Biosynth_Inj_', ' Humulin_30_70_Cartridge', ' Humulin_40_60_Cartridge', ' Humulin_40_60_Inj', ' Humulin_50_50', ' Humulin_50_50_Cartridge', ' Humulin_50_50_Inj', ' Humulin_70_30', ' Humulin_70_30_70_30', ' Humulin_70_30_KwikPen', ' Humulin_L', ' Humulin_N', ' Humulin_N__cartridge_', ' Humulin_N__kwikpen_', ' Humulin_R', ' Humulin_R__kwikpen_', ' Humulin_R_Cartridge', ' Humulin_R_U_500', ' Humulin_R_U_500_KwikPen', ' Humulin_U', ' Insulatard', ' Insulatard_Flexpen', ' Insulatard_Innolet', ' Insulatard_Penfill', ' Insulin_Human_Winthrop_Basal', ' Insulin_Human_Winthrop_Comb_15', ' Insulin_Human_Winthrop_Comb_25', ' Insulin_Human_Winthrop_Comb_30', ' Insulin_Human_Winthrop_Comb_50', ' Insulin_Human_Winthrop_Infusat', ' Insulin_Human_Winthrop_Rapid', ' Insuman_Basal', ' Insuman_Comb_15', ' Insuman_Comb_25', ' Insuman_Comb_30', ' Insuman_Comb_50', ' Insuman_Implantable', ' Insuman_Infusat', ' Insuman_Rapid', ' Myxredlin', ' Novolin_70_30', ' Novolin_Ge_10_90_Penfill_Inj_Sus', ' Novolin_Ge_20_80_Penfill_Inj_Sus', ' Novolin_Ge_30_70', ' Novolin_Ge_30_70_Penfill', ' Novolin_Ge_40_60_Penfill', ' Novolin_Ge_50_50_Penfill', ' Novolin_Ge_Nph', ' Novolin_Ge_Nph_Penfill', ' Novolin_Ge_Toronto', ' Novolin_Ge_Toronto_Penfill', ' Novolin_N', ' Novolin_R', ' Novolin70_30_70_30', ' NovolinN_N', ' Novolinset_Ge_30_70_Inj_Sus', ' Novolinset_Ge_Nph_Inj_Sus_100u_ml', ' Novolinset_Ge_Toronto_Inj_Liq_100u_ml', ' Velosulin'], 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00030', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Insulin_human', 'SYNONYM': ['High_molecular_weight_insulin_human', ' Human_insulin', ' human_insulin__rDNA_', ' Insulin__human_', ' Insulin_human', ' Insulin_human__rDNA_origin_', ' Insulin_Human_Regular', ' Insulin_human_regular__rDNA_', ' Insulin_human__rDNA_origin', ' Insulin_recombinant_human', ' Insulin_recombinant_purified_human', ' Insulin_regular', ' Insulin__human', ' Regular_Insulin__human'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'e': {'SPECIES': 'HUMAN', 'GO_BP': ['GO:0070330', ' GO:0034875', ' GO:0032451', ' GO:0009055', ' GO:0019899', ' GO:0101020', ' GO:0101021', ' GO:0020037', ' GO:0005506', ' GO:0016829', ' GO:0004497', ' GO:0016491', ' GO:0016712'], 'NAME': 'CP1A2', 'FAMILY': ['IPR001128', ' IPR002401', ' IPR008066'], 'RELATED_KEYWORD': ['3D-structure', ' Direct protein sequencing', ' Endoplasmic reticulum', ' Fatty acid metabolism', ' Glycoprotein', ' Heme', ' Iron', ' Lipid metabolism', ' Lyase', ' Membrane', ' Metal-binding', ' Microsome', ' Monooxygenase', ' Oxidoreductase', ' Polymorphism', ' Reference proteome', ' Steroid metabolism', ' Sterol metabolism'], 'PROTEIN_EXPRESSED_IN': ['Liver__Hepatocytes', ' Liver'], 'OTHER_ID': ['Q6NWU3', ' Q6NWU5', ' Q9BXX7', ' Q9UK49'], 'RELATED_PUBMED_ID': ['pubmed:3755823', ' pubmed:3462722', ' pubmed:2575218', ' pubmed:3681487', ' pubmed:11207026', ' pubmed:15489334', ' pubmed:3517618', ' pubmed:3000715', ' pubmed:9435160', ' pubmed:10681376', ' pubmed:11555828', ' pubmed:12865317', ' pubmed:14725854', ' pubmed:19965576', ' pubmed:21068195', ' pubmed:21576599', ' pubmed:24275569', ' pubmed:26988023', ' pubmed:17311915', ' pubmed:9884316', ' pubmed:11295848', ' pubmed:14563787', ' pubmed:15469410', ' pubmed:15770072', ' pubmed:15643613', ' pubmed:16522833'], 'FULL_NAME': ['Cytochrome P450 1A2 {ECO:0000303|PubMed:14725854}', ' CYPIA2', ' Cholesterol 25-hydroxylase {ECO:0000305|PubMed:21576599}', ' Cytochrome P(3)450', ' Cytochrome P450 4', ' Cytochrome P450-P3', ' Hydroperoxy icosatetraenoate dehydratase {ECO:0000305|PubMed:21068195}'], 'CONSERVED_SITE': 'IPR017972', 'GO_CC': ['GO:0005789', ' GO:0043231'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036396', 'id': 'P05177', 'GO_MF': ['GO:0009820', ' GO:0045333', ' GO:0071276', ' GO:0071280', ' GO:0008203', ' GO:0018894', ' GO:0042737', ' GO:0017144', ' GO:0019373', ' GO:0008210', ' GO:0042738', ' GO:0046483', ' GO:0050665', ' GO:0042759', ' GO:0030324', ' GO:0032259', ' GO:0032787', ' GO:0016098', ' GO:0097267', ' GO:0055114', ' GO:0071615', ' GO:0070989', ' GO:0006778', ' GO:0009791', ' GO:0010468', ' GO:0032355', ' GO:0035902', ' GO:0032496', ' GO:0042572', ' GO:0006706', ' GO:0009403', ' GO:0006805'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}, 'ds': {'DISEASE_SUPERGRP': ['C23.300.070', ' C23.300', ' C23'], 'id': 'D001284', 'TYPE': 'DISEASE', 'NAME': 'Atrophy'}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['23512415', ' 11118018', ' 10841002', ' 14578244', ' 18715209', ' 27574374'], 'DRUG_ATC_CODE': ['ATC:A10AE01', ' ATC:A10AD', ' ATC:A10AB01', ' ATC:A', ' ATC:A10AE', ' ATC:A10AC', ' ATC:A10AD01', ' ATC:A10', ' ATC:A10AF', ' ATC:A10AB', ' ATC:A10A', ' ATC:A10AF01', ' ATC:A10AC01'], 'PRODUCT': ['Actraphane_30', ' Actraphane_30_Flexpen', ' Actraphane_30_Innolet', ' Actraphane_30_Penfill', ' Actraphane_40_Penfill', ' Actraphane_50_Penfill', ' Actrapid', ' Actrapid_Flexpen', ' Actrapid_Innolet', ' Actrapid_Penfill', ' Afrezza', ' Entuzity_Kwikpen', ' Exubera', ' Humalog_70_30', ' Humulin_10_90_Cartridge', ' Humulin_10_90_Inj', ' Humulin_20_80_Cartridge', ' Humulin_20_80_Inj', ' Humulin_30_70__insulin_Human_Biosynth_Inj_', ' Humulin_30_70_Cartridge', ' Humulin_40_60_Cartridge', ' Humulin_40_60_Inj', ' Humulin_50_50', ' Humulin_50_50_Cartridge', ' Humulin_50_50_Inj', ' Humulin_70_30', ' Humulin_70_30_70_30', ' Humulin_70_30_KwikPen', ' Humulin_L', ' Humulin_N', ' Humulin_N__cartridge_', ' Humulin_N__kwikpen_', ' Humulin_R', ' Humulin_R__kwikpen_', ' Humulin_R_Cartridge', ' Humulin_R_U_500', ' Humulin_R_U_500_KwikPen', ' Humulin_U', ' Insulatard', ' Insulatard_Flexpen', ' Insulatard_Innolet', ' Insulatard_Penfill', ' Insulin_Human_Winthrop_Basal', ' Insulin_Human_Winthrop_Comb_15', ' Insulin_Human_Winthrop_Comb_25', ' Insulin_Human_Winthrop_Comb_30', ' Insulin_Human_Winthrop_Comb_50', ' Insulin_Human_Winthrop_Infusat', ' Insulin_Human_Winthrop_Rapid', ' Insuman_Basal', ' Insuman_Comb_15', ' Insuman_Comb_25', ' Insuman_Comb_30', ' Insuman_Comb_50', ' Insuman_Implantable', ' Insuman_Infusat', ' Insuman_Rapid', ' Myxredlin', ' Novolin_70_30', ' Novolin_Ge_10_90_Penfill_Inj_Sus', ' Novolin_Ge_20_80_Penfill_Inj_Sus', ' Novolin_Ge_30_70', ' Novolin_Ge_30_70_Penfill', ' Novolin_Ge_40_60_Penfill', ' Novolin_Ge_50_50_Penfill', ' Novolin_Ge_Nph', ' Novolin_Ge_Nph_Penfill', ' Novolin_Ge_Toronto', ' Novolin_Ge_Toronto_Penfill', ' Novolin_N', ' Novolin_R', ' Novolin70_30_70_30', ' NovolinN_N', ' Novolinset_Ge_30_70_Inj_Sus', ' Novolinset_Ge_Nph_Inj_Sus_100u_ml', ' Novolinset_Ge_Toronto_Inj_Liq_100u_ml', ' Velosulin'], 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00030', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Insulin_human', 'SYNONYM': ['High_molecular_weight_insulin_human', ' Human_insulin', ' human_insulin__rDNA_', ' Insulin__human_', ' Insulin_human', ' Insulin_human__rDNA_origin_', ' Insulin_Human_Regular', ' Insulin_human_regular__rDNA_', ' Insulin_human__rDNA_origin', ' Insulin_recombinant_human', ' Insulin_recombinant_purified_human', ' Insulin_regular', ' Insulin__human', ' Regular_Insulin__human'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'e': {'SPECIES': 'HUMAN', 'GO_BP': ['GO:0070330', ' GO:0034875', ' GO:0032451', ' GO:0009055', ' GO:0019899', ' GO:0101020', ' GO:0101021', ' GO:0020037', ' GO:0005506', ' GO:0016829', ' GO:0004497', ' GO:0016491', ' GO:0016712'], 'NAME': 'CP1A2', 'FAMILY': ['IPR001128', ' IPR002401', ' IPR008066'], 'RELATED_KEYWORD': ['3D-structure', ' Direct protein sequencing', ' Endoplasmic reticulum', ' Fatty acid metabolism', ' Glycoprotein', ' Heme', ' Iron', ' Lipid metabolism', ' Lyase', ' Membrane', ' Metal-binding', ' Microsome', ' Monooxygenase', ' Oxidoreductase', ' Polymorphism', ' Reference proteome', ' Steroid metabolism', ' Sterol metabolism'], 'PROTEIN_EXPRESSED_IN': ['Liver__Hepatocytes', ' Liver'], 'OTHER_ID': ['Q6NWU3', ' Q6NWU5', ' Q9BXX7', ' Q9UK49'], 'RELATED_PUBMED_ID': ['pubmed:3755823', ' pubmed:3462722', ' pubmed:2575218', ' pubmed:3681487', ' pubmed:11207026', ' pubmed:15489334', ' pubmed:3517618', ' pubmed:3000715', ' pubmed:9435160', ' pubmed:10681376', ' pubmed:11555828', ' pubmed:12865317', ' pubmed:14725854', ' pubmed:19965576', ' pubmed:21068195', ' pubmed:21576599', ' pubmed:24275569', ' pubmed:26988023', ' pubmed:17311915', ' pubmed:9884316', ' pubmed:11295848', ' pubmed:14563787', ' pubmed:15469410', ' pubmed:15770072', ' pubmed:15643613', ' pubmed:16522833'], 'FULL_NAME': ['Cytochrome P450 1A2 {ECO:0000303|PubMed:14725854}', ' CYPIA2', ' Cholesterol 25-hydroxylase {ECO:0000305|PubMed:21576599}', ' Cytochrome P(3)450', ' Cytochrome P450 4', ' Cytochrome P450-P3', ' Hydroperoxy icosatetraenoate dehydratase {ECO:0000305|PubMed:21068195}'], 'CONSERVED_SITE': 'IPR017972', 'GO_CC': ['GO:0005789', ' GO:0043231'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036396', 'id': 'P05177', 'GO_MF': ['GO:0009820', ' GO:0045333', ' GO:0071276', ' GO:0071280', ' GO:0008203', ' GO:0018894', ' GO:0042737', ' GO:0017144', ' GO:0019373', ' GO:0008210', ' GO:0042738', ' GO:0046483', ' GO:0050665', ' GO:0042759', ' GO:0030324', ' GO:0032259', ' GO:0032787', ' GO:0016098', ' GO:0097267', ' GO:0055114', ' GO:0071615', ' GO:0070989', ' GO:0006778', ' GO:0009791', ' GO:0010468', ' GO:0032355', ' GO:0035902', ' GO:0032496', ' GO:0042572', ' GO:0006706', ' GO:0009403', ' GO:0006805'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}, 'ds': {'DISEASE_SUPERGRP': ['C06.130', ' C06', ' C06.130.120', ' C06.130.120.135'], 'id': 'D002779', 'TYPE': 'DISEASE', 'NAME': 'Cholestasis'}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['23512415', ' 11118018', ' 10841002', ' 14578244', ' 18715209', ' 27574374'], 'DRUG_ATC_CODE': ['ATC:A10AE01', ' ATC:A10AD', ' ATC:A10AB01', ' ATC:A', ' ATC:A10AE', ' ATC:A10AC', ' ATC:A10AD01', ' ATC:A10', ' ATC:A10AF', ' ATC:A10AB', ' ATC:A10A', ' ATC:A10AF01', ' ATC:A10AC01'], 'PRODUCT': ['Actraphane_30', ' Actraphane_30_Flexpen', ' Actraphane_30_Innolet', ' Actraphane_30_Penfill', ' Actraphane_40_Penfill', ' Actraphane_50_Penfill', ' Actrapid', ' Actrapid_Flexpen', ' Actrapid_Innolet', ' Actrapid_Penfill', ' Afrezza', ' Entuzity_Kwikpen', ' Exubera', ' Humalog_70_30', ' Humulin_10_90_Cartridge', ' Humulin_10_90_Inj', ' Humulin_20_80_Cartridge', ' Humulin_20_80_Inj', ' Humulin_30_70__insulin_Human_Biosynth_Inj_', ' Humulin_30_70_Cartridge', ' Humulin_40_60_Cartridge', ' Humulin_40_60_Inj', ' Humulin_50_50', ' Humulin_50_50_Cartridge', ' Humulin_50_50_Inj', ' Humulin_70_30', ' Humulin_70_30_70_30', ' Humulin_70_30_KwikPen', ' Humulin_L', ' Humulin_N', ' Humulin_N__cartridge_', ' Humulin_N__kwikpen_', ' Humulin_R', ' Humulin_R__kwikpen_', ' Humulin_R_Cartridge', ' Humulin_R_U_500', ' Humulin_R_U_500_KwikPen', ' Humulin_U', ' Insulatard', ' Insulatard_Flexpen', ' Insulatard_Innolet', ' Insulatard_Penfill', ' Insulin_Human_Winthrop_Basal', ' Insulin_Human_Winthrop_Comb_15', ' Insulin_Human_Winthrop_Comb_25', ' Insulin_Human_Winthrop_Comb_30', ' Insulin_Human_Winthrop_Comb_50', ' Insulin_Human_Winthrop_Infusat', ' Insulin_Human_Winthrop_Rapid', ' Insuman_Basal', ' Insuman_Comb_15', ' Insuman_Comb_25', ' Insuman_Comb_30', ' Insuman_Comb_50', ' Insuman_Implantable', ' Insuman_Infusat', ' Insuman_Rapid', ' Myxredlin', ' Novolin_70_30', ' Novolin_Ge_10_90_Penfill_Inj_Sus', ' Novolin_Ge_20_80_Penfill_Inj_Sus', ' Novolin_Ge_30_70', ' Novolin_Ge_30_70_Penfill', ' Novolin_Ge_40_60_Penfill', ' Novolin_Ge_50_50_Penfill', ' Novolin_Ge_Nph', ' Novolin_Ge_Nph_Penfill', ' Novolin_Ge_Toronto', ' Novolin_Ge_Toronto_Penfill', ' Novolin_N', ' Novolin_R', ' Novolin70_30_70_30', ' NovolinN_N', ' Novolinset_Ge_30_70_Inj_Sus', ' Novolinset_Ge_Nph_Inj_Sus_100u_ml', ' Novolinset_Ge_Toronto_Inj_Liq_100u_ml', ' Velosulin'], 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00030', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Insulin_human', 'SYNONYM': ['High_molecular_weight_insulin_human', ' Human_insulin', ' human_insulin__rDNA_', ' Insulin__human_', ' Insulin_human', ' Insulin_human__rDNA_origin_', ' Insulin_Human_Regular', ' Insulin_human_regular__rDNA_', ' Insulin_human__rDNA_origin', ' Insulin_recombinant_human', ' Insulin_recombinant_purified_human', ' Insulin_regular', ' Insulin__human', ' Regular_Insulin__human'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'e': {'SPECIES': 'HUMAN', 'GO_BP': ['GO:0070330', ' GO:0034875', ' GO:0032451', ' GO:0009055', ' GO:0019899', ' GO:0101020', ' GO:0101021', ' GO:0020037', ' GO:0005506', ' GO:0016829', ' GO:0004497', ' GO:0016491', ' GO:0016712'], 'NAME': 'CP1A2', 'FAMILY': ['IPR001128', ' IPR002401', ' IPR008066'], 'RELATED_KEYWORD': ['3D-structure', ' Direct protein sequencing', ' Endoplasmic reticulum', ' Fatty acid metabolism', ' Glycoprotein', ' Heme', ' Iron', ' Lipid metabolism', ' Lyase', ' Membrane', ' Metal-binding', ' Microsome', ' Monooxygenase', ' Oxidoreductase', ' Polymorphism', ' Reference proteome', ' Steroid metabolism', ' Sterol metabolism'], 'PROTEIN_EXPRESSED_IN': ['Liver__Hepatocytes', ' Liver'], 'OTHER_ID': ['Q6NWU3', ' Q6NWU5', ' Q9BXX7', ' Q9UK49'], 'RELATED_PUBMED_ID': ['pubmed:3755823', ' pubmed:3462722', ' pubmed:2575218', ' pubmed:3681487', ' pubmed:11207026', ' pubmed:15489334', ' pubmed:3517618', ' pubmed:3000715', ' pubmed:9435160', ' pubmed:10681376', ' pubmed:11555828', ' pubmed:12865317', ' pubmed:14725854', ' pubmed:19965576', ' pubmed:21068195', ' pubmed:21576599', ' pubmed:24275569', ' pubmed:26988023', ' pubmed:17311915', ' pubmed:9884316', ' pubmed:11295848', ' pubmed:14563787', ' pubmed:15469410', ' pubmed:15770072', ' pubmed:15643613', ' pubmed:16522833'], 'FULL_NAME': ['Cytochrome P450 1A2 {ECO:0000303|PubMed:14725854}', ' CYPIA2', ' Cholesterol 25-hydroxylase {ECO:0000305|PubMed:21576599}', ' Cytochrome P(3)450', ' Cytochrome P450 4', ' Cytochrome P450-P3', ' Hydroperoxy icosatetraenoate dehydratase {ECO:0000305|PubMed:21068195}'], 'CONSERVED_SITE': 'IPR017972', 'GO_CC': ['GO:0005789', ' GO:0043231'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036396', 'id': 'P05177', 'GO_MF': ['GO:0009820', ' GO:0045333', ' GO:0071276', ' GO:0071280', ' GO:0008203', ' GO:0018894', ' GO:0042737', ' GO:0017144', ' GO:0019373', ' GO:0008210', ' GO:0042738', ' GO:0046483', ' GO:0050665', ' GO:0042759', ' GO:0030324', ' GO:0032259', ' GO:0032787', ' GO:0016098', ' GO:0097267', ' GO:0055114', ' GO:0071615', ' GO:0070989', ' GO:0006778', ' GO:0009791', ' GO:0010468', ' GO:0032355', ' GO:0035902', ' GO:0032496', ' GO:0042572', ' GO:0006706', ' GO:0009403', ' GO:0006805'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}, 'ds': {'DISEASE_SUPERGRP': ['C23.550', ' C23.550.393', ' C23'], 'id': 'D006130', 'TYPE': 'DISEASE', 'NAME': 'Growth Disorders'}}, {'d': {'NumRotatableBonds': nan, 'DRUG_INDICATION_ASSOCIATION': nan, 'LogP': nan, 'CLASS': 'Carboxylic_Acids_and_Derivatives', 'DRUG_SIDEEFFECT_ASSOCIATION': nan, 'DRUGCOMP': nan, 'COMPCLAS': nan, 'SUBSTITUENT': nan, 'PUBMED_ARTICLE': ['23512415', ' 11118018', ' 10841002', ' 14578244', ' 18715209', ' 27574374'], 'DRUG_ATC_CODE': ['ATC:A10AE01', ' ATC:A10AD', ' ATC:A10AB01', ' ATC:A', ' ATC:A10AE', ' ATC:A10AC', ' ATC:A10AD01', ' ATC:A10', ' ATC:A10AF', ' ATC:A10AB', ' ATC:A10A', ' ATC:A10AF01', ' ATC:A10AC01'], 'PRODUCT': ['Actraphane_30', ' Actraphane_30_Flexpen', ' Actraphane_30_Innolet', ' Actraphane_30_Penfill', ' Actraphane_40_Penfill', ' Actraphane_50_Penfill', ' Actrapid', ' Actrapid_Flexpen', ' Actrapid_Innolet', ' Actrapid_Penfill', ' Afrezza', ' Entuzity_Kwikpen', ' Exubera', ' Humalog_70_30', ' Humulin_10_90_Cartridge', ' Humulin_10_90_Inj', ' Humulin_20_80_Cartridge', ' Humulin_20_80_Inj', ' Humulin_30_70__insulin_Human_Biosynth_Inj_', ' Humulin_30_70_Cartridge', ' Humulin_40_60_Cartridge', ' Humulin_40_60_Inj', ' Humulin_50_50', ' Humulin_50_50_Cartridge', ' Humulin_50_50_Inj', ' Humulin_70_30', ' Humulin_70_30_70_30', ' Humulin_70_30_KwikPen', ' Humulin_L', ' Humulin_N', ' Humulin_N__cartridge_', ' Humulin_N__kwikpen_', ' Humulin_R', ' Humulin_R__kwikpen_', ' Humulin_R_Cartridge', ' Humulin_R_U_500', ' Humulin_R_U_500_KwikPen', ' Humulin_U', ' Insulatard', ' Insulatard_Flexpen', ' Insulatard_Innolet', ' Insulatard_Penfill', ' Insulin_Human_Winthrop_Basal', ' Insulin_Human_Winthrop_Comb_15', ' Insulin_Human_Winthrop_Comb_25', ' Insulin_Human_Winthrop_Comb_30', ' Insulin_Human_Winthrop_Comb_50', ' Insulin_Human_Winthrop_Infusat', ' Insulin_Human_Winthrop_Rapid', ' Insuman_Basal', ' Insuman_Comb_15', ' Insuman_Comb_25', ' Insuman_Comb_30', ' Insuman_Comb_50', ' Insuman_Implantable', ' Insuman_Infusat', ' Insuman_Rapid', ' Myxredlin', ' Novolin_70_30', ' Novolin_Ge_10_90_Penfill_Inj_Sus', ' Novolin_Ge_20_80_Penfill_Inj_Sus', ' Novolin_Ge_30_70', ' Novolin_Ge_30_70_Penfill', ' Novolin_Ge_40_60_Penfill', ' Novolin_Ge_50_50_Penfill', ' Novolin_Ge_Nph', ' Novolin_Ge_Nph_Penfill', ' Novolin_Ge_Toronto', ' Novolin_Ge_Toronto_Penfill', ' Novolin_N', ' Novolin_R', ' Novolin70_30_70_30', ' NovolinN_N', ' Novolinset_Ge_30_70_Inj_Sus', ' Novolinset_Ge_Nph_Inj_Sus_100u_ml', ' Novolinset_Ge_Toronto_Inj_Liq_100u_ml', ' Velosulin'], 'PRODUCT_STAGE': ['approved', ' investigational'], 'KINGDOM': 'Organic_Compounds', 'DRUGBANK_NAME': nan, 'id': 'DB00030', 'TYPE': 'DRUG', 'MolecularWeight': nan, 'SUBCLASS': 'Amino_Acids__Peptides__and_Analogues', 'SUPERCLASS': 'Organic_Acids', 'TPSA': nan, 'NumHDonors': nan, 'NAME': 'Insulin_human', 'SYNONYM': ['High_molecular_weight_insulin_human', ' Human_insulin', ' human_insulin__rDNA_', ' Insulin__human_', ' Insulin_human', ' Insulin_human__rDNA_origin_', ' Insulin_Human_Regular', ' Insulin_human_regular__rDNA_', ' Insulin_human__rDNA_origin', ' Insulin_recombinant_human', ' Insulin_recombinant_purified_human', ' Insulin_regular', ' Insulin__human', ' Regular_Insulin__human'], 'LipinskiViolationCount': nan, 'SOURCE': 'BIOKG', 'NumHAcceptors': nan, 'DRUGINKE': nan, 'DIRECT_PARENT': 'Peptides', 'ALTERNATIVE_PARENT': nan}, 'e': {'SPECIES': 'HUMAN', 'GO_BP': ['GO:0070330', ' GO:0034875', ' GO:0032451', ' GO:0009055', ' GO:0019899', ' GO:0101020', ' GO:0101021', ' GO:0020037', ' GO:0005506', ' GO:0016829', ' GO:0004497', ' GO:0016491', ' GO:0016712'], 'NAME': 'CP1A2', 'FAMILY': ['IPR001128', ' IPR002401', ' IPR008066'], 'RELATED_KEYWORD': ['3D-structure', ' Direct protein sequencing', ' Endoplasmic reticulum', ' Fatty acid metabolism', ' Glycoprotein', ' Heme', ' Iron', ' Lipid metabolism', ' Lyase', ' Membrane', ' Metal-binding', ' Microsome', ' Monooxygenase', ' Oxidoreductase', ' Polymorphism', ' Reference proteome', ' Steroid metabolism', ' Sterol metabolism'], 'PROTEIN_EXPRESSED_IN': ['Liver__Hepatocytes', ' Liver'], 'OTHER_ID': ['Q6NWU3', ' Q6NWU5', ' Q9BXX7', ' Q9UK49'], 'RELATED_PUBMED_ID': ['pubmed:3755823', ' pubmed:3462722', ' pubmed:2575218', ' pubmed:3681487', ' pubmed:11207026', ' pubmed:15489334', ' pubmed:3517618', ' pubmed:3000715', ' pubmed:9435160', ' pubmed:10681376', ' pubmed:11555828', ' pubmed:12865317', ' pubmed:14725854', ' pubmed:19965576', ' pubmed:21068195', ' pubmed:21576599', ' pubmed:24275569', ' pubmed:26988023', ' pubmed:17311915', ' pubmed:9884316', ' pubmed:11295848', ' pubmed:14563787', ' pubmed:15469410', ' pubmed:15770072', ' pubmed:15643613', ' pubmed:16522833'], 'FULL_NAME': ['Cytochrome P450 1A2 {ECO:0000303|PubMed:14725854}', ' CYPIA2', ' Cholesterol 25-hydroxylase {ECO:0000305|PubMed:21576599}', ' Cytochrome P(3)450', ' Cytochrome P450 4', ' Cytochrome P450-P3', ' Hydroperoxy icosatetraenoate dehydratase {ECO:0000305|PubMed:21068195}'], 'CONSERVED_SITE': 'IPR017972', 'GO_CC': ['GO:0005789', ' GO:0043231'], 'HOMOLOGOUS_SUPERFAMILY': 'IPR036396', 'id': 'P05177', 'GO_MF': ['GO:0009820', ' GO:0045333', ' GO:0071276', ' GO:0071280', ' GO:0008203', ' GO:0018894', ' GO:0042737', ' GO:0017144', ' GO:0019373', ' GO:0008210', ' GO:0042738', ' GO:0046483', ' GO:0050665', ' GO:0042759', ' GO:0030324', ' GO:0032259', ' GO:0032787', ' GO:0016098', ' GO:0097267', ' GO:0055114', ' GO:0071615', ' GO:0070989', ' GO:0006778', ' GO:0009791', ' GO:0010468', ' GO:0032355', ' GO:0035902', ' GO:0032496', ' GO:0042572', ' GO:0006706', ' GO:0009403', ' GO:0006805'], 'ORGANISM_CLASS': ['Eukaryota', ' Metazoa', ' Chordata', ' Craniata', ' Vertebrata', ' Euteleostomi', ' Mammalia', ' Eutheria', ' Euarchontoglires', ' Primates', ' Haplorrhini', ' Catarrhini', ' Hominidae', ' Homo.']}, 'ds': {'DISEASE_SUPERGRP': ['C23.300.775', ' C06', ' C23.300.775.525', ' C23', ' C06.552.416', ' C23.300', ' C06.552'], 'id': 'D006529', 'TYPE': 'DISEASE', 'NAME': 'Hepatomegaly'}}]\u001b[0m\n",
      "Attempt 1 failed with error: Error code: 413 - {'error': {'message': 'Request too large for model `llama3-70b-8192` in organization `org_01j9yv18bffqksnp3w7k6gzqk0` on tokens per minute (TPM): Limit 6000, Requested 7883, please reduce your message size and try again. Visit https://console.groq.com/docs/rate-limits for more information.', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}\n",
      "Attempt 2 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d1:Drug {NAME:\"Pegfilgrastim\"})-[:DRUG_ENZYME]->(p1:Protein {NAME:\"ELNE\"})-[:PROTEIN_DISEASE_ASSOCIATION]->(disease1),\n",
      "(d1)-[:DRUG_DISEASE_ASSOCIATION]->(disease1)\n",
      "WHERE disease1.NAME IN [\"Cyclic neutropenia\", \"Neutropenia\", \"Severe Congenital Autosomal Dominant 1\"]\n",
      "MATCH (d2:Drug {NAME:\"Insulin_human\"})-[:DRUG_ENZYME]->(p2:Protein {NAME:\"CP1A2\"})-[:PROTEIN_DISEASE_ASSOCIATION]->(disease2),\n",
      "(d2)-[:DRUG_DISEASE_ASSOCIATION]->(disease2)\n",
      "WHERE disease2.NAME IN [\"Atrophy\", \"Carcinoma Squamous Cell\", \"Cholestasis\", \"Diabetes Mellitus Type 2\", \"Growth Disorders\", \"Carcinoma Hepatocellular\", \"Hepatomegaly\"]\n",
      "RETURN d1, p1, disease1, d2, p2, disease2\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "context: Here are the important terms: Pegfilgrastim, ELNE, Cyclic neutropenia, Neutropenia, Severe Congenital Autosomal Dominant 1, Insulin_human, CP1A2, Atrophy, Carcinoma Squamous Cell, Cholestasis, Diabetes Mellitus Type 2, Growth Disorders, Carcinoma Hepatocellular, and Hepatomegaly.\n",
      "Attempt 1 for format_context2...\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new GraphCypherQAChain chain...\u001b[0m\n",
      "Generated Cypher:\n",
      "\u001b[32;1m\u001b[1;3mMATCH (d:Disease) WHERE d.NAME IN [\"Cyclic neutropenia\", \"Neutropenia\", \"Severe Congenital Autosomal Dominant 1\", \"Atrophy\", \"Carcinoma Squamous Cell\", \"Cholestasis\", \"Diabetes Mellitus Type 2\", \"Growth Disorders\", \"Carcinoma Hepatocellular\", \"Hepatomegaly\"] \n",
      "OPTIONAL MATCH (d)-[:DRUG_DISEASE_ASSOCIATION]->(drug:Drug) \n",
      "OPTIONAL MATCH (drug)-[:DRUG_TARGET]->(protein:Protein) \n",
      "WHERE drug.NAME = \"Pegfilgrastim\" OR drug.NAME = \"ELNE\" \n",
      "OPTIONAL MATCH (protein)-[:PROTEIN_DISEASE_ASSOCIATION]->(d) \n",
      "RETURN d, drug, protein;\u001b[0m\n",
      "Full Context:\n",
      "\u001b[32;1m\u001b[1;3m[{'d': {'DISEASE_SUPERGRP': nan, 'id': 'C536227', 'TYPE': 'SCR_DISEASE', 'NAME': 'Cyclic neutropenia'}, 'drug': None, 'protein': None}, {'d': {'DISEASE_SUPERGRP': ['C23.300.070', ' C23.300', ' C23'], 'id': 'D001284', 'TYPE': 'DISEASE', 'NAME': 'Atrophy'}, 'drug': None, 'protein': None}, {'d': {'DISEASE_SUPERGRP': ['C06.130', ' C06', ' C06.130.120', ' C06.130.120.135'], 'id': 'D002779', 'TYPE': 'DISEASE', 'NAME': 'Cholestasis'}, 'drug': None, 'protein': None}, {'d': {'DISEASE_SUPERGRP': ['C23.550', ' C23.550.393', ' C23'], 'id': 'D006130', 'TYPE': 'DISEASE', 'NAME': 'Growth Disorders'}, 'drug': None, 'protein': None}, {'d': {'DISEASE_SUPERGRP': ['C23.300.775', ' C06', ' C23.300.775.525', ' C23', ' C06.552.416', ' C23.300', ' C06.552'], 'id': 'D006529', 'TYPE': 'DISEASE', 'NAME': 'Hepatomegaly'}, 'drug': None, 'protein': None}, {'d': {'DISEASE_SUPERGRP': ['C15.378.553.546', ' C15.378.553', ' C15.378.553.546.184.564', ' C15.378.553.546.184', ' C15.378', ' C15'], 'id': 'D009503', 'TYPE': 'DISEASE', 'NAME': 'Neutropenia'}, 'drug': None, 'protein': None}]\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "CHATBOT: Here are the important terms that match the provided information: Cyclic neutropenia, Atrophy, Cholestasis, Growth Disorders, Hepatomegaly, and Neutropenia.\n",
      "You:  exit\n",
      "Goodbye!\n"
     ]
    }
   ],
   "source": [
    "chatbot.chat() # calling the chatbot \n",
    "# correcting the word enxymes "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
